{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7f33ada0-6685-46cf-abe3-cd2101806b4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import pybedtools\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "\n",
    "pd.set_option(\"display.max_columns\", 1000)\n",
    "pd.set_option(\"display.max_rows\", 100)\n",
    "\n",
    "import statsmodels\n",
    "import statsmodels.api as sm\n",
    "import warnings\n",
    "\n",
    "import seaborn as sns\n",
    "import palettable\n",
    "import matplotlib.pyplot as plt\n",
    "from scripts import manuscript_colors as mc\n",
    "\n",
    "from scripts import helper\n",
    "\n",
    "bold_10 = palettable.cartocolors.qualitative.Bold_10.mpl_colors\n",
    "\n",
    "from scripts import aesthetics, sv_plot\n",
    "\n",
    "aesthetics.activate_paper_rcParams('illustrator')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fbca85d-a331-409c-823b-c30f672c9a51",
   "metadata": {},
   "source": [
    "# Targeted gene testing\n",
    "\n",
    "In this notebook, we examine specific genes that have been implicated in the pathogenesis of these pediatric cancers. Essentially, this is a focused look to determine if SVs might affect these genes in cases vs. controls. The coding approach taken here is very high level, and it generates results for a predetermined list of disease-gene pairings. As a result, it can be a bit difficult to parse. The only part of the manuscript where these results are shown is in Figure S10.\n",
    "\n",
    "This notebook also uses the language of \"discovery\" vs \"validation\" to refer to GMKF and St. Jude. This is an older approach to thinking about these datasets that is not used in the paper."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0067ee62-52e0-4f90-a417-f06165af63d1",
   "metadata": {},
   "source": [
    "# Load the data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91808b0d-2382-42f1-a3fb-2d966cfe9ecd",
   "metadata": {},
   "source": [
    "## Define paths to our data\n",
    "We have 5 pieces of data.\n",
    "\n",
    "Two of these files depends on whether we're analyzing the discovery or validation dataset:\n",
    "1. Our SV file (in bed format)\n",
    "2. Our dosage file (a dense matrix of SV genotypes)\n",
    "\n",
    "Three files do not depend on this:\n",
    "\n",
    "3. Our metadata files (which define which samples are cases, which are controls, etc)\n",
    "4. Our list of genes (which defines which regions we analyze)\n",
    "5. Our gene gtf (which defines the locations of our regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2f3f6e30-f428-4abd-85d9-6aeb76d0a695",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define the folder name for all our results\n",
    "folder_name = \"targeted-genes\"\n",
    "\n",
    "# define SVs and dosages for discovery and validation\n",
    "svs_discovery_path = \"gs://vanallen-pedsv-analysis/beds/PedSV.v2.5.4.case_control_cohort.analysis_samples.sites.bed.gz\"\n",
    "svs_validation_path = \"gs://vanallen-pedsv-analysis/beds/PedSV.v2.5.4.trio_cohort.analysis_samples.sites.bed.gz\"\n",
    "\n",
    "dosages_discovery_path = \"gs://vanallen-pedsv-analysis/beds/PedSV.v2.5.4.case_control_cohort.analysis_samples.allele_dosages.bed.gz\"\n",
    "dosages_validation_path = \"gs://vanallen-pedsv-analysis/beds/PedSV.v2.5.4.trio_cohort.analysis_samples.allele_dosages.bed.gz\"\n",
    "\n",
    "# define metadata\n",
    "metadata_path = \"gs://vanallen-pedsv-analysis/sample_info/PedSV.v2.5.4.cohort_metadata.w_control_assignments.tsv.gz\"\n",
    "\n",
    "discovery_samples_path = \"gs://vanallen-pedsv-analysis/sample_info/PedSV.v2.5.4.case_control_analysis_cohort.samples.list\"\n",
    "validation_samples_path = \"gs://vanallen-pedsv-analysis/sample_info/PedSV.v2.5.4.trio_analysis_cohort.samples.list\"\n",
    "\n",
    "# define which genes are tested\n",
    "genes_to_test_path = \"data/genes-for-analysis.csv\"\n",
    "\n",
    "# define reference for gene locations - GENCODE v43 (hg38)\n",
    "gtf_path = \"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_43/gencode.v43.basic.annotation.gtf.gz\"\n",
    "\n",
    "# define the window size for our analysis (centered on the gene's center) - use 1 Mb\n",
    "window_size = 1e6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9e92f32e-ca2f-44e4-a448-43f6bbafb43e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "      <th>disease</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EWSR1</td>\n",
       "      <td>EWSR1-206</td>\n",
       "      <td>ewing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FLI1</td>\n",
       "      <td>FLI1-205</td>\n",
       "      <td>ewing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MYCN</td>\n",
       "      <td>MYCN-201</td>\n",
       "      <td>neuroblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TERT</td>\n",
       "      <td>TERT-201</td>\n",
       "      <td>neuroblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ALK</td>\n",
       "      <td>ALK-201</td>\n",
       "      <td>neuroblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TP53</td>\n",
       "      <td>TP53-201</td>\n",
       "      <td>osteosarcoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>RB1</td>\n",
       "      <td>RB1-201</td>\n",
       "      <td>osteosarcoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ATRX</td>\n",
       "      <td>ATRX-201</td>\n",
       "      <td>neuroblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SHANK2</td>\n",
       "      <td>SHANK2-214</td>\n",
       "      <td>neuroblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1-203</td>\n",
       "      <td>neuroblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>CDKN2A-201</td>\n",
       "      <td>ewing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>DLG2</td>\n",
       "      <td>DLG2-203</td>\n",
       "      <td>osteosarcoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>TP53</td>\n",
       "      <td>TP53-201</td>\n",
       "      <td>neuroblastoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>TP53</td>\n",
       "      <td>TP53-201</td>\n",
       "      <td>ewing</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene  transcript        disease\n",
       "0    EWSR1   EWSR1-206          ewing\n",
       "1     FLI1    FLI1-205          ewing\n",
       "2     MYCN    MYCN-201  neuroblastoma\n",
       "3     TERT    TERT-201  neuroblastoma\n",
       "4      ALK     ALK-201  neuroblastoma\n",
       "5     TP53    TP53-201   osteosarcoma\n",
       "6      RB1     RB1-201   osteosarcoma\n",
       "7     ATRX    ATRX-201  neuroblastoma\n",
       "8   SHANK2  SHANK2-214  neuroblastoma\n",
       "9    BRCA1   BRCA1-203  neuroblastoma\n",
       "10  CDKN2A  CDKN2A-201          ewing\n",
       "11    DLG2    DLG2-203   osteosarcoma\n",
       "12    TP53    TP53-201  neuroblastoma\n",
       "13    TP53    TP53-201          ewing"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_to_test = pd.read_csv(genes_to_test_path)\n",
    "genes_to_test"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dac6eac3-e55f-4eca-add4-a295bbaad1aa",
   "metadata": {},
   "source": [
    "## Load the data into memory"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e279969-4afb-4c27-8816-4adc0e153a5e",
   "metadata": {},
   "source": [
    "First, load the SVs and dosages (takes a bit). We use a dictionary so we can simply refer to keys later (because discovery and validation are processed the same way downstream)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3c278d65-c906-4419-9b70-ddc8b121ca12",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jettcrowdis/opt/anaconda3/envs/pedsv-test-conda/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3457: DtypeWarning: Columns (10,12,13,34,35,37,39,41,45,46,47,48,72,73,87,88,106,107,121,122,136,137,155,156,170,171,185,186,204,205,219,220,234,235,253,254,268,269,283,284,302,303,366,367,381,382,400,401,415,416,430,431,449,450,464,465,479,480,498,499,513,514,528,529,547,548,562,563,577,578,596,597,645,646,647,648,649,650,690,691,692,693,694,695,713,714,728,729,743,744,758,759,773,774,788,789,803,804,818,819,833,834,848,849,863,864,878,879,893,894,908,909,923,924,931,937,943) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "/Users/jettcrowdis/opt/anaconda3/envs/pedsv-test-conda/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3457: DtypeWarning: Columns (34,35,37,39,41,45,46,72,73,87,88,106,107,121,122,136,137,155,156,170,171,185,186,204,205,219,220,234,235,253,254,268,269,283,284,302,303,366,367,381,382,400,401,415,416,430,431,449,450,464,465,479,480,498,499,513,514,528,529,547,548,562,563,577,578,596,597,645,646,647,648,649,650,690,691,692,693,694,695,713,714,728,729,743,744,758,759,773,774,788,789,803,804,818,819,833,834,848,849,863,864,878,879,893,894,908,909,923,924,931,937,943) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    }
   ],
   "source": [
    "###############\n",
    "##### SVs #####\n",
    "###############\n",
    "svs = {}\n",
    "\n",
    "svs[\"discovery\"] = pd.read_csv(\n",
    "    svs_discovery_path,\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "\n",
    "svs[\"validation\"] = pd.read_csv(\n",
    "    svs_validation_path,\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "###############\n",
    "### Dosages ###\n",
    "###############\n",
    "dosages = {}\n",
    "\n",
    "dosages[\"discovery\"] = pd.read_csv(\n",
    "    dosages_discovery_path,\n",
    "    sep=\"\\t\",\n",
    "    index_col=False,\n",
    ")\n",
    "\n",
    "dosages[\"validation\"] = pd.read_csv(\n",
    "    dosages_validation_path,\n",
    "    sep=\"\\t\",\n",
    "    index_col=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65d2ecaa-b5b8-48f2-b228-34ff0135813a",
   "metadata": {},
   "source": [
    "Next we load metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e0786514-4fc1-4af2-ab3e-ff5b33f0eff3",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(\n",
    "    metadata_path,\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "\n",
    "# add a sex label to metadata\n",
    "metadata[\"sex\"] = (metadata[\"chrX_CopyNumber\"].round() < 2).astype(int)\n",
    "\n",
    "###############\n",
    "### Samples ###\n",
    "###############\n",
    "samples = {}\n",
    "\n",
    "samples[\"discovery\"] = pd.read_csv(\n",
    "    discovery_samples_path,\n",
    "    header=None,\n",
    ")[0].to_list()\n",
    "\n",
    "samples[\"validation\"] = pd.read_csv(\n",
    "    validation_samples_path,\n",
    "    header=None,\n",
    ")[0].to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83f037b0-16f8-4a4f-8a64-42505d20886e",
   "metadata": {},
   "source": [
    "And finally, load the gene data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8b620b3c-1a1a-43ba-b993-db598fc558f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_to_test = pd.read_csv(genes_to_test_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a466b669-f765-4f40-a41c-2c61f6eac4ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = pd.read_csv(\n",
    "    gtf_path,\n",
    "    comment=\"#\",\n",
    "    sep=\"\\t\",\n",
    "    header=None,\n",
    "    names=[\n",
    "        \"seqname\",\n",
    "        \"source\",\n",
    "        \"feature\",\n",
    "        \"start\",\n",
    "        \"end\",\n",
    "        \"score\",\n",
    "        \"strand\",\n",
    "        \"frame\",\n",
    "        \"attribute\",\n",
    "    ],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d7dca2a-459a-4e4e-8fe0-be632ebb7dd4",
   "metadata": {},
   "source": [
    "## Make data folders\n",
    "\n",
    "Here, we define folders that will contain all the data.\n",
    "\n",
    "__Note: To avoid confusion, this code will erase any pre-existing runs with the same name. Use with caution.__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c500efc9-293c-41d8-9f06-fc2c06bce7a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "if os.path.isdir(f\"data/{folder_name}\"):\n",
    "    shutil.rmtree(f\"data/{folder_name}\")\n",
    "\n",
    "# make the data directory\n",
    "os.mkdir(f\"data/{folder_name}\")\n",
    "\n",
    "for index, row in genes_to_test.iterrows():\n",
    "    g = row[\"gene\"]\n",
    "    disease = row[\"disease\"]\n",
    "    os.mkdir(f\"data/{folder_name}/{g}-{disease}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30446f71-fb21-4b56-9cfa-f6bd07da0a5d",
   "metadata": {},
   "source": [
    "# Extract gene locations\n",
    "\n",
    "Here, we extract from the gtf the locations of our genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2644789e-da72-4cb8-96fb-3ad7401bd7b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_to_transcript = dict(genes_to_test[[\"gene\", \"transcript\"]].values)\n",
    "\n",
    "gene_locs = []\n",
    "for index, row in genes_to_test.iterrows():\n",
    "    g, t, disease = row.values\n",
    "\n",
    "    # extract all elements of the gene and store them\n",
    "    g_gtf = gtf[(gtf[\"attribute\"].str.contains(f\"transcript_name \\\"{t}\\\"\"))]\n",
    "    g_gtf.to_csv(f\"data/{folder_name}/{g}-{disease}/{g}-gtf.csv\", index=False)\n",
    "    \n",
    "    # we also define the start and end of the gene here\n",
    "    gene_loc = gtf[(gtf[\"attribute\"].str.contains(f\"gene_name \\\"{g}\\\"\")) & (gtf['feature'] == 'gene')]\n",
    "    assert len(gene_loc) == 1\n",
    "    chrom, start, end = gene_loc[['seqname', 'start', 'end']].values[0]\n",
    "    \n",
    "    # store gene location\n",
    "    gene_locs.append([g, chrom, start, end])\n",
    "    \n",
    "gene_locs = pd.DataFrame(gene_locs, columns = ['gene', 'chrom', 'start', 'end'])\n",
    "\n",
    "# drop duplicates - happens with the same gene and different diseases\n",
    "gene_locs = gene_locs.drop_duplicates()\n",
    "\n",
    "# and expand the window according to our window size\n",
    "gene_locs['window_start'] = (gene_locs['start'] - window_size / 2).astype(int)\n",
    "gene_locs['window_end'] = (gene_locs['end'] + window_size / 2).astype(int)\n",
    "\n",
    "# merge if we haven't already\n",
    "if 'chrom' not in genes_to_test.columns:\n",
    "    genes_to_test = genes_to_test.merge(gene_locs, on = 'gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ffd97820-3d82-4389-8fab-1e0c156a6ed9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "      <th>disease</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EWSR1</td>\n",
       "      <td>EWSR1-206</td>\n",
       "      <td>ewing</td>\n",
       "      <td>chr22</td>\n",
       "      <td>29268009</td>\n",
       "      <td>29300525</td>\n",
       "      <td>28768009</td>\n",
       "      <td>29800525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FLI1</td>\n",
       "      <td>FLI1-205</td>\n",
       "      <td>ewing</td>\n",
       "      <td>chr11</td>\n",
       "      <td>128686535</td>\n",
       "      <td>128813267</td>\n",
       "      <td>128186535</td>\n",
       "      <td>129313267</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene transcript disease  chrom      start        end  window_start  \\\n",
       "0  EWSR1  EWSR1-206   ewing  chr22   29268009   29300525      28768009   \n",
       "1   FLI1   FLI1-205   ewing  chr11  128686535  128813267     128186535   \n",
       "\n",
       "   window_end  \n",
       "0    29800525  \n",
       "1   129313267  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_to_test.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "acf74a7d-e7d0-47ad-9b8b-4eb8af64d46d",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_to_test.to_csv(f'data/{folder_name}/gene-locations.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bacdf66-3868-4693-a339-0f46a0d6925e",
   "metadata": {},
   "source": [
    "# Define the SVs of interest for each analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd75f170-1f9e-4236-82ef-15ead15e2282",
   "metadata": {},
   "source": [
    "Doing this is somewhat complicated, as we need to juggle both discovery and validation cohorts. We do the following:\n",
    "\n",
    "1. Define the cases and controls for the analysis based on the disease type\n",
    "2. Define the genomic region of interest, based on the gene data defined above\n",
    "3. Subset the SVs down to SVs in this region\n",
    "4. Subset the dosage down to those SVs and the samples of interest\n",
    "\n",
    "We use a kind of round-about disease-based approach because then we only need to make three copies of large dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "419f4503-dac9-404a-8f0c-9835e0dde9e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ewing discovery\n",
      "ewing validation\n",
      "osteosarcoma discovery\n",
      "osteosarcoma validation\n",
      "neuroblastoma discovery\n",
      "neuroblastoma validation\n"
     ]
    }
   ],
   "source": [
    "for disease in ['ewing', 'osteosarcoma', 'neuroblastoma']:\n",
    "    \n",
    "    # define the number of genes we're looking at in this disease\n",
    "    disease_genes = genes_to_test[genes_to_test['disease'] == disease]\n",
    "    if len(disease_genes) == 0:\n",
    "        continue\n",
    "    \n",
    "    # we subset discovery and validation datasets to relevant samples and SVs\n",
    "    for cohort in [\"discovery\", \"validation\"]:\n",
    "        print(disease, cohort)\n",
    "        \n",
    "        # osteosarcoma has no validation datasets, so we skip\n",
    "        if disease == 'osteosarcoma' and cohort == 'validation':\n",
    "            continue\n",
    "        \n",
    "        ########################\n",
    "        #### DEFINE SAMPLES ####\n",
    "        ########################\n",
    "        \n",
    "        # this term for cohort is used in the metadata\n",
    "        study_label = \"case_control\" if cohort == \"discovery\" else \"trio\"\n",
    "\n",
    "        # define the samples\n",
    "        cohort_samples = samples[cohort]\n",
    "\n",
    "        # get the cases - must belong to disease, must be right cohort, and must be a case included for analysis\n",
    "        cases = metadata[\n",
    "            (metadata[\"study_phase\"] == study_label)\n",
    "            & (metadata[f\"{disease}_case\"])\n",
    "        ][\"entity:sample_id\"].to_list()\n",
    "        cases = sorted(set(cohort_samples) & set(cases))\n",
    "\n",
    "        # get the controls - must be a control, right cohort, and a control for that analysis\n",
    "        controls = metadata[\n",
    "            (metadata[\"study_phase\"] == study_label)\n",
    "            & (metadata[f\"{disease}_control\"])\n",
    "        ][\"entity:sample_id\"].to_list()\n",
    "        controls = sorted(set(cohort_samples) & set(controls))\n",
    "\n",
    "        # some quick checks\n",
    "        assert set(cases) & set(controls) == set()\n",
    "        \n",
    "        ################################\n",
    "        #### DEFINE SVS AND DOSAGES ####\n",
    "        ################################\n",
    "        \n",
    "        # fetch our cohort data (discovery or validation)\n",
    "        cohort_svs = svs[cohort].copy()\n",
    "        cohort_dosages = dosages[cohort].copy()\n",
    "        \n",
    "        # subset the dosage matrix down to the cases and controls for our disease\n",
    "        cohort_dosages = cohort_dosages[['#chr', 'start', 'end', 'ID'] + cases + controls].copy()\n",
    "        \n",
    "        # drop any SVs that have no alternative genotypes in these samples (i.e. aren't present)\n",
    "        cohort_sv_counts = cohort_dosages.iloc[:, 4:].fillna(0).sum(axis = 1)\n",
    "\n",
    "        # subset down SVs and dosages to just SVs with nonzero dosages\n",
    "        cohort_svs = cohort_svs[cohort_sv_counts != 0]\n",
    "        cohort_dosages = cohort_dosages[cohort_sv_counts != 0]\n",
    "        \n",
    "        #########################################\n",
    "        #### SUBSET SVS and DOSAGES TO GENES ####\n",
    "        #########################################\n",
    "        gene_locs = disease_genes[['chrom', 'window_start', 'window_end', 'gene']]\n",
    "        \n",
    "        # bedtools expects specific column names and ordering\n",
    "        gene_locs.columns = ['chrom', 'start', 'end', 'gene']\n",
    "        \n",
    "        # create bedtools object for our genes\n",
    "        gene_bed = pybedtools.BedTool(\n",
    "            gene_locs.to_csv(sep=\"\\t\", index=False, header=False), from_string=True\n",
    "        )\n",
    "        \n",
    "        # create bedtool object for svs\n",
    "        sv_bed = pybedtools.BedTool(\n",
    "            cohort_svs[[\"#chrom\", \"start\", \"end\", \"name\"]].to_csv(sep=\"\\t\", index=False, header=False),\n",
    "            from_string=True,\n",
    "        )\n",
    "\n",
    "        svs_in_regions = sv_bed.intersect(gene_bed, wo=True)\n",
    "        svs_in_regions = svs_in_regions.to_dataframe()\n",
    "\n",
    "        # drop unnecessary columns\n",
    "        svs_in_regions = svs_in_regions.iloc[:, [0, 1, 2, 3, 7]].copy()\n",
    "        svs_in_regions.columns = [\"chrom\", \"start\", \"end\", \"name\", \"gene\"]\n",
    "        \n",
    "        ############################\n",
    "        #### ADD IN COMPLEX SVS ####\n",
    "        ############################\n",
    "        \n",
    "        # This step is not completely necessary, but I include it for completeness.\n",
    "        # The idea is that basing our SVs on strict overlap isn't good enough--some SVs\n",
    "        # may involve in the region as part of a complex rearrangement.\n",
    "        cohort_complex_svs = cohort_svs[(cohort_svs[\"svtype\"] == \"CPX\")][\n",
    "                                    [\n",
    "                                        \"name\",\n",
    "                                        \"#chrom\",\n",
    "                                        \"start\",\n",
    "                                        \"end\",\n",
    "                                        \"svtype\",\n",
    "                                        \"CHR2\",\n",
    "                                        \"CPX_INTERVALS\",\n",
    "                                        \"CPX_TYPE\",\n",
    "                                        \"END\",\n",
    "                                    ]\n",
    "                                ].copy()\n",
    "        \n",
    "        # split each component SV--represented by CPX_INTERVALS--into their own rows\n",
    "        cohort_complex_svs = cohort_complex_svs.assign(\n",
    "            intervals=cohort_complex_svs[\"CPX_INTERVALS\"].str.split(\",\")\n",
    "        ).explode(\"intervals\")\n",
    "\n",
    "        # convert this data to a proper bed for intersection\n",
    "        cohort_complex_sv_bed = cohort_complex_svs[[\"name\", \"intervals\"]].copy()\n",
    "        cohort_complex_sv_bed[\"chrom\"] = cohort_complex_sv_bed[\"intervals\"].apply(\n",
    "            lambda interv: interv.split(\":\")[0].split(\"_\")[-1]\n",
    "        )\n",
    "        cohort_complex_sv_bed[\"start\"] = cohort_complex_sv_bed[\"intervals\"].apply(\n",
    "            lambda interv: interv.split(\":\")[-1].split(\"-\")[0]\n",
    "        )\n",
    "        cohort_complex_sv_bed[\"end\"] = cohort_complex_sv_bed[\"intervals\"].apply(\n",
    "            lambda interv: interv.split(\":\")[-1].split(\"-\")[1]\n",
    "        )\n",
    "\n",
    "        # convert to a bedtool\n",
    "        cohort_complex_sv_bed = pybedtools.BedTool(\n",
    "            cohort_complex_sv_bed[[\"chrom\", \"start\", \"end\", \"name\"]].to_csv(\n",
    "                sep=\"\\t\", index=False, header=False\n",
    "            ),\n",
    "            from_string=True,\n",
    "        )\n",
    "\n",
    "        # do our intersection\n",
    "        complex_svs_in_regions = cohort_complex_sv_bed.intersect(gene_bed, wo=True)\n",
    "        complex_svs_in_regions = complex_svs_in_regions.to_dataframe()\n",
    "\n",
    "        # drop unnecessary columns\n",
    "        if len(complex_svs_in_regions):\n",
    "            complex_svs_in_regions = complex_svs_in_regions.iloc[:, [0, 1, 2, 3, 7]].copy()\n",
    "            complex_svs_in_regions.columns = [\"chrom\", \"start\", \"end\", \"name\", \"gene\"]\n",
    "        else:\n",
    "            complex_svs_in_regions = pd.DataFrame(\n",
    "                columns=[\"chrom\", \"start\", \"end\", \"name\", \"gene\"]\n",
    "            )\n",
    "        \n",
    "        ######################################################\n",
    "        #### DEFINE AND FILTER OUR SVS IN THESE REGIONS ######\n",
    "        ######################################################\n",
    "        \n",
    "        # combine normal and complex SVs together\n",
    "        svs_in_regions_names = pd.concat(\n",
    "            [svs_in_regions[[\"name\", \"gene\"]], complex_svs_in_regions[[\"name\", \"gene\"]]]\n",
    "        )\n",
    "\n",
    "        # drop duplicates (only happens if an SV is simple but also part of a complex SV)\n",
    "        svs_in_regions_names = svs_in_regions_names.drop_duplicates(subset = ['name'])\n",
    "        \n",
    "        # subset SVs\n",
    "        subset_svs = cohort_svs.set_index(\"name\").loc[svs_in_regions_names[\"name\"]].reset_index()\n",
    "        subset_svs.insert(0, \"gene\", svs_in_regions_names[\"gene\"].values)\n",
    "\n",
    "        # convert AF to float. We extract an AF for CNVs as well\n",
    "        subset_svs['AF'] = subset_svs.apply(lambda row: float(row['AF']) if row['svtype'] != 'CNV' else row['CN_NONREF_FREQ'], axis = 1)\n",
    "\n",
    "        # only include SVs that pass the FILTER and are in our dosage matrix\n",
    "        svs_to_keep = subset_svs[\n",
    "            (\n",
    "                (subset_svs[\"FILTER\"].isin([\"PASS\", \"MULTIALLELIC\"]))\n",
    "                & (subset_svs[\"name\"].isin(cohort_dosages[\"ID\"].to_list()))\n",
    "            )\n",
    "        ].copy()\n",
    "\n",
    "        svs_to_keep[\"rare\"] = svs_to_keep[\"AF\"] < 0.01\n",
    "        \n",
    "        # subset dosage matrix\n",
    "        subset_dosages = cohort_dosages.set_index('ID').loc[svs_to_keep[\"name\"].to_list()].reset_index()\n",
    "\n",
    "        # we add in the information about which gene is affected (the first column is just the SV name, so we drop it)\n",
    "        subsetted_gene_dosages = (\n",
    "            svs_to_keep[[\"name\", \"gene\", \"AF\", \"rare\"]]\n",
    "            .merge(subset_dosages, right_on=\"ID\", left_on=\"name\")\n",
    "            .drop(columns = ['ID'])\n",
    "        )\n",
    "\n",
    "        #########################\n",
    "        #### EXPORT OUR DATA ####\n",
    "        #########################\n",
    "        \n",
    "        # extract cohort metadata\n",
    "        cohort_metadata = metadata[metadata['entity:sample_id'].isin(cases + controls)]\n",
    "        \n",
    "        # we export everything to reconstruct whatever analyses we want.\n",
    "        for g in gene_locs['gene']:\n",
    "            \n",
    "            export_prefix = f'data/{folder_name}/'\n",
    "\n",
    "            # export sample info\n",
    "            cohort_metadata.to_csv(export_prefix + f'{g}-{disease}/{cohort}-metadata.csv', index=False)\n",
    "            \n",
    "            gene_svs = svs_to_keep[svs_to_keep['gene'] == g]\n",
    "            gene_dosages = subsetted_gene_dosages[subsetted_gene_dosages['gene'] == g]\n",
    "            \n",
    "            gene_svs.to_csv(export_prefix + f'{g}-{disease}/{cohort}-svs.csv', index=False)\n",
    "            gene_dosages.to_csv(export_prefix + f'{g}-{disease}/{cohort}-dosages.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4289b5b-bf66-4e7c-ab44-19018455e997",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Calculate allele fractions\n",
    "\n",
    "While we could have done this above, I think it's worth separating this out for compartmentalization.\n",
    "\n",
    "We devise two metrics for cases and controls--one listing the allele frequency of each SV and another listing the number of samples with an SV in the gene. We need to convert the dosage (allele count) to allele frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "807f8047-7e48-4657-94ed-73e58b1b0f79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EWSR1 ewing\n",
      "FLI1 ewing\n",
      "MYCN neuroblastoma\n",
      "TERT neuroblastoma\n",
      "ALK neuroblastoma\n",
      "TP53 osteosarcoma\n",
      "TP53 neuroblastoma\n",
      "TP53 ewing\n",
      "RB1 osteosarcoma\n",
      "ATRX neuroblastoma\n",
      "SHANK2 neuroblastoma\n",
      "BRCA1 neuroblastoma\n",
      "CDKN2A ewing\n",
      "DLG2 osteosarcoma\n"
     ]
    }
   ],
   "source": [
    "for index, row in genes_to_test.iterrows():\n",
    "    gene, disease = row[['gene', 'disease']].values\n",
    "    print(gene, disease)\n",
    "    \n",
    "    file_prefix = f'data/{folder_name}/{gene}-{disease}/'\n",
    "    for cohort in ['discovery', 'validation']:\n",
    "\n",
    "        # osteosarcoma has no validation datasets, so we skip\n",
    "        if disease == 'osteosarcoma' and cohort == 'validation':\n",
    "            continue\n",
    "    \n",
    "        ######################\n",
    "        #### READ IN DATA ####\n",
    "        ######################\n",
    "        analysis_dosages = pd.read_csv(file_prefix + f'{cohort}-dosages.csv')\n",
    "        analysis_svs = pd.read_csv(file_prefix + f'{cohort}-svs.csv')\n",
    "        analysis_metadata = pd.read_csv(file_prefix + f'{cohort}-metadata.csv')\n",
    "\n",
    "        # get the cases and controls\n",
    "        analysis_cases = analysis_metadata.query(f'disease == \"{disease}\"')['entity:sample_id'].to_list()\n",
    "        analysis_controls = analysis_metadata.query(f'disease == \"control\"')['entity:sample_id'].to_list()\n",
    "        \n",
    "        ###########################\n",
    "        #### EXTRACT SV COUNTS ####\n",
    "        ###########################\n",
    "        \n",
    "        case_control_counts = []\n",
    "        case_control_samples_with_svs = []\n",
    "\n",
    "        for sample_group, label in zip(\n",
    "            [analysis_cases, analysis_controls], [\"cases\", \"controls\"]\n",
    "        ):\n",
    "\n",
    "            # subset to these samples\n",
    "            sample_dosages = analysis_dosages[\n",
    "                [\"name\", \"gene\", \"#chr\", \"start\", \"end\", \"AF\", \"rare\"] + sample_group\n",
    "            ]\n",
    "\n",
    "            # first, count the SVs. This function counts the number of samples with each genotype.\n",
    "            sv_counts = (\n",
    "                analysis_dosages[sample_group]\n",
    "                .apply(lambda row: row.dropna().value_counts(), axis=1)\n",
    "                .fillna(0)\n",
    "                .astype(int)\n",
    "            )\n",
    "\n",
    "            # name columns to genotypes\n",
    "            sv_counts.columns = f\"{label}_\" + sv_counts.columns.astype(int).astype(str)\n",
    "            sv_counts[\"name\"] = sample_dosages[\"name\"]\n",
    "\n",
    "            case_control_counts.append(sv_counts)\n",
    "\n",
    "            # now count the number of samples with at least one SV around each gene.\n",
    "            # we groupby gene, then determine whether each sample has an SV in that gene, then\n",
    "            # count these samples up per gene.\n",
    "            num_samples_with_svs = sample_dosages.groupby([\"gene\"]).apply(\n",
    "                lambda group: (group[sample_group] > 0).any(axis=0).sum()\n",
    "            )\n",
    "\n",
    "            num_samples_with_svs.name = label + \"_all\"\n",
    "            num_samples_with_svs = num_samples_with_svs / len(sample_group)\n",
    "            case_control_samples_with_svs.append(num_samples_with_svs)\n",
    "\n",
    "            # repeat for rare SVs\n",
    "            num_samples_with_svs_rare = (\n",
    "                sample_dosages.query(\"rare == True\")\n",
    "                .groupby([\"gene\"])\n",
    "                .apply(lambda group: (group[sample_group] > 0).any(axis=0).sum())\n",
    "            )\n",
    "\n",
    "            num_samples_with_svs_rare.name = label + \"_rare\"\n",
    "            num_samples_with_svs_rare = num_samples_with_svs_rare / len(sample_group)\n",
    "            case_control_samples_with_svs.append(num_samples_with_svs_rare)\n",
    "\n",
    "        # merge cases and controls together\n",
    "        case_control_sv_counts = sample_dosages.iloc[:, :7].reset_index(drop=True)\n",
    "        for df in case_control_counts:\n",
    "            case_control_sv_counts = case_control_sv_counts.merge(df)\n",
    "\n",
    "        case_control_sv_counts[\"midpoint\"] = np.mean(\n",
    "            case_control_sv_counts[[\"start\", \"end\"]], axis=1\n",
    "        )\n",
    "\n",
    "        # and total up the sample counts\n",
    "        case_control_sample_counts = pd.DataFrame(case_control_samples_with_svs).T\n",
    "        \n",
    "        ######################################\n",
    "        #### CALCULATE ALLELE FREQUENCIES ####\n",
    "        ######################################\n",
    "        \n",
    "        # Remove CNVs and calculate separately if present\n",
    "        if case_control_sv_counts['name'].str.contains('_CNV_').sum() > 0:\n",
    "            cnv_counts = case_control_sv_counts[case_control_sv_counts['name'].str.contains('CNV')].copy()\n",
    "            \n",
    "            for label in ['cases', 'controls']:\n",
    "                columns = [col for col in cnv_counts.columns if col.startswith(label)]\n",
    "                genotypes = [int(c.split('_')[-1]) for c in columns]\n",
    "                alleles = (cnv_counts[columns] * np.array(genotypes))\n",
    "                average_cn = alleles.sum(axis = 1) / cnv_counts[columns].sum(axis = 1)\n",
    "\n",
    "                # overwrite the value with cases and controls\n",
    "                cnv_counts[f'average_CN_{label}'] = average_cn\n",
    "            \n",
    "            # export\n",
    "            cnv_counts.to_csv(file_prefix + f'{cohort}-cnv-genotypes.csv', index=False)\n",
    "        \n",
    "            # and remove any trace of CNVs from our genotypes\n",
    "            cnv_columns = [c for c in case_control_sv_counts.columns if '_' in c and c.split('_')[-1] not in ['0', '1', '2']]\n",
    "            case_control_sv_counts = case_control_sv_counts.drop(index = cnv_counts.index, columns = cnv_columns).reset_index(drop = True)\n",
    "            \n",
    "        # now we do our SVs. We calculate their allele fraction\n",
    "        for label in [\"cases\", \"controls\"]:\n",
    "\n",
    "            # a quick check to see if there are non-0/1/2 genotypes\n",
    "            columns = [col for col in case_control_sv_counts.columns if col.startswith(label)]\n",
    "            assert columns == [f\"{label}_{i}\" for i in range(3)]\n",
    "\n",
    "            alt = (case_control_sv_counts[columns] * np.array([0, 1, 2])).sum(axis=1)\n",
    "            ref = case_control_sv_counts[columns].sum(axis=1) * 2\n",
    "            case_control_sv_counts[f\"AF_{label}\"] = alt / ref\n",
    "            \n",
    "        # export\n",
    "        case_control_sv_counts.to_csv(file_prefix + f'{cohort}-sv-genotypes.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "324aacdf-0865-4bee-bce2-fc8b9f7f236a",
   "metadata": {},
   "source": [
    "# Calculate associations\n",
    "\n",
    "Up next, we'll calculate the different statistical tests, e.g. individual SV tests and burden testing.\n",
    "\n",
    "We begin by doing burden testing. We do it in a few different ways:\n",
    "\n",
    "1. `logit(case_status) ~ count_of_rare/common_SVs + ancestry_PCs + sex`\n",
    "2. `logit(case_status) ~ count_of_rare/common_SVs + sex`\n",
    "3. `logit(case_status) ~ count_of_rare/common_SVs`\n",
    "\n",
    "We store these all in the same file, and we calculate for both discovery and validation. In the analysis without ancestry, we subset to individuals of predominantly European ancestry."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3e750628-49f7-453e-afe8-cd83044ce0a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAyYAAAEdCAYAAAD5FXN1AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/NK7nSAAAACXBIWXMAAA9hAAAPYQGoP6dpAABUjklEQVR4nO3deVxU5f4H8M/AwLDvKAgIsiduuZRLJmIupWKm2cXrQipYEmlo98b1V+It42aKLaKpLWh5MZdM01Jb3FLTLNNcEREFJGXfGbbz+4PrJLHLzJw5M5/36zWvnLM930OHh/M951lkgiAIICIiIiIiEpGR2AEQERERERExMSEiIiIiItExMSEiIiIiItExMSEiIiIiItExMSEiIiIiItExMSEiIiIiItExMSEiIiIiItExMSEiIiIiItExMSEiIiIiItExMSEiIiIiItHJxQ6AiAzX3r17kZ6ejoKCAvzf//2f2OEQERGRiPjGhIhEM3bsWHh5eaGsrEzsUIiIiEhkTEyISDSXLl3C2LFjUVdXB0EQxA6HiIiIRMSmXEQkmt9++w07d+6EnZ0dZDKZ2OEQERGRiGSCxB9TPv/881i7dq3YYWhFUFAQAODChQsiR0LUNllZWZgyZQo2b94MhUKB+fPnw9XVFdbW1njjjTfafJzk5GQkJyc3ua68vBzfffedukImiWM9aVgM6R4A4PVN+k/yb0yysrLEDkFrEhISxA6BqM2Ki4vx9ttvw87ODgCwbt06REREYOTIkZgxYwbS09Ph5eXVpmOFhYUhLCysyXWhoaFqipj0AetJw2JI9wAAr2/Sf5JLTJRKJZRKpep7XV2diNFo1+jRo8UOgajNbGxs8M477yA8PBwAkJ2drUpE3N3dkZWV1ebEhKitWE+SPuP1TfpOcp3f4+PjYWtrq/pcvXpV7JCIqA08PT2RkZEBAMjMzISbm9t9HUepVKK4uFj1MaSHE0RERPpMcolJbGwsioqKVB8/Pz+xQ9KaoKAgVftSIqmJjIzEhg0bEBUVBV9f3/t+W8KHE9QS1pOkz3h9k76TXFMuhUIBhUKh+m5kJLncisigJCUlqf7dXAf29oiNjUVMTIzq+9SpUzt8TCIiIhKf5BITQ+5jQkR8OEFERKSvJJeYxMfHY+nSparv/v7+IkZDRNrGhxNEuiM7O7vJYcCbWv7666/j119/ha2tLYYNG4aZM2di7ty5sLKyQl5eHtatWwdzc3ORz4iIxCS5R42G3MeEiNjHhEiX3B0G/N1338XNmzeRnp7e7PIjR46gS5cuqKurw8CBA3HkyBF06tQJq1atwrBhw7B161ZxT4aIRCe5NyZsxkFk2NjHhEh3NDcMeFPLY2NjERwcjJycHMyYMQPh4eGqbTw8PPDzzz83WUZLE6xqYh4TZXkVinNK4ezpgCvHrqHodgkeeqqP2sshosYkl5iIIW7onxMaxR2NaWFLzdq1a5doZRPpCj6coJZosp6MG5og6t8AXXR3GHA/P78Gw4D/dXmXLl2wefNmhISEwNbWVrXNvn37AAAZGRlwd3dvsgxtT7C6+61vceHgFbzw2bP4bt2PyL2Rj95jukNhYar2stqL9wGk7ySXmGizffm9CYku8PX1FTsEItGxjwm1RBP1pK79LdAlkZGRiI6Oxo4dO+Dr64tt27Zh7NixjZZ369YNxsbGiIyMhFKpRFxcHAYNGoTPPvsMCxYsQEFBATZs2CD26QAA/AZ1Q+rJ6/jxv6cwfNYgFOeUNpmUCIKAS4evwr1HF9g4WWklNt4HkL6TXGJiyJ3fU1NTAbBiIsNmyHUAtU6d9SQTktY5OTk128zqr8tjY2MbbbNmzRqNxNVexTklAAAbZ2t0H+aHQx8fh1Ar4NsPjqLojxIMmNgbxnLjBvvc/P0Wtr66B73HdMfExWO0EifvA0jfSS4xMeT25RMmTAAAXLhwQeRIiMRjyHUAtY71JN2P9RH/BQRg0a65MDU3wYKtcwAAFw6moKywvFFSAgBugZ0x7NmBCBjio7U4eX2TvpNcYiJ2+/K7T9DYzphIHGLXAWTY+DdAP/UL7QkIjZcHDW/+jazcVI7hswZrMCoiw8O/6PeJr/iJiIj0w/BZgzF8dtuSjOLcUuz6zwHcTsvVcFREhkdyb0yIyLCx8ztpAx8+6bfMS3/g2zVH4D/YG0PC+je5TUleKfa9dwj2bnZ4LPIR1fIbv2XizN7zsHO1QWdvJ22FTGQQJJeYaOOmhH+QiHQXO78TUUd9FrMDlaVKmFkpGiUmN3+/hcNJP+HaqXQAgEwGOLrb48EnglBXJyAoJACWdhYwszFDdsptfL/hGPqH9kLgUHZIJ+ooySUmvCkhMmzs/E5EHTVi7iOorqzBQ0/1brD81pXb2PbaHpTklqqWCQKwZ8W3OLP3PDLO30KvUQ9g9AvDsHzcWhjJjVBXU4eqimomJkRqILnExJBvSjgKBxE7v1PLWE9SWwx4sjd+3XseJz7/BWe+uYDC7GL0HdcDvx+4DGV51Z8bygAIQG11HTIu3IJQJ+DsvovoMSIAjl3tYaIwQe7NfFg7Wmolbl7fpO8kl5jwpoSIiIjao6qiGkc/O4UeIwLQ2dsJFSWV2P2fAwAAmbEMQq2A01+ea7yjAFVyItTWD9vl4G6HX3adQ48RgSi4VYjK0kp0CXRR7VJbU4cdS/fixtkszNs0E5Z25lo4QyL9ILnERJdoe9jI/fv3AwBGjx6tlfKIdBE7v1NLOlpPtrWPIYcNlpYb5zJxdNNJVJZUYmzMCJhbm2H8K6Nw/tvLuP7LzZZ3/sswwvmZhcjPLMTlH6+pln237ijO7ruI4bMH43ZaLi4eugprJ0sYGcvUeh68DyB9x8REQu42YeOrXDJk7GdGLWE9SU3xGeCFp/89DjadrPDv4FWATAb7LrbIu1mgluMLtQLupOVi+9K9qKupg1dfD4xbOAJfvf0dcq7nYsBTffDQxD4dLofXN+k7toMiIkmJjY1FUVGR6uPn5yd2SESko3JvFqC8uAJGRjIEDffHqS/Ooq5WQF1NHWpratVeXm11LYQ6AR49uyD3Rj4uHkxBTno+vnnnIOrqmpjB8R43zmbilz2/I/1MhtrjIpIKyb0xYTMOIsPGfmZE1BalBeVYM3MjPHu7Y+Y7kwEAzl4O8Ozthhtns1B4q1j9hQqAY1d7HN14EsOeHYiJ/zcGtTW1sOtsC1kLrbrKiyqQ9OI2CHUCrJ0ssXDnXPXHRjqlOLcUlnYWMJbzb9i9JPfTiI+Ph62trepz9epVsUMiIiIiHWNubYYHnwhCz5GBAOqHAv5h/TGY25pDZlSfJdh2tlZ7uV17dkH/J3vBu78njv33NE5s+QUO7rZ4a+wafLXiO1w8fBW1NXU48/UFpPx0HaX5ZbCwNceQqQPQ87FAhL4ySu0xkW75IzUHqyZtwP7Vh8QORedI7o2JpocL5uSKRERE0mcsN8L4l0eqvnf2cULwrEEws1LALbAzjm05jaLbJWovd8i0h/HxvGRkXsjGnbRcyIxl+HBuMgAg90Yeftl1DmNeDMa+9w7ByFgGm042WLB1Nm5fy8HVE9cRPHuQ2mMi3WLtaIlufT3g0aOL2KHoHMklJrrYjENbo7P4+Pho9PhERFLHepKaYyw3xoAne+PtCR/A7QEXVFfWaKSc1VM/BgCUF1QAqO8YX1pQDrmpHGZWZujWzx0/bv4ZFnbm8Ozjjs7eTgDq51Zx8nSAnYtNs8du7/VdV1sHZXkVzK3N7vNsSBMs7S0wY9VkscPQSZJLTAzZ7t27xQ6BiEinsZ6klljaW2DAxD44u/8iaqvU3/kdQKPhhQEg4BFvXDmahis/XoPMSAahToCR3AgjnxsKBzc7AID/YG/4D/Zu8pCfRG+F3MS40fWdl1mAvIwC+A9qer+9q37Ab99cQNSnM+HQxa7BurTTN1F0pwQPPhHU7lMk0hTxXzcQEbWDUqlEcXGx6sMBMIioPVx8nFBVVtX6hmpiam6CK0fT/lwgA2AEzP1oGq4cS0NeZutDFlcrq1GtbPyGZ8+K7/Hff3yJgltFTe7n4uMMtwdcYGalaLTum/cOYtd/9qOyTNnEntpTVVGNhEkbsG3JXlHjIN3ANyYSkpiYCACIiooSORIi8XAeE2rJ/daT7F9oODx6umm1vE6+Trh16TbqauofotydQX7P298i43x9P5QRkUOw6z8H4OBuhzHRwSjNK8NP239F+plMuPp3RuT6vwNofH0/OuMh+AzoCjvXppt/DZjYGwMm9m5y3ZOxo1GaXwYzy8ZJizbJjGSwsDWDubW4cZBu4BsTCVmzZg3WrFkjdhhEouI8JtQS1pPUkmunb+DM1+fh4uestTIzf89WJSW4Z8jgjPPZeHjygxgWPhCHk37C1RPXcXL7GVSWKvH7d5dxbPNpZF38A2f2/K7a597ru6ywAsV3SvHAMD/cvpbb7rjcHnBBwBDx+2SZKOR47uPpGLfosWa3KSuswPnvr6Culm/I9R3fmKhR3NAEjXeAJzJ0ujgABhFJw+kvz+HS4auw62IrWgzmNgrUVNWiurIGp774DRkXslVznIyYOxTm1mboF9oLFnYWECDA8X99UO4S/jdR47H//ozjyadh62KDktxSvPJ1FEzNTbR8NtpxZNNPOLntDP4WPwGBj4ifTJHmMDEhIiIigzB24Qj0Ht0dWxbvEicAAagqr4b/YG+knkyHTWdr3Lr0Bzp5OwIALh5MweBn+sLU3AR9xnRvtHt1ZTUK/yjGia2/ov+EXjA1N4GlgwVKc8tw+WgqTMzkuHX5NuxcbdFvfE9tn53G9J/QGwoLU3Tr6yF2KKRhkktMOPM7ERHpCr4pl4brv2Zg13/2Y9QLw1BZXNnkyFnaUltbB6++XdFrTHf8sOEYAOBOWh6M5EYYHNYPK55cB0t7C1jammPCv0Y3GE3LyNgIclM57LvYwsHNDsHP/jnnyesh78Lc2gwVJZVw8nRoMjG5eCgFNdW16DXyAY2fpzo5ezogZM4QscMgLZBcYsKOr0RERNQeZ/aeR2F2Mc7tvwS7Ls3PE6IVAnDm6/P4I+UOAMDY1Bg2naxRkFmIbxOPwtLOArnp+cgFsP/9wwiZMxjffnAUj80dCmMTY9h3sUWNsgafvfwFnvq/x2Fhaw4AmLZiIuQKORQWplA006F9b8IPqKqoklxiQoZDcomJpmd+12WjRo0SOwQiIp3GelK7srOzMX/+fLi6usLa2hpvvPFGs8vXrFmDc+fOobq6Gv3798fzzz+PcePGwcmpfoLBuXPnYtAgzcx6Pm7RY3APckV+ZiFOf3EOMKqfcFFjc5m0Qib7sxe8UFuHuupamNuYobK0EiX5pQDqZ66/8uM1lOSU4taV23jgUT8MeWgoTM1NkP5bJlJPpuOPq3fg3d8TAODZxwM3z2bCwd8echNjAMD1X2/CwtYcnX3qO/s/sywUd67n4rsPjuLRmQMl2SeltKAcF364ggfH9oCpmfTip5ZJrteoQqGAjY2N6mNIHV9XrVqFVatWiR0GEZHOYj2pXevWrUNERATeffdd3Lx5E+np6c0ut7KywgcffID169fj888/h1KpREpKCszMzGBpaYmgIM1N9GdqboKHnuqD2tpamFkr4DfQG3XV4iQlAJB95TYAwNzGDEEhAfAf4o1/7p0HY1NjCLUCHhzfE7X/G8nr1pXbCF89BZ693eF2LQCBxX0R+KgvjOXG2LPie9y5ngcA+P3AJSS9uA0/bf0VALA/8TA2LtiO7XFfq8rt2rMLsq/cxo+bf0bmhWwtn7V6nNpxBt+8cxAXD10VOxTSAMm9MSEiw8Z+ZkS6Izs7G15eXgAAd3d3ZGVlwcvLq8nlM2bMQGVlJWJiYvDKK6+gpqYGSUlJGDx4MHbv3o2VK1c2aKp9V3JyMpKTk5ssPysrq9UYq5XVOPzJTwgc6ouxL43A4y8Ox1tj10AQAJkRIIhYhVQUV+LKj9dQVVmNzAvZqCqvBgAUZBVgdHQwDiQehqmFCbx6u+PzxbshCAJybxTgu7VH4DOgK1KOX8cXr3+N5z6eDmsny/q+Kfb1Tbt+++YCzKwUGDH3kQZlhswZAr+B3dCtX8OO5HV1Ai4dvgrPPu6wsrfAuW8v4eimU3hmWSicuto32LamqgbXz2TAu58njOXafUA84MneMLMywwOP+mq1XNIOJiYS8tJLLwEAnwaSQWM/M2oJ60nt8vT0REZGBvz8/JCZmQk3N7dml6ekpGDp0qX417/+haCgIKSlpSEtLQ2DBw+Gk5MTqqqano09LCwMYWFhTa4LDQ1tNcZbV+7gx80/oySvDO5BrjAyNsKIyCHIzy7GT1t+uf+TVwOZsQxVFfXJSPaVO4Csfl4P775dMWhKX9i72sDYtL5Zlv8QH6xOfrc+fpOx+ONqDoJG+KP3qPrRu6oqa1BWUA7l/2a1f+7j6TAyksHayapBmZb2FlCWV2HV5A/x97cnorN3fVO667/cxLbX9qDfhF6oKK5E6sl0VFdWo7K0slHcP207g+8+OIrQV0ah79geGvv5NMXayQqD/9avw8dRllfB1NykQbM6Eh8TEwk5cOCA2CEQic6Q+5lR61hPaldkZCSio6OxY8cO+Pr6Ytu2bRg7dmyj5R4eHvD390fPnj0RHx8PBwcHrFy5EgcOHMAvv/yC/Px8LF++XCMxevZyw9S3nkSXB1wAALkZ+fjmnUMQxBya63/uzgIP1M+AbiSXIeqzcNi51HfQDxz651uBB58IwpU7FwFBwHCXx6GwNEXv0d3hP8i7fttHfPDS9gjYdrYGANh2sm6yTGV5FUrzylBWWI4aZY1quUfPLnh05sPoHuyPM3vPw6mrPaatfAoWNuaNjhE41BcF2UXwGeDZ8R/C/xTeLoaNszWMjDSfKOTezMfa8E/x0FN9MPqFYRovj9qOiQkRSQonWCTSHU5OTs02s/rr8mvXrjXaZtOmTRqJ614leWX4Mn4/eo/ujtEvDMOm+dshkwFyEzmq77kx1yYjudGfs5j/LzcR6gRY2lli9fQkTFw8BkHBjd8GW9lboLa6Ft36euCx54fCLbA+2SrOLcXG+dvQb3yvVt8mfDxvC0rzy7H42+gG9aepmYlqSN7H5w9v8RhOXe0x8rmh2PbaHjwQ7If+ob3aeupNSvvlJjYt2I7hcwZj2MyBHTpWa+rqBBibGKGzjxMc/9JEjcTHxOR/4oYmiB0CERERqZmRkQwKS1OYmNXf8lSWKVFXJyB45sM4+NHxBm8ttKXufx3bbV1sUFFUAXNrBTx6uuH891cAAL989Tu8+njA0q7h2wpzGzOUF1bg+q8ZKPqjWJWY1FTVoDinFOWF5c2WWVtTh5vnMiEzksG7f9cOP9SpKK7E9TMZsHay6nBiYu9qC6++Hqrz0aQdS7/GtVPpeHHLLNVQy6Q7mJgQERGR3rK0t8D8LbNV34dOfxjfr/sRhz85AVNzEyhLm+7bog3dHnRH0e0S3Dx/C0/EhKD7cD9sfW0P0k7dwC+7z+HRGQ/jxNZfUJhdjMeeGwqgfoSxHiMC4NbdVXUchy52iP0mCkbGTScbBbeKsCZ8E+w6WyMnPR/j/zGyw7Hbd7FFzBeRMLdues6U9h4r/N2nO3yctnDytEdJbinkprwF1kX8v0JEREQGY0hYf5zcfgY1VTWqjuKikAEPDA+Aq68zqiurcGTjSbj4OquadvkP8UZZQTn2v38YANC1V/3AAnKFHJPjxjY6XHNJCQDITY1h28ka3fp3BWQyVBQ17tB+P6zsLdRynKYU3CpC4e1idHvQo/WN22H4rMEYPkuthyQ1YmIiIfPmzRM7BCIincZ6klpjZGyERV/OxYfPJyPzfP1cHjJjQNDytCaOXe2R/M+dsHGyQk1NDcoL/kwW5ApjbHxxG17e/Rx6j+mO9DMZyE65g9H9n0Dh7WJcOXYNAUN8GhyvvLgC//3nl+gREoCBT/dtsM7ayQovfBaOaz/fwKntv+HO9Vz4PuzVKKaKkkocSDyM3qO7w0vNCUFryosqkHMjH57/S8B2vrkPN89lIeaLSNj8ZWQx0l9MTNTsbl+VuKMxrWzZflFRUWo/JhGRPmE9Sc2pqxPw3doj6OTthKCQANRW1cLK0RLKsirIjIGayhrUabG/Sd6NAgBAcU6paplnb3dUK6tx6/Jt1ChrkX4mE9WV1SjOKcXJ7WfgauEDm7JypJ5Mb5SYVFVU44+rOarhf5viM8ATi3bNhZWDZZPrb1/LxZm9FyBXyNWSmJTklQEArB2bLu9e37x7EL9/exlzP54GV79OeOTvA5A9wLNN+5L+kFxiwsnViIhInTj4iWFQlinx07YzcPF3xv7Ew/Ds5Yac9DxY2JqrkoOAod5IOZ6m0bcnclNj1FTVF+A9wAPXf8mAUAfYdbHB7bQc+A3shluXb8PRwx6efdxwLPlnCHUCJi15As5eDhDqBBTnlGLf+4cQEjEEpmYmAAC7zjZ4efdzMPnf9+Y0l5QAgFcfd8z+4G/o1K355KY91kdsBgQBC3fObXXbPk8EwdzaTDWZo/8gb9VQyGQ4JJeYGPLkancnktq9e7fIkRAR6SbWk9Qcc2szzP14GkzN5Pjyzf1w6uqArr3ccOlIKroEuuDy0VRcOZqmsfIt7c1RWaKEk6cDeo4MxNn9F2FqrkD3YD9c/yUThX8UA3X1bz78h3hjbMwIGMuN8eiMh1GSV4bqymo8O28mAGBqj9m4fPQaugf7oWtPN1UZCgvTFmO4dCQVTl3t4ezl2Ow2HkFdOnyuJz7/Bcf+exqBj/rAzMqsTfv49PeET3/1zYtC0iS5xMSQJ1dragx4IiL6kxj1pCab8JJ63W3m9OzqZwAAZQXl+H79MbgGdIK9my0KsopgYi5HbVXdn/OMqElZQQUAoDS/HD4DPPHtmqO4cy0PQP2Qxo/NfQR5NwpQUaJEyvE01FbXv1UxtzFDzvU8XP8148/ru0f9JITXfr6B07vOYfzLj8FE0fKbksI/ivH54t3w7OOOwKG+6OzjBO9+XVuNWxAE7E34AfauNhgydUCbzrWutg61NbUYNKUfHD04Vwi1neQSE06uRkREROpgaW+BCbGj8MW/v4FcIYeRiREs7Sxg7WyNjHNZai3rbhOu0rwy7FnxvWq5mbUCDu52eGTqQwDq51kpySmFg5sdAKBTNyfM3zobVg6WiN1Yv09IxBB06+eB67/exJUf0zB81mDYd7FtsXzbztYY82Iw0k7fwP73D8Gtuwu817X+cLemqhbn9l+Eo4d9mxOTIVMHtHlbontJLjEhIsPGfmZEpE6evd3h85An+oX2xO20PBz++AR6jAhE9pXbMDU3QXlhRYfLMDY1hoOHPe5cywUAWHeyBC7Ur+vWtyueeWM8AODXvedRfKcEwc8OarC/XWcb7Hv/EApuFcHO1QadujmiUzdH9Hk8CCW5Za0mJQAgk8kw8Om+yE65AzMrBR5fENKm2E0UckR9Gg4Tc94ykubxdQMRSUp8fDxsbW1Vn6tXr4odEhFJmG0na0xfOQmW9pY4/PEJdB/uh2plDUzM5GpJSuy62KC2qhZ3ruXCyKT+tuvaTzdgZCwDZECnbo4oKyjH7Ws5+HrVDziy6SRqqmoaHaesoLy+edk9A4cpLExVncXbauLiMfjn1/Pg/kDbZ1m37WwNCxvNzpJ+cvsZfDRvCypK1DPHCkkTExMikpTY2FgUFRWpPn5+fmKHREQSVq2sxifRW5FyPA29H++OtJ9vIu30DSgsTBH0WECHjm3T2RqFt4qB+i4hMJYbwcLOHBZ25jC3MYORkREOJ/2Er985iPysQtQoazAkrD/kpnIIgoCa6j+HB5v02hNw9LCHzEjWoZiA+rcnuuZ2Wi6yr9xGZamy9Y1Jb/G9nIZoojNkQgKHtCRiPzNqSXvqSQ4TTEB9H4rbqTmwtDPHk/8agzvXcuHdvytGPv8ovnn3YMs7GwFmVgpUFishMwKEe1qWKixNYWapQDFK4OBuh/yMQlRX1KC6ogblRRUInjUIx7f8gprKGvQL7Qnvfl0R80UEbJytAQA7/v01rp26gejkZ1VvK/T5PmDcoscwKupRmFkqWt+Y9BYTEwkZPXq02CEQEek01pPUXubWZojZGYmKogpUlimhsFLg5y/PobyoEqOiHoWlvQV+2HAMzl4OMDE3wa1Lt1X7Ons6Iic9D0P+3h+l+WU4+80l1TplWRVKcuvnRxnwZG98v/4YFBamqKqoQnVlDa7/koGw+AkQagV49+sKQRCQ/Mou2Lna4pk3xsO+ix0c3IsgN/nzVk2fr28jIxmTEmJiQkRERIZNJpMhccYm2LvZIi+jAEZGRlBYmMLc2gwDn+6Lw5+cgACg77ieuHXpNmTGgLFcjpzrebB1scFP2840OSxuRXElugR2RuaFbLj4OOPO9VwEPOIDG2drPDzpQVSWKuHgbgcAKLpdguyUO8i9WT8j/IiIIRgRMUSLP4XGinNLYaKQw9y6bXOREHUUExMJCQoKAgBcuHBB5EiIiHQT60m6H3JTYwQN94eDmx36P9kLpuYmMJYbAwBMzU3Qf2JvFN0uwYHVhyE3k8N/YDcEhQQg7fQN2LnaQlmqRNovNwHU9yNxC3KBkbExMn6/heKcUmRfvYN5G2egJLcMmxZsR+8x3XFy+xkc33Ia3fp7YOaqp2HnYoPJS8fC0b1hglNTXQu5SX0s2ry+a6pqsPrvSXB0t8Pcj6ZpvDwigIkJERERGTiZTIYJr4xqdv3jLw7HqS9+w81zWaitqUVZYQUChnjjize+gZWDJawcLHDrcn0TLwd3O5QVVKDv+J6oqapB155d0Hdcr/oZ1z0dEfrPkdib8D36jusJMysFvPv9Odt5j5CGne1/3nkW37x7EDPfnQzP3u6aOflmGJsYo+fIQNh1ttFquWTYmJgQERERteKhp/rgoaf6oDi3FEZGMtTVCaitrkXRH8Uo+qNYtZ3foG44nvwL0k7dwJy1YY2O49zNEebW5ujWryvGxoxosUxLe3PYudpA0Ya+F+m/ZSL1ZDqCZw1SvWHpCJlMhvGLHuvwcYjaw+ATE47KQkRERABQUVKJK8fS0CPEH3LTP2+R0n/LxE/bfsUTC4bj05gvUHynBLW1dfB7uBuu/nQdCgtTmJiboO/YIASFBOC3fRfxwLCmhzLvEtAZi3bNBQDc/P0Wftx8Co/PHw5718aTJHYP9kf3YP82xX7qi99w8WAKgkL84erX6T7Onkh8HGdTw5j4EBEZhrihCazzJe7k9jP4ctk+/P7t5QbL007fwOUjqbidlgufhzzh4ucMcxsz9H6iOwKH+mDq8gkozStDdkoOOvs44x+7n0f/Cb0aHX9vwvf4zxNrUHCrEACQfiYDKcfScDs1p8OxP/5iMKateEovkpK62rrWNyK9ZPBvTIiIiEiaDh48qOoQfq9du3bB19cXqampmDBhQpP73u1Avn//fsTE1M85VldTh4rSSmxd9CF81/li9+7dAIDvz+7D55lbsG3WRw3L/9fX6C9/FL3HBCHT6xI+/2oD3vz6/xpsM2/ePERFRQEA3vw0DldTriKxx1uqNyS11bUYVB6EQPgCQJPnc/ecAKCqqqrZbZo6p3v5+PiozikxMRFr1qxptM2oUaOwatUqAMBLL72EAwcONNrm3nMKDQ3FtWvXGm2TkJCgGt64pXP66/+n2ppaFGQVwdzGDJb2FnpxTn/VnnMyNExMJORupURERE1jPUkdYSQ3gqWdRYNl+VmFOPP1BVQUV0JhYQoAEAQBlaVK5BYWoEdEAAIe8YH1TisYGbfcEMXB3R6mmaaqEb+A+k7mbbVr1y7cuHEDL774YjvOSlpkMhmMjI1a/VmSfpIJgiCIHURHhIaGdiir1MZrd3XO/k5EDXW0DiDDpan6n3W+9mjj918QBJza8Ru6BHaGR48uAIDtS/fiwg8p8BvYDY/OfBju3V2RdfkPnN13EYPD+sNEYQJLO3ONxkWkj/jGREJSU1MBAL6+viJHQkSkm1hPald2djbmz58PV1dXWFtb44033mh2+ZYtW7Bv3z7U1NTgueeew+DBgzF37lxYWVkhLy8P69atg7m57t3My2QyPDz5wQbLAob4AAIQ+s9RMDU3AQAc/uQEUo5fx+/fXobMSIaXdz8HmUzWrrK2/GsXyosqMSvxmSbX8/omfSd6YpKYmIiioiJUVVUhLi5O7HB02t22ipw4jIioaawntWvdunWIiIjAyJEjMWPGDKSnp8PLy6vJ5YmJiThy5AgqKiowadIk/POf/0SnTp2wbNkyfPTRR9i6dStmzpwp9im1Sc/HAtHzscAGywZM7IOsS7fh2ccN5jbm7U5KAKCyrAoVJZXNruf1TfpO9MTkySefhIuLCxYtWiR2KERERNQO2dnZ8PLyAgC4u7sjKysLXl5eTS6Xy+WQyWSwsLBAeXl5g208PDzw888/N1lGcnIykpOTm1yXlZWl7lMCUD/recaFbHj2doeRUesJRkleGVz8OuHl3c91qNzwd5/u0P5EUid6YmJnZ4fY2FjMnz9f7FCIiIioHTw9PZGRkQE/Pz9kZmbCzc2t2eVyuRyCIKCiogLm5ubw9PTEvn37AAAZGRlwd296ZvOwsDCEhTWeqBCo72OiCce3/IIfNhzDpCVPNHoz0pQNkZsh1AlYuHOuRuIhMhSiD3kwa9YsCIKAL7/8UuxQiIiIqB0iIyOxYcMGREVFwdfXF9u2bcPFixcbLffy8sK8efMwa9YsPPvss3j99dcxePBg5ObmYsGCBThy5EizyYcYAoZ4o8/jQfDs7dam7XuP7o7eo7trOCoi/afxNyZZWVmYMmUKNm/eDIVC0agz3Oeff97qMcR4jUtE2sF+ZkTS5eTk1Ozf578unzhxIiZOnNhgWVNzTuiCzj7OePJfo9u8/YjIRzQYDZHh0GhiUlxcjLfffht2dnYAmu8k1xoxXuMSkXawnxkR6ZtLR67C3NoMXg96iB2KWpQXVeDbtUfQd1xP1ZDJRJqg0cTExsYG77zzDsLDwwE030mO2oajcJA+Yj8zUqe21JPamL+KDFdNdS22vroHdq42mL9ltlqPLdZ9wK0rt3Fm7wWYWZkxMSGN0mofk7ud4QA06CTXHkqlEsXFxapPXV2dusMkIi1iPzMi0kWCIGDHv7/Gtx8cbXJ9TVUNTu86h5Lc0gbL5SbGmBw3Fg9PehCl+WWqYx1IPIwze89rPG5N8H3IC88mPoPhsweLHQrpOa2OyhUZGYno6Gjs2LFD1RmuveLj47F06VLVd39/fzVGqNv2798PABg9uu3tXonExn5mpE2sJ0ld6moFXDt1A3auNk2uv3IsDXtWfIe8jH4Y/cKwBus6dXPEttf2IP1MBv725gQoy6rw0/YzcPXvhAfH9rjvmMS8vj17tf9hMlF7yQRBEMQOoj2USiWUSqXq+9SpU7Fnz577Pp42X+nHHY3p0P5BQUEA2KSLpKO4uBivvfYarl69isTERCQlJWHIkCGqfmb//ve/O9ycMzQ0FLt371ZPwCR5baknNV3vd7Sup7bT9O9/RUkljOXGqtnd71VVWY2fd55F0HB/2Lk0TF5qqmvxw/of4T3AE74PeQEA7lzPhbm1GaydrO47Ht4HkL4TfR6T9lIoFFAoFKrvRkaij3hMRM3QRD+zvz6cYHNOItIUc2uzBt+rldW4cz0PboEuMDUzwZCw/o32uXkuC87dHDEq6q9vUZw0GiuRPpDcXT37mBBJlzr6mcXHx8PW1lb1uXr1qrrDJCIDd+lIKnb95wCqKqsbLP9+/TFsiPgvrv96s8n9bl25jY+jPsc37x7URphEekdyiQlvSoikq6lJ19orNjYWRUVFqo+fn5/6AyUig3b++ys48/V5FN0uabA88BEf9BwZiE7ezqirE3DhYArKCitU682sFeg3oRd6j+Fki0T3Q3JNuWJjYxET82f73alTp4oYDRG1RVJSkurfzXVibys25yQiTRv/8mMYOv0hOHs6NFju9aCHam6SlONp2PbaHjw0qQ+eWBCCguwirP57EnqMCIBPf0/VPuVFFTC1MIXcxFir50AkRZJLTAz5psTHx0fsEIhExz4m1BLWk6QOZlYKuPg6t7iNZx93DJ3xMHo+FgAAsLA1h3e/ruh6z+hVpfllePeZj+A3sBumvD6+w3Hx+iZ9J7nExJBvSjjyEJFhDxlOrWutntTGSIxxQxM4MpcBUFiYYkTEkAbfp614qsE2pham8OztDvcg9UxKyPsA0neSe90g5T4mcUMTOOMwUQexjwkR6aLKUiXWzNyE79b9qFpmamaCaSuewuC/9RMxMiLpkFxiYsg3JYmJiUhMTBQ7DCJRKRQK2NjYqD6G1JyTWsd6ksRSV1uHkrxSlBWUa6wMXt+k7yTXlMuQ+5isWbMGABAVFSVyJETiMeTmnNQ61pMklrKCcrj4dUK/8T01Vgavb9J3hnNXT0R6QcrNOYlIf92+lovrp28i69IfYodCJFmSS0w4wSKRYTPk5pxEpLt6jAjAC5ufhcLKFNuX7m00OSMRtU5yiQmflhIZNvYxISJd5dTVHqd3ncP576+gNL9M7HCIJEdyf9H5tJSIiIh0VWleGWQyGWw7WYsdCpHksPO7hIwaNUrsEIhEx87v1BLWkyS2Z5aFQlmqhLFc/TO9N3d9lxaUY+P8begzpjuGTB2g9nKJtEVyiYk6SW1OkVWrVokdApHoOMEitYT1JInN1a+Txo7d3PVdo6xB0R/FKMlj8zGSNsN53UBEeoHNOYmIGrJzscE/v47CmOhgsUMh6hDJvTEx5GYcL730EgA+ESTDZsjNOal1rCdJn7V0fRvLWReS9EkuMTHkZhwHDhwQOwQiIp3GepL0Ga9v0neSS6/ZjIOIiEg3XLx4EVOmTMHzzz+P999/v9V1S5YsQXR0NKZPn46dO3cCAAYOHIjw8HCEh4frxRQAd67nolpZI3YYRJIkuTcmbMZBZNgMuTknka5Zvnw5li1bBj8/P4wePRoREREwMzNrct2cOXPg5+eHadOmITc3FxEREejXrx9KSkpU8xN169ZN5DPqmIzzt/DR81vw0KQ+eGJBiNjhEEmO5BITIjJshtyck0hsSUlJSEpKUn0/fPgwNmzYAABwdHREXl4e3NzcAADZ2dnw8vJSrcvPz8e0adNQWFiIBQsW4NVXX4WFhQU2b96MPn364L333sOnn36KZ599tkGZycnJSE5ObjKerKws9Z9kBzi426P7cH/4DZJ2gkUkFiYmRCQpsbGxiImJUX2fOnWqiNEQGZa7Ta7uioyMREZGBry9vZGTkwMnJyfVOk9Pz0brTp48idWrVyM+Ph4eHh44ffo07ty5gz59+sDJyQklJSWNygwLC0NYWFiT8YSGhqr9HDvC0s4cU/49TuwwiCSLiYmEzJs3T+wQiETH5pzUEtaT2rVo0SIsXrwY1tbWmDx5MhQKBRYuXIjFixc3WldWVobx48cjJCQEsbGxCAgIQHR0NN566y18//33KCoqwnvvvSf2Kek0Xt+k7ySXmBhy+/KoqCixQyAi0mmsJ7XL39+/UTOrlStXAgAcHBwarbtz506jY2zbtk1zAeoZXt+k7yT3qDE+Ph62traqjxRH8JDajPNERETUcee/v4Lcm/lih0GksySXmBjycMGhoaE6156WiEiXsJ4kXXXneh62x+3FvvcO3fcxeH2TvpNcUy5Dbl9+7do1sUMgEp0hN+ek1rGeJE2rralFTno+XHyd27Wfs5cDxrwYDPcg1/sum9c36TvDuavXMXFDE9iki+g+6ENzTiKSrqOfnsIHz36KlONpTa6/+tN1xI9Zjas/XW+wXCaTYeDTfeHe/f4TEyJ9x8SEiCTFkJtzEpH4vPt1ReCjvujc0hsTmQyQaS8mIn0huaZcRGTYDLk5JxGJr2svN3Tt5dbser+B3RD7DUfPIrof/ItORERERESia/Mbkxs3bsDT0xMAcP78efTo0UNjQVHTEhLYJ4XEwzqApID1JOkzXt+k71pNTDIzM/HRRx/h4MGDCAkJgSAI+O6773D06FFtxEf3GD16tNghkAFiHUBS0lw9ycFGSB/wPoD0XauJiZubG4YPH47KykoEBwcDAKZPn67puJqljqFC+QeKqO10rQ4gIiIi/dRqHxOZTIZHH30UkyZNwuHDh3Ho0CF89tln2oitSYY8VGhQUBCCgoLEDoMMjK7VAUQtYT3Zdl9++SUmTZqEs2fPAgCWLVsmckTUGl7fpO/a3Pn97bffxsCBAzFs2DAMGzZMkzG1iEOFEolDV+oApVKJ4uJi1YcTLBLdn507d2L9+vVYsWIFsrKycOXKFbFDIiID1+bO73379sXIkSM1GUubcKhQInHoSh0QHx+PpUuXqr77+/uLGA1R0+42GY47GiNyJM2ztbWFo6MjVq9ejejoaNTU1IgdEhGpgSAIuPn7LeRnFsLB3Q5de3aBTCaNiXXanJh89NFH2LFjBywtLSGTyfDDDz9oMi4i0jG6UgfExsYiJubPm72pU6eKEgeR1Dk4OOC///0vpk6diqioKISGhoodEhF1UOEfxfh04Q4UZhfBWG6M2ppa2LnaYvrKSbBzsWnXsZKSkrB+/Xp4e3sDAKytrXHp0iUcOnRItU1wcDAOHTqEnj17YsiQIQCAvLw8DB8+HPPmzWt3/G1OTFJSUtp9cCLSH7pSB/CtKZF6xMXFqf790EMP4fLly+IFQ0QdJggCPl24A/lZhRBqBdRW1zd1zs8qxGcLv0DUZzPb/eYkMjIS4eHhqu93B8H5K0dHR3zwwQcAgPLycowcOVKzicnw4cMhk8lQWFgIOzs7vjEhMjCsA4j0y65du5Ceno758+cjIiICTz75JMaNGyd2WJJQW1OLW1duwyOoi9ihEKnc/P0WCrOLIdQKDZYLtQIKsgtx8/db8Ozl1q5jfvjhh6o3JH369Gl2u9zcXISHhyMzMxMymQwrVqxob/gA2pGYHDx4EED98LyLFi26r8Kosfa0Q961a5emwyFqFusAkgLWk223adMmbN26FQCwdu1aTJkyhYlJGx3772n8sOEYnv73OAQN114/N17f1JL8zEIYy41QW13baJ2R3Bj5mYXtTkzmzJnT4I3Jrl27UFpaCisrK5SVlakGoHFyckJSUhLKysowadKk+z6HNicmN2/eBABUVVVx5A6R+Pr6ih0CGTDWASQFrCfbztHREcbGxgAAExMT2NvbixyRdPg+7IU7ablw7+6q3XJ5fVMLHNztUFvTOCkBgLqaWji427X7mOvXr8d3332n+r5gwQI89dRTcHR0RHFxMV599dUG21taWmLTpk0YN24cvvrqK3Tu3Lld5bU5MVmyZAlkMhnkcvl9tRmjjktNTQXAionEwTqApID1ZNuVlZVBqVRCoVCgqKgIubm5YockGV0COmNy3Fitl8vrm1rStWcX2LnaqvqY3CUzlsHetX50rvYIDw9v8LbkrgkTJjRadm+H+E6dOuHUqVPtKuuuNvcaXbZsGZydnWFvb4/+/fvfV2HUMRMmTGjyYiDSBtYBJAWsJ9suJiYGgwcPxtNPP41x48bhX//6l9ghUSt4fVNLZDIZpq+cBAc3OxibGMHE3ATGJkZwdLPHtISnJDFkcJvfmLz88st44YUXIAgCFi9ejA8//FCTcRGRjmEdQKRf1q1bB0dHRwwYMAAvv/yyJG5aiKhldi42eOGzcP2fx8TZ2RmDBg0CAHz55Zeaisdg3e0ED+j2hFxkuFgHEOmX8vJyHDhwALNmzZLMTQsRtU4mk8Gzl1u7O7rrgjYnJvn5+UhOTgYA3L59W2MBtUapVEKpVKq+3x0NgIg0S1fqACJSj7sd35mUEJGuaHNiIpfLkZGRAQsLC3Tq1EmTMbUoPj4eS5cuVX3399feMH1EhkxX6gAiUo/U1FRs2rQJKSkp2LRpEwBgxowZ7TrGxYsXERcXB0dHR3Tv3h3R0dEtrpszZw4qKyshl8sxadIkhISEIDIyEp06dUJFRQXWrl3LRInIgLW587uVlRX+8Y9/4IUXXkBFRYUmY2pRbGwsioqKVB8/Pz/RYiEyJLpSByiVShQXF6s+fGtKdH8iIyMhCALmzJkDQRAgCELrO/3F8uXLsWzZMqxduxZ79uxBZWVli+tOnz4NW1tbyOVy9OvXD9u2bUNwcDBWrVoFZ2dnHDlyRJ2nSEQS0+Y3JiUlJTh16hSMjIxQUFCgyZhapFAooFAoVN+NjNqcW0nehQsXxA6BDJiu1AF8a0otYT3ZdjNnzmz3PklJSUhKSlJ9P3z4MDZs2ACgfl6UvLw8uLnVt2vPzs6Gl5dXg3UJCQkICQnBuXPnEBsbi8DAQNUofx4eHsjMzGxUZnJysqoZ6V9lZWW1+xykjNc36bs2Jybx8fFITExEVVUVli1bpsmYiEgH6UodEBsbi5iYPweImDp1qmixEBmav85rEBkZiYyMDHh7eyMnJwdOTk6qdZ6eng3WWVlZ4fz58wgJCYGTkxOqqqpU2wBARkYGHnvssUZlhoWFISwsrMl4QkND1XuCRCSqNicmLi4ueP311zUZC7Vi//79AIDRo0eLHAkZIl2pAwz5rSm1jvWkdi1atAiLFy+GtbU1Jk+eDIVCgYULF2Lx4sWN1tna2iI1NRXR0dEoKipCXFwcvLy8EBERgbNnz6Kurg7Dhg0T+5TuS8rxNFSWKtFr1AMaLYfXN+m7NicmJL67T4n5KpeIqGl/rSfvHYqd1M/f379RM6uVK1cCABwcHBqte++99xod427Heynbu+oHFOeUoMeIABgZa+5hCe8DSN8xMSEiIiLqgMlxT6CqvFqjSQmRIWBiQkRERNQBHkFdxA6BSC8wtSciIiIiItExMSEiIiIiItGxKZeE+Pj4iB0CEZFOYz1J+ozXN+k7JiYSsnv3brFDICLSaffWkxyRi3RFRUklzK3NOnwc3geQvmNTLiIiIiINST2ZjrfGrsGpL34TOxQincfEREISExORmJgodhhERDqL9STpGmsnS7j4OsPBza7Dx+L1TfqOTbl00N3mB3FHYxosX7NmDQAgKipK6zEREUkB60nSNZ19nPHcx9PVcixe36Tv+MaEiIiIiIhEx8SEiIiIiIhEx8SEiIiIiIhEJ3piUlpaipkzZ+Ly5ctih0JERERERCIRvfN7VlYW+vTpI3YYkjBq1CixQyAi0mmsJ0mf8fomfSd6YhIQEIATJ06IHYYkrFq1SuwQiNSutLQUUVFRiI2NRWBgoNjhkMSxniR9xuub9J3oTbmIyLDxrSnps7ihCZyBnoiojTT+xiQrKwtTpkzB5s2boVAoMH/+fLi6usLa2hpvvPEGACA8PLzFYyQnJyM5ObnZ4xuKl156CQCfmJB+4VtTUifWk6TPeH2TvtNoYlJcXIy3334bdnZ2AIB169YhIiICI0eOxIwZM5Ceng4vL69WjxMWFoawsLAm14WGhqoxYt124MABsUMgItJprCdJbDXVtSjIKoSzl6Paj83rm/SdRhMTGxsbvPPOO6o3ItnZ2apExN3dHVlZWW1KTIhIuvjWlIgMyXcfHMVPW39F8OxBOLPnPP4WPwGufp3EDotIErTa+d3T0xMZGRnw8/NDZmYm3Nzc2n0MpVIJpVKp+l5XV6fOEIlIjfjWlIgMjc9Dnij8oxgmChOU5JWhuqJa7JCIJEOrnd8jIyOxYcMGREVFwdfX977elsTHx8PW1lb1uXr1qvoD1RHsMElSd/etqbOzM4Cm35q2l1KpRHFxserDhxNEpEv8Hu6Gvy0LxZCw/vi/7+eja6/2P4QlMlRaeWOSlJSk+ndzzTHaKjY2FjExMarvU6dO7dDxiEh71PHWND4+HkuXLlV99/f3V2eIpAdy0vP4YId0gpGRTOwQiCRF9HlM2kuhUEChUKi+GxkZzojH8+bNEzsEog6JjIxEdHQ0duzYcd9vTflwgloyb9487E34XuwwiDSC9wGk7ySXmBhyH5OoqCixQyC6L+p8a2rIDyeodVFRUcjZomx9Q1KLixcvIi4uDo6OjujevTuio6ObXTdjxgzMnz8fAJCeng43Nzd88skn6NOnDx566CEA9cPg2tvbi3Iu2pCfVQhzGzOYW5vd1/68DyB9J7m/6IbUx6Q5nLCLDBn7mBDpjuXLl2PZsmVYu3Yt9uzZg8rKymbXKRQKJCUlITExEY6Ojli3bh1++ukn2NjYwMTEBP7+/nqdlJQWlGP1tCRsX7JX7FCIdJbk3pgYcjOOu6MP9UWwuIEQiYh9TKgloaGhuHLxGsK6zxY7FL2UlJTU4A3o4cOHsWHDBgCAo6Mj8vLyVH3H7h3s4t51CQkJePHFF2FlZQUvLy9s2rQJ/v7+iImJwcGDBzF8+PAGZerLcOHm1gr0GvUAPHp0ue9j3L0P2L17t7rCItIpkktMDLkZx7Vr1wAAfR2CxQ2ESESG/HCCWnft2jUUVOaJHYbeCg8PbzDvUGRkJDIyMuDt7Y2cnBw4OTmp1t0d7OLedTU1NTh+/DheffVVAMClS5dUw4k7OTmhqqqqUZn6Mly4sdwYT8aO7tAx7t4HEOkrySUmhtzHhIgM++EEka5ZtGgRFi9eDGtra0yePBkKhQILFy7E4sWLm1z322+/wdvbW7V/YGAgXnrpJXh5eaG2thajRo0S8WyISGySS0wMrRnH3b4kcUdj1LIdkdTx4QSR7vD392/UzGrlypUAAAcHh0br+vTpg8TERNV3T09PfPHFF5oPlIgkQXKPGmNjY1FUVKT6+Pn5iR0SEWkRB8AgIl1UV1uHA2uO4OKhFLFDIZIsySUmCoUCNjY2qg+bcRAZFj6cICJdVFpQjhOf/4Kfd54VOxQiyZJcUy5DlpBQ31zrxBsXRI6ESDzsY0ItSUhIwKcLd4gdBhmQqopqyBVy2DhZYc4HYbBxttJYWXfvA4j0leQSE0NuXz56dP1oHkxMiIiaNnr0aNaRpDXFuaV4f+onCAr2x5P/Gg23B1w0Wt7d+wAifSW5xKQjnd/1YVLCjp4DO8mT1Bnywwki0i2m5iZw8XGGk6eD2KEQ6QXJtYEw5PblQUFBWPPrcrHDIBIVO79TS1hPkiYU55Y2WlZWUA5TMxPMXvs3PPL3AairE5rct/B2MSpLlU2ua6+goCAEBQWp5VhEukhyiQk7vxMZNkN+OEFE2nf1p+tImLgeJz7/RbXszvVcrJy4HvtXHwYAXDiYgjdC3kXKibQG+1aUVGL11E+wZTFnaidqC8k15SIiw8bO70SkTXautvDo0QWdfP6c1d7CzgIePbvANaAzAEBhYQorR0uYmps22NfU3BRBIQHoEthZqzETSRUTEyIiIqJmOHs6YPbavzVYZmVvgWffn6L67vuwF2J2RDTa11huhImLx2g8RiJ9IbnEhB1fiYiIiIj0j+TaQNxvx1epj8gVNzQBOel5YodBJDqlUoni4mLVhw8niEibaqprcfHwVVQrq8UOhUjvSC4xMeSOr397YBb+9sAsscMgEhVH5aKW7Nq1i/UkadSZveex9f++EmWG9127dmHXrl1aL5dIWyTXlMuQO746mDu1vhGRnouNjUVMzJ/z8EydOlXEaEjX+Pr6sq4kjQp4xAd5GQXoHtz2edTUxdfXV+tlEmmT5BITQ5ZfkQuACQoZNkN+OEGtS01NRX5FLutJ0hgbJyuMiQ4WpezU1FQA+pWgVFVWw9TMROwwSEfwL7qEbLn0MbZc+ljsMIiIdNaECRNYT5LemjBhAiZMmCB2GGpz41wW4sesxk/bfxU7FNIRTEyIiIiISOvMrc3g6G4PG2drsUMhHSG5plwcLpiIiIhI+jp1c8QLn4WLHQbpEMm9MeGIPH+KG5og+WGQidqLwwUTERHpJ8klJoY8XDAR6ffDCT5saOzen0dLP5+76zjfE+mC8uIKAED21Tt4c/RqrP57En7/7rLIURHpPsklJgqFAjY2NqoPR+QhMix8OEFEuiz9t0wsH7cWP27+GTKZDEbGMuRm5CPzQrbYoRHpPMn1MTFk8/r+Q+wQiETH4YKpJawnSWzWjpboEtAZTp4OcPF1xitfR6E4txSWdhYdPvaFCxfUECGR7mJiQkRERKQmjh72iNzw9wbLbJysOnTMY8mnkXXpD0x67XEYy407dCwiXcZHjRKSWnAZqQVso0pE1BzWkySGojslKM0v09jx036+gasn0rD3q6+xf/9+jZVDJDa+MZGQA9d3AwB87QNFjoSISDexntSuixcvIi4uDo6OjujevTuio6MbrL9y5QomTJiAy5frk8WEhASkpKSguLgYr732Gjw8PBAZGYlOnTqhoqICa9euhUwmE+NU7psgCFgzcxMsbc0R9Vk4jOX1z3zPf38F1coaPPhEUIfLeObNUFSVV+HhoQ8BYJMu0l98Y0JERET3Zfny5Vi2bBnWrl2LPXv2oLKyUrUuLS0NGzduhJVVfTOmyspK7N27Fx988AFef/11vPXWW9i2bRuCg4OxatUqODs748iRI2Kdyn2TyWR48PHuKMkrw+Z/7FQt3/feIexd+Z1ayjA1M4GVg6VajkWky/jGhIiIiNokKSkJSUlJqu+HDx/Ghg0bAACOjo7Iy8uDm5sbAMDb2xtvvvkmjh8/DgAoKCiAi4sLAMDDwwOZmZnIzs5G//79Gyz7q+TkZCQnJzcZT1ZWltrOrSNGzhuGO+n56NTNUbUs7K0JqK1uep6l0vwynDtwCf1Ce0FhYaqtMIl0nuQSE878TmTYWAcQiSc8PBzh4eGq75GRkcjIyIC3tzdycnLg5OTU7L7Ozs7Iy6ufZyYjIwPu7u7w9PRERkaGatljjz3WaL+wsDCEhYU1eczQ0NAOnE3HVZRUYt97h9Br9AOYkTCpwTq3QJdm9zu96xwOfXwCZlYK9B3XU9NhEkmG5BKT+Ph4LF26VPXd399fxGikhRO3kT5obx1w97qPOxrT5LqmlqtbSzGQ4fhrHawP18OiRYuwePFiWFtbY/LkyVAoFFi4cCEWL14MBweHBtvK5XJMmDABzz33HHJzc7FixQq4uroiIiICZ8+eRV1dHYYNGybSmdyfnPR8nN13ESZmJvDp79nm/QZM7ANzGzP0GMG+UET3klxiEhsbi5iYPyvzqVOnihiNdtmbOba+EZGeM+Q6gFrHelK7/P39GzWzWrlyZYPvhw4dUv37+eefb3SMTZs2aSQ2bejaswsiP/w7nLo6tL7xPSztzPHwpAfbXZ6Pj0+79yGSEsklJoY8uVpY99lih0AkOkOuA6h1rCdJ27oEdNZaWbt379ZaWURi4F90IiIiIiISHRMTCTl160ecuvWj2GEQEeks1pMktmplDX748BhuXbmt9mMnJiYiMTFR7ccl0hVMTCTk9B/HcfqP42KHQUSks1hPkthuXf4DRzaexKkdvzVa9+ve87j5+637PvaaNWuwZs2aDkRHpNuYmBARERGpiWdvdzyzLBQj5j7SYHlJXhl2/+cA9q8+JE5gRBIguc7vRERERLrM60F3mFubNVhm7WiJyUvHwsHNTpygiCSAb0yIiIiI1OToZ6ewfNxa3DzXeFb6HiEBWh3Fi0hqmJgQERERqYmjhz1cfJ1haW8hdihEksOmXBLibRcgdghERDqN9SSJrfswP3Qf5qeRY48aNUojxyXSFUxMJGSM9wSxQyAi0mmsJ0mfrVq1SuwQiDRKcomJUqmEUqlUfa+rqxMxGiLSNtYBRERE+klyiUl8fDyWLl2q+u7v7y9iNNq1L20XgLY/EYwbmlD/36Mx97Xu3vVxQxOa3FYdmiqPqDm6Xgfcez0D2rmmNfn72V7tiaWt9VBLZfx1u/bWk0RS8tJLLwHgmxPSX5Lr/B4bG4uioiLVx89PM+04dVFa4RWkFV4ROwwiURlyHUCtYz1J+uzAgQM4cOCA2GEQaYzk3pgoFAooFArVdyMjyeVWRNQBrAOIiIj0E/+iExERERGR6JiYEBERERGR6JiYEBERERGR6CTXx8SQ9XcZLHYIREQ6jfUk6bN58+aJHQKRRjExkZCHujwidghERDqN9STpihOf/4Irx9MQFj8BCgtTtRwzKipKLcch0lVsykVERESkZtkpd5B5IRvK8iqxQyGSDL4xkZDkix8BAMK6zxY5EiIi3cR6knTFhNjReHzBcJhbm6ntmKGhoQCA3bt3q+2YRLqEiYmEFFTmiR0CEZFOYz1JusJYbqTWpAQArl27ptbjEekaNuUiIiIi0iGVpUoIgiB2GERax8SEiIiI7svFixcxZcoUPP/883j//fcbrb9y5QoCAwNV35csWYLo6GhMnz4dO3fuBAAMHDgQ4eHhCA8Px9WrV7UWu646/8MV/OeJRBz8+LjYoQAALv94DduW7EFFSaXYoZABYGJCRERE92X58uVYtmwZ1q5diz179qCy8s+b17S0NGzcuBFWVlYAAEEQ4Ofnh/fffx+rVq3Cpk2bcPPmTZSUlEChUMDZ2RndunUT61R0xq+7fwcEwNTMROxQAABXT6Thwg8pKLhVJHYoZADYx4SIiIjaJCkpCUlJSarvhw8fxoYNGwAAjo6OyMvLg5ubGwDA29sbb775Jo4fr3/yL5PJMG3aNBQWFmLBggV49dVXYWFhgc2bN6NPnz5477338Omnn+LZZ59tUGZycjKSk5ObjCcrK0sDZ6leRbdL8NXb32LQM/3gM8Cz1e1HRw9DxoVs9BvfUwvRtW7Mi8EY+HRfOHs5ih0KGQAmJhIyqluo2CEQEek01pOadbfJ1V2RkZHIyMiAt7c3cnJy4OTk1OL+J0+exOrVqxEfHw8PDw+cPn0ad+7cQZ8+feDk5ISSkpJG+4SFhSEsLKzJ490dpUqX5WUWIPVkOroEdm5TYtLZxxmdfZybXJeQkKDu8FplojBhUkJaw8REQnztA1vfiIjIgLGe1K5FixZh8eLFsLa2xuTJk6FQKLBw4UIsXrwYDg4ODbbNz8/H+PHjERISgtjYWAQEBCA6OhpvvfUWvv/+exQVFeG9994T6Uw0x7tfV0QnPwt7V9sOH2v06NFqiIhIdzExISIiovvi7+/fqJnVypUrG3w/dOgQAMDBwQF37txpdIxt27ZpLD5d4ehuL3YIRJIgemLy1Vdf4fr166iursbChQvFDkenrfl1OQBgXt9/iBwJkfqwDiB1Yj1J+iwoKAgAcOHCBZEjIdIM0UflOnbsGF588cUmn6IQkf5jHUBEUlatrBE7BCK9IXpiIpfXv7SRyWQiR0JEYmAdQERSlX4mA2+Ofh+nvvit0brTu85h9bQkFP5RrP3AiCRK9MSkX79+eOedd+Dq6ip2KEQkAtYBRCRV5jZmcHS3h7WTZaN1JXmlKMwuQlVFVbP7K8ur8MOHx5CTnqfJMIkkQ+N9TLKysjBlyhRs3rwZCoUC8+fPh6urK6ytrfHGG29g4sSJrR6jI2OYxw3V/tB62nbvOXbkfJvat7VlcUdjWt2nqW3aWl7c0RjVsrvHiRua0OwxW9Na7Nrw1/PRd2LXAU1p6dpr7f9Lk/vec222tay/7nNvufcuU/f1/9d923q89pTb2jk1t70+M4RzNESdfZzxwmfhTa4bPmswHp3xMIzlxs3un34mA0c2noSyrAqPzx+uoSiJpEOjiUlxcTHefvtt2NnZAQDWrVuHiIgIjBw5EjNmzEB6ejq8vLxaPY7UxzAnMlSsA4jIkLWUlACA38BumBw3Ft36ddVSRES6TaOJiY2NDd555x3VZEzZ2dmqmxB3d3dkZWW16aaE6v3tgVlih0DULqwDSNtYT5KUGBkboceIgDZvv2vXLg1GQyQ+rQ4X7OnpiYyMDPj5+SEzMxNubm7tPoZSqYRSqVR9r6urU2eIOs3BvOUZdYl0HesA0jTWk6TPfH19xQ6BSKO02vk9MjISGzZsQFRUFHx9fe/rSWl8fDxsbW1Vn6tXr6o/UB2VX5GL/IpcscMgum+sA0jTWE+SPktNTUVqaqrYYRBpjFbemCQlJan+3VwH1raKjY1FTMyfnSenTp3aoeNJyZZLHwPgxGEkPawDSFtYT5I+mzBhAgBOsEj6S/SZ39tLoVBAoVCovhsZiT7iMRFpEesAIiIi/SS5xITty4kMG+sAIiIi/SS5R41sX05k2FgHEBER6SfJJSaxsbEoKipSffz8/MQOiYi0iHUAERGRfpJcUy62LycybKwDiIiI9JNMEARB7CA6IigoCD4+Pm3aNisr677mTWgvfStHm2WxnD+5ublh7dq1ao5I/7SnDtAmbf5+qhPj1q6W4mYd0Lr2/v5Loe435HK0WZYUyjHIOkAwIOPHj2c5Ol4WyyF9IdX/94xbu6Qat1TpW92vb+Vosyx9K0dfsA0EERERERGJjokJERERERGJjokJERERERGJjokJERERERGJjokJERERERGJzqASk7CwMJaj42WxHNIXUv1/z7i1S6pxS5W+1f36Vo42y9K3cvSF5OcxISIiIiIi6TOoNyZERERERKSbmJgQEREREZHomJgQEREREZHo5GIHoAkXL15EXFwcHB0d0b17d0RHR7e47t1330VWVhby8/MxbNgwTJ8+XSPlbN26FYcPH0ZdXR369u2LiIgIjZQDANevX8fAgQNx+/btNpVxP+UsWbIE+fn5KCwsxFNPPYWJEye2WkZ2djbmz58PV1dXWFtb44033mh2+ZYtW7Bv3z7U1NTgueeewyOPPNLmc2lPOWvWrMG5c+dQXV2N/v374/nnn29zOe0tCwDy8/MxePBg7Nu3D15eXu0qi3RPS783AHDlyhVMmDABly9fBgAkJCQgJSUFxcXFeO211xAYGChG2O3+fZ8zZw4qKyshl8sxadIkjB8/XqvxaqvuEDPu119/Hb/++itsbW0xbNgwPPvss6LFLXW8D+B9gLbuA3gPoGaCHpo5c6aQkpIiCIIgjBo1SqioqGhx3aOPPio8+eSTwuDBg4VDhw5ppJzy8nJh1KhRwssvvyzMmDFDuHz5ssbOp6CgQJg9e7YwcODANpdxP+fz6aefCoIgCDk5OcKTTz7ZpjKWLFkiHDhwQBAEQZg+fbpw/fr1Zpc/8sgjQl1dnVBWViaMGTOmXefSnnI2btwoCIIg1NTUCMOGDWtXOe0tS6lUCnPmzBHGjx+v2o6kraXfm2vXrgmxsbFCv379BEEQhIqKCiEkJEQQBEFITU0VwsPDtR/w/7S3Xundu7cwb948Yfbs2UJWVpbW49VW3aFu7Yn7scceE+bNmydMnz5duHjxolgh6wXeB/A+QFv3AbwHUC+9aMqVlJSE4OBg1Wfjxo2qLNTR0RF5eXmqbbOzsxuti4qKwvbt27F9+3YsX75cI+XcuXMH58+fR1xcHFasWIGXXnpJI+Xcvn0bMTExiI+Ph0Kh0NjPLT8/H9OmTUNhYSEWLFiAV199tcWymjqOu7s7srKyml0ul8shk8lgYWGB8vLyNh3/fsqZMWMGKisrER0djVdeeaVd5bS3rEWLFuHll1+Gg4NDu8sh3dCe3xtvb2+8+eabsLKyAgAUFBTAxcUFAODh4YHMzEydjLupejIhIQGJiYl48cUXERsbq7W4m4pJk3WHurUn7tjYWLz//vt4++23sWDBAnEClijeB/y5jvcB7S+nI/cBvAdQL71ITMLDw3Ho0CHVJyIiAhkZGQCAnJwcODk5qbb19PRssE4ul2P9+vUwNjaGo6MjlEqlRspxcXGBjY0NzMzMYGdnh6qqKo2U89tvv6GgoACvvvoqUlJSsGzZMo2U4+TkhJMnTyI6Ohrx8fHo27dvs+Xc697jZGZmws3NrdnlcrkcgiCgvLwc5ubmbTr+/ZSTkpKC2bNnIyoqCmPGjGlXOe0py9raGteuXUNCQgJOnDjR4v8b0l3t+b35K2dnZ9Uf+oyMDLi7u2slZqBjv+9WVlY4f/48AMDJyanF+ktTtFV3qFtb4+7SpQuOHz8OIyMj2NraihavVPE+gPcBHSmnI/cBvAdQL72cxyQlJQVLliyBtbU1+vXrh7lz52LhwoVYvHgxcnNzm1yXn5+PmpoazJ49G8HBwRopZ/369Th8+DAA4Pnnn29zW8n2lnNXcHAwDh06pJGf29NPP43AwECEhIRALpcjICCgTU9LcnNzER0dDQcHB3Tu3Bnm5uYYO3YsOnXq1GD5a6+9hp07d2L37t0oLy/HokWLMGDAgDafS1vLWbx4Mfz9/dGzZ09YWVnBwcEB7733XpvLae853RUeHo64uDi2L9UDLf3e3H0qdu/v4tq1a3H27Fnk5uZixYoVol0D7a1XXnzxRQiCgKKiIixevBgBAQFajVdbdYeYccfHx+P69etQKpV47rnnMGjQINHiljreB9TjfYDm7wN4D6BeepmYEBERERGRtOhFUy4iIiIiIpI2JiZERERERCQ6JiZERERERCQ6JiZERERERCQ6JiZERERERCQ6JiZERERERCQ6udgBkG5JSkrC+vXr4ePjg9raWlhaWmLu3LlISEiApaUlcnNzMWPGDEycOBGCIOCTTz7B8ePH8eGHH4odOhGpQXvqgCVLliA/Px+FhYV46qmnMHHiRLHDJ6IO4n0AiYmJCTUSGRmJ8PBwAMDIkSMRERGB48ePw9zcHNXV1Vi9ejUAYOPGjaoJqYhIf7SlDhAEAX5+fpg2bRpyc3MRERHBxIRIT/A+gMTCxIQa+fDDD3H48GEIgoARI0bg559/hrm5OQDAxMQEL730EoD6mUvT09Nx/vx5McMlIjVrax0wbdo0FBYWYsGCBW2a8ZmIpIH3ASQWJibUyJw5c1RPSurq6jB06FAUFhbCzs4O1dXVeO6557B+/XoYGxuLGygRaURb64DTp09j9erViI+Ph4eHh7hBE5Ha8D6AxMLEhFpkZGSEt99+G9OnT4ejoyMKCgowa9YsVkZEBqK5OqCoqAjjx49HSEgIYmNjERAQwLcmRHqI9wGkTTJBEASxgyAiIiIiIsPG4YKJiIiIiEh0TEyIiIiIiEh0TEyIiIiIiEh0TEyIiIiIiEh0TEyIiIiIiEh0TEyIiIiIiEh0TEyIiIiIiEh0TEyIiIiIiEh0TEyIiIiIiEh0/w9vvwFG3rIuCQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 900x300 with 3 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# select european samples\n",
    "fig, axes = plt.subplots(1, 3, figsize = (9, 3), gridspec_kw = {'wspace': 0.4})\n",
    "\n",
    "euro_samples = metadata.query('ancestry_inferred_by_SVs == \"EUR\"')\n",
    "euro_metadata = metadata.copy()\n",
    "\n",
    "cutoffs = []\n",
    "for i, comp in enumerate(['PC1', 'PC2']):\n",
    "    \n",
    "    # determine cutoffs\n",
    "    skew = euro_samples[comp].mean() - euro_samples[comp].median()\n",
    "    \n",
    "    # if left skewed, use 2.5% percentile\n",
    "    if skew < 0:\n",
    "        cut = np.percentile(euro_samples[comp], [1])[0]\n",
    "        euro_metadata = euro_metadata[euro_metadata[comp] > cut]\n",
    "        \n",
    "    else:\n",
    "        cut = np.percentile(euro_samples[comp], [99])[0]\n",
    "        euro_metadata = euro_metadata[euro_metadata[comp] < cut]\n",
    "        \n",
    "    cutoffs.append(cut)\n",
    "    \n",
    "    axes[i].hist(metadata.query('ancestry_inferred_by_SVs == \"EUR\"')[comp], bins = 100);\n",
    "    axes[i].set_yscale('log')\n",
    "    aesthetics.strip_axis(ax = axes[i], x = 'ignore', y = 'ignore')\n",
    "    axes[i].set_ylabel('count')\n",
    "    axes[i].set_xlabel(comp)\n",
    "    \n",
    "    # add cutoff\n",
    "    axes[i].axvline(x = cut, color = 'black', linestyle = 'dashed')\n",
    "    \n",
    "sns.scatterplot(x = 'PC1', y = 'PC2', hue = 'ancestry_inferred_by_SVs', ax = axes[2], data = euro_samples, s = 2, ec = None)\n",
    "axes[2].axhline(cutoffs[1], color = 'black', linestyle = 'dashed')\n",
    "axes[2].axvline(cutoffs[0], color = 'black', linestyle = 'dashed')\n",
    "sns.despine(ax = axes[2])\n",
    "axes[2].legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "fig.savefig(f'data/{folder_name}/european-samples.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "427221f9-034c-4347-a215-5c7faf6f640c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10590 total samples\n",
      "6346 European samples\n"
     ]
    }
   ],
   "source": [
    "print(metadata.shape[0], 'total samples')\n",
    "print(euro_metadata.shape[0], 'European samples')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "873a0d55-da74-4097-bac3-3e16fd65e920",
   "metadata": {},
   "outputs": [],
   "source": [
    "# here, we define the models that will be run\n",
    "covariates = {'no-covariates': [],\n",
    "              'sex': [\"sex\"],\n",
    "              'sex-ancestry': ['sex', 'PC1', 'PC2', 'PC3']}\n",
    "\n",
    "# any analysis without ancestry must be subset to european samples\n",
    "euro_samples = euro_metadata['entity:sample_id'].to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5332c522-0432-4d3f-810c-239680a44e37",
   "metadata": {},
   "source": [
    "In the following code block, we run our association tests."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2c0c9bba-f061-4759-b7bf-ad863fd09438",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EWSR1 ewing\n",
      "FLI1 ewing\n",
      "MYCN neuroblastoma\n",
      "TERT neuroblastoma\n",
      "ALK neuroblastoma\n",
      "TP53 osteosarcoma\n",
      "TP53 neuroblastoma\n",
      "TP53 ewing\n",
      "RB1 osteosarcoma\n",
      "ATRX neuroblastoma\n",
      "SHANK2 neuroblastoma\n",
      "BRCA1 neuroblastoma\n",
      "CDKN2A ewing\n",
      "DLG2 osteosarcoma\n"
     ]
    }
   ],
   "source": [
    "for index, row in genes_to_test.iterrows():\n",
    "    gene, disease = row[['gene', 'disease']].values\n",
    "    print(gene, disease)\n",
    "    \n",
    "    file_prefix = f'data/{folder_name}/{gene}-{disease}/'\n",
    "    \n",
    "    gene_disease_results = []\n",
    "    for cohort in ['discovery', 'validation']:\n",
    "\n",
    "        # osteosarcoma has no validation datasets, so we skip\n",
    "        if disease == 'osteosarcoma' and cohort == 'validation':\n",
    "            continue\n",
    "            \n",
    "        #######################\n",
    "        #### LOAD OUR DATA ####\n",
    "        #######################\n",
    "        \n",
    "        analysis_dosages = pd.read_csv(file_prefix + f'{cohort}-dosages.csv')\n",
    "        analysis_metadata = pd.read_csv(file_prefix + f'{cohort}-metadata.csv')\n",
    "\n",
    "        # get the cases and controls\n",
    "        analysis_cases = analysis_metadata.query(f'disease == \"{disease}\"')['entity:sample_id'].to_list()\n",
    "        analysis_controls = analysis_metadata.query(f'disease == \"control\"')['entity:sample_id'].to_list()\n",
    "        \n",
    "        # note that we have to drop CNVs here (their AFs are inaccurate)\n",
    "        analysis_dosages = analysis_dosages[~analysis_dosages['name'].str.contains('_CNV_')]\n",
    "        \n",
    "        # load the model data. We next add dosages and burdens to this data.\n",
    "        model_data = metadata.set_index('entity:sample_id').loc[analysis_cases + analysis_controls]\n",
    "        model_data[\"disease\"] = model_data[\"disease\"].replace({disease: 1, \"control\": 0})\n",
    "        \n",
    "        # keep track of all our analyses\n",
    "        exog_cols = []\n",
    "        \n",
    "        ########################\n",
    "        #### ADD SV DOSAGES ####\n",
    "        ########################\n",
    "        \n",
    "        # define an n x m matrix, where n is sv, m is sample, and element is dosage\n",
    "        temp_dosages = analysis_dosages.set_index(\"name\")[analysis_cases + analysis_controls]\n",
    "        model_data.loc[temp_dosages.columns, list(temp_dosages.index)] = temp_dosages.T\n",
    "        \n",
    "        # all these SVs are analyzed independently\n",
    "        exog_cols += list(temp_dosages.index)\n",
    "        \n",
    "        #######################\n",
    "        #### ADD SV BURDEN ####\n",
    "        #######################\n",
    "        \n",
    "        # add three columns to model data: rare, common, and combined\n",
    "        # these describe how many of each SV is present in that sample around this gene\n",
    "        for rare in [True, False]:\n",
    "\n",
    "            # define a label for the analysis\n",
    "            label = {True: 'rare-burden', False: 'common-burden'}[rare]\n",
    "            exog_cols.append(label)\n",
    "            \n",
    "            # subset to rare or common variants\n",
    "            subanalysis_dosages = analysis_dosages[analysis_dosages['rare'] == rare]\n",
    "\n",
    "            # count the number of samples with at least one non-ref allele\n",
    "            sv_counts = (subanalysis_dosages[analysis_cases + analysis_controls] > 0).sum(axis=0)\n",
    "\n",
    "            model_data.loc[sv_counts.index, label] = sv_counts.astype(int)\n",
    "            \n",
    "        # add rare + common burden\n",
    "        model_data['combined-burden'] = model_data['rare-burden'] + model_data['common-burden']\n",
    "        exog_cols.append('combined-burden')\n",
    "        \n",
    "        # we also try a one-hot encoding\n",
    "        for col in ['rare-burden', 'common-burden', 'combined-burden']:\n",
    "            onehot_col = col + '-onehot'\n",
    "            model_data[onehot_col] = (model_data[col] > 0).astype(int)\n",
    "            exog_cols.append(onehot_col)\n",
    "            \n",
    "        model_data.reset_index().to_csv(file_prefix + f'{cohort}-model-data.csv', index=False)\n",
    "\n",
    "        ####################\n",
    "        #### RUN MODELS ####\n",
    "        ####################\n",
    "        \n",
    "        # store model results\n",
    "        modeling_results = []\n",
    "        \n",
    "        # for each set of covariates, run all our models\n",
    "        for covar_label, covar in covariates.items():\n",
    "            \n",
    "            # if ancestry isn't included as a covariate, we subset to euro samples\n",
    "            if 'ancestry' not in covar_label:\n",
    "                parent_covar_data = model_data.loc[set(model_data.index) & set(euro_samples)].copy()\n",
    "            else:\n",
    "                parent_covar_data = model_data.copy()\n",
    "            \n",
    "            # Run each SV and burden analysis\n",
    "            for col in exog_cols:\n",
    "                \n",
    "                analysis_type = 'burden' if 'burden' in col else 'sv'\n",
    "                \n",
    "                # define data\n",
    "                submodel_data = parent_covar_data[['disease', col] + covar]\n",
    "                \n",
    "                # we drop samples that aren't defined for our variable of interest\n",
    "                # this is usually SVs that can't be genotyped\n",
    "                submodel_data_nonna = submodel_data.dropna(subset=[col]).copy()\n",
    "                \n",
    "                # this has the potential to introduce bias, so we quantify what was dropped\n",
    "                old_counts = submodel_data['disease'].value_counts().loc[[0, 1]]\n",
    "                new_counts = submodel_data_nonna['disease'].value_counts().loc[[0, 1]]\n",
    "                dropped = old_counts - new_counts\n",
    "                \n",
    "                # define x and y for the model\n",
    "                Y = submodel_data_nonna[\"disease\"]\n",
    "                x = sm.tools.add_constant(submodel_data_nonna, has_constant = 'add')\n",
    "                x = x.drop(columns = ['disease'])\n",
    "                \n",
    "                # suppress some warnings\n",
    "                with warnings.catch_warnings():\n",
    "                    warnings.simplefilter(\"ignore\")\n",
    "                    logit_model = sm.Logit(\n",
    "                    Y,\n",
    "                    x,\n",
    "                    ).fit(disp=0, method=\"bfgs\", maxiter=1000)\n",
    "                \n",
    "                # store model results as a row\n",
    "                row = [analysis_type, covar_label, col, dropped.loc[0], dropped.loc[1]]\n",
    "                \n",
    "                # extract p and coeff (change coef to nan if p is nan)\n",
    "                # this happens whenever the model is singular\n",
    "                p = logit_model.pvalues.loc[col]\n",
    "                coef = logit_model.params.loc[col] if not np.isnan(p) else np.nan\n",
    "\n",
    "                row += [coef, p]\n",
    "                    \n",
    "                modeling_results.append(row)\n",
    "        \n",
    "        # combine into one dataframe\n",
    "        modeling_results = pd.DataFrame(modeling_results, columns = ['type', 'covar', 'model', 'controls_dropped', 'cases_dropped', 'coef', 'p'])\n",
    "    \n",
    "        ################################\n",
    "        #### BONFERRONI CORRECT SVS ####\n",
    "        ################################\n",
    "        if cohort == 'discovery':\n",
    "            for covar_label in covariates.keys():\n",
    "\n",
    "                covar_results = modeling_results[(modeling_results['type'] == 'sv') & \n",
    "                                                 (modeling_results['covar'] == covar_label)\n",
    "                                                ]\n",
    "\n",
    "                pvals = covar_results['p']\n",
    "                p_bonferroni = statsmodels.stats.multitest.multipletests(pvals, alpha=0.05, method='bonferroni')[1]\n",
    "\n",
    "                modeling_results.loc[covar_results.index, 'p_bonferroni'] = p_bonferroni\n",
    "        \n",
    "        ################\n",
    "        #### EXPORT ####\n",
    "        ################\n",
    "\n",
    "        modeling_results.to_csv(file_prefix + f'{cohort}-model-results.csv', index=False)\n",
    "    \n",
    "    ####################################\n",
    "    #### DEFINE SIGNIFICANT RESULTS ####\n",
    "    ####################################\n",
    "    \n",
    "    discovery = pd.read_csv(file_prefix + 'discovery-model-results.csv')\n",
    "    \n",
    "    if disease == 'osteosarcoma':\n",
    "        cohorts = [discovery]\n",
    "    else:\n",
    "        validation = pd.read_csv(file_prefix + 'validation-model-results.csv')\n",
    "        cohorts = [discovery, validation]\n",
    "        \n",
    "    # subset results in each cohort down to those that are significant\n",
    "    sig_results = []\n",
    "    for i, cohort in enumerate(cohorts):\n",
    "        cohort_sig_results = []\n",
    "        \n",
    "        for index, row in cohort.iterrows():\n",
    "            \n",
    "            # define our p value cutoff\n",
    "            if i == 0 and row['type'] == 'sv':\n",
    "                p_col = 'p_bonferroni'\n",
    "            else:\n",
    "                p_col = 'p'\n",
    "            \n",
    "            if row[p_col] < 0.05:\n",
    "                cohort_sig_results.append(index)\n",
    "            # get all models, which use different covariates\n",
    "                \n",
    "        # subset the cohort to these results\n",
    "        cohort = cohort.loc[cohort_sig_results]\n",
    "        \n",
    "        sig_results.append(cohort)\n",
    "    \n",
    "    # combine results\n",
    "    if len(cohorts) == 1:\n",
    "        sig_results = sig_results[0].copy()\n",
    "    else:\n",
    "        sig_results = sig_results[0].merge(sig_results[1], on = ['type', 'covar', 'model'], \n",
    "                                           suffixes = ['_discovery', '_validation'], how = 'inner')\n",
    "\n",
    "    if len(sig_results):\n",
    "        sig_results.to_csv(file_prefix + 'significant-results.csv', index=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2293f8f-68d9-4ff9-be0d-a4179c9e9ede",
   "metadata": {},
   "source": [
    "# Plot results\n",
    "\n",
    "Here, we define and make some plots for visual inspection. We visualize ~1 Mb around the gene of interest and plot all SVs, denoting which ones were significant in our association testing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e1ad023e-cab8-4a33-9b86-c004e15a3579",
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_colors = {\"DEL\": \"blue\", \"DUP\": \"red\", \"INS\": bold_10[0]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e6da4011-25ce-4353-86fa-6f6b8960e08f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jettcrowdis/opt/anaconda3/envs/pedsv-test-conda/lib/python3.7/site-packages/ipykernel_launcher.py:126: UserWarning: FixedFormatter should only be used together with FixedLocator\n"
     ]
    }
   ],
   "source": [
    "for index, row in genes_to_test.iterrows():\n",
    "    gene, disease = row[['gene', 'disease']].values\n",
    "    \n",
    "    file_prefix = f'data/{folder_name}/{gene}-{disease}/'\n",
    "    \n",
    "    # plot discovery and validation, split by common and rare on same plot\n",
    "    num_plots = 2 if disease == 'osteosarcoma' else 4\n",
    "    gene_fig, gene_axes = plt.subplots(num_plots, 1, figsize=(7, num_plots), gridspec_kw = {'hspace': 0.5})\n",
    "    \n",
    "    # figure for gene burden\n",
    "    burden_fig, burden_axes = plt.subplots(num_plots, 1, figsize=(6, num_plots * 3), gridspec_kw = {'hspace': 0.5})\n",
    "    \n",
    "    plot_num = 0\n",
    "    for sv_type, cohort in itertools.product(['common', 'rare'], ['discovery', 'validation']):\n",
    "\n",
    "        # osteosarcoma has no validation datasets, so we skip\n",
    "        if disease == 'osteosarcoma' and cohort == 'validation':\n",
    "            continue\n",
    "            \n",
    "        #######################\n",
    "        #### LOAD OUR DATA ####\n",
    "        #######################\n",
    "        \n",
    "        # load counts, and subset to rare or common\n",
    "        analysis_counts = pd.read_csv(file_prefix + f\"{cohort}-sv-genotypes.csv\")\n",
    "        analysis_counts = analysis_counts[analysis_counts['rare'] == (sv_type == \"rare\")].reset_index()\n",
    "        \n",
    "        # load model data, strictly for burden testing\n",
    "        model_data = pd.read_csv(file_prefix + f\"{cohort}-model-data.csv\")\n",
    "\n",
    "        # significance results - based on the model with both sex and ancestry\n",
    "        model_results = pd.read_csv(file_prefix + f\"{cohort}-model-results.csv\")\n",
    "        model_results = model_results.query('covar == \"sex-ancestry\"').set_index(\"model\")\n",
    "        \n",
    "        # gene-related data\n",
    "        g_gtf = pd.read_csv(file_prefix + f\"{gene}-gtf.csv\")\n",
    "        \n",
    "        ########################\n",
    "        #### PLOT SV BURDEN ####\n",
    "        ########################\n",
    "        ax = burden_axes[plot_num]\n",
    "\n",
    "        max_count = 0\n",
    "        for label in ['cases', 'controls']:\n",
    "            ind = label == \"cases\"\n",
    "            data = model_data.query(f\"disease == {ind}\")[f\"{sv_type}-burden\"].value_counts()\n",
    "            \n",
    "            if max(data.index) > max_count:\n",
    "                max_count = max(data.index)\n",
    "\n",
    "            for i in range(max(data.index)):\n",
    "                if i not in data.index:\n",
    "                    data.loc[i] = 0\n",
    "\n",
    "            data = data.sort_index()\n",
    "            data = data / data.sum()\n",
    "\n",
    "            base = -0.2 if label == 'cases' else 0.2\n",
    "            x = np.arange(base, max(data.index) + 1 + base, 1)\n",
    "\n",
    "            ax.bar(x=x, height=data, width=0.4, label=label)\n",
    "\n",
    "        aesthetics.strip_axis(ax=ax, x='label', y = 'ignore')\n",
    "        \n",
    "        if plot_num == 1:\n",
    "            ax.legend(bbox_to_anchor = (1, 0))\n",
    "\n",
    "        ax.set_xlabel('SV count')\n",
    "        ax.set_ylabel('Density')\n",
    "        cohort_name = 'GMKF' if cohort == 'discovery' else 'St. Jude'\n",
    "        ax.text(x = 0.5, y = 1.05, s = f'{cohort_name} SVs', \n",
    "                transform = ax.transAxes, ha = 'center', va = 'bottom')\n",
    "        ax.set_xticks(range(max_count + 1))\n",
    "        \n",
    "        if sv_type == 'rare':\n",
    "            ax.set_yscale('log')\n",
    "            \n",
    "        # add p value\n",
    "        coef, p = model_results.loc[f'{sv_type}-burden', ['coef', 'p']].values\n",
    "        s = f\" = {round(coef, 2)}\\np = {p:.2}\"\n",
    "        ax.text(x = 1, y = 0.9, s = s, \n",
    "                transform = ax.transAxes, ha = 'right', va = 'top')\n",
    "        \n",
    "        #######################\n",
    "        #### PLOT THE GENE ####\n",
    "        #######################\n",
    "        ax = gene_axes[plot_num]\n",
    "        gene_data = genes_to_test[(genes_to_test['gene'] == gene) &\n",
    "                                  (genes_to_test['disease'] == disease)].iloc[0]\n",
    "        \n",
    "        # we weirdly have to do this now because the code that plots\n",
    "        # the genes does so with relative sizes\n",
    "        ylim = [-0.02, 0.02] if sv_type == 'rare' else [-1, 1]\n",
    "        yticks = [-0.02, 0, 0.02] if sv_type == 'rare' else [-1, 0, 1]\n",
    "\n",
    "        # plot the genetic elements\n",
    "        color = {'neuroblastoma': mc.palette['neuroblastoma']['main'], 'ewing': mc.palette['ewing']['main'], 'osteosarcoma': mc.palette['osteosarcoma']['main']}[disease]\n",
    "        feature_colors = {\"exon\": color, \"CDS\": color}\n",
    "        window = gene_data.loc[['window_start', 'window_end']].values\n",
    "        feature_span = {\"exon\": ylim[-1] / 4, \"CDS\": ylim[-1] / 2}\n",
    "\n",
    "        # custom code that plots a genomic element\n",
    "        ax = sv_plot.plot_genomic_elements(\n",
    "            ax,\n",
    "            g_gtf,\n",
    "            window=window,\n",
    "            feature_span=feature_span,\n",
    "            feature_colors=feature_colors,\n",
    "            label=False,\n",
    "        )\n",
    "        aesthetics.strip_axis(ax=ax, x=\"label\", y=\"ignore\")\n",
    "\n",
    "        # plot the range of interest\n",
    "        ax.set_xlim(\n",
    "            window + np.array([-10000, 10000])\n",
    "        )\n",
    "        ax.spines[\"bottom\"].set_bounds(window)\n",
    "\n",
    "        # add labels\n",
    "        ax.text(s=\"cases\", x=-0.03, y=0.75, transform=ax.transAxes, ha=\"right\", va='center')\n",
    "        ax.text(s=\"controls\", x=-0.03, y=0.25, transform=ax.transAxes, ha=\"right\", va='center')\n",
    "        ax.set_ylabel('AF', labelpad = 15)\n",
    "\n",
    "        # switch to Mb\n",
    "        contig = gene_data.loc[\"chrom\"]\n",
    "        ax.set_xticklabels(ax.get_xticks() / 1e6)\n",
    "        ax.set_xlabel(f\"{contig} (Mb)\")\n",
    "        \n",
    "        ax.set_yticks(yticks)\n",
    "        ax.set_yticklabels(np.abs(yticks))\n",
    "        ax.set_ylim(ylim)\n",
    "        \n",
    "        # add title for gene\n",
    "        # ax.text(x = gene_data.loc[['start', 'end']].mean(), y = ylim[-1], s = gene, \n",
    "        #         fontstyle = 'italic', ha = 'center', fontsize = 8)\n",
    "        \n",
    "        # label for plot\n",
    "        cohort_name = 'GMKF' if cohort == 'discovery' else 'St. Jude'\n",
    "        ax.text(x = 0.01, y = 1, s = f'{cohort_name} SVs', transform = ax.transAxes, ha = 'left')\n",
    "        \n",
    "        ######################\n",
    "        #### PLOT THE SVS ####\n",
    "        ######################\n",
    "        # its much easier to plot everything at once with plt.stem, but that function is poor\n",
    "        # so we do it manually. We scale the frequency to fill the plot\n",
    "        case_y = analysis_counts['AF_cases']\n",
    "        control_y = analysis_counts['AF_controls']\n",
    "\n",
    "        # add the points\n",
    "        ax.scatter(x=analysis_counts[\"midpoint\"], y=case_y, color=\"black\", s=3, zorder=3)\n",
    "        ax.scatter(x=analysis_counts[\"midpoint\"], y=-control_y, color=\"black\", s=3, zorder=3)\n",
    "        \n",
    "        ################################\n",
    "        #### DENOTE SIGNIFICANT SVS ####\n",
    "        ################################\n",
    "        \n",
    "        # add the line denoting the SV\n",
    "        legend_handles = {}\n",
    "        for i in range(len(analysis_counts)):\n",
    "            \n",
    "            loc = analysis_counts[\"midpoint\"][i]\n",
    "            \n",
    "            # sometimes SVs are shown that are way outside the boundaries\n",
    "            if not (window[0] <= loc <= window[1]):\n",
    "                continue\n",
    "            \n",
    "            name = analysis_counts[\"name\"][i]\n",
    "\n",
    "            case, control = case_y[i], control_y[i]\n",
    "\n",
    "            ax.plot(\n",
    "                [loc, loc],\n",
    "                [case, -control],\n",
    "                color=\"black\",\n",
    "                linewidth=0.5,\n",
    "                zorder=3,\n",
    "            )\n",
    "            \n",
    "            # look up the p value of the sv, if it exists\n",
    "            if name in model_results.index:\n",
    "                p_col = 'p_bonferroni' if cohort == 'discovery' else \"p\"\n",
    "                p_val, coeff = model_results.loc[name, [p_col, \"coef\"]]\n",
    "\n",
    "                if p_val < 0.05:\n",
    "\n",
    "                    y = case if coeff > 0 else -control\n",
    "                    text_y = y + 0.3 * ylim[-1] if coeff > 0 else y - 0.3 * ylim[-1]\n",
    "\n",
    "                    sv_type = name.split('_')[1]\n",
    "                    color = sv_colors.get(sv_type)\n",
    "\n",
    "                    scatter = ax.scatter(\n",
    "                        [loc], [y], color=color, s=10, zorder=3, label=sv_type\n",
    "                    )\n",
    "\n",
    "                    legend_handles[sv_type] = scatter\n",
    "\n",
    "                    # add text for coeff\n",
    "                    ax.text(\n",
    "                        x=loc,\n",
    "                        y=text_y,\n",
    "                        s=f\" = {round(coeff, 2)}, p = {p_val:.1}\",\n",
    "                        ha=\"center\",\n",
    "                        va=\"center\",\n",
    "                    )\n",
    "        \n",
    "        plot_num += 1\n",
    "        \n",
    "    gene_fig.savefig(file_prefix + 'sv-plot.pdf')\n",
    "    plt.close(gene_fig)\n",
    "    \n",
    "    burden_fig.savefig(file_prefix + 'sv-burden.png')\n",
    "    plt.close(burden_fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a0944c7-1721-4edf-bb41-2726a57ce576",
   "metadata": {},
   "source": [
    "Note that while several of these figures depict common SVs (AF > 5%) that appear plausible, these have been individually reviewed and can all be explained by either technical factors."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47785e15-2034-4e60-97bd-f42f2dcd0542",
   "metadata": {},
   "source": [
    "# Quality control"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c60af273-bd67-475f-ac7f-627b354043e4",
   "metadata": {},
   "source": [
    "Here, we briefly extract some QC metrics for our \"significant\" SVs, and we also batch these SVs together. This allows us to manually review QC metrics for SVs that are deemed to be significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e7f5f448-4819-4afd-9d74-29f41bab2e91",
   "metadata": {},
   "outputs": [],
   "source": [
    "qc_cols = [line.strip() for line in open('data/qc-columns-order.txt')]\n",
    "\n",
    "sv_cols = ['name', 'svtype', 'ALGORITHMS', 'EVIDENCE', 'SL_MEAN', 'SL_MAX', 'NCR', 'HG38_REF_PATCH_LOCUS']\n",
    "sv_cols += [c for c in svs['discovery'].columns if c.startswith('gnomad') and '_CN_' not in c]\n",
    "\n",
    "ucsc_tracks = [\"simpleRepeat\", \"genomicSuperDups\", \"rmsk\", \"microsat\"]\n",
    "\n",
    "df_cols = qc_cols + sv_cols + ucsc_tracks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5404fd86-cc84-4ba6-9234-ba372e3491c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "significant_results_qc = []\n",
    "for index, row in genes_to_test.iterrows():\n",
    "    gene, disease = row[['gene', 'disease']].values\n",
    "    \n",
    "    file_prefix = f'data/{folder_name}/{gene}-{disease}/'\n",
    "    \n",
    "    ######################\n",
    "    #### READ IN DATA ####\n",
    "    ######################\n",
    "    \n",
    "    # begin by loading the significant results\n",
    "    if os.path.exists(file_prefix + 'significant-results.csv'):\n",
    "        significant_results = pd.read_csv(file_prefix + 'significant-results.csv')\n",
    "    else:\n",
    "        continue\n",
    "        \n",
    "    significant_svs = significant_results.query('type == \"sv\" & covar == \"sex-ancestry\"')['model'].to_list()\n",
    "\n",
    "    if len(significant_svs) == 0:\n",
    "        continue\n",
    "    \n",
    "    #################################################\n",
    "    #### COMBINE SV GENOTYPES WITH MODEL RESULTS ####\n",
    "    #################################################\n",
    "    \n",
    "    # we load in discovery and validation separately, to better enable side by side comparison\n",
    "    cohort_significant_results = []\n",
    "    for cohort in ['discovery', 'validation']:\n",
    "        if disease == 'osteosarcoma' and cohort == 'validation':\n",
    "            continue\n",
    "\n",
    "        cohort_results = pd.read_csv(file_prefix + f'{cohort}-model-results.csv')\n",
    "        cohort_results = cohort_results[(cohort_results['model'].isin(significant_svs)) &\n",
    "                                        (cohort_results['covar'] == \"sex-ancestry\")]\n",
    "        \n",
    "        # some reformating of the model results\n",
    "        cohort_results = cohort_results.drop(columns = ['type', 'covar'])\n",
    "        cohort_results = cohort_results.rename(columns = {'model': 'name'})\n",
    "        cohort_results.insert(1, 'cohort', cohort)\n",
    "        \n",
    "        cohort_genotypes = pd.read_csv(file_prefix + f'{cohort}-sv-genotypes.csv')\n",
    "        cohort_genotypes_condensed = cohort_genotypes[['name', 'gene', '#chr', 'start', 'end']].copy()\n",
    "        cohort_genotypes_condensed[['cases_gt']] = list(zip(cohort_genotypes[['cases_0', 'cases_1', 'cases_2']].values))\n",
    "        cohort_genotypes_condensed[['controls_gt']] = list(zip(cohort_genotypes[['controls_0', 'controls_1', 'controls_2']].values))\n",
    "        cohort_genotypes_condensed[['AF_cases', 'AF_controls']] = cohort_genotypes[['AF_cases', 'AF_controls']]\n",
    "        \n",
    "        merged_sv_data = cohort_results.merge(cohort_genotypes_condensed, on = ['name'])\n",
    "        \n",
    "        cohort_significant_results.append(merged_sv_data)\n",
    "        \n",
    "    cohort_significant_results = pd.concat(cohort_significant_results)\n",
    "    \n",
    "    cohort_significant_results['disease'] = disease\n",
    "    \n",
    "    #############################################\n",
    "    #### ADD SV QUALITY METRICS FROM GATK-SV ####\n",
    "    #############################################\n",
    "    \n",
    "    sv_data = pd.read_csv(file_prefix + 'discovery-svs.csv')\n",
    "    sv_char = sv_data[sv_cols]\n",
    "    \n",
    "    cohort_significant_results = cohort_significant_results.merge(sv_char, on = 'name')\n",
    "    \n",
    "    ##########################################################\n",
    "    #### ADD IN CHARACTERISTICS ABOUT THIS GENOMIC REGION ####\n",
    "    ##########################################################\n",
    "    \n",
    "    for track in ucsc_tracks:\n",
    "        \n",
    "        track_results = []\n",
    "        copy = 1 if disease == 'osteosarcoma' else 2\n",
    "        for index, row in cohort_significant_results.drop_duplicates(subset = 'name').iterrows():\n",
    "            \n",
    "            chrom, start, end = row[['#chr', 'start', 'end']].values\n",
    "            \n",
    "            if start == end:\n",
    "                end += 1\n",
    "            \n",
    "            overlap = helper.fetch_track_overlap(track, (chrom, start, end))\n",
    "            \n",
    "            track_results += [overlap] * copy\n",
    "        \n",
    "        cohort_significant_results[track] = track_results\n",
    "        \n",
    "    cohort_significant_results = cohort_significant_results[df_cols]\n",
    "\n",
    "    cohort_significant_results.to_csv(file_prefix + 'significant-results-qc.csv', index=False)\n",
    "        \n",
    "    significant_results_qc.append(cohort_significant_results)\n",
    "    \n",
    "significant_results_qc = pd.concat(significant_results_qc).reset_index(drop = True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pedsv-test-conda",
   "language": "python",
   "name": "pedsv-test-conda"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
