# Gameplan

Coming back to devise a formulated gameplan after some discussion with Riaz and Ryan. See my plan that I sent them in the slack below:

```
- EWSR1/FLI1 in Ewing’s: Not much here in the 2.1 callset, unfortunately. The only two potential hits in my eyes are a common duplication favoring controls near EWSR1 and a rare variant within EWSR1 (but p = 0.04 in discovery). There are no CNVs.
    - Do a quick check in the validation cohort for the above two SVs
    - The joint SV model, while interesting, doesn’t seem worth pursuing further.

- MYCN in neuroblastoma: majority of validated signal are in a cluster of common SVs.
    - Need to examine these regions and genotype quality
    - I’m equivocal on whether these SVs are actually in a somatic rearrangement “hotspot”--I don’t think the somatic data is high enough quality to know just yet.
    - Since there’s RNA here, I think it makes sense to try to examine if these SVs (either collectively or individually) lead to higher expression.
    - If that’s the case, can do a deeper dive on the somatic-germline breakpoint interaction, probably using the somatic breakpoint data that Riaz has. Don’t think the CNV data will be good enough, but I can check.

- TERT/ALK in neuroblastoma: lots of activity near TERT. Nothing incredibly convincing in the common SV space, but some rare variants could validate. Nothing passes genome-wide significance in discovery.
    - I think these SVs are worth quickly checking in the validation cohort
    - If any validate, follow-up with RNA.

More general thoughts:
- These will likely take me a bit to get through but not so long. I can munch on these questions and then potentially join for more CWAS/category focused efforts, particularly if we want to intersect with RNA later.
- Since I’ll likely be working with the neuroblastoma RNA a bit and spinning up code for it, if there are any focused questions or loci that seem interesting for that dataset, let me know.
- I’ll just need to know more about the neuroblastoma RNA (how the data were generated, if we know what processing/normalization/filtering was done, if we have covariates for the tumors, etc)
```

Unfortunately, in digging into some of these results in the validation cohort, it's become clear that there may not be much in terms of specific dives... which is disappointing. Will continue to munch on the above, but I'm pretty sure that my time will be better spent getting acquainted with the RNA and joining in with Riaz on some of his efforts.