{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7faf7828-a5a8-4a0c-bab2-1583e9b9bf26",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "from collections import defaultdict\n",
    "\n",
    "import gseapy as gp\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1727894e-e3fa-4c1d-9d87-e0ba4a25a790",
   "metadata": {},
   "outputs": [],
   "source": [
    "# efficient fishers tests\n",
    "import pyranges\n",
    "import statsmodels.stats.multitest"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ff50573-913c-4d4a-85d2-2adcf20872ae",
   "metadata": {},
   "source": [
    "# CWAS Gene Set Enrichment\n",
    "\n",
    "In this notebook, we implement the gene set enrichment analysis for categories. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03d4797b-4907-4f7a-872c-4dc2aae28825",
   "metadata": {},
   "source": [
    "## How is gene set enrichment defined generally?\n",
    "\n",
    "Gene set enrichment (or \"GO term analysis\"--the terms are interchangeable in this context) generally attempts to determine if groups of genes are overrepresented in a particular context. Gene set enrichment always requires three components:\n",
    "\n",
    "1. A gene set (a list of genes)\n",
    "2. A background \"reference\" from which genes are \"drawn\" under the null model\n",
    "3. An \"observed\" set of genes (e.g. from an experiment)\n",
    "\n",
    "There are a variety of tests that can be performed on the above data, but the most common is the Fisher's exact test, which tests if membership in the observed set of genes is associated with membership in the gene set. The classic case in which this is used is differential expression with RNA sequencing. Consider an example below:\n",
    "\n",
    "1. A gene set corresponding to DNA damage repair: `BRCA1, BRCA2, TP53, MSH1, MSH2`\n",
    "2. An RNA-sequencing experiment in which all ~19k genes are sequenced (and are thus \"eligible\" to be differentially expressed genes),\n",
    "3. A list of differentially expressed genes (`BRCA1, BRCA2, TTN`)\n",
    "\n",
    "A gene set enrichment analysis might then be a Fisher's exact table of the following formulation:\n",
    "\n",
    "```\n",
    "number of genes in gene set and differentially expressed (n = 2) | number of genes in gene set and not differentially expressed (n = 4)\n",
    "number of genes not in gene set and differentially expressed (n = 1)| number of genes not in gene set and not differentially expressed (n = 18993)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b6438262-c67e-4fc3-be02-d512cd59b01d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.4928560531097303e-07\n"
     ]
    }
   ],
   "source": [
    "# fisher exact test for the above contigency table\n",
    "stat, p = scipy.stats.fisher_exact(np.array([[2, 4], [1, 18993]]))\n",
    "print(p)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "169c94f9-b23d-4d09-9755-934e8852dc9e",
   "metadata": {},
   "source": [
    "Another way to conceptualize the Fisher's test is in terms of \"true positives\", \"true negatives\", etc. In this way, the above contigency table would be:\n",
    "\n",
    "```\n",
    "True positives (n = 2) | False positives (n = 4)\n",
    "False negatives (n = 1)| True negatives (n = 18993)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e243b200-aac4-4914-bd1f-3749a04aa6f6",
   "metadata": {},
   "source": [
    "## How is gene set enrichment defined in this context?\n",
    "\n",
    "In the context of this project, we are interested in the categories that are demonstrated to have higher burden in cases vs. controls. If specific gene sets are enriched in these categories, this may reflect an underlying biological process that is targeted by the increased SV burden.\n",
    "\n",
    "Rather than unique genes, the process we define makes use of the fact that the Fisher's exact test is well defined for _counts_ generally--in this case, SV counts. Traditional packages for GO term enrichment analysis do not allow for gene duplicates (e.g. a gene hit multiple times by an SV), but that is valid here. We use the following Fisher's contigency table, for a given category and given gene set:\n",
    "\n",
    "```\n",
    "Count of gene-set genes affected by category SVs | Count of non-gene-set genes affected by category SVs\n",
    "Count of gene-set genes affected by non-category SVs | Count of non-gene-set genes affected by non-category SVs\n",
    "```\n",
    "\n",
    "We add a continuity correct of 1 to all cells, and the \"background\" of non-category SVs will be defined as all rare SVs that affect genes. This \"background\" will be further filtered such that:\n",
    "\n",
    "1) the background SV class matches the category of interest (coding vs. noncoding)\n",
    "2) SVs affecting more than 10 genes are removed (across coding and noncoding effects)\n",
    "3) SVs that do not have a coding/noncoding consequence detailed in the CWAS are removed\n",
    "4) noncoding SVs farther than 500 kb from their nearest gene for the noncoding consequence “INTERGENIC” are removed.\n",
    "\n",
    "## A sidenote on baseline enrichment\n",
    "\n",
    "Some categories, particular those that restrict to regions of the genome like \"adrenally expressed\", will be definition be enriched for particular gene sets. In these cases, the calculation of enrichment is particularly important in controls vs. cases."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13dd9c2d-e951-42c6-9e3f-0f769f4ec9dc",
   "metadata": {},
   "source": [
    "# Load in the data\n",
    "\n",
    "Here, we load in the SVs and associated sample dosages."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f5ac656-9484-4cb2-b493-3dbae9c63129",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Read in SVs, metadata, dosages, and reference genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3fbde6e9-7298-45c9-a6b8-7596f5d4c033",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define SVs and dosages for discovery and validation\n",
    "sv_path = \"gs://vanallen-pedsv-analysis/beds/PedSV.v2.5.3.full_cohort.analysis_samples.sites.bed.gz\"\n",
    "dosages_path = \"gs://vanallen-pedsv-analysis/beds/PedSV.v2.5.3.full_cohort.analysis_samples.allele_dosages.bed.gz\"\n",
    "\n",
    "# define metadata\n",
    "metadata_path = \"gs://vanallen-pedsv-analysis/sample_info/PedSV.v2.5.3.cohort_metadata.w_control_assignments.tsv.gz\"\n",
    "samples_path = \"gs://vanallen-pedsv-analysis/sample_info/PedSV.v2.5.3.final_analysis_cohort.samples.list\"\n",
    "\n",
    "gene_ref_path = \"data/updated-cwas/genes/gencode_hg38_protein_coding_genes_for_annotation_7_31_23 (1).txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "690438a4-3874-464c-8ef7-cecfb7364742",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19201"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_ref = pd.read_csv(gene_ref_path)\n",
    "\n",
    "# ENSG00 genes are removed\n",
    "gene_ref = gene_ref[~gene_ref[\"value\"].str.startswith(\"ENSG00\")]\n",
    "gene_ref = sorted(set(gene_ref[\"value\"].tolist()))\n",
    "len(gene_ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9643b8dc-53ed-49b7-98b0-0559502b52fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "intergenic_sv_to_gene_distances = pd.read_csv('data/updated-cwas/intergenic-sv-to-gene-distances.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8085c096-1ea6-4a62-ae31-87ea9afde972",
   "metadata": {},
   "source": [
    "Load metadata and SVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "258b57d3-0c78-46c7-801f-62ac188508af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "neuroblastoma case 688\n",
      "neuroblastoma control 4830\n",
      "ewing case 773\n",
      "ewing control 4574\n",
      "osteosarcoma case 284\n",
      "osteosarcoma control 4805\n"
     ]
    }
   ],
   "source": [
    "metadata = pd.read_csv(\n",
    "    metadata_path,\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "\n",
    "# add a sex label to metadata\n",
    "metadata[\"sex\"] = (metadata[\"chrX_CopyNumber\"].round() < 2).astype(int)\n",
    "\n",
    "###############\n",
    "### Samples ###\n",
    "###############\n",
    "samples = defaultdict(dict)\n",
    "\n",
    "total_samples = []\n",
    "for disease in [\"neuroblastoma\", \"ewing\", \"osteosarcoma\"]:\n",
    "    for cohort in [\"case\", \"control\"]:\n",
    "        disease_cohort_samples = metadata[(metadata[f\"{disease}_{cohort}\"] == True)][\n",
    "            \"entity:sample_id\"\n",
    "        ].tolist()\n",
    "\n",
    "        samples[disease][cohort] = disease_cohort_samples\n",
    "        total_samples += disease_cohort_samples\n",
    "\n",
    "        print(disease, cohort, len(disease_cohort_samples))\n",
    "\n",
    "total_samples = sorted(set(total_samples))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f7209b0-7630-4502-a796-5eb7faac61c1",
   "metadata": {},
   "source": [
    "Now we load the SVs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c3304748-d6f6-4e28-aea1-75fc7ddf24bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jettcrowdis/opt/anaconda3/envs/pediatric-germline-svs-3.7.13/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3457: DtypeWarning: Columns (34,35,37,39,41,45,46,47,48) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    }
   ],
   "source": [
    "###############\n",
    "##### SVs #####\n",
    "###############\n",
    "\n",
    "# only use the first 49 columns of the SV file, which includes affected genes\n",
    "svs = pd.read_csv(sv_path, sep=\"\\t\", usecols=range(49))\n",
    "\n",
    "###############\n",
    "### Dosages ###\n",
    "###############=\n",
    "\n",
    "# we only need the dosages of our samples in question\n",
    "dosage_head = pd.read_csv(dosages_path, sep=\"\\t\", index_col=False, nrows=1)\n",
    "\n",
    "cols = [3] + [i for i, c in enumerate(dosage_head.columns) if c in total_samples]\n",
    "\n",
    "dosages = pd.read_csv(dosages_path, sep=\"\\t\", index_col=False, usecols=cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4997797c-d7d6-46c2-9a47-75d22eac9146",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>svtype</th>\n",
       "      <th>AC</th>\n",
       "      <th>AF</th>\n",
       "      <th>ALGORITHMS</th>\n",
       "      <th>AN</th>\n",
       "      <th>BOTHSIDES_SUPPORT</th>\n",
       "      <th>...</th>\n",
       "      <th>PREDICTED_INV_SPAN</th>\n",
       "      <th>PREDICTED_LOF</th>\n",
       "      <th>PREDICTED_MSV_EXON_OVERLAP</th>\n",
       "      <th>PREDICTED_NEAREST_TSS</th>\n",
       "      <th>PREDICTED_NONCODING_BREAKPOINT</th>\n",
       "      <th>PREDICTED_NONCODING_SPAN</th>\n",
       "      <th>PREDICTED_PARTIAL_EXON_DUP</th>\n",
       "      <th>PREDICTED_PROMOTER</th>\n",
       "      <th>PREDICTED_TSS_DUP</th>\n",
       "      <th>PREDICTED_UTR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12000</td>\n",
       "      <td>30001</td>\n",
       "      <td>PedSV.2.5.2_CNV_chr1_1</td>\n",
       "      <td>CNV</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>depth</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ewing_chromHMM15_Quies,neuroblastoma_chromHMM1...</td>\n",
       "      <td>neuroblastoma_chromHMM15_EnhG_conserved,neurob...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12000</td>\n",
       "      <td>40001</td>\n",
       "      <td>PedSV.2.5.2_DUP_chr1_1</td>\n",
       "      <td>DUP</td>\n",
       "      <td>356</td>\n",
       "      <td>0.027423</td>\n",
       "      <td>depth</td>\n",
       "      <td>12982</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ewing_chromHMM15_Quies,neuroblastoma_chromHMM1...</td>\n",
       "      <td>neuroblastoma_chromHMM15_EnhG_conserved,neurob...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  #chrom  start    end                    name svtype   AC        AF  \\\n",
       "0   chr1  12000  30001  PedSV.2.5.2_CNV_chr1_1    CNV    0       NaN   \n",
       "1   chr1  12000  40001  PedSV.2.5.2_DUP_chr1_1    DUP  356  0.027423   \n",
       "\n",
       "  ALGORITHMS     AN  BOTHSIDES_SUPPORT  ... PREDICTED_INV_SPAN  PREDICTED_LOF  \\\n",
       "0      depth      0              False  ...                NaN            NaN   \n",
       "1      depth  12982              False  ...                NaN            NaN   \n",
       "\n",
       "  PREDICTED_MSV_EXON_OVERLAP PREDICTED_NEAREST_TSS  \\\n",
       "0                        NaN                 OR4F5   \n",
       "1                        NaN                 OR4F5   \n",
       "\n",
       "                      PREDICTED_NONCODING_BREAKPOINT  \\\n",
       "0  ewing_chromHMM15_Quies,neuroblastoma_chromHMM1...   \n",
       "1  ewing_chromHMM15_Quies,neuroblastoma_chromHMM1...   \n",
       "\n",
       "                            PREDICTED_NONCODING_SPAN  \\\n",
       "0  neuroblastoma_chromHMM15_EnhG_conserved,neurob...   \n",
       "1  neuroblastoma_chromHMM15_EnhG_conserved,neurob...   \n",
       "\n",
       "  PREDICTED_PARTIAL_EXON_DUP  PREDICTED_PROMOTER  PREDICTED_TSS_DUP  \\\n",
       "0                        NaN                 NaN                NaN   \n",
       "1                        NaN                 NaN                NaN   \n",
       "\n",
       "   PREDICTED_UTR  \n",
       "0            NaN  \n",
       "1            NaN  \n",
       "\n",
       "[2 rows x 49 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "svs.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7c87ebcd-96a7-4188-8779-d1c9fc51b5b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "dosages = dosages.set_index(\"ID\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ddae480-1a7e-4b4c-bdee-791c0d2b92f7",
   "metadata": {},
   "source": [
    "Add the intergenic distances, if relevant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ef2e8fdd-6b6c-413d-bf1c-40ae0219b67b",
   "metadata": {},
   "outputs": [],
   "source": [
    "svs = svs.merge(intergenic_sv_to_gene_distances[['name', 'distance']].rename(columns = {'distance': 'distance_to_nearest_gene'}), \n",
    "          on = 'name', how = 'left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b5a9510a-2846-4df4-9e5f-baf5fa11383a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(228887, 50)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "svs.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bc80f52-735d-46f2-bb38-2032ea169030",
   "metadata": {},
   "source": [
    "## Read in the category SVs and category results\n",
    "\n",
    "Here, we read in the category SVs (which SVs are in each category) and the category results (case vs. control burdens)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8d5a4d3e-cde0-49ad-bbd9-f6d668a22cbd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SV</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>category</th>\n",
       "      <th>disease</th>\n",
       "      <th>sv_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PedSV.2.5.2_DUP_chr1_455</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6720524</td>\n",
       "      <td>6722070</td>\n",
       "      <td>DUP.RARE.PREDICTED_NONCODING_BREAKPOINT.ewing_...</td>\n",
       "      <td>ewing</td>\n",
       "      <td>noncoding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PedSV.2.5.2_DUP_chr1_898</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23649481</td>\n",
       "      <td>23651544</td>\n",
       "      <td>DUP.RARE.PREDICTED_NONCODING_BREAKPOINT.ewing_...</td>\n",
       "      <td>ewing</td>\n",
       "      <td>noncoding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         SV chrom     start       end  \\\n",
       "0  PedSV.2.5.2_DUP_chr1_455  chr1   6720524   6722070   \n",
       "1  PedSV.2.5.2_DUP_chr1_898  chr1  23649481  23651544   \n",
       "\n",
       "                                            category disease sv_category  \n",
       "0  DUP.RARE.PREDICTED_NONCODING_BREAKPOINT.ewing_...   ewing   noncoding  \n",
       "1  DUP.RARE.PREDICTED_NONCODING_BREAKPOINT.ewing_...   ewing   noncoding  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###########\n",
    "### SVs ###\n",
    "###########\n",
    "category_svs = []\n",
    "for file in glob.glob(\"data/updated-cwas/svs-in-categories/*.txt\"):\n",
    "    disease = file.split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    sv_category = \"noncoding\" if \"noncoding\" in file else \"coding\"\n",
    "\n",
    "    cat_svs = pd.read_csv(file, sep=\"\\t\")\n",
    "    cat_svs[[\"disease\", \"sv_category\"]] = [disease, sv_category]\n",
    "    category_svs.append(cat_svs)\n",
    "\n",
    "category_svs = pd.concat(category_svs)\n",
    "category_svs.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88851a48-2725-42b4-bf0c-66060e79a088",
   "metadata": {},
   "source": [
    "Read in category burden testing:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "401f854f-6749-4333-9acb-36da395e9d93",
   "metadata": {},
   "outputs": [],
   "source": [
    "#########################\n",
    "### FRAMEWORK RESULTS ###\n",
    "#########################\n",
    "columns = ['category_name', 'point_estimate', 'std_error', 'z_score', 'p_value']\n",
    "framework_results = []\n",
    "for file in glob.glob(\"data/updated-cwas/summary-stats/*.txt\"):\n",
    "\n",
    "    disease = file.split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    sv_category = \"noncoding\" if \"noncoding\" in file else \"coding\"\n",
    "\n",
    "    data = pd.read_csv(file, sep=\"\\t\", usecols = columns)\n",
    "    data[[\"disease\", \"sv_category\"]] = [\n",
    "        disease,\n",
    "        sv_category,\n",
    "    ]\n",
    "    framework_results.append(data)\n",
    "\n",
    "framework_results = pd.concat(framework_results)\n",
    "framework_results['negative_log10_p_value'] = -np.log10(framework_results['p_value'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "38dd1721-aadb-4b31-828e-0c2f088317ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>point_estimate</th>\n",
       "      <th>std_error</th>\n",
       "      <th>z_score</th>\n",
       "      <th>p_value</th>\n",
       "      <th>category_name</th>\n",
       "      <th>disease</th>\n",
       "      <th>sv_category</th>\n",
       "      <th>negative_log10_p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.219704</td>\n",
       "      <td>0.587820</td>\n",
       "      <td>0.373761</td>\n",
       "      <td>0.708582</td>\n",
       "      <td>DUP.RARE.PREDICTED_NONCODING_BREAKPOINT.ewing_...</td>\n",
       "      <td>ewing</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>0.149610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.140260</td>\n",
       "      <td>0.584964</td>\n",
       "      <td>-0.239775</td>\n",
       "      <td>0.810504</td>\n",
       "      <td>DUP.RARE.PREDICTED_NONCODING_BREAKPOINT.ewing_...</td>\n",
       "      <td>ewing</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>0.091245</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   point_estimate  std_error   z_score   p_value  \\\n",
       "0        0.219704   0.587820  0.373761  0.708582   \n",
       "1       -0.140260   0.584964 -0.239775  0.810504   \n",
       "\n",
       "                                       category_name disease sv_category  \\\n",
       "0  DUP.RARE.PREDICTED_NONCODING_BREAKPOINT.ewing_...   ewing   noncoding   \n",
       "1  DUP.RARE.PREDICTED_NONCODING_BREAKPOINT.ewing_...   ewing   noncoding   \n",
       "\n",
       "   negative_log10_p_value  \n",
       "0                0.149610  \n",
       "1                0.091245  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "framework_results.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4acb3782-3d5d-4d05-ae97-175b96ae9b5e",
   "metadata": {},
   "source": [
    "Here, we import a file that depicts all the possible filters that can be applied to categories:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ab6d937b-f522-419b-b24d-20647f00ba8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "framework_schema = defaultdict(dict)\n",
    "for file in glob.glob(\"data/updated-cwas/schema/*.txt\"):\n",
    "\n",
    "    sv_category = \"noncoding\" if \"noncoding\" in file else \"coding\"\n",
    "    suffix = file.split(\"/\")[-1]\n",
    "    if sv_category == \"coding\":\n",
    "        disease = suffix.split('_')[2]\n",
    "    else:\n",
    "        disease = suffix.split('_')[3]\n",
    "\n",
    "    data = pd.read_csv(file, sep=\"\\t\")\n",
    "    framework_schema[disease][sv_category] = data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ce700ab9-de05-45c3-bb97-4e49143a3437",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sv_type</th>\n",
       "      <th>frequency</th>\n",
       "      <th>genic_relationship</th>\n",
       "      <th>constraint</th>\n",
       "      <th>expression</th>\n",
       "      <th>gene_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DUP</td>\n",
       "      <td>RARE</td>\n",
       "      <td>PREDICTED_COPY_GAIN</td>\n",
       "      <td>lof_constrained</td>\n",
       "      <td>expressed_in_adrenal_gland</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DEL</td>\n",
       "      <td>SINGLETON</td>\n",
       "      <td>PREDICTED_INTRAGENIC_EXON_DUP</td>\n",
       "      <td>missense_constrained</td>\n",
       "      <td>ANY</td>\n",
       "      <td>cosmic_cancer_genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CPX_or_INV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PREDICTED_LOF_or_PREDICTED_PARTIAL_EXON_DUP</td>\n",
       "      <td>unconstrained</td>\n",
       "      <td>NaN</td>\n",
       "      <td>germline_CPGs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>INS_ALL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ANY</td>\n",
       "      <td>ANY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>base_excision_repair_genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ANY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chromatin_organization_genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dna_damage_bypass_genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dna_damage_reversal_genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dna_DSB_repair_genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dna_DSB_response_genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dna_repair_genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>fanconi_genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mismatch_repair_genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nucleotide_excision_repair_genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>oncogenic_MAPK_signaling_genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>signaling_transduction_by_receptor_tyrosine_ki...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       sv_type  frequency                           genic_relationship  \\\n",
       "0          DUP       RARE                          PREDICTED_COPY_GAIN   \n",
       "1          DEL  SINGLETON                PREDICTED_INTRAGENIC_EXON_DUP   \n",
       "2   CPX_or_INV        NaN  PREDICTED_LOF_or_PREDICTED_PARTIAL_EXON_DUP   \n",
       "3      INS_ALL        NaN                                          ANY   \n",
       "4          ANY        NaN                                          NaN   \n",
       "5          NaN        NaN                                          NaN   \n",
       "6          NaN        NaN                                          NaN   \n",
       "7          NaN        NaN                                          NaN   \n",
       "8          NaN        NaN                                          NaN   \n",
       "9          NaN        NaN                                          NaN   \n",
       "10         NaN        NaN                                          NaN   \n",
       "11         NaN        NaN                                          NaN   \n",
       "12         NaN        NaN                                          NaN   \n",
       "13         NaN        NaN                                          NaN   \n",
       "14         NaN        NaN                                          NaN   \n",
       "\n",
       "              constraint                  expression  \\\n",
       "0        lof_constrained  expressed_in_adrenal_gland   \n",
       "1   missense_constrained                         ANY   \n",
       "2          unconstrained                         NaN   \n",
       "3                    ANY                         NaN   \n",
       "4                    NaN                         NaN   \n",
       "5                    NaN                         NaN   \n",
       "6                    NaN                         NaN   \n",
       "7                    NaN                         NaN   \n",
       "8                    NaN                         NaN   \n",
       "9                    NaN                         NaN   \n",
       "10                   NaN                         NaN   \n",
       "11                   NaN                         NaN   \n",
       "12                   NaN                         NaN   \n",
       "13                   NaN                         NaN   \n",
       "14                   NaN                         NaN   \n",
       "\n",
       "                                           gene_group  \n",
       "0                                      protein_coding  \n",
       "1                                 cosmic_cancer_genes  \n",
       "2                                       germline_CPGs  \n",
       "3                          base_excision_repair_genes  \n",
       "4                        chromatin_organization_genes  \n",
       "5                             dna_damage_bypass_genes  \n",
       "6                           dna_damage_reversal_genes  \n",
       "7                                dna_DSB_repair_genes  \n",
       "8                              dna_DSB_response_genes  \n",
       "9                                    dna_repair_genes  \n",
       "10                                      fanconi_genes  \n",
       "11                              mismatch_repair_genes  \n",
       "12                   nucleotide_excision_repair_genes  \n",
       "13                     oncogenic_MAPK_signaling_genes  \n",
       "14  signaling_transduction_by_receptor_tyrosine_ki...  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "framework_schema[\"neuroblastoma\"][\"coding\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c227743a-e276-4132-848f-0be688d3ad3f",
   "metadata": {},
   "source": [
    "# Defining the \"background reference\" of gene counts\n",
    "\n",
    "Here, we define all the genes affected by SVs in the dataset that go into the \"denominator\" of the Fisher exact test. Later, we simply extract the genes affected by SVs in categories to form the \"numerator\" from this reference.\n",
    "\n",
    "As mentioned earlier, the following filters are also applied to this background:\n",
    "\n",
    "1) the background SV class matches the category of interest (coding vs. noncoding)\n",
    "2) SVs affecting more than 10 genes are removed (across coding and noncoding effects)\n",
    "3) SVs that do not have a coding/noncoding consequence detailed in the CWAS are removed\n",
    "4) noncoding SVs farther than 500 kb from their nearest gene for the noncoding consequence “INTERGENIC” are removed."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c00e3a58-e9a2-430d-9a3c-35b7a6aee025",
   "metadata": {},
   "source": [
    "## How an SV affecting a gene appears in the VCF\n",
    "\n",
    "An example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d0f01fc6-81c9-4796-b8d2-ffdfb5ad5cd8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>svtype</th>\n",
       "      <th>AC</th>\n",
       "      <th>AF</th>\n",
       "      <th>ALGORITHMS</th>\n",
       "      <th>AN</th>\n",
       "      <th>BOTHSIDES_SUPPORT</th>\n",
       "      <th>...</th>\n",
       "      <th>PREDICTED_LOF</th>\n",
       "      <th>PREDICTED_MSV_EXON_OVERLAP</th>\n",
       "      <th>PREDICTED_NEAREST_TSS</th>\n",
       "      <th>PREDICTED_NONCODING_BREAKPOINT</th>\n",
       "      <th>PREDICTED_NONCODING_SPAN</th>\n",
       "      <th>PREDICTED_PARTIAL_EXON_DUP</th>\n",
       "      <th>PREDICTED_PROMOTER</th>\n",
       "      <th>PREDICTED_TSS_DUP</th>\n",
       "      <th>PREDICTED_UTR</th>\n",
       "      <th>distance_to_nearest_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>chr1</td>\n",
       "      <td>958900</td>\n",
       "      <td>983001</td>\n",
       "      <td>PedSV.2.5.2_DEL_chr1_100</td>\n",
       "      <td>DEL</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000307</td>\n",
       "      <td>depth</td>\n",
       "      <td>13032</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>KLHL17,NOC2L,PERM1,PLEKHN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ewing_H3K27Ac_peak,ewing_H3K27Ac_peak_conserve...</td>\n",
       "      <td>ewing_ABC_MAX_enhancer,ewing_H3K27Ac_peak_cons...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   #chrom   start     end                      name svtype  AC        AF  \\\n",
       "78   chr1  958900  983001  PedSV.2.5.2_DEL_chr1_100    DEL   4  0.000307   \n",
       "\n",
       "   ALGORITHMS     AN  BOTHSIDES_SUPPORT  ...               PREDICTED_LOF  \\\n",
       "78      depth  13032              False  ...  KLHL17,NOC2L,PERM1,PLEKHN1   \n",
       "\n",
       "    PREDICTED_MSV_EXON_OVERLAP PREDICTED_NEAREST_TSS  \\\n",
       "78                         NaN                   NaN   \n",
       "\n",
       "                       PREDICTED_NONCODING_BREAKPOINT  \\\n",
       "78  ewing_H3K27Ac_peak,ewing_H3K27Ac_peak_conserve...   \n",
       "\n",
       "                             PREDICTED_NONCODING_SPAN  \\\n",
       "78  ewing_ABC_MAX_enhancer,ewing_H3K27Ac_peak_cons...   \n",
       "\n",
       "    PREDICTED_PARTIAL_EXON_DUP PREDICTED_PROMOTER  PREDICTED_TSS_DUP  \\\n",
       "78                         NaN                NaN                NaN   \n",
       "\n",
       "    PREDICTED_UTR  distance_to_nearest_gene  \n",
       "78            NaN                       NaN  \n",
       "\n",
       "[1 rows x 50 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "svs[svs['name'] == 'PedSV.2.5.2_DEL_chr1_100']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ef593ca-6d14-40de-af11-1de82ab64ff5",
   "metadata": {},
   "source": [
    "This SV is a coding SV that affects 4 genes with a predicted loss of function consequence: `KLHL17,NOC2L,PERM1,PLEKHN1`. Thus it would contribute a count to each of these genes in our reference dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2cfda92-bc5f-4c46-9709-5ee40903910c",
   "metadata": {},
   "source": [
    "## Adding an annotation for coding/non-coding/multi effects\n",
    "\n",
    "We add this annotation to filter out SVs that affect many genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2bcc660a-c71d-4d49-a1a3-936bf90a2ce1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# an svs is True for \"multiple-genes-coding\" if it affects more than one gene across all coding genic relationships.\n",
    "# The same is true for multiple-genes-noncoding.\n",
    "coding_cols = [\n",
    "    \"PREDICTED_LOF\",\n",
    "    \"PREDICTED_PARTIAL_EXON_DUP\",\n",
    "    \"PREDICTED_INTRAGENIC_EXON_DUP\",\n",
    "    \"PREDICTED_COPY_GAIN\",\n",
    "]\n",
    "noncoding_cols = [\n",
    "    \"PREDICTED_NEAREST_TSS\",\n",
    "    \"PREDICTED_INTRONIC\",\n",
    "    \"PREDICTED_PROMOTER\",\n",
    "    \"PREDICTED_UTR\",\n",
    "]\n",
    "\n",
    "# for each class of consequences (noncoding, coding, and \"both\", we asssess the genes that are affected by the SV)\n",
    "for label, cols in zip(\n",
    "    [\"coding\", \"noncoding\", \"both\"],\n",
    "    [coding_cols, noncoding_cols, coding_cols + noncoding_cols],\n",
    "):\n",
    "\n",
    "    # we define all the genes that are \"affected\" by an SV. We\n",
    "    # do this by looking at successive genic relationships\n",
    "    affected_genes = np.array(svs[cols[0]].fillna(\"\").str.split(\",\"))\n",
    "    for c in cols[1:]:\n",
    "        affected_genes += np.array(svs[c].fillna(\"\").str.split(\",\"))\n",
    "\n",
    "    # remove blanks and duplicate genes - if an SV affects the same gene in more than one way, that's fine\n",
    "    affected_genes = [[g for g in set(g_list) if g != \"\"] for g_list in affected_genes]\n",
    "    affected_genes_col = [\",\".join(g_list) for g_list in affected_genes]\n",
    "\n",
    "    # extract out how many unique genes are affected by each SV\n",
    "    num_unique_affected_genes = [len(set(g_list)) for g_list in affected_genes]\n",
    "    multiple_genes = np.array(num_unique_affected_genes) > 1\n",
    "\n",
    "    # add a column to our SVs\n",
    "    svs[f\"num_genes_{label}\"] = num_unique_affected_genes\n",
    "    svs[f\"genes_{label}\"] = affected_genes_col"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "49f15160-4a04-4b33-b3c9-fd0d248a65f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>genes_coding</th>\n",
       "      <th>PREDICTED_LOF</th>\n",
       "      <th>PREDICTED_PARTIAL_EXON_DUP</th>\n",
       "      <th>PREDICTED_INTRAGENIC_EXON_DUP</th>\n",
       "      <th>PREDICTED_COPY_GAIN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PedSV.2.5.2_CNV_chr1_1</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PedSV.2.5.2_DUP_chr1_1</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PedSV.2.5.2_DUP_chr1_4</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PedSV.2.5.2_CNV_chr1_2</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     name genes_coding PREDICTED_LOF  \\\n",
       "0  PedSV.2.5.2_CNV_chr1_1                        NaN   \n",
       "1  PedSV.2.5.2_DUP_chr1_1                        NaN   \n",
       "2  PedSV.2.5.2_DUP_chr1_4        OR4F5           NaN   \n",
       "3  PedSV.2.5.2_CNV_chr1_2                        NaN   \n",
       "\n",
       "  PREDICTED_PARTIAL_EXON_DUP PREDICTED_INTRAGENIC_EXON_DUP PREDICTED_COPY_GAIN  \n",
       "0                        NaN                           NaN                 NaN  \n",
       "1                        NaN                           NaN                 NaN  \n",
       "2                        NaN                           NaN               OR4F5  \n",
       "3                        NaN                           NaN                 NaN  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# as an example, the 3rd SV in this dataset affects ORF45 - a copy gain\n",
    "svs[['name', 'genes_coding'] + coding_cols].head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b170cc06-20c6-4d87-886d-f976e686e071",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((228887, 56), 64)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multigene_svs = svs.query(\"num_genes_both > 10\")[\"name\"].tolist()\n",
    "\n",
    "# only 64 SVs affect more than 10 genes\n",
    "svs.shape, len(multigene_svs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80254619-168f-4662-a21d-bc22a2037b1a",
   "metadata": {},
   "source": [
    "We drop these SVs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "51e2ba7a-5b6b-45cb-8ff9-a1d2fa36c629",
   "metadata": {},
   "outputs": [],
   "source": [
    "svs = svs.query(\"num_genes_both <= 10\").copy()\n",
    "category_svs = category_svs[~(category_svs[\"SV\"].isin(multigene_svs))].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58cb5a16-2341-4115-8687-6be2128a3beb",
   "metadata": {},
   "source": [
    "## Drop intergenic SVs that are far from their gene\n",
    "\n",
    "We only drop an SV in this manner if it's the only way it affects a gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "6c4c80cc-1713-4921-88dd-f6557b56c383",
   "metadata": {},
   "outputs": [],
   "source": [
    "only_intergenic_svs = svs[(svs['PREDICTED_NEAREST_TSS'] == svs['genes_both']) & (svs['num_genes_coding'] == 0)].copy()\n",
    "dropped_intergenic_svs = only_intergenic_svs.query('distance_to_nearest_gene > 5e5')['name'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a6bff05c-af9e-446a-9ad6-2cee7755f562",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(228823, 56)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "svs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f9e2ef94-9bc7-4e8d-98ff-33aba6aa6219",
   "metadata": {},
   "outputs": [],
   "source": [
    "svs = svs[~(svs['name'].isin(dropped_intergenic_svs))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4e7dae75-3d95-4f0e-a050-9f1a436a73ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(206533, 56)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# we are left with 206533 SVs\n",
    "svs.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "905d39cc-e259-4e87-b1ba-71069f5f664f",
   "metadata": {},
   "source": [
    "## Create the reference for gene counts\n",
    "\n",
    "These references will also serve as the lookup point for our actual categories. That way, we don't have to recalculate counts, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "269da104-e3a9-461e-aad9-3420ade8c20a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "neuroblastoma coding\n",
      "neuroblastoma noncoding\n",
      "ewing coding\n",
      "ewing noncoding\n"
     ]
    }
   ],
   "source": [
    "# identify rare SVs first - this is the \"background\" (multigene SVs and noncoding SVs far from their gene have already been removed)\n",
    "rare_svs = svs[svs[\"AF\"] < 0.01].copy().set_index(\"name\")\n",
    "\n",
    "ref_counts = []\n",
    "for disease in [\"neuroblastoma\", \"ewing\"]:\n",
    "\n",
    "    sub_reference = rare_svs.copy()\n",
    "    \n",
    "    # fetch the samples we have the for this disease\n",
    "    disease_samples = samples[disease]\n",
    "\n",
    "    # identify SVs that affect genes in coding or non-coding fashion\n",
    "    for sv_category in [\"coding\", \"noncoding\"]:\n",
    "        sub_reference = rare_svs[rare_svs[f\"num_genes_{sv_category}\"] > 0]\n",
    "\n",
    "        # now build the reference gene counts - for cases and controls\n",
    "        gene_counts = {}\n",
    "        for cohort in [\"case\", \"control\"]:\n",
    "\n",
    "            # get the case or control samples, and then subset the dosage matrix to these samples\n",
    "            cohort_samples = disease_samples[cohort]\n",
    "            sub_dosages = dosages.loc[sub_reference.index, cohort_samples]\n",
    "\n",
    "            # drop samples with bad genotyping and fill in na's as 0 (effectively ignored)\n",
    "            kept_samples = np.where(\n",
    "                pd.isnull(sub_dosages).sum(axis=0) / len(sub_dosages) < 0.05\n",
    "            )[0]\n",
    "            sub_dosages = sub_dosages.iloc[:, kept_samples].fillna(0).astype(int)\n",
    "\n",
    "            # finally, we create the gene list with counts. First get the genes affected by each SV\n",
    "            sv_genes = sub_reference[f\"genes_{sv_category}\"]\n",
    "            \n",
    "            # and expand them into their own rows (in case an SV affects more than one gene)\n",
    "            sv_genes = sv_genes.str.split(',').explode()\n",
    "            \n",
    "            # next, determine the number of times the SV is present in our samples (case or control)\n",
    "            sv_names = sv_genes.index\n",
    "            sv_counts = sub_dosages.loc[sv_names].sum(axis=1)\n",
    "\n",
    "            # construct the dataframe containing the SV names, the genes they affect (one each row),\n",
    "            # and their dosage in the dataset (e.g. number of times this gene is affected)\n",
    "            subreference_counts = pd.DataFrame(\n",
    "                [sv_names, sv_counts, sv_genes], index=[\"name\", \"dose\", \"gene\"]\n",
    "            ).T\n",
    "\n",
    "            # remove genes not in our reference (shouldn't happen, but just in case)\n",
    "            subreference_counts = subreference_counts[\n",
    "                subreference_counts[\"gene\"].isin(gene_ref)\n",
    "            ]\n",
    "\n",
    "            # add relevant metadata\n",
    "            subreference_counts[\"disease\"] = disease\n",
    "            subreference_counts[\"sv_category\"] = sv_category\n",
    "            subreference_counts[\"cohort\"] = cohort\n",
    "            \n",
    "            # store\n",
    "            ref_counts.append(subreference_counts)\n",
    "            \n",
    "            # if neuroblastoma cases and coding SVs, do a small subanalysis\n",
    "            if disease == \"neuroblastoma\" and cohort == \"case\" and sv_category == \"coding\":\n",
    "                \n",
    "                # get the samples affected by the SVs\n",
    "                sv_to_samples = {sv: ','.join(sub_dosages.columns[sub_dosages.loc[sv] > 0].tolist()) for sv in sub_dosages.index}\n",
    "                \n",
    "                # add this info\n",
    "                nbl_coding_singleton_export = subreference_counts.copy()\n",
    "                nbl_coding_singleton_export['samples'] = nbl_coding_singleton_export['name'].apply(lambda sv: sv_to_samples[sv])\n",
    "\n",
    "        print(disease, sv_category)\n",
    "\n",
    "ref_counts = pd.concat(ref_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "c1c2d153-5d94-4439-b4c2-a288f7e028b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_dict = {sv: sub_dosages.columns[sub_dosages.loc[sv] > 0].tolist() for sv in sub_dosages.index}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "c68f05af-1409-4811-b680-aad8b6593a64",
   "metadata": {},
   "outputs": [],
   "source": [
    "nbl_coding_singleton_export.query('dose == 1').to_csv('nbl-singletons.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "7b989566-582a-4eb9-b983-90004d103cb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "!open ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "1614c9ea-d4a5-48d2-a3c1-7112d4bfb11e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "name\n",
       "PedSV.2.5.2_DEL_chr1_2        OR4F5\n",
       "PedSV.2.5.2_INS_chr1_2        OR4F5\n",
       "PedSV.2.5.2_DEL_chr1_3        OR4F5\n",
       "PedSV.2.5.2_INS_chr1_3        OR4F5\n",
       "PedSV.2.5.2_DEL_chr1_9        OR4F5\n",
       "                              ...  \n",
       "PedSV.2.5.2_DEL_chrX_9256      IL9R\n",
       "PedSV.2.5.2_DEL_chrX_9259      IL9R\n",
       "PedSV.2.5.2_DEL_chrX_9264      IL9R\n",
       "PedSV.2.5.2_DEL_chrY_59       TSPY2\n",
       "PedSV.2.5.2_DEL_chrY_440     EIF1AY\n",
       "Name: genes_noncoding, Length: 179975, dtype: object"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sv_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "859a768f-cee7-4de7-8f78-172a1bd51515",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>dose</th>\n",
       "      <th>gene</th>\n",
       "      <th>disease</th>\n",
       "      <th>sv_category</th>\n",
       "      <th>cohort</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PedSV.2.5.2_DEL_chr1_2</td>\n",
       "      <td>2</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ewing</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PedSV.2.5.2_INS_chr1_2</td>\n",
       "      <td>1</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ewing</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PedSV.2.5.2_DEL_chr1_3</td>\n",
       "      <td>2</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ewing</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PedSV.2.5.2_INS_chr1_3</td>\n",
       "      <td>1</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ewing</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PedSV.2.5.2_DEL_chr1_9</td>\n",
       "      <td>1</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ewing</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179970</th>\n",
       "      <td>PedSV.2.5.2_DEL_chrX_9256</td>\n",
       "      <td>0</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>ewing</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179971</th>\n",
       "      <td>PedSV.2.5.2_DEL_chrX_9259</td>\n",
       "      <td>0</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>ewing</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179972</th>\n",
       "      <td>PedSV.2.5.2_DEL_chrX_9264</td>\n",
       "      <td>1</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>ewing</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179973</th>\n",
       "      <td>PedSV.2.5.2_DEL_chrY_59</td>\n",
       "      <td>25</td>\n",
       "      <td>TSPY2</td>\n",
       "      <td>ewing</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179974</th>\n",
       "      <td>PedSV.2.5.2_DEL_chrY_440</td>\n",
       "      <td>17</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>ewing</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>179664 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             name dose    gene disease sv_category   cohort\n",
       "0          PedSV.2.5.2_DEL_chr1_2    2   OR4F5   ewing   noncoding  control\n",
       "1          PedSV.2.5.2_INS_chr1_2    1   OR4F5   ewing   noncoding  control\n",
       "2          PedSV.2.5.2_DEL_chr1_3    2   OR4F5   ewing   noncoding  control\n",
       "3          PedSV.2.5.2_INS_chr1_3    1   OR4F5   ewing   noncoding  control\n",
       "4          PedSV.2.5.2_DEL_chr1_9    1   OR4F5   ewing   noncoding  control\n",
       "...                           ...  ...     ...     ...         ...      ...\n",
       "179970  PedSV.2.5.2_DEL_chrX_9256    0    IL9R   ewing   noncoding  control\n",
       "179971  PedSV.2.5.2_DEL_chrX_9259    0    IL9R   ewing   noncoding  control\n",
       "179972  PedSV.2.5.2_DEL_chrX_9264    1    IL9R   ewing   noncoding  control\n",
       "179973    PedSV.2.5.2_DEL_chrY_59   25   TSPY2   ewing   noncoding  control\n",
       "179974   PedSV.2.5.2_DEL_chrY_440   17  EIF1AY   ewing   noncoding  control\n",
       "\n",
       "[179664 rows x 6 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subreference_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "6290adbf-16e5-4764-8c5a-ab45f1bdddd8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TPMCCDG10002</th>\n",
       "      <th>TPMCCDG10012</th>\n",
       "      <th>TPMCCDG10017</th>\n",
       "      <th>TPMCCDG10018</th>\n",
       "      <th>TPMCCDG10028</th>\n",
       "      <th>TPMCCDG10029</th>\n",
       "      <th>TPMCCDG10041</th>\n",
       "      <th>TPMCCDG10042</th>\n",
       "      <th>TPMCCDG10043</th>\n",
       "      <th>TPMCCDG10044</th>\n",
       "      <th>...</th>\n",
       "      <th>ssi_25815</th>\n",
       "      <th>ssi_25837</th>\n",
       "      <th>ssi_25849</th>\n",
       "      <th>ssi_25985</th>\n",
       "      <th>ssi_26055</th>\n",
       "      <th>ssi_26060</th>\n",
       "      <th>ssi_26301</th>\n",
       "      <th>ssi_26305</th>\n",
       "      <th>ssi_26369</th>\n",
       "      <th>ssi_26393</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PedSV.2.5.2_DEL_chr1_2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PedSV.2.5.2_INS_chr1_2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PedSV.2.5.2_DEL_chr1_3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PedSV.2.5.2_INS_chr1_3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PedSV.2.5.2_DEL_chr1_9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PedSV.2.5.2_DEL_chrX_9256</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PedSV.2.5.2_DEL_chrX_9259</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PedSV.2.5.2_DEL_chrX_9264</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PedSV.2.5.2_DEL_chrY_59</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PedSV.2.5.2_DEL_chrY_440</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>179278 rows × 4573 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           TPMCCDG10002  TPMCCDG10012  TPMCCDG10017  \\\n",
       "name                                                                  \n",
       "PedSV.2.5.2_DEL_chr1_2                0             0             0   \n",
       "PedSV.2.5.2_INS_chr1_2                0             0             0   \n",
       "PedSV.2.5.2_DEL_chr1_3                0             0             0   \n",
       "PedSV.2.5.2_INS_chr1_3                0             0             0   \n",
       "PedSV.2.5.2_DEL_chr1_9                0             0             0   \n",
       "...                                 ...           ...           ...   \n",
       "PedSV.2.5.2_DEL_chrX_9256             0             0             0   \n",
       "PedSV.2.5.2_DEL_chrX_9259             0             0             0   \n",
       "PedSV.2.5.2_DEL_chrX_9264             0             0             0   \n",
       "PedSV.2.5.2_DEL_chrY_59               0             0             0   \n",
       "PedSV.2.5.2_DEL_chrY_440              0             0             0   \n",
       "\n",
       "                           TPMCCDG10018  TPMCCDG10028  TPMCCDG10029  \\\n",
       "name                                                                  \n",
       "PedSV.2.5.2_DEL_chr1_2                0             0             0   \n",
       "PedSV.2.5.2_INS_chr1_2                0             0             0   \n",
       "PedSV.2.5.2_DEL_chr1_3                0             0             0   \n",
       "PedSV.2.5.2_INS_chr1_3                0             0             0   \n",
       "PedSV.2.5.2_DEL_chr1_9                0             0             0   \n",
       "...                                 ...           ...           ...   \n",
       "PedSV.2.5.2_DEL_chrX_9256             0             0             0   \n",
       "PedSV.2.5.2_DEL_chrX_9259             0             0             0   \n",
       "PedSV.2.5.2_DEL_chrX_9264             0             0             0   \n",
       "PedSV.2.5.2_DEL_chrY_59               0             0             0   \n",
       "PedSV.2.5.2_DEL_chrY_440              0             0             0   \n",
       "\n",
       "                           TPMCCDG10041  TPMCCDG10042  TPMCCDG10043  \\\n",
       "name                                                                  \n",
       "PedSV.2.5.2_DEL_chr1_2                0             0             0   \n",
       "PedSV.2.5.2_INS_chr1_2                0             0             0   \n",
       "PedSV.2.5.2_DEL_chr1_3                0             0             0   \n",
       "PedSV.2.5.2_INS_chr1_3                0             0             0   \n",
       "PedSV.2.5.2_DEL_chr1_9                0             0             0   \n",
       "...                                 ...           ...           ...   \n",
       "PedSV.2.5.2_DEL_chrX_9256             0             0             0   \n",
       "PedSV.2.5.2_DEL_chrX_9259             0             0             0   \n",
       "PedSV.2.5.2_DEL_chrX_9264             0             0             0   \n",
       "PedSV.2.5.2_DEL_chrY_59               0             0             0   \n",
       "PedSV.2.5.2_DEL_chrY_440              0             0             0   \n",
       "\n",
       "                           TPMCCDG10044  ...  ssi_25815  ssi_25837  ssi_25849  \\\n",
       "name                                     ...                                    \n",
       "PedSV.2.5.2_DEL_chr1_2                0  ...          0          0          0   \n",
       "PedSV.2.5.2_INS_chr1_2                0  ...          0          0          0   \n",
       "PedSV.2.5.2_DEL_chr1_3                0  ...          0          0          0   \n",
       "PedSV.2.5.2_INS_chr1_3                0  ...          0          0          0   \n",
       "PedSV.2.5.2_DEL_chr1_9                0  ...          0          0          0   \n",
       "...                                 ...  ...        ...        ...        ...   \n",
       "PedSV.2.5.2_DEL_chrX_9256             0  ...          0          0          0   \n",
       "PedSV.2.5.2_DEL_chrX_9259             0  ...          0          0          0   \n",
       "PedSV.2.5.2_DEL_chrX_9264             0  ...          0          0          0   \n",
       "PedSV.2.5.2_DEL_chrY_59               0  ...          0          0          0   \n",
       "PedSV.2.5.2_DEL_chrY_440              0  ...          0          0          0   \n",
       "\n",
       "                           ssi_25985  ssi_26055  ssi_26060  ssi_26301  \\\n",
       "name                                                                    \n",
       "PedSV.2.5.2_DEL_chr1_2             0          0          0          0   \n",
       "PedSV.2.5.2_INS_chr1_2             0          0          0          0   \n",
       "PedSV.2.5.2_DEL_chr1_3             0          0          0          0   \n",
       "PedSV.2.5.2_INS_chr1_3             0          0          0          0   \n",
       "PedSV.2.5.2_DEL_chr1_9             0          0          0          0   \n",
       "...                              ...        ...        ...        ...   \n",
       "PedSV.2.5.2_DEL_chrX_9256          0          0          0          0   \n",
       "PedSV.2.5.2_DEL_chrX_9259          0          0          0          0   \n",
       "PedSV.2.5.2_DEL_chrX_9264          0          0          0          0   \n",
       "PedSV.2.5.2_DEL_chrY_59            0          0          0          0   \n",
       "PedSV.2.5.2_DEL_chrY_440           0          0          0          0   \n",
       "\n",
       "                           ssi_26305  ssi_26369  ssi_26393  \n",
       "name                                                        \n",
       "PedSV.2.5.2_DEL_chr1_2             0          0          0  \n",
       "PedSV.2.5.2_INS_chr1_2             0          0          0  \n",
       "PedSV.2.5.2_DEL_chr1_3             0          0          0  \n",
       "PedSV.2.5.2_INS_chr1_3             0          0          0  \n",
       "PedSV.2.5.2_DEL_chr1_9             0          0          0  \n",
       "...                              ...        ...        ...  \n",
       "PedSV.2.5.2_DEL_chrX_9256          0          0          0  \n",
       "PedSV.2.5.2_DEL_chrX_9259          0          0          0  \n",
       "PedSV.2.5.2_DEL_chrX_9264          0          0          0  \n",
       "PedSV.2.5.2_DEL_chrY_59            0          0          0  \n",
       "PedSV.2.5.2_DEL_chrY_440           0          0          0  \n",
       "\n",
       "[179278 rows x 4573 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sub_dosages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "dd5e6331-f328-4b3e-95c0-f83e037704e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4574, 4573)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cohort_samples), len(kept_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "63d91706-5a50-46aa-937e-ab32bd00f51b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>dose</th>\n",
       "      <th>gene</th>\n",
       "      <th>disease</th>\n",
       "      <th>sv_category</th>\n",
       "      <th>cohort</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PedSV.2.5.2_DEL_chr1_80</td>\n",
       "      <td>0</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>coding</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PedSV.2.5.2_DEL_chr1_100</td>\n",
       "      <td>1</td>\n",
       "      <td>KLHL17</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>coding</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       name dose    gene        disease sv_category cohort\n",
       "0   PedSV.2.5.2_DEL_chr1_80    0  SAMD11  neuroblastoma      coding   case\n",
       "1  PedSV.2.5.2_DEL_chr1_100    1  KLHL17  neuroblastoma      coding   case"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_counts.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32373966-5bb1-4b46-883b-15a608459445",
   "metadata": {},
   "source": [
    "For example, if we wanted to see how often `TTN` is affected in neuroblastoma cases by coding SVs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "d3f0a4dc-d5f6-4100-8762-00b584a93975",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>dose</th>\n",
       "      <th>gene</th>\n",
       "      <th>disease</th>\n",
       "      <th>sv_category</th>\n",
       "      <th>cohort</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1451</th>\n",
       "      <td>PedSV.2.5.2_DUP_chr2_3218</td>\n",
       "      <td>1</td>\n",
       "      <td>TTN</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>coding</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1454</th>\n",
       "      <td>PedSV.2.5.2_DEL_chr2_11838</td>\n",
       "      <td>1</td>\n",
       "      <td>TTN</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>coding</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24078</th>\n",
       "      <td>PedSV.2.5.2_DEL_chr2_11830</td>\n",
       "      <td>2</td>\n",
       "      <td>TTN</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24079</th>\n",
       "      <td>PedSV.2.5.2_DEL_chr2_11831</td>\n",
       "      <td>1</td>\n",
       "      <td>TTN</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24080</th>\n",
       "      <td>PedSV.2.5.2_DEL_chr2_11835</td>\n",
       "      <td>2</td>\n",
       "      <td>TTN</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24081</th>\n",
       "      <td>PedSV.2.5.2_INS_chr2_5134</td>\n",
       "      <td>2</td>\n",
       "      <td>TTN</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24083</th>\n",
       "      <td>PedSV.2.5.2_DEL_chr2_11839</td>\n",
       "      <td>1</td>\n",
       "      <td>TTN</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             name dose gene        disease sv_category cohort\n",
       "1451    PedSV.2.5.2_DUP_chr2_3218    1  TTN  neuroblastoma      coding   case\n",
       "1454   PedSV.2.5.2_DEL_chr2_11838    1  TTN  neuroblastoma      coding   case\n",
       "24078  PedSV.2.5.2_DEL_chr2_11830    2  TTN  neuroblastoma   noncoding   case\n",
       "24079  PedSV.2.5.2_DEL_chr2_11831    1  TTN  neuroblastoma   noncoding   case\n",
       "24080  PedSV.2.5.2_DEL_chr2_11835    2  TTN  neuroblastoma   noncoding   case\n",
       "24081   PedSV.2.5.2_INS_chr2_5134    2  TTN  neuroblastoma   noncoding   case\n",
       "24083  PedSV.2.5.2_DEL_chr2_11839    1  TTN  neuroblastoma   noncoding   case"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_counts.query('gene == \"TTN\" & disease == \"neuroblastoma\" & cohort == \"case\" & dose > 0')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d26f3e17-1b79-4f9d-a840-98b9587aff8e",
   "metadata": {},
   "source": [
    "So the total dosage here in cases is 1 + 1 + 2 + 1 + 2 + 2 + 1 = 10"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b7a6cf4-27ef-4964-b9be-2ea4779d6720",
   "metadata": {
    "tags": []
   },
   "source": [
    "# An example\n",
    "\n",
    "To illustrate the process we'll systematically undertake later, consider a simple example gene set:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0d04c5d3-8a70-4b40-81c6-0025125f9c78",
   "metadata": {},
   "outputs": [],
   "source": [
    "# a contrived \"gene set\" designed explicitly to be higher in cases than controls\n",
    "test_gs = ['MSL1', 'WIZ', 'ACTC1', 'NFAT5', 'STAG1', 'MRC2', 'NANOGP8', 'CLTA', 'TTN']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab627b98-f232-4fdd-b9e7-c7cc0bfe77a9",
   "metadata": {},
   "source": [
    "And a specific category: singleton lof deletions in lof_constrainted genes expressed in adrenal tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "50a93ff6-dadc-4a11-b869-5e724f2614ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_cat = \"DEL.SINGLETON.PREDICTED_LOF_or_PREDICTED_PARTIAL_EXON_DUP.lof_constrained.expressed_in_adrenal_gland.protein_coding\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3c5a2c9-74bd-4a99-92fe-92f7ca1835cd",
   "metadata": {},
   "source": [
    "We begin by identifying the SVs that are part of that category."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "6d329232-5d40-428d-89a2-56fac470e615",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "249"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_cat_svs = category_svs[\n",
    "    (category_svs[\"category\"] == test_cat)\n",
    "    & (category_svs[\"disease\"] == \"neuroblastoma\")\n",
    "][\"SV\"].tolist()\n",
    "\n",
    "len(test_cat_svs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78d3044c-210e-4461-b784-744a7e6f3a98",
   "metadata": {},
   "source": [
    "So 249 SVs are part of this category. We next extract all the genes affected by these SVs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "11cca458-b324-45b7-a82e-3e9f4ef7d6c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>dose</th>\n",
       "      <th>gene</th>\n",
       "      <th>disease</th>\n",
       "      <th>sv_category</th>\n",
       "      <th>cohort</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>PedSV.2.5.2_DEL_chr1_401</td>\n",
       "      <td>1</td>\n",
       "      <td>GNB1</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>coding</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>PedSV.2.5.2_DEL_chr1_5766</td>\n",
       "      <td>1</td>\n",
       "      <td>NFIA</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>coding</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          name dose  gene        disease sv_category cohort\n",
       "46    PedSV.2.5.2_DEL_chr1_401    1  GNB1  neuroblastoma      coding   case\n",
       "397  PedSV.2.5.2_DEL_chr1_5766    1  NFIA  neuroblastoma      coding   case"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# category SVs\n",
    "test_cat_counts = ref_counts[\n",
    "    (ref_counts[\"name\"].isin(test_cat_svs))\n",
    "    & (ref_counts[\"disease\"] == \"neuroblastoma\")\n",
    "    & (ref_counts[\"sv_category\"] == \"coding\")\n",
    "    & (ref_counts[\"dose\"] > 0)\n",
    "]\n",
    "\n",
    "\n",
    "# non-category SVs\n",
    "test_non_cat_counts = ref_counts[\n",
    "    (~ref_counts[\"name\"].isin(test_cat_svs))\n",
    "    & (ref_counts[\"disease\"] == \"neuroblastoma\")\n",
    "    & (ref_counts[\"sv_category\"] == \"coding\")\n",
    "    & (ref_counts[\"dose\"] > 0)\n",
    "]\n",
    "\n",
    "test_cat_counts.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24b40795-ab63-43f5-a918-e0031b8aa2de",
   "metadata": {},
   "source": [
    "This includes all cohorts (case and controls). Now let's pull up a hallmark gene set and get cracking."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8cc7b17-2eb9-4b9f-b04e-c68a04add883",
   "metadata": {},
   "source": [
    "Next, for both cases and controls, we calculate the 4 elements of our contingency table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "93d94118-19a2-47e0-82b3-7122784dd5f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "case\n",
      "[[   8   45]\n",
      " [   2 3926]]\n",
      "(256.10869565217394, 1.958690806922406e-15)\n",
      "\n",
      "control\n",
      "[[    4   178]\n",
      " [   41 26443]]\n",
      "(17.58712423516893, 1.79514162402813e-05)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for cohort in [\"case\", \"control\"]:\n",
    "\n",
    "    # the counts ascribed to the category within the cohort\n",
    "    cohort_counts_in_cat = test_cat_counts[test_cat_counts[\"cohort\"] == cohort]\n",
    "    \n",
    "    # the counts ascribed to non-category SVs within the cohort\n",
    "    cohort_counts_in_non_cat = test_non_cat_counts[test_non_cat_counts['cohort'] == cohort]\n",
    "    \n",
    "    # number of genes affected by category SVs and in gene set (top left)\n",
    "    true_positives = cohort_counts_in_cat[cohort_counts_in_cat['gene'].isin(test_gs)]['dose'].sum()\n",
    "    \n",
    "    # number of genes affected by category SVs and not gene set (top right)\n",
    "    false_positives = cohort_counts_in_cat[~cohort_counts_in_cat['gene'].isin(test_gs)]['dose'].sum()\n",
    "    \n",
    "    # number of genes affected by non-category SVs and in the gene set (bottom left)\n",
    "    false_negatives = cohort_counts_in_non_cat[cohort_counts_in_non_cat['gene'].isin(test_gs)]['dose'].sum()\n",
    "    \n",
    "    # number of genes affected by category SVs and not gene set (top right)\n",
    "    true_negatives = cohort_counts_in_non_cat[~cohort_counts_in_non_cat['gene'].isin(test_gs)]['dose'].sum()\n",
    "    \n",
    "    # make the contingency table\n",
    "    cont_table = np.array(\n",
    "            [\n",
    "                [true_positives, false_positives],\n",
    "                [false_negatives, true_negatives],\n",
    "            ]\n",
    "        )\n",
    "    \n",
    "    print(cohort)\n",
    "    print(cont_table)\n",
    "\n",
    "    # pseudocount of 1\n",
    "    print(scipy.stats.fisher_exact(cont_table + 1))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9101834a-60f4-47ce-a5c6-14a0a75cac2f",
   "metadata": {},
   "source": [
    "So the odds ratio of this gene set from the Fisher's exact test is much higher in cases compared to controls (though it is significant in both)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a7ab9b1-a1fc-40b6-8b3a-17eb30ae3b36",
   "metadata": {},
   "source": [
    "# Generalized gene set enrichment\n",
    "\n",
    "Here, we apply the above process to all gene sets and a subset of categories\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33170e35-ca3c-45b8-88c9-a87a2fa09c0d",
   "metadata": {},
   "source": [
    "## Define the gene sets\n",
    "\n",
    "These GO terms were taken directly from the GO website via the GO API. We only analyze gene sets that have between 30 and 1000 genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "3b18a037-9aa1-468b-8046-29952d597e42",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_sets = {}\n",
    "with open(\"data/updated-cwas/go-gene-sets.txt\") as gs_in:\n",
    "    for line in gs_in:\n",
    "        comp = line.strip().split(\"\\t\")\n",
    "        gs = comp[0]\n",
    "        genes = comp[1:]\n",
    "\n",
    "        if len(genes) >= 30 and len(genes) <= 1000:\n",
    "            gene_sets[gs] = genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "721e7906-f4f7-4807-9776-635cef8cc0c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2725\n"
     ]
    }
   ],
   "source": [
    "# 2725 out of 12558 gene sets are analyzed\n",
    "print(len(gene_sets))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45f3a9ac-4727-40bc-9f4a-2d5b405e8b1b",
   "metadata": {},
   "source": [
    "## Define the categores\n",
    "\n",
    "We only analyze categories that are at least nominally significant. Additionally, we do not evaluate categories that are not `protein_coding`, as the others are already prefiltered to specific gene sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "9346cf4e-fda6-47f7-a537-807fd82bf5b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "810"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "categories_to_analyze = framework_results[(framework_results['p_value'] < 0.05) &\n",
    "                                          (framework_results['category_name'].str.contains('protein_coding'))]\n",
    "categories_to_analyze = categories_to_analyze.sort_values(by = ['disease', 'sv_category']).reset_index(drop=True)\n",
    "len(categories_to_analyze)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "92774f31-bbaf-4ce2-94d5-82b5f38a36f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>point_estimate</th>\n",
       "      <th>std_error</th>\n",
       "      <th>z_score</th>\n",
       "      <th>p_value</th>\n",
       "      <th>category_name</th>\n",
       "      <th>disease</th>\n",
       "      <th>sv_category</th>\n",
       "      <th>negative_log10_p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.589401</td>\n",
       "      <td>0.278940</td>\n",
       "      <td>2.113004</td>\n",
       "      <td>0.034600</td>\n",
       "      <td>ANY.SINGLETON.PREDICTED_COPY_GAIN.lof_constrai...</td>\n",
       "      <td>ewing</td>\n",
       "      <td>coding</td>\n",
       "      <td>1.460918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.541190</td>\n",
       "      <td>0.227879</td>\n",
       "      <td>2.374903</td>\n",
       "      <td>0.017554</td>\n",
       "      <td>ANY.SINGLETON.PREDICTED_INTRAGENIC_EXON_DUP.AN...</td>\n",
       "      <td>ewing</td>\n",
       "      <td>coding</td>\n",
       "      <td>1.755634</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   point_estimate  std_error   z_score   p_value  \\\n",
       "0        0.589401   0.278940  2.113004  0.034600   \n",
       "1        0.541190   0.227879  2.374903  0.017554   \n",
       "\n",
       "                                       category_name disease sv_category  \\\n",
       "0  ANY.SINGLETON.PREDICTED_COPY_GAIN.lof_constrai...   ewing      coding   \n",
       "1  ANY.SINGLETON.PREDICTED_INTRAGENIC_EXON_DUP.AN...   ewing      coding   \n",
       "\n",
       "   negative_log10_p_value  \n",
       "0                1.460918  \n",
       "1                1.755634  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "categories_to_analyze.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a120a6b-f5c7-4e1e-ba50-9853179f1c55",
   "metadata": {},
   "source": [
    "## Convert to matrix algebra\n",
    "\n",
    "The entire process of the fisher exact test can be implemented using matrix algebra (simple dot products). To achieve this, we construct a sparse matrix representing which genes are in each gene set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "9e74ee58-b53f-4461-82d9-dbfe1fce52ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define lookup dictionaries for indices\n",
    "gene_to_idx = {gene: i for i, gene in enumerate(gene_ref)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "b544442e-6f1e-47e9-ac9a-a7c7c90c2aac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert the gene sets to a sparse matrix\n",
    "gs_to_idx = {}\n",
    "\n",
    "values = []\n",
    "row_indices = []\n",
    "column_indices = []\n",
    "for i, (gs, genes) in enumerate(gene_sets.items()):\n",
    "    gs_to_idx[gs] = i\n",
    "\n",
    "    column_indices += [gene_to_idx[g] for g in genes]\n",
    "    row_indices += [i] * len(genes)\n",
    "    values += [1] * len(genes)\n",
    "\n",
    "values = np.array(values)\n",
    "row_indices = np.array(row_indices)\n",
    "column_indices = np.array(column_indices)\n",
    "\n",
    "gs_gene_sparse_mtx = scipy.sparse.csr_matrix(\n",
    "    (values, (row_indices, column_indices)), shape=(len(gs_to_idx), len(gene_to_idx))\n",
    ").T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "96523cfb-f334-48f3-badb-5497a77ec41d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<19201x2725 sparse matrix of type '<class 'numpy.int64'>'\n",
       "\twith 396034 stored elements in Compressed Sparse Column format>"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gs_gene_sparse_mtx"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd614678-31a9-4aa9-b478-03eaa2c0cdfa",
   "metadata": {},
   "source": [
    "## Run the gene set enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "1cc839c5-5aca-461d-b652-e95ca1205872",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "neuroblastoma    539\n",
       "ewing            271\n",
       "Name: disease, dtype: int64"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "categories_to_analyze['disease'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "ca8a5142-e39d-48c7-9e2c-162e971691d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200, 210, 220, 230, 240, 250, 260, 270, 280, 290, 300, 310, 320, 330, 340, 350, 360, 370, 380, 390, 400, 410, 420, 430, 440, 450, 460, 470, 480, 490, 500, 510, 520, 530, 540, 550, 560, 570, 580, 590, 600, 610, 620, 630, 640, 650, 660, 670, 680, 690, 700, 710, 720, 730, 740, 750, 760, 770, 780, 790, 800, "
     ]
    }
   ],
   "source": [
    "fisher_results = []\n",
    "\n",
    "for i, (index, row) in enumerate(categories_to_analyze.iterrows()):\n",
    "    \n",
    "    #######################################\n",
    "    ### STEP 1 - DEFINE OUR GENE COUNTS ###\n",
    "    #######################################\n",
    "    \n",
    "    # This process is the exact same for all categories - later steps will subset our gene counts\n",
    "    if i % 10 == 0:\n",
    "        print(i, end = ', ')\n",
    "    cat_name = row['category_name']\n",
    "    sv_category = row['sv_category']\n",
    "    disease = row['disease']\n",
    "    \n",
    "    # next, we pull out our SVs in this category\n",
    "    svs_in_category = category_svs[(category_svs[\"category\"] == cat_name) &\n",
    "                                   (category_svs[\"disease\"] == disease)]['SV'].tolist()\n",
    "    \n",
    "    ###############################\n",
    "    ### EXTRACT CATEGORY COUNTS ###\n",
    "    ###############################\n",
    "    \n",
    "    # get the counts of genes affected by the category SVs\n",
    "    category_counts = ref_counts[\n",
    "        (ref_counts[\"name\"].isin(svs_in_category))\n",
    "        & (ref_counts[\"disease\"] == disease)\n",
    "        & (ref_counts[\"sv_category\"] == sv_category)\n",
    "        & (ref_counts[\"dose\"] > 0)\n",
    "    ]\n",
    "    \n",
    "    ###############################\n",
    "    ### DEFINE REFERENCE COUNTS ###\n",
    "    ###############################\n",
    "    \n",
    "    # get the counts of genes affected by the NON category SVs\n",
    "    ref_category_counts = ref_counts[\n",
    "        (~ref_counts[\"name\"].isin(svs_in_category))\n",
    "        & (ref_counts[\"disease\"] == disease)\n",
    "        & (ref_counts[\"sv_category\"] == sv_category)\n",
    "        & (ref_counts[\"dose\"] > 0)\n",
    "    ]\n",
    "    \n",
    "    ##############################\n",
    "    ### RUN ANALYSES BY COHORT ###\n",
    "    ###############################\n",
    "        \n",
    "    for cohort in ['case', 'control']:\n",
    "        \n",
    "        cohort_category_counts = category_counts[category_counts['cohort'] == cohort]\n",
    "        cohort_ref_category_counts = ref_category_counts[ref_category_counts['cohort'] == cohort]\n",
    "        \n",
    "        # transform gene counts into arrays\n",
    "        cohort_gene_counts = cohort_category_counts.groupby(['gene'])['dose'].sum()\n",
    "        cohort_ref_gene_counts = cohort_ref_category_counts.groupby(['gene'])['dose'].sum()\n",
    "        \n",
    "        # fill in missing gene counts (with 0)\n",
    "        analysis_gene_counts = cohort_gene_counts.reindex(gene_ref, fill_value = 0).values.reshape(1, -1).astype(int)\n",
    "        analysis_ref_gene_counts = cohort_ref_gene_counts.reindex(gene_ref, fill_value = 0).values.reshape(1, -1).astype(int)\n",
    "        \n",
    "        ###################################\n",
    "        ### RUN THE FISHER'S EXACT TEST ###\n",
    "        ###################################\n",
    "                \n",
    "        # convert to sparse\n",
    "        analysis_gene_counts = scipy.sparse.csr_matrix(analysis_gene_counts)\n",
    "        analysis_ref_gene_counts = scipy.sparse.csr_matrix(analysis_ref_gene_counts)\n",
    "            \n",
    "        # build our contingency table using matrix math\n",
    "        cat_and_gs = analysis_gene_counts.dot(gs_gene_sparse_mtx)\n",
    "        cat_and_not_gs = analysis_gene_counts.sum(axis = 1) - cat_and_gs\n",
    "        not_cat_and_gs = analysis_ref_gene_counts.dot(gs_gene_sparse_mtx)\n",
    "        not_cat_and_not_gs = analysis_ref_gene_counts.sum(axis = 1) - not_cat_and_gs\n",
    "\n",
    "        # contingency table\n",
    "        tp = cat_and_gs.todense().A1.astype(int)\n",
    "        fp = cat_and_not_gs.A1.astype(int)\n",
    "        fn = not_cat_and_gs.todense().A1.astype(int)\n",
    "        tn = not_cat_and_not_gs.A1.astype(int)\n",
    "\n",
    "        # calculate p values\n",
    "        results = pyranges.statistics.fisher_exact(tp, fp, fn, tn, pseudocount = 1)\n",
    "\n",
    "        results['ref_freq'] = pd.Series(fn.astype(str)) + '/' + pd.Series((fn + tn).astype(str))\n",
    "        results['cat_freq'] = pd.Series((tp).astype(str)) + '/' + pd.Series((tp + fp).astype(str))\n",
    "        results['gs'] = gene_sets.keys()\n",
    "        results['category'] = cat_name\n",
    "        results['cohort'] = cohort\n",
    "        results['sv_category'] = sv_category\n",
    "        results['disease'] = disease\n",
    "            \n",
    "        # add in a feature for unique number of genes in the overlap\n",
    "        unique_gene_overlaps = analysis_gene_counts.astype(bool).multiply(gs_gene_sparse_mtx.T.astype(bool))\n",
    "        unique_gene_overlaps = unique_gene_overlaps.todense().A.sum(axis = 1)\n",
    "        results['num_unique_genes_in_overlap'] = unique_gene_overlaps\n",
    "\n",
    "        results = results[['disease', 'category', 'sv_category', 'cohort', 'gs', \n",
    "                           'ref_freq', 'cat_freq', 'num_unique_genes_in_overlap', 'OR', 'P']]\n",
    "        results = results.rename(columns = {'OR': 'odds_ratio', 'P': 'p'})\n",
    "\n",
    "        # here, we do FDR correction.\n",
    "        fdr_p = statsmodels.stats.multitest.multipletests(results['p'].to_list(), method='fdr_bh')[1]\n",
    "        results['fdr_p'] = fdr_p\n",
    "        results['bonf_p'] = results['p'] * len(gene_sets)\n",
    "\n",
    "        fisher_results.append(results)\n",
    "                \n",
    "gse_results = pd.concat(fisher_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "9cb7cec1-0bbc-4027-bcfa-d736ba050175",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease</th>\n",
       "      <th>category</th>\n",
       "      <th>sv_category</th>\n",
       "      <th>cohort</th>\n",
       "      <th>gs</th>\n",
       "      <th>ref_freq</th>\n",
       "      <th>cat_freq</th>\n",
       "      <th>num_unique_genes_in_overlap</th>\n",
       "      <th>odds_ratio</th>\n",
       "      <th>p</th>\n",
       "      <th>fdr_p</th>\n",
       "      <th>bonf_p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ewing</td>\n",
       "      <td>ANY.SINGLETON.PREDICTED_COPY_GAIN.lof_constrai...</td>\n",
       "      <td>coding</td>\n",
       "      <td>case</td>\n",
       "      <td>regulation of protein secretion (GO:0050708)</td>\n",
       "      <td>54/3909</td>\n",
       "      <td>1/50</td>\n",
       "      <td>1</td>\n",
       "      <td>2.804364</td>\n",
       "      <td>0.505341</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1377.053474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ewing</td>\n",
       "      <td>ANY.SINGLETON.PREDICTED_COPY_GAIN.lof_constrai...</td>\n",
       "      <td>coding</td>\n",
       "      <td>case</td>\n",
       "      <td>negative regulation of fibroblast proliferatio...</td>\n",
       "      <td>3/3909</td>\n",
       "      <td>0/50</td>\n",
       "      <td>0</td>\n",
       "      <td>19.151961</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2725.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  disease                                           category sv_category  \\\n",
       "0   ewing  ANY.SINGLETON.PREDICTED_COPY_GAIN.lof_constrai...      coding   \n",
       "1   ewing  ANY.SINGLETON.PREDICTED_COPY_GAIN.lof_constrai...      coding   \n",
       "\n",
       "  cohort                                                 gs ref_freq cat_freq  \\\n",
       "0   case       regulation of protein secretion (GO:0050708)  54/3909     1/50   \n",
       "1   case  negative regulation of fibroblast proliferatio...   3/3909     0/50   \n",
       "\n",
       "   num_unique_genes_in_overlap  odds_ratio         p  fdr_p       bonf_p  \n",
       "0                            1    2.804364  0.505341    1.0  1377.053474  \n",
       "1                            0   19.151961  1.000000    1.0  2725.000000  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gse_results.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a2bc59d-c493-4810-af76-488f996a8304",
   "metadata": {},
   "source": [
    "Compress the p values for export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "4a46d435-e1e2-4647-b583-479214e6c217",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gse_results[\"p\"] = gse_results[\"p\"].round(5)\n",
    "# gse_results[\"fdr_p\"] = gse_results[\"fdr_p\"].round(5)\n",
    "# gse_results[\"odds_ratio\"] = gse_results[\"odds_ratio\"].round(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "692fd950-c8bc-4c8f-a8ce-f61dc868da9e",
   "metadata": {},
   "source": [
    "We now reorganize this dataframe so that the two cohorts (`cases`, `controls`) are all on one row."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "c08752df-8ffe-49f4-a4a3-f5a7e7b80091",
   "metadata": {},
   "outputs": [],
   "source": [
    "cases = gse_results.query('cohort == \"case\"')\n",
    "controls = gse_results.query('cohort == \"control\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1a7f848-d14b-4ba4-9892-88a857e729be",
   "metadata": {},
   "source": [
    "Each cohort should has the exact same order of columns, so we can just add new columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "3f18dbb5-54a7-42c3-98ac-2ea66626c89a",
   "metadata": {},
   "outputs": [],
   "source": [
    "overlap_cols = [\n",
    "    \"disease\",\n",
    "    \"category\",\n",
    "    \"sv_category\",\n",
    "    \"gs\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "11b4bbd7-0cf3-46de-b364-24ca97074724",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "disease        True\n",
       "category       True\n",
       "sv_category    True\n",
       "gs             True\n",
       "dtype: bool"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check that the overlaps are the same\n",
    "(cases[overlap_cols] == controls[overlap_cols]).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "d6672001-e061-4008-86fa-a34867853688",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_gse_results = cases.drop(columns=[\"cohort\"]).copy()\n",
    "\n",
    "# rename some columns\n",
    "combined_gse_results.columns = [\n",
    "    c + \"_cases\" if c not in overlap_cols else c for c in combined_gse_results.columns\n",
    "]\n",
    "\n",
    "d_temp = controls.copy().drop(columns=overlap_cols + ['cohort'])\n",
    "d_temp.columns = [c + '_controls' for c in d_temp.columns]\n",
    "combined_gse_results[d_temp.columns] = d_temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "07f9fdf6-f972-4a49-9950-3659ec64cd57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease</th>\n",
       "      <th>category</th>\n",
       "      <th>sv_category</th>\n",
       "      <th>gs</th>\n",
       "      <th>ref_freq_cases</th>\n",
       "      <th>cat_freq_cases</th>\n",
       "      <th>num_unique_genes_in_overlap_cases</th>\n",
       "      <th>odds_ratio_cases</th>\n",
       "      <th>p_cases</th>\n",
       "      <th>fdr_p_cases</th>\n",
       "      <th>bonf_p_cases</th>\n",
       "      <th>ref_freq_controls</th>\n",
       "      <th>cat_freq_controls</th>\n",
       "      <th>num_unique_genes_in_overlap_controls</th>\n",
       "      <th>odds_ratio_controls</th>\n",
       "      <th>p_controls</th>\n",
       "      <th>fdr_p_controls</th>\n",
       "      <th>bonf_p_controls</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ewing</td>\n",
       "      <td>ANY.SINGLETON.PREDICTED_COPY_GAIN.lof_constrai...</td>\n",
       "      <td>coding</td>\n",
       "      <td>regulation of protein secretion (GO:0050708)</td>\n",
       "      <td>54/3909</td>\n",
       "      <td>1/50</td>\n",
       "      <td>1</td>\n",
       "      <td>2.804364</td>\n",
       "      <td>0.505341</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1377.053474</td>\n",
       "      <td>270/22714</td>\n",
       "      <td>0/192</td>\n",
       "      <td>0</td>\n",
       "      <td>0.429134</td>\n",
       "      <td>0.179149</td>\n",
       "      <td>0.513875</td>\n",
       "      <td>488.181684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ewing</td>\n",
       "      <td>ANY.SINGLETON.PREDICTED_COPY_GAIN.lof_constrai...</td>\n",
       "      <td>coding</td>\n",
       "      <td>negative regulation of fibroblast proliferatio...</td>\n",
       "      <td>3/3909</td>\n",
       "      <td>0/50</td>\n",
       "      <td>0</td>\n",
       "      <td>19.151961</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2725.000000</td>\n",
       "      <td>7/22714</td>\n",
       "      <td>1/192</td>\n",
       "      <td>1</td>\n",
       "      <td>29.567708</td>\n",
       "      <td>0.065132</td>\n",
       "      <td>0.288123</td>\n",
       "      <td>177.483892</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  disease                                           category sv_category  \\\n",
       "0   ewing  ANY.SINGLETON.PREDICTED_COPY_GAIN.lof_constrai...      coding   \n",
       "1   ewing  ANY.SINGLETON.PREDICTED_COPY_GAIN.lof_constrai...      coding   \n",
       "\n",
       "                                                  gs ref_freq_cases  \\\n",
       "0       regulation of protein secretion (GO:0050708)        54/3909   \n",
       "1  negative regulation of fibroblast proliferatio...         3/3909   \n",
       "\n",
       "  cat_freq_cases  num_unique_genes_in_overlap_cases  odds_ratio_cases  \\\n",
       "0           1/50                                  1          2.804364   \n",
       "1           0/50                                  0         19.151961   \n",
       "\n",
       "    p_cases  fdr_p_cases  bonf_p_cases ref_freq_controls cat_freq_controls  \\\n",
       "0  0.505341          1.0   1377.053474         270/22714             0/192   \n",
       "1  1.000000          1.0   2725.000000           7/22714             1/192   \n",
       "\n",
       "   num_unique_genes_in_overlap_controls  odds_ratio_controls  p_controls  \\\n",
       "0                                     0             0.429134    0.179149   \n",
       "1                                     1            29.567708    0.065132   \n",
       "\n",
       "   fdr_p_controls  bonf_p_controls  \n",
       "0        0.513875       488.181684  \n",
       "1        0.288123       177.483892  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_gse_results.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "628569ac-2186-4ab8-84c8-5b10a96efc36",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_gse_results.to_csv(\n",
    "    \"data/updated-cwas/results/cwas-gene-set-enrichment-results.csv\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4207bc8c-98e7-4864-a577-bcc695284177",
   "metadata": {},
   "source": [
    "And export the reference counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "c0d13d1d-693b-480d-9e01-d2f8525bb669",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_counts.to_csv('data/updated-cwas/reference-counts.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c7b9723-c5f7-4120-bcad-a43390d9129c",
   "metadata": {},
   "source": [
    "This notebook is getting much too large, so I'm going to analyze these results in a different notebook."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pediatric-germline-svs-3.7.13",
   "language": "python",
   "name": "pediatric-germline-svs-3.7.13"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
