{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "7f33ada0-6685-46cf-abe3-cd2101806b4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "import pybedtools\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "\n",
    "pd.set_option(\"display.max_columns\", 1000)\n",
    "pd.set_option(\"display.max_rows\", 100)\n",
    "\n",
    "import statsmodels\n",
    "import statsmodels.api as sm\n",
    "import warnings\n",
    "\n",
    "import seaborn as sns\n",
    "import palettable\n",
    "import matplotlib.pyplot as plt\n",
    "from scripts import manuscript_colors as mc\n",
    "\n",
    "from scripts import helper\n",
    "\n",
    "bold_10 = palettable.cartocolors.qualitative.Bold_10.mpl_colors\n",
    "\n",
    "from scripts import aesthetics, sv_plot\n",
    "\n",
    "aesthetics.activate_paper_rcParams('illustrator')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fbca85d-a331-409c-823b-c30f672c9a51",
   "metadata": {},
   "source": [
    "# Targeted gene testing\n",
    "\n",
    "In the months leading up to the creation of this notebook, we've generated separate notebooks for processing SVs around individual genes.\n",
    "\n",
    "We're starting to test enough genes to where this is becoming cumbersome. In particular, the SVs and dosage files are so large that loading them--across all the notebooks--takes several minutes.\n",
    "\n",
    "This notebook is intended to handle preprocessing for these genes, such that we can simply load in the SVs and begin processing quickly."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0067ee62-52e0-4f90-a417-f06165af63d1",
   "metadata": {},
   "source": [
    "# Load the data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91808b0d-2382-42f1-a3fb-2d966cfe9ecd",
   "metadata": {},
   "source": [
    "## Define paths to our data\n",
    "We have 5 pieces of data.\n",
    "\n",
    "Two of these files depends on whether we're analyzing the discovery or validation dataset:\n",
    "1. Our SV file (in bed format)\n",
    "2. Our dosage file (a dense matrix of SV genotypes)\n",
    "\n",
    "Three files do not depend on this:\n",
    "\n",
    "3. Our metadata files (which define which samples are cases, which are controls, etc)\n",
    "4. Our list of genes (which defines which regions we analyze)\n",
    "5. Our gene gtf (which defines the locations of our regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2f3f6e30-f428-4abd-85d9-6aeb76d0a695",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define the folder name for all our results\n",
    "folder_name = \"processed-data-v2.5.3\"\n",
    "\n",
    "# define SVs and dosages for discovery and validation\n",
    "svs_discovery_path = \"gs://vanallen-pedsv-analysis/beds/PedSV.v2.5.3.case_control_cohort.analysis_samples.sites.bed.gz\"\n",
    "svs_validation_path = \"gs://vanallen-pedsv-analysis/beds/PedSV.v2.5.3.trio_cohort.analysis_samples.sites.bed.gz\"\n",
    "\n",
    "dosages_discovery_path = \"gs://vanallen-pedsv-analysis/beds/PedSV.v2.5.3.case_control_cohort.analysis_samples.allele_dosages.bed.gz\"\n",
    "dosages_validation_path = \"gs://vanallen-pedsv-analysis/beds/PedSV.v2.5.3.trio_cohort.analysis_samples.allele_dosages.bed.gz\"\n",
    "\n",
    "# define metadata\n",
    "metadata_path = \"gs://vanallen-pedsv-analysis/sample_info/PedSV.v2.5.3.cohort_metadata.w_control_assignments.tsv.gz\"\n",
    "\n",
    "discovery_samples_path = \"gs://vanallen-pedsv-analysis/sample_info/PedSV.v2.5.3.case_control_analysis_cohort.samples.list\"\n",
    "validation_samples_path = \"gs://vanallen-pedsv-analysis/sample_info/PedSV.v2.5.3.trio_analysis_cohort.samples.list\"\n",
    "\n",
    "# define which genes are tested\n",
    "genes_to_test_path = \"ref/genes-for-analysis.csv\"\n",
    "\n",
    "# define reference for gene locations - GENCODE v43 (hg38)\n",
    "gtf_path = \"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_43/gencode.v43.basic.annotation.gtf.gz\"\n",
    "\n",
    "# define the window size for our analysis (centered on the gene's center) - use 1 Mb\n",
    "window_size = 1e6"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dac6eac3-e55f-4eca-add4-a295bbaad1aa",
   "metadata": {},
   "source": [
    "## Load the data into memory"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e279969-4afb-4c27-8816-4adc0e153a5e",
   "metadata": {},
   "source": [
    "First, load the SVs and dosages (takes a bit). We use a dictionary so we can simply refer to keys later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3c278d65-c906-4419-9b70-ddc8b121ca12",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jettcrowdis/opt/anaconda3/envs/pediatric-germline-svs-3.7.13/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3457: DtypeWarning: Columns (10,12,13,34,35,37,39,41,45,46,47,48,72,73,87,88,106,107,121,122,136,137,155,156,170,171,185,186,204,205,219,220,234,235,253,254,268,269,283,284,302,303,366,367,381,382,400,401,415,416,430,431,449,450,464,465,479,480,498,499,513,514,528,529,547,548,562,563,577,578,596,597,645,646,647,648,649,650,690,691,692,693,694,695,713,714,728,729,743,744,758,759,773,774,788,789,803,804,818,819,833,834,848,849,863,864,878,879,893,894,908,909,923,924,931,937,943) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "/Users/jettcrowdis/opt/anaconda3/envs/pediatric-germline-svs-3.7.13/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3457: DtypeWarning: Columns (34,35,37,39,41,45,46,72,73,87,88,106,107,121,122,136,137,155,156,170,171,185,186,204,205,219,220,234,235,253,254,268,269,283,284,302,303,366,367,381,382,400,401,415,416,430,431,449,450,464,465,479,480,498,499,513,514,528,529,547,548,562,563,577,578,596,597,645,646,647,648,649,650,690,691,692,693,694,695,713,714,728,729,743,744,758,759,773,774,788,789,803,804,818,819,833,834,848,849,863,864,878,879,893,894,908,909,923,924,931,937,943) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    }
   ],
   "source": [
    "###############\n",
    "##### SVs #####\n",
    "###############\n",
    "svs = {}\n",
    "\n",
    "svs[\"discovery\"] = pd.read_csv(\n",
    "    svs_discovery_path,\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "\n",
    "svs[\"validation\"] = pd.read_csv(\n",
    "    svs_validation_path,\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "###############\n",
    "### Dosages ###\n",
    "###############\n",
    "dosages = {}\n",
    "\n",
    "dosages[\"discovery\"] = pd.read_csv(\n",
    "    dosages_discovery_path,\n",
    "    sep=\"\\t\",\n",
    "    index_col=False,\n",
    ")\n",
    "\n",
    "dosages[\"validation\"] = pd.read_csv(\n",
    "    dosages_validation_path,\n",
    "    sep=\"\\t\",\n",
    "    index_col=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65d2ecaa-b5b8-48f2-b228-34ff0135813a",
   "metadata": {},
   "source": [
    "Next we load metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e0786514-4fc1-4af2-ab3e-ff5b33f0eff3",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(\n",
    "    metadata_path,\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "\n",
    "# add a sex label to metadata\n",
    "metadata[\"sex\"] = (metadata[\"chrX_CopyNumber\"].round() < 2).astype(int)\n",
    "\n",
    "###############\n",
    "### Samples ###\n",
    "###############\n",
    "samples = {}\n",
    "\n",
    "samples[\"discovery\"] = pd.read_csv(\n",
    "    discovery_samples_path,\n",
    "    header=None,\n",
    ")[0].to_list()\n",
    "\n",
    "samples[\"validation\"] = pd.read_csv(\n",
    "    validation_samples_path,\n",
    "    header=None,\n",
    ")[0].to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83f037b0-16f8-4a4f-8a64-42505d20886e",
   "metadata": {},
   "source": [
    "And finally, load the gene data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8b620b3c-1a1a-43ba-b993-db598fc558f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_to_test = pd.read_csv(genes_to_test_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a466b669-f765-4f40-a41c-2c61f6eac4ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = pd.read_csv(\n",
    "    gtf_path,\n",
    "    comment=\"#\",\n",
    "    sep=\"\\t\",\n",
    "    header=None,\n",
    "    names=[\n",
    "        \"seqname\",\n",
    "        \"source\",\n",
    "        \"feature\",\n",
    "        \"start\",\n",
    "        \"end\",\n",
    "        \"score\",\n",
    "        \"strand\",\n",
    "        \"frame\",\n",
    "        \"attribute\",\n",
    "    ],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d7dca2a-459a-4e4e-8fe0-be632ebb7dd4",
   "metadata": {},
   "source": [
    "## Make data folders\n",
    "\n",
    "Here, we define folders that will contain all the data.\n",
    "\n",
    "__Note: To avoid confusion, this code will erase any pre-existing runs with the same name. Use with caution.__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c500efc9-293c-41d8-9f06-fc2c06bce7a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "if os.path.isdir(f\"data/{folder_name}\"):\n",
    "    shutil.rmtree(f\"data/{folder_name}\")\n",
    "\n",
    "# make the data directory\n",
    "os.mkdir(f\"data/{folder_name}\")\n",
    "\n",
    "for index, row in genes_to_test.iterrows():\n",
    "    g = row[\"gene\"]\n",
    "    disease = row[\"disease\"]\n",
    "    os.mkdir(f\"data/{folder_name}/{g}-{disease}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30446f71-fb21-4b56-9cfa-f6bd07da0a5d",
   "metadata": {},
   "source": [
    "# Extract gene locations\n",
    "\n",
    "Here, we extract from the gtf the locations of our genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2644789e-da72-4cb8-96fb-3ad7401bd7b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_to_transcript = dict(genes_to_test[[\"gene\", \"transcript\"]].values)\n",
    "\n",
    "gene_locs = []\n",
    "for index, row in genes_to_test.iterrows():\n",
    "    g, t, disease = row.values\n",
    "\n",
    "    # extract all elements of the gene and store them\n",
    "    g_gtf = gtf[(gtf[\"attribute\"].str.contains(f\"transcript_name \\\"{t}\\\"\"))]\n",
    "    g_gtf.to_csv(f\"data/{folder_name}/{g}-{disease}/{g}-gtf.csv\", index=False)\n",
    "    \n",
    "    # we also define the start and end of the gene here\n",
    "    gene_loc = gtf[(gtf[\"attribute\"].str.contains(f\"gene_name \\\"{g}\\\"\")) & (gtf['feature'] == 'gene')]\n",
    "    assert len(gene_loc) == 1\n",
    "    chrom, start, end = gene_loc[['seqname', 'start', 'end']].values[0]\n",
    "    \n",
    "    # store gene location\n",
    "    gene_locs.append([g, chrom, start, end])\n",
    "    \n",
    "gene_locs = pd.DataFrame(gene_locs, columns = ['gene', 'chrom', 'start', 'end'])\n",
    "\n",
    "# drop duplicates - happens with the same gene and different diseases\n",
    "gene_locs = gene_locs.drop_duplicates()\n",
    "\n",
    "# and expand the window according to our window size\n",
    "gene_locs['window_start'] = (gene_locs['start'] - window_size / 2).astype(int)\n",
    "gene_locs['window_end'] = (gene_locs['end'] + window_size / 2).astype(int)\n",
    "\n",
    "# merge if we haven't already\n",
    "if 'chrom' not in genes_to_test.columns:\n",
    "    genes_to_test = genes_to_test.merge(gene_locs, on = 'gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ffd97820-3d82-4389-8fab-1e0c156a6ed9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "      <th>disease</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>window_start</th>\n",
       "      <th>window_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EWSR1</td>\n",
       "      <td>EWSR1-206</td>\n",
       "      <td>ewing</td>\n",
       "      <td>chr22</td>\n",
       "      <td>29268009</td>\n",
       "      <td>29300525</td>\n",
       "      <td>28768009</td>\n",
       "      <td>29800525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FLI1</td>\n",
       "      <td>FLI1-205</td>\n",
       "      <td>ewing</td>\n",
       "      <td>chr11</td>\n",
       "      <td>128686535</td>\n",
       "      <td>128813267</td>\n",
       "      <td>128186535</td>\n",
       "      <td>129313267</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene transcript disease  chrom      start        end  window_start  \\\n",
       "0  EWSR1  EWSR1-206   ewing  chr22   29268009   29300525      28768009   \n",
       "1   FLI1   FLI1-205   ewing  chr11  128686535  128813267     128186535   \n",
       "\n",
       "   window_end  \n",
       "0    29800525  \n",
       "1   129313267  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_to_test.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "acf74a7d-e7d0-47ad-9b8b-4eb8af64d46d",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_to_test.to_csv(f'data/{folder_name}/gene-locations.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bacdf66-3868-4693-a339-0f46a0d6925e",
   "metadata": {},
   "source": [
    "# Define the SVs of interest for each analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd75f170-1f9e-4236-82ef-15ead15e2282",
   "metadata": {},
   "source": [
    "Doing this is somewhat complicated, as we need to juggle both discovery and validation cohorts. We do the following:\n",
    "\n",
    "1. Define the cases and controls for the analysis based on the disease type\n",
    "2. Define the genomic region of interest, based on the gene data defined above\n",
    "3. Subset the SVs down to SVs in this region\n",
    "4. Subset the dosage down to those SVs and the samples of interest\n",
    "\n",
    "We use a kind of round-about disease-based approach because then we only need to make three copies of large dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "419f4503-dac9-404a-8f0c-9835e0dde9e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ewing discovery\n",
      "ewing validation\n",
      "osteosarcoma discovery\n",
      "osteosarcoma validation\n",
      "neuroblastoma discovery\n",
      "neuroblastoma validation\n"
     ]
    }
   ],
   "source": [
    "for disease in ['ewing', 'osteosarcoma', 'neuroblastoma']:\n",
    "    \n",
    "    # define the number of genes we're looking at in this disease\n",
    "    disease_genes = genes_to_test[genes_to_test['disease'] == disease]\n",
    "    if len(disease_genes) == 0:\n",
    "        continue\n",
    "    \n",
    "    # we subset discovery and validation datasets to relevant samples and SVs\n",
    "    for cohort in [\"discovery\", \"validation\"]:\n",
    "        print(disease, cohort)\n",
    "        \n",
    "        # osteosarcoma has no validation datasets, so we skip\n",
    "        if disease == 'osteosarcoma' and cohort == 'validation':\n",
    "            continue\n",
    "        \n",
    "        ########################\n",
    "        #### DEFINE SAMPLES ####\n",
    "        ########################\n",
    "        \n",
    "        # this term for cohort is used in the metadata\n",
    "        study_label = \"case_control\" if cohort == \"discovery\" else \"trio\"\n",
    "\n",
    "        # define the samples\n",
    "        cohort_samples = samples[cohort]\n",
    "\n",
    "        # get the cases - must belong to disease, must be right cohort, and must be a case included for analysis\n",
    "        cases = metadata[\n",
    "            (metadata[\"study_phase\"] == study_label)\n",
    "            & (metadata[f\"{disease}_case\"])\n",
    "        ][\"entity:sample_id\"].to_list()\n",
    "        cases = sorted(set(cohort_samples) & set(cases))\n",
    "\n",
    "        # get the controls - must be a control, right cohort, and a control for that analysis\n",
    "        controls = metadata[\n",
    "            (metadata[\"study_phase\"] == study_label)\n",
    "            & (metadata[f\"{disease}_control\"])\n",
    "        ][\"entity:sample_id\"].to_list()\n",
    "        controls = sorted(set(cohort_samples) & set(controls))\n",
    "\n",
    "        # some quick checks\n",
    "        assert set(cases) & set(controls) == set()\n",
    "        \n",
    "        ################################\n",
    "        #### DEFINE SVS AND DOSAGES ####\n",
    "        ################################\n",
    "        \n",
    "        # fetch our cohort data (discovery or validation)\n",
    "        cohort_svs = svs[cohort].copy()\n",
    "        cohort_dosages = dosages[cohort].copy()\n",
    "        \n",
    "        # subset the dosage matrix down to the cases and controls for our disease\n",
    "        cohort_dosages = cohort_dosages[['#chr', 'start', 'end', 'ID'] + cases + controls].copy()\n",
    "        \n",
    "        # drop any SVs that have no alternative genotypes in these samples (i.e. aren't present)\n",
    "        cohort_sv_counts = cohort_dosages.iloc[:, 4:].fillna(0).sum(axis = 1)\n",
    "\n",
    "        # subset down SVs and dosages to just SVs with nonzero dosages\n",
    "        cohort_svs = cohort_svs[cohort_sv_counts != 0]\n",
    "        cohort_dosages = cohort_dosages[cohort_sv_counts != 0]\n",
    "        \n",
    "        #########################################\n",
    "        #### SUBSET SVS and DOSAGES TO GENES ####\n",
    "        #########################################\n",
    "        gene_locs = disease_genes[['chrom', 'window_start', 'window_end', 'gene']]\n",
    "        \n",
    "        # bedtools expects specific column names and ordering\n",
    "        gene_locs.columns = ['chrom', 'start', 'end', 'gene']\n",
    "        \n",
    "        # create bedtools object for our genes\n",
    "        gene_bed = pybedtools.BedTool(\n",
    "            gene_locs.to_csv(sep=\"\\t\", index=False, header=False), from_string=True\n",
    "        )\n",
    "        \n",
    "        # create bedtool object for svs\n",
    "        sv_bed = pybedtools.BedTool(\n",
    "            cohort_svs[[\"#chrom\", \"start\", \"end\", \"name\"]].to_csv(sep=\"\\t\", index=False, header=False),\n",
    "            from_string=True,\n",
    "        )\n",
    "\n",
    "        svs_in_regions = sv_bed.intersect(gene_bed, wo=True)\n",
    "        svs_in_regions = svs_in_regions.to_dataframe()\n",
    "\n",
    "        # drop unnecessary columns\n",
    "        svs_in_regions = svs_in_regions.iloc[:, [0, 1, 2, 3, 7]].copy()\n",
    "        svs_in_regions.columns = [\"chrom\", \"start\", \"end\", \"name\", \"gene\"]\n",
    "        \n",
    "        ############################\n",
    "        #### ADD IN COMPLEX SVS ####\n",
    "        ############################\n",
    "        \n",
    "        # This step is not completely necessary, but I include it for completeness.\n",
    "        # The idea is that basing our SVs on strict overlap isn't good enough--some SVs\n",
    "        # may involve in the region as part of a complex rearrangement.\n",
    "        \n",
    "        cohort_complex_svs = cohort_svs[(cohort_svs[\"svtype\"] == \"CPX\")][\n",
    "                                    [\n",
    "                                        \"name\",\n",
    "                                        \"#chrom\",\n",
    "                                        \"start\",\n",
    "                                        \"end\",\n",
    "                                        \"svtype\",\n",
    "                                        \"CHR2\",\n",
    "                                        \"CPX_INTERVALS\",\n",
    "                                        \"CPX_TYPE\",\n",
    "                                        \"END\",\n",
    "                                    ]\n",
    "                                ].copy()\n",
    "        \n",
    "        # split each component SV--represented by CPX_INTERVALS--into their own rows\n",
    "        cohort_complex_svs = cohort_complex_svs.assign(\n",
    "            intervals=cohort_complex_svs[\"CPX_INTERVALS\"].str.split(\",\")\n",
    "        ).explode(\"intervals\")\n",
    "\n",
    "        # convert this data to a proper bed for intersection\n",
    "        cohort_complex_sv_bed = cohort_complex_svs[[\"name\", \"intervals\"]].copy()\n",
    "        cohort_complex_sv_bed[\"chrom\"] = cohort_complex_sv_bed[\"intervals\"].apply(\n",
    "            lambda interv: interv.split(\":\")[0].split(\"_\")[-1]\n",
    "        )\n",
    "        cohort_complex_sv_bed[\"start\"] = cohort_complex_sv_bed[\"intervals\"].apply(\n",
    "            lambda interv: interv.split(\":\")[-1].split(\"-\")[0]\n",
    "        )\n",
    "        cohort_complex_sv_bed[\"end\"] = cohort_complex_sv_bed[\"intervals\"].apply(\n",
    "            lambda interv: interv.split(\":\")[-1].split(\"-\")[1]\n",
    "        )\n",
    "\n",
    "        # convert to a bedtool\n",
    "        cohort_complex_sv_bed = pybedtools.BedTool(\n",
    "            cohort_complex_sv_bed[[\"chrom\", \"start\", \"end\", \"name\"]].to_csv(\n",
    "                sep=\"\\t\", index=False, header=False\n",
    "            ),\n",
    "            from_string=True,\n",
    "        )\n",
    "\n",
    "        # do our intersection\n",
    "        complex_svs_in_regions = cohort_complex_sv_bed.intersect(gene_bed, wo=True)\n",
    "        complex_svs_in_regions = complex_svs_in_regions.to_dataframe()\n",
    "\n",
    "        # drop unnecessary columns\n",
    "        if len(complex_svs_in_regions):\n",
    "            complex_svs_in_regions = complex_svs_in_regions.iloc[:, [0, 1, 2, 3, 7]].copy()\n",
    "            complex_svs_in_regions.columns = [\"chrom\", \"start\", \"end\", \"name\", \"gene\"]\n",
    "        else:\n",
    "            complex_svs_in_regions = pd.DataFrame(\n",
    "                columns=[\"chrom\", \"start\", \"end\", \"name\", \"gene\"]\n",
    "            )\n",
    "        \n",
    "        ######################################################\n",
    "        #### DEFINE AND FILTER OUR SVS IN THESE REGIONS ######\n",
    "        ######################################################\n",
    "        \n",
    "        # combine normal and complex SVs together\n",
    "        svs_in_regions_names = pd.concat(\n",
    "            [svs_in_regions[[\"name\", \"gene\"]], complex_svs_in_regions[[\"name\", \"gene\"]]]\n",
    "        )\n",
    "\n",
    "        # drop duplicates (only happens if an SV is simple but also part of a complex SV)\n",
    "        svs_in_regions_names = svs_in_regions_names.drop_duplicates(subset = ['name'])\n",
    "        \n",
    "        # subset SVs\n",
    "        subset_svs = cohort_svs.set_index(\"name\").loc[svs_in_regions_names[\"name\"]].reset_index()\n",
    "        subset_svs.insert(0, \"gene\", svs_in_regions_names[\"gene\"].values)\n",
    "\n",
    "        # convert AF to float. We extract an AF for CNVs as well\n",
    "        subset_svs['AF'] = subset_svs.apply(lambda row: float(row['AF']) if row['svtype'] != 'CNV' else row['CN_NONREF_FREQ'], axis = 1)\n",
    "\n",
    "        # only include SVs that pass the FILTER and are in our dosage matrix\n",
    "        svs_to_keep = subset_svs[\n",
    "            (\n",
    "                (subset_svs[\"FILTER\"].isin([\"PASS\", \"MULTIALLELIC\"]))\n",
    "                & (subset_svs[\"name\"].isin(cohort_dosages[\"ID\"].to_list()))\n",
    "            )\n",
    "        ].copy()\n",
    "\n",
    "        svs_to_keep[\"rare\"] = svs_to_keep[\"AF\"] < 0.01\n",
    "        \n",
    "        # subset dosage matrix\n",
    "        subset_dosages = cohort_dosages.set_index('ID').loc[svs_to_keep[\"name\"].to_list()].reset_index()\n",
    "\n",
    "        # we add in the information about which gene is affected (the first column is just the SV name, so we drop it)\n",
    "        subsetted_gene_dosages = (\n",
    "            svs_to_keep[[\"name\", \"gene\", \"AF\", \"rare\"]]\n",
    "            .merge(subset_dosages, right_on=\"ID\", left_on=\"name\")\n",
    "            .drop(columns = ['ID'])\n",
    "        )\n",
    "\n",
    "        #########################\n",
    "        #### EXPORT OUR DATA ####\n",
    "        #########################\n",
    "        \n",
    "        # extract cohort metadata\n",
    "        cohort_metadata = metadata[metadata['entity:sample_id'].isin(cases + controls)]\n",
    "        \n",
    "        # we export everything to reconstruct whatever analyses we want.\n",
    "        for g in gene_locs['gene']:\n",
    "            \n",
    "            export_prefix = f'data/{folder_name}/'\n",
    "\n",
    "            # export sample info\n",
    "            cohort_metadata.to_csv(export_prefix + f'{g}-{disease}/{cohort}-metadata.csv', index=False)\n",
    "            \n",
    "            gene_svs = svs_to_keep[svs_to_keep['gene'] == g]\n",
    "            gene_dosages = subsetted_gene_dosages[subsetted_gene_dosages['gene'] == g]\n",
    "            \n",
    "            gene_svs.to_csv(export_prefix + f'{g}-{disease}/{cohort}-svs.csv', index=False)\n",
    "            gene_dosages.to_csv(export_prefix + f'{g}-{disease}/{cohort}-dosages.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4289b5b-bf66-4e7c-ab44-19018455e997",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Calculate allele fractions\n",
    "\n",
    "While we could have done this above, I think it's worth separating this out for compartmentalization. It means we need to read in exported data, but I think the tradeoff is worth it.\n",
    "\n",
    "We devise two metrics for cases and controls--one listing the allele frequency of each SV and another listing the number of samples with an SV in the gene. We need to convert the dosage (allele count) to allele frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "807f8047-7e48-4657-94ed-73e58b1b0f79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EWSR1 ewing\n",
      "FLI1 ewing\n",
      "MYCN neuroblastoma\n",
      "TERT neuroblastoma\n",
      "ALK neuroblastoma\n",
      "TP53 osteosarcoma\n",
      "TP53 neuroblastoma\n",
      "TP53 ewing\n",
      "RB1 osteosarcoma\n",
      "ATRX neuroblastoma\n",
      "SHANK2 neuroblastoma\n",
      "BRCA1 neuroblastoma\n",
      "CDKN2A ewing\n",
      "DLG2 osteosarcoma\n"
     ]
    }
   ],
   "source": [
    "for index, row in genes_to_test.iterrows():\n",
    "    gene, disease = row[['gene', 'disease']].values\n",
    "    print(gene, disease)\n",
    "    \n",
    "    file_prefix = f'data/{folder_name}/{gene}-{disease}/'\n",
    "    for cohort in ['discovery', 'validation']:\n",
    "\n",
    "        # osteosarcoma has no validation datasets, so we skip\n",
    "        if disease == 'osteosarcoma' and cohort == 'validation':\n",
    "            continue\n",
    "    \n",
    "        ######################\n",
    "        #### READ IN DATA ####\n",
    "        ######################\n",
    "        analysis_dosages = pd.read_csv(file_prefix + f'{cohort}-dosages.csv')\n",
    "        analysis_svs = pd.read_csv(file_prefix + f'{cohort}-svs.csv')\n",
    "        analysis_metadata = pd.read_csv(file_prefix + f'{cohort}-metadata.csv')\n",
    "\n",
    "        # get the cases and controls\n",
    "        analysis_cases = analysis_metadata.query(f'disease == \"{disease}\"')['entity:sample_id'].to_list()\n",
    "        analysis_controls = analysis_metadata.query(f'disease == \"control\"')['entity:sample_id'].to_list()\n",
    "        \n",
    "        ###########################\n",
    "        #### EXTRACT SV COUNTS ####\n",
    "        ###########################\n",
    "        \n",
    "        case_control_counts = []\n",
    "        case_control_samples_with_svs = []\n",
    "\n",
    "        for sample_group, label in zip(\n",
    "            [analysis_cases, analysis_controls], [\"cases\", \"controls\"]\n",
    "        ):\n",
    "\n",
    "            # subset to these samples\n",
    "            sample_dosages = analysis_dosages[\n",
    "                [\"name\", \"gene\", \"#chr\", \"start\", \"end\", \"AF\", \"rare\"] + sample_group\n",
    "            ]\n",
    "\n",
    "            # first, count the SVs. This function counts the number of samples with each genotype.\n",
    "            sv_counts = (\n",
    "                analysis_dosages[sample_group]\n",
    "                .apply(lambda row: row.dropna().value_counts(), axis=1)\n",
    "                .fillna(0)\n",
    "                .astype(int)\n",
    "            )\n",
    "\n",
    "            # name columns to genotypes\n",
    "            sv_counts.columns = f\"{label}_\" + sv_counts.columns.astype(int).astype(str)\n",
    "            sv_counts[\"name\"] = sample_dosages[\"name\"]\n",
    "\n",
    "            case_control_counts.append(sv_counts)\n",
    "\n",
    "            # now count the number of samples with at least one SV around each gene.\n",
    "            # we groupby gene, then determine whether each sample has an SV in that gene, then\n",
    "            # count these samples up per gene.\n",
    "            num_samples_with_svs = sample_dosages.groupby([\"gene\"]).apply(\n",
    "                lambda group: (group[sample_group] > 0).any(axis=0).sum()\n",
    "            )\n",
    "\n",
    "            num_samples_with_svs.name = label + \"_all\"\n",
    "            num_samples_with_svs = num_samples_with_svs / len(sample_group)\n",
    "            case_control_samples_with_svs.append(num_samples_with_svs)\n",
    "\n",
    "            # repeat for rare SVs\n",
    "            num_samples_with_svs_rare = (\n",
    "                sample_dosages.query(\"rare == True\")\n",
    "                .groupby([\"gene\"])\n",
    "                .apply(lambda group: (group[sample_group] > 0).any(axis=0).sum())\n",
    "            )\n",
    "\n",
    "            num_samples_with_svs_rare.name = label + \"_rare\"\n",
    "            num_samples_with_svs_rare = num_samples_with_svs_rare / len(sample_group)\n",
    "            case_control_samples_with_svs.append(num_samples_with_svs_rare)\n",
    "\n",
    "        # merge cases and controls together\n",
    "        case_control_sv_counts = sample_dosages.iloc[:, :7].reset_index(drop=True)\n",
    "        for df in case_control_counts:\n",
    "            case_control_sv_counts = case_control_sv_counts.merge(df)\n",
    "\n",
    "        case_control_sv_counts[\"midpoint\"] = np.mean(\n",
    "            case_control_sv_counts[[\"start\", \"end\"]], axis=1\n",
    "        )\n",
    "\n",
    "        # and total up the sample counts\n",
    "        case_control_sample_counts = pd.DataFrame(case_control_samples_with_svs).T\n",
    "        \n",
    "        ######################################\n",
    "        #### CALCULATE ALLELE FREQUENCIES ####\n",
    "        ######################################\n",
    "        \n",
    "        # Remove CNVs and calculate separately if present\n",
    "        if case_control_sv_counts['name'].str.contains('_CNV_').sum() > 0:\n",
    "            cnv_counts = case_control_sv_counts[case_control_sv_counts['name'].str.contains('CNV')].copy()\n",
    "            \n",
    "            for label in ['cases', 'controls']:\n",
    "                columns = [col for col in cnv_counts.columns if col.startswith(label)]\n",
    "                genotypes = [int(c.split('_')[-1]) for c in columns]\n",
    "                alleles = (cnv_counts[columns] * np.array(genotypes))\n",
    "                average_cn = alleles.sum(axis = 1) / cnv_counts[columns].sum(axis = 1)\n",
    "\n",
    "                # overwrite the value with cases and controls\n",
    "                cnv_counts[f'average_CN_{label}'] = average_cn\n",
    "            \n",
    "            # export\n",
    "            cnv_counts.to_csv(file_prefix + f'{cohort}-cnv-genotypes.csv', index=False)\n",
    "        \n",
    "            # and remove any trace of CNVs from our genotypes\n",
    "            cnv_columns = [c for c in case_control_sv_counts.columns if '_' in c and c.split('_')[-1] not in ['0', '1', '2']]\n",
    "            case_control_sv_counts = case_control_sv_counts.drop(index = cnv_counts.index, columns = cnv_columns).reset_index(drop = True)\n",
    "            \n",
    "        # now we do our SVs. We calculate their allele fraction\n",
    "        for label in [\"cases\", \"controls\"]:\n",
    "\n",
    "            # a quick check to see if there are non-0/1/2 genotypes\n",
    "            columns = [col for col in case_control_sv_counts.columns if col.startswith(label)]\n",
    "            assert columns == [f\"{label}_{i}\" for i in range(3)]\n",
    "\n",
    "            alt = (case_control_sv_counts[columns] * np.array([0, 1, 2])).sum(axis=1)\n",
    "            ref = case_control_sv_counts[columns].sum(axis=1) * 2\n",
    "            case_control_sv_counts[f\"AF_{label}\"] = alt / ref\n",
    "            \n",
    "        # export\n",
    "        case_control_sv_counts.to_csv(file_prefix + f'{cohort}-sv-genotypes.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "324aacdf-0865-4bee-bce2-fc8b9f7f236a",
   "metadata": {},
   "source": [
    "# Calculate associations\n",
    "\n",
    "Up next, we'll calculate the different statistical tests, e.g. individual SV tests and burden testing.\n",
    "\n",
    "We begin by doing burden testing. We do it in a few different ways:\n",
    "\n",
    "1. `logit(case_status) ~ count_of_rare/common_SVs + ancestry_PCs + sex`\n",
    "2. `logit(case_status) ~ count_of_rare/common_SVs + sex`\n",
    "3. `logit(case_status) ~ count_of_rare/common_SVs`\n",
    "\n",
    "We store these all in the same file, and we calculate for both discovery and validation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3e750628-49f7-453e-afe8-cd83044ce0a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA24AAAE0CAYAAABU22DfAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/NK7nSAAAACXBIWXMAAA9hAAAPYQGoP6dpAAB5xElEQVR4nO3deVhU9f4H8PfMAMM6oIAwIIuKiuIWiqKCQJrmkqapmbuZuXAr96W6N7VMK0W9mTdzL60sy6VSs3JJVHC3BNxQURBUFEG2EZjz+4Mfk+MM+6zM+/U8PI+c8z3nfM545sv5nPNdRIIgCCAiIiIiIiKTJTZ2AERERERERFQxJm5EREREREQmjokbERERERGRiWPiRkREREREZOKYuBEREREREZk4Jm5EREREREQmjokbERERERGRiWPiRkREREREZOKYuBEREREREZk4Jm5EZPaKiorwyiuvIDw8HM8//zwyMzONHRIRERGRTjFxIyKzt23bNnh7e+PIkSMYNmwYPvroI2OHRERERKRTVsYOgIiotkaOHIlhw4YBAFJTU1GvXj0jR0RERESkW0zciKhOsLKyQt++fXHy5En89ttvxg6HiIiISKdEgiAIxg6CiEhXrl69ir59++LSpUvGDoWIiIhIZ9jHjYjM3hdffIFVq1YBABwdHSEWs2ojIiKiuoV3N0RkchQKBVq1aoVDhw6plhUWFmL8+PFwcXGBXC7HsmXLVOuGDh2K/fv3IyIiAkOGDMEXX3xhhKiJiIiI9Id93J6Qn5+PixcvIjAwEPb29sYOh8giFRYWYvjw4UhISFBbPmvWLJw6dQoHDhxASkoKxowZAz8/PwwePBguLi7YvXt3lY+Rnp6O9PR0jeUFBQXIzc1FeHg46wAiC8T7ACIyZUzcnnDx4kW0b98ep0+fRnBwsLHD0bugoCAA0LhBJjKWxMREDB8+HE93vc3Ly8O6deuwd+9eBAcHIzg4GAkJCVi1ahUGDx5c7eOsWbMGCxYsKHe9pdQBVDnWk5aF9wFEZMqYuFmwmJgYY4dApObw4cOIiorCokWL4ODgoFp+/vx5FBUVoUuXLqplYWFhWLRoEZRKZbX7tE2cOBH9+/fXWJ6UlISRI0fW/ASozmE9SXUZr28i82LxiZtCoYBCoQAA5ObmGjkaw+rVq5exQyBSM3nyZK3L09PT4ebmBhsbG9UyDw8PFBYW4v79+3B3d6/WceRyOeRyea1iJcvAepLqMl7fRObF4gcnWbx4MZydneHs7IyIiAhjh0NEWuTn50MqlaotK/u97MFLTSkUCuTk5CAnJ8fiHt4QERGR+bD4xG3evHnIzs5GdnY2Dh8+bOxwDCooKEjVvp3IlNna2mokaGW/13YAAT68oYqwnqydikaDfdrZs2fRqVMn2NvbIyQkBKdPn1Zb7+LiApFIpPZT9rClOsehf/D6JjIvFp+4SaVSyGQyyGQyODo6GjscItLC29sbmZmZKC4uVi3LyMiAnZ0dXFxcarVvS354Q6RvT44Gu3r1aixYsADbt2/XKJeXl4c+ffogPDwcp0+fRpcuXdC3b1/k5eUBANLS0pCdnY3k5GTVqLDp6emqvrBVPQ4RkTljHzcL7uNGZC7atWsHa2trxMXFISwsDAAQGxuLkJCQWk+2LZVKVc0u+fCGSHeqMxrstm3bYGdnh08++QQikQgrVqzAnj178P3332Ps2LFISkqCXC5H48aNa3UcIiJzZvFv3NhMisj02dvbY8yYMZg0aRJOnjyJnTt3YunSpXjrrbdqvW/2cSPSj/JGg42Pj4dSqVQrW/ZQRiQSAQBEIhG6du2K48ePAyidKqRZs2a1Pg4RkTmz+MSNzaSIzENMTAzat2+PqKgoREdHY8GCBRg0aFCt98uHN0T6UdlosE+X9fLyUlvm4eGB1NRUAKVTdeTn5yMyMhJyuRx9+vTB5cuXq32cJ4935swZjZ+kpCSdnDsRkT5YfFNJNpMiMk1PT8Jtb2+PzZs3Y/PmzTo9zrx58zB9+nQAwLlz55i8EelIdUaDLa9sWbmLFy/iwYMH+PDDDyGTyfDRRx+he/fuSExMrNGos2vWrMGCBQtqfnLVUPy4GJk3s+AZ4I4b51KRmnAbXV4JgVgsMsjxiajusPjEzRjmh/8z4eX8I9ONFseuXbuMdmwiU8GHN1QRfdaT88NjjPo3QN+qMxpseWXLyu3btw9FRUWq7+jWrVvh4+ODn376qUajzk6cOBH9+/fXWJ6UlISRI0dW9RSr5I8vjuL4ttMY+98hOPLVCSSfTEGLbk3h6lNPp8epCd4HEJkXi0/cDDk4yZMJmykICAgwdghERscBiqgi+qgnTe1vgb48ORqslVXp7UZ5o8F6e3sjIyNDbVlGRgbkcjkA9QcsQGmi16hRI6SlpaFr165VPk4ZuVyu2re+NQ1thHN7E3Bkywl0HRECnzbe5SZtl49fg5tvfdT3djFIbLwPIDIvFt/HzZL7t1y9ehVXr141dhhERmXJdQBVTpf15PzwGItJ2gD10WDLlDcabGhoKI4dO6ZqIi0IAo4ePYrQ0FAIgoAmTZpg06ZNqvJ5eXm4cuUKAgMDq3UcQ8nLyseD2w8BAI07+ELWwAmCUsDhjcdxaP0xZN99pLHNg9sP8fXsnfh56e8Gi5P3AUTmxeITN0senGTAgAEYMGCAscMgMipLrgOocqwna66y0WAzMjJQUFAAABg8eDAePnyIqVOnIjExEVOnTkVeXh6GDh0KkUiEvn374r333sOhQ4eQkJCAUaNGoWHDhujTp49eR52tqa2zd2D16C/xuKAIADB54yiMXj4Y4aM7ofvEMMjcNZtl15M7o8ekMISP7mSwOHl9E5kXi28qaez+LWVPX+tyPwciU2bsOoAsW13/GxATE4PJkycjKioKzs7OaqPByuVybNy4EWPHjoVMJsPPP/+MSZMm4YsvvkCbNm2wZ88e1QTbH3/8MaytrTF8+HBkZ2fj2WefxZ49eyCRSCo9jjG07dUS91OzYG2rfpsV0NEfAR39tW4jEokQNqKjAaIjInNl8YmbqajrndSJiMjyVDQa7NMjx3bs2BFnzpzRuh9bW1ssW7YMy5Ytq/ZxjKHT4GeqXLYwT4Hf/3cErXu2gF8bbz1GRUTmjokbEVk0Dk5ChmBJfdss0f1bWdiz4gA8A9zx3ORuWss8up+Hw5viUFJcggFzeqqWZ1y5h1O7/gIAJm5EVCGLT9wMcdPGP9hEpmvx4sUGm8+JiOqmb9/ejXs37iM3qwDPTVZfl3H1Hv78Mg6JB6+olrn51kPXV0KgVArwb9cQYz8dCltHKdKv3sPhjcfRrHMjBPdrbeCzICJTZ/GJG2/aiCwbJ+AmotrqNqYT7t/MQpdXOqgtz7z5ANvn/4LMlAdqy39bfQTXTt/EtRMpCAhthKHv98PiXp9BqVQCAnDvxn0mbkSkweITN0u+aUtISDB2CERGx8FJqCKsJ6kqWvcIROLhKzj+3WlcOX4daUkZaP9Ca1w+fg05d59ozSMC8P9d+5LjUwAAV45fx4XfL8Er0AP52QXIzy6Ag4udQeLm9U1kXiw+ceNNGxEREVVHSbESsVtOoFGwD3z/v1/ajwv3oPhxCSTWYghKQdVvTY2A0omYlP8scnJ3xIWDl9EsrDHuJmci48o9eLf4Z3JwQRDw87I/8PdvFzFx3YhyJ+8morrP4hM3U2LoYaF//fVXAECvXr0McjwiU8TBSagita0nq9rHua5PC1DXZKbcx8H1x3D7UhNV4jZ4YT+c/OEckk+mVLyxUv3XR/dy8eheLpLjbzyx/we4duYmOg5sBwA4vesv2DnZQmyl2+l3eR9AZF6YuFmwsiaibCpBloz9XKkirCdJG48m7nhlyQA4e8jwQff/oqS4BA1bynHrwm2dHePOlXv49dNDeFxQhIZBnug7ozuOfBmPlL/S0Pq5QESO7VzrY/D6JjIvun10Q0RkZubNm4fs7GxkZ2fj8OHDxg6HiEzYw4wcPLqfBwBo3rUJLh65iuLHxRCUAgpzC3V+vCJFMQDA3d8NRYXFOPPzBdy/mYVD64/jUWbFLQRuX7qDs3su4NLRazqPi4iMw+LfuLGZFJFlYz9XIqoKZYkSq0dvhqyBE/61ZSwAwMnVAb5tvHHzrzTcu/Gg4h3UgKAU4OTmgLO/XIBYIsKQhf3wuOAx7GR2cHR1qHDbDdHbUFJUAoiAf//xFsQSPquvy3Kz8mHrYAMrG4u/ta/TLP5/l82kiIiIqDJiiRgdBrSBQz17AKUTav/0ye/waeUFiY0EJY9LIPNwQs6dRzo9rntjV7SIbIrmXZvg4PpjeJSZi3GrhmH54HXwat4AbXq2QEBoI1w5fh1isQhegZ5w9nBC1PjOSEu6g5aRTZm01XGP7udh5dB1aNq5MV7+4AVjh0N6ZPGJm76nA+Dk20RERHVDz+h/7hGcXB3w3JRuEIlFaBnVDMe+PqHzpA0AIsZ2xs5F+5B8IgX3b2ZBJAbWTdyK4sdK5OcU4rt//4xnJ3TFwfXHIBKLYGUtwZw9U/Ag9SESD15Gx0HtdB4TmRZbRxs0DvGDXztvY4dCembxiZspNpMy1OhiTZo00ev+iYjMHetJqkiXYe3xQY//ls67JhLp5Rgbp2xT+10QgNwH+RBLxLB1lKJRex+c2nUeUnsbNAnxg0N9e0isJGjbqyXEVmLImzUod9/Vvb4FQUDBo0LYywwzzxxVjbXUGsOXvGjsMMgALD5xs2S7d+82dghERCaN9SRVRCQSIXxUJ8R9dxqFjxSGOagANOnoi+QTN3H56DWIxCIISgEikQihQ4Ph08oLAODbxls1VcHTti/4BZkpWdjx405InphiICczF6kXbqNlZDOt2/25OR6HNh7Hq5+9rDpOmbSkDKQmpqPjoHYQ6SmJJbJ0bPRMRBZNoVAgJycHOTk5HKCIiKqtYUtPwyVtACTWEiSfuKn6XYAAABi1fBAyrt5D+uW7le6jWFGMIkUR8P/blvn98yP47t8/4+ZfaVq3c/evD6/mHnCsrzkwyoG1R7F3xUFk3syqxtnox+oxX2JD9LbKCxKZGb5xs2CfffYZACA6OtrIkRAZDwcooorUtJ5k/2bLIW/mYdDj1fdxxsPbOaVTBQhQTej9S8xB3L/5AAGd/DFkYT/sXrIfVlIJXpjdE8WKYhz75hRS/kqDg4sdhn04AIDm9R06JBj1vJzhFaj9nFpGNiv3bdzzb0Yi4+o9uPvV1+0J14C9sy2kDlJjh0Gkc0zcLNjq1asBMHEjy6bvAYrIvLGepIqkXczA+b2JCAj1x9W4GwY55r1r/0w7UNZMEgDu33yAZ/oGofPL7XFyx3kkHLwMAAgb2QkZV+7izy/jVdsp8h9Dam+jdn0/LijC3WuZaPt8EDKu3kPDlvJqxeXu7wp3f9fanp5OjP3v0ArXK/If49LRZLTo1hTWUt4Kk/ng1WrC5ofH6H2AEiJLZ4oDFBGRefjr1ySc+PFchQOA6JNIIoJIBEgdpCjIKcTZXxKQfuVu6WApADq82AbufvVR39sZypLnIbYSw9ZBCqm9jWofgiBAEASc+flv7PvvIbj51UfmzQd469vxqOflbJTz0rdTO8/jt/8dQd/pjxEysK2xwyGqMvZxIyIiIr0oLCzE+PHj4eLiArlcjmXLlpVb9uzZs+jUqRPs7e0REhKC06dPq9YJgoCPPvoIjRo1gkwmQ/fu3ZGYmKi2rUgkUvvp0KGDXs8NAKLGd8GwxQNw51qm3o+ljbJICWWJgNbPBUJiLYZ7Y1dkXL6HB6nZEEvEuPhnMkqKlRCJxWjbqyVadw9E09BGqu1LikqQmZKF31b/iVbdmyNibCjCRoQgdEgwMq7exbm9CTj6zSnEbj1plPPTl9bPBSJsZEe0iGxq7FCIqsXi37gpFAooFKWdijkwARERGVNda2kxa9YsnDp1CgcOHEBKSgrGjBkDPz8/DB48WK1cXl4e+vTpgxEjRmDTpk34/PPP0bdvXyQnJ8PBwQFr1qzB0qVLsXHjRjRr1gwff/wxevfujaSkJNjb2yMxMRHt2rXD3r17Vfu0trbW23ndu3EfX8/dic4vt0eDRq5QFiv1dqxKiQFbRyle/rA/Dm+KAwBk3c4GAESM64RPh28ARCI4uNih7/Tu8Gr+T/81kVgEKxsJXH3rw7G+A6LGd1GtWzF0PbLv5sBOZgsIQNiIEI1DJ59MwYO0hwh50bzeWsncndBjYpixwyCqNotP3DgwARERke7l5eVh3bp12Lt3L4KDgxEcHIyEhASsWrVKI3Hbtm0b7Ozs8Mknn0AkEmHFihXYs2cPvv/+e4wdOxabNm3CzJkz0a9fPwDA//73P9SrVw9Hjx7Fc889h6SkJLRo0QKenp4GObcLf1xCVlo2zu1JQKsezQ1yzHIpgaQ/r+LPzaV92CRWYrh4OeP+zSzsX30EHk3ckHohHQ9vZ+OXZb/jpfl9sXfFQXR5pQPEEjHqeTnDRS7Dl9O244VZz6maRw5Z0BeFeQrUkztDEASth/7tf0eQcfUugp5txrndiAzA4ptKzps3D9nZ2cjOzsbhw4eNHY5B9ezZEz179jR2GEREJov1ZM2dP38eRUVF6NLln7c4YWFhiI+Ph1Kp/oYqLi4OYWFhqvm/RCIRunbtiuPHjwMAli5dihEjRqjKi0QiCIKA7OzSN0uJiYlo1kz7aIf60G1MJ/Sd0R3Nw5rg4NpjAAArqcRgx3+a8MSw/la2Vsi5mwvH+g4QiUVITUgHAIglIqQl3cG383bjyvHrSDmXivDQbogIi0RaYgaunbqJlPOpqv14t/CEtY0VHOs7wLVhPQBAakK6an8A8OLbvdB/9nM48tUJ5GXlG+hsdUuR/xjx288iP7vA2KEQVcri37hZ8sAEy5cvN3YIREQmjfVkzaWnp8PNzQ02Nv8MhOHh4YHCwkLcv38f7u7uamWDgoLUtvfw8MCFCxcAlCZ8T1q3bh2Ki4tVy5OSkqBUKtG6dWtkZ2ejd+/e+OSTTyCTycqNLT09XWN5UlJSlc5NYiVByItt8eeX8ZBYS9A8rAkSD12p0rb6kHm9dKRJK6kV2vRsgfzsQgyZ3xcrh23A47zHCB7QCmd2lX6Wd69l4uVF/eHfriGOfHUC9i52kDf3gJ2zLfZ9egju/q7wbuGJa6dv4sup29FpyDPo/WYUjn17GgfWxkJiY4V5e0tHWfUMcMfFI1dx/NvT8GjijnbPtzTWR1Bjf/92EXtXHsTjgiKEj+po7HCIKmTxiRsRWTb2cyXSj/z8fNWD0TJlv5d95yor+3Q5AIiPj8eMGTMwa9YseHp6oqioCMnJyWjUqBE2btyIrKwsTJs2DaNGjcKuXbu0xrZmzZoad5NQKgX8uTkODYPk6Da6E7qN7oRPh2+EssSI/dz+X7GiGIkHryAvKx+ZKQ9QkFP6FinlbBr6zuyBPz4/gsJcBZp2boTfVv+JkqISPLqXi1+W/Y4mHfyQdPgKts7egdk/TYZDfXs4ujrAwbm0CeT5fQkQBKDXv7qpHbOsn1/zsCYa8Vw6mowGjd1QT+6M5BMp2LPyAF58uxd8grzUyilLlEg+lQL/dg1hLdVf30RtWnVvjscFj9G2l/klnWR5mLhZsGnTpgHgE2WybOznShVhPVlztra2GolX2e/29vZVKvt0uePHj6N3797o3bs3Fi5cCKB0EJLMzEzY2dmpBiTZvHkzOnTogNu3b8PLSz1JAICJEyeif//+GsuTkpIwcuTICs8rOyMHhzYcR+MOvgjo6A8A6DY2FHeuZSL++9MoeWy8BE4sEauaLN65eg9AadPJRs/4IGRAG7g2dEFBTiGsrCVo3MEPiz9dBEEpoJ/sJVz44xKadWmEVj0CAQDFhcXIy8pHYd5jAMDo5YNRpCiGi6f6W0ypvQ1s7G2wYuh6DPp3bzR6xgcAcPf6fXwzdxeadW2M+l4uOLX7LxQripGXpdkk8cIfl/Dj+3sRNb4LIsaG6u3z0cbWUYouw2o/AunjwiJYWUsgllh8LyTSIyZuFmz//v3GDoHI6DgBN1WE9WTNeXt7IzMzE8XFxbCyKr3dyMjIgJ2dHVxcXDTKZmRkqC3LyMiAXP7PJNCHDh1Cv3790LNnT3zzzTcQi/+5QX66SWSLFi0AAGlpaVoTN7lcrrbv6qjn5YzRywejfsPSc3iUmYt9Kw+iIKewRvvTJbW3fmIAAvDqqpdVI0k2bu+rWt28a2NcvpMIpVKJ7t59YW1njRYRzdDmudLPzruFJ6b9MAGO9UqTZ4d66kl0mSJFER5l5iHvQT4e5xeplrv51kPUa13g/4wPbp5Pg4unDCM+GYh6cs254RoF+6DDgDZoqcPh+XPuPYJDPXtIrPTf9zDvYQH+O2w9moY2wuD5ffV+PLJcfCxARBZNKpVCJpNBJpNZXD9XIn1q164drK2tERcXp1oWGxuLkJAQtaQLAEJDQ3Hs2DHV6IWCIODo0aMIDS19+3LhwgX0798fvXv3xnfffac21H9iYiKcnJxw/fp11bJz587BysoKAQEBOj+vIkURdn+8H7/97wgA4Ou5u1CQUwg7Z1udH6uqRBIRxBIRxFZPfK5KwMnVARuityHu+zNat3N0tYetky0ad/DF0IX98Eyf0n6GivzH+PzVLYj77kylb5C2vfMT9qw4gJm7J6J518aq5WKJGBFjQuHXxhvhozriX1vGak3aAMDJzRG9p0bhj7VH8eeX8dU8e013r9/H8iHrsHfFwVrvqzKCIEBQCvAIcIe7v6vej0eWjYmbnswPj8H88Bhjh0FERGQU9vb2GDNmDCZNmoSTJ09i586dWLp0Kd566y0ApW/UCgpKm80NHjwYDx8+xNSpU5GYmIipU6ciLy8PQ4cOBVDatNHHxwcxMTHIzMxERkaGavvAwEAEBARgwoQJuHDhAmJjYzFhwgRMmDAB9erV08OZiUqbB9qVJo/52aVNE1v3CISV1DgNmYQSAcoSAY717GFjbwPH+vZo1aM58h4WoFhRjPO/JuJhRo7GdlIHKaylVrh26ibup2aplitLlMi59wi5D/LKPaayRIn0y3dRkFMI/7YNYWNnU27ZqihSFOPaqZu4cS618sKVcHS1R+P2vvBt413rfVVm/2d/YsWQteg/u6fBm3mS5WHiRkRERHoRExOD9u3bIyoqCtHR0ViwYAEGDRoEoLS54rZt2wCUNnX8+eefceTIEbRv3x5xcXHYs2cPHBwckJGRgWPHjiExMRG+vr6qZo5l24vFYuzevRsymQzh4eEYMGAAunfvrrd+idZSK0zeNBoD5pZOE9Hj9XAAwMkfz8OxvvbmhIbiLHdCYHgT5GblI2JsZ4z7dCgkNhKkX7qLo1+fBACc/zUBP33yGwoelTbttLa1QtCzzdWaUdo52WLmzokY9G5vrccpeFSIpS+uwY/v70FaUgba928NK+vaNUm0dZBi6nfj8cpizb6H1WUvs8OoZS+hTc8Wtd5XZep7u8Dd3xVSh9olrkRVwT5uREREpBf29vbYvHkzNm/erLHu6UmdO3bsiDNnNJv0eXp6ljsBdBkfHx/8+OOPtQu2hlo/F4jDX8bhYXq20ecCa/d8EJqGNkLEmFCc25uABo1cIZSUfnaNQ/wAADs/3A9BKcDNr7RZn8RKgiELNPtlVdREUmIlgbOHDPLmDSCWiKHIf6yT+O2d9TeJ96P7eci4chdNQxvpdL8hA9siZGBbne6TqDxM3CzYlClTjB0CEZFJYz1JVfGvr8Zi27s/Iemw8eZyc3J3xC/LDsDGzhoOrva4n5IFiAAIgLWdFX5csAdvfTceXYa1x/lfk1CQU4Du7Z5H/sMCnN2ToOrfVqa4qARfz94Br0BP9JioPo+ejZ01Jq4bgTvJ9/D5LxeQfvmu1uH0ix8X49dVh9G4gx9adNN9f8OKPC4oQmpiOhoF+0AkEuHXTw/hwh+XMGHtcHgHeho0FiJdYeJm4sr6yc0/Ml3n+46Ojtb5PomI6hLWk1SRw5vjYGNrjZBB7VD8uBj2znZ4XPgYVjYSPM4vNujcbo/ulc5DWZirQGFu6dQKDVt6QoAIaQmlk41fPn4dJUVK5N7Pw7k9CXB/6AtliRIXj1zVSNxKikqQfvkuJBU0gfRo4o6ZOybCTqZ9YJbsu49wcud5PMrM1Unilp9TgMcFRXDx0D6x+pMObTyOY9+cwvCPXkSzLo3RcVA7uMid4dHYrdZxEBmLxSdunHyXiIh0iQNTWY5j35yCraMU8T+chb2LHWzsbaDIf4zC3MeAAPg/0xCpSRkoLizWWwxiiViVIHoFeiDj6j0oi5WQeTgiMyULLSKbIi0hHfb17BEU1Qy3/k4DADz7ehjkzRrA2tYKeVn5+GX5H4gc1wUOLqXNFaX2Npi2fUKFiRtQ/jQBAODasB5eXztCY+63mto6cwfuXs/ErN2TVYPDlKdV9+ZQ5CnQMKh02gffNt4GGayESJ8sPnGz5Ml3yyYf3b17t5EjISIyTawnqSKvrx0BsZUYe1cehL2zHfzaNsTZPRdg72yPi7FXcONsKkRikV6ObWNvDWWxEnbOdoga3wXx35+Bo6sDWvnWw+Xj15GfVYjix8W4f/MBAkL98dzkbpDa26DzsPalo0FmF2DC1FcBABOjpuLkj+fh29oLrXv8M6BHZclR8skU2DpJK2x6WDaHXG38/ftF7FlxAG2eawHP/082K+PV3ANes56r9bGJTInFJ26WPPlucnKysUMgIjJpxqgn9dlEnnTL1ad0uoHhS14EUDpE/r7/HoRDPXv4Bnnh5t+3IZaIIIhFUBbrttlk2WTXRQWP0bClJ3YnZ+JOcmbpShHQY1I3ZFy+A1tHKU7t+gvhozoBAOyd7ZGZ8gDXTt1UXd+iqNLN7iRnIunPn9FvRvdKBwopflyMLTN/RD0vZ4SNCIGdzK7KzSH/WHsUyuISPDe5W5XKK0uUKClSIujZ5vBtrTmhOpGlsPjETSqVQiqVAgAn3yUiIqIaE0vEGPbhAHw5bTvyHuRDLBHDzsUO8oAGuHL8euU7qAZrWysUFRajMPcxts/fo1pu6ySFraMUoUOegZW1BEWKInR4sS08A9wBAI717PHmtvGwdZTiwz3vAgC6DO8AF7kzHqRlIfHgZXQeGlxp4mZlY4X+s5/DjXOp2P3Rb5A1cKpy4vbXr4koLqp64ta2V0utg58QWRqLT9yIyLKxnysR6ZK8WQM07dwIgeEBKH5cjD0xB+ASJoOtkxRiiRj5D2s/ZYBYIoJ7I1fcTroDAKjvLcPda6Vv2+o3dMHrX4wAACT9eQW3/r6N7hPD1baXuTkidutJPEh9CBe5DC4eMnQZ1h5FimJ0HNQODRpVbQCPZ/q2wv3Uh7CWWqHv9GerHP+EL4ajkhkeiEgLTsBNRBZt8eLFcHZ2hrOzs0U1lSYi/bB1lGL4khfh364h9iw/AL92DSFr4ARliVIn87zZyWwhALiddAdiSWn/uRvn0iC2Kr2l82ruiew7j5B95xF2Ld6PY9tOI/d+nsZ+CnIKoCxRqiVQ1lKrKidtZXpMDMPbv72B5l2bVHkbx/oOcHJ1qNZxquvCH5fwxYStyErP1utxiAyJiRsRWbR58+YhOzsb2dnZOHz4sLHDIaI64Ou5OxG79SQ6DmqHzJsPcPaXC5Da2+CZvq0gktR8sBInNwcUKYpUk2qLRCLYu9jB3rn0RyQCTu08jx8W7kFO5iMU5irQrncQnD2cAJT2Syvz3ORucPOrD4lV7W8FRSL9DMBSG5k3HyD9yl2dvOEkMhVsKmkm9NFZPSaGQ1YTsZ8rVaQ69SSnASAAUCoFZFy5C1kDJ4xfPQx3r2XC2UOGge88j9itJ1VJV3lsZVIU5iggEgPCE+OZSB1sILGWoFhRAmcPJ2TfeYSSYiXyHxYg/2EBOgxsi+QTKchKe4jgF1rDJ8gLM3a8Dof6pW+29n92GCd2nMfkTaPg2rB0UJW6fB8QMTYUnQY/Azsn7XPMEZkjJm4WrFevXsYOgYjIpLGepOoSi0V44+tXUfCoEHlZ+bCT2eLS0WRsmfkjXl70AqztrLF3+QE4uTpAHuiBy0evqbZ1b1wf9649QIcBbeDqWx+/fnpItU6R9xiCsjTpa9urBU7sOA8IAMRAYY4C10/dxOD5ffDoXi4Cw0sHCdmxaB9Kiksw7tOX4ewhg2tDF9jY2aj2WZevb5FIxKSN6hwmbkREREQ6ZC21wmejvkXx42IoSwSUFCshdZDCWmqNToPa4dD6Yyh6XIz2L7TG5aPXYONgjaLCYty79gDu/vVxavdfaPSMj8Z+HxcUQR7ogduX78KvbUNcibuOVs82h53MFiED20KpFNCovS8AoKS4BNdO3YRILIJSKaDT4GfQafAzhv4o1ORl5UMQBDjW12//NqK6iombBQsKCgIAJCQkGDkSIiLTxHqSaiooqhmUJQIixoVCYiWGtfSfyaw7DX4Gd67ew/b5v8Da1hryph7oOrwDLv55FW5+9fEwPQdZGQ8BACKxCA2auMK5gQzXT99Czt1HSL94B6+ufhmR40Kx9vVv4NeuIW6cTcXPS3+Hm399/OursZBYSTDik4GwtrWG+IlJwIuLSmBlLSmN0cDX99qJX6NIUYxZuyYZ5HhEdQ0TNyIiIiIdq2iOsshxnZF4+Aqun7kFGztrPMrMRbPOjbFr8X5c+OMSGof44sqxGwAAmbsjhBKgQWM3lBSVwMnVAV1e6aAa/XHoBy/ghwW/oL63M2ydpGj8/2/cAKBpaCO14yb9eRXf/fsnDJ7fF0FRzXR/0pVo3SMQxY9LDH5corqCiRsRERGRgbWMaIqWEU2R9zAfRYrS0R4V+Y9RrCjGxT+TVeUCw5sg/odzAIApm0dr7Mfdrz7sZLbwaOKOF2Y9V+Ex7WS2cPGUwU5Wed+vjKv3cH5fIiLGhsLWUVqNMytf99fDdLIfIkvFxE3HOKoYERERAUCRohgJBy4hMDxALfm5k3wPB9YdQ/fXw7Dv00O4eT4VglJAYERTJPxxCTZ21rC2tUKLyOboPLQ9kv68ipaRTbUew9lThuk/vA4AuHfjPn773xFEvdYF8qYNNMr6t2uIt7aNr1LsZ/dcQPz3Z9Eo2AfNujSu/skTkc5xHjczw8SQiMgyzA+PYZ1v5v7+LQk7P/wV8T+cVVt+8+/buBSbjNTEdDQK9oG8uQfsZLZo3qUxAsMDMGLZIBQ8UuDW32lwkTtj+g+vI3JcZ439//llPD7s9SlSE9MBAGlJGbh87BpuXbhd69gjx3XGK0sGoGnnRpUXNnHKEmXlhYjMQJ1541ZYWIjRo0fjzp07UCgUWLFiBUJDQ40dFhERkcUqLCxEdHQ0fvjhB9jZ2WHmzJmYMWOG1rJnz57FpEmT8PfffyMoKAiff/452rdvr1r/zTff4N1330V6ejp69eqFtWvXws2ttJ+XIAiYN28e1q9fj5KSErz22mtYsmQJxOKaPZ8ePHgw7Ozs1Jbt2rULAQEBuHr1KgYMGKB1u7JBPn799VdMnz4dglJAfk4Bti1eC/HHYjRp0gS7d++Gs4cTHrfNwqjZL6tt/+mYjyAX+6Jxh5WYsGY4Fi6dj6CgJRrHmTJlCqKjo+Hk6oCv/1qH1aEfw9WndG624scl+P7d9Yixj1EN9182CMnTHj9+DBsbm2qd09PKzgkAPvvsM6xevVqjTM+ePbF8+XIAwLRp07B///5yzwkA+vfvj+TkZI0yMTGVn9PT/0+CUsD91IeQ2lvDyc2xTpyTNtU5JzJfdSZx27BhAwIDA/Hdd9/h0qVLGDt2LI4fP27ssEzarl27jB0CEZFJYz1ZO7NmzcKpU6dw4MABpKSkYMyYMfDz88PgwYPVyuXl5aFPnz4YMWIENm3ahM8//xx9+/ZFcnIyHBwccOLECYwfPx6ff/452rVrhzfffBNjx47Fzz//DKD05vfrr7/Gjh07UFRUhJEjR6JBgwaYOXOmMU5bRSQWwcHFXm3Z48IifD1nJ64W3PhnoVDav60gpwDOzZwQ/EJrWFlLYOtQcd+yZ/q2gqyBI/IyH6mWWdlIqhzf559/Dj8/vyqXN0siQGIlhljCRmZk/kSCIAjGDkIXHj16BJFIBEdHRyQkJGDKlCk4fPhwtfZx5swZtG/fHqdPn0ZwcHCN4jBEs5b5RzSfpBBR7emiDiDLpa/631zr/Ly8PLi5uWHv3r2IjIwEAHzwwQf4/fffcejQIbWyGzZswAcffIDk5GSIRCIIgoBmzZrhnXfewdixYzF69GiIxWJs2rQJAHDr1i34+fkhOTkZjRo1gq+vLxYuXIixY8cCALZs2YJ3330XN27cqFbMhqoDzv5yAU5ujgjo5A8A2P/ZYRz79jSadPRD6NBgNO3UCPdvZeH4d2fQeWgwbOz+eVtERJarzjx+cHJygqOjI+7du4dRo0bhP//5j7FDMnlXr17F1atXjR0GEZHJYj1Zc+fPn0dRURG6dOmiWhYWFob4+Hgolep9juLi4hAWFgaRqHS+MZFIhK5du6pazsTFxaFbt3+G1/fx8YGvry/i4uJw+/Zt3Lp1S219WFgYUlJSkJ6ers9TrLFn+rZSJW0A0DjED4HhTTDwnefRtFNpn7L47Wdxaud5bIj+Fv99ZQMKHhVW+zi/xPyB1WM243Fhkdb1vL6JzEudaSoJAJcvX8bgwYOxaNEidO/e3djhmLyydtKcWJbqAvZzJX1gPVlz6enpcHNzg42NjWqZh4cHCgsLcf/+fbi7u6uVfbp/j4eHBy5cuKBa7+XlpbE+NTVVlZw9ud7DwwMAkJqaCrlcrjU2bUldUlJSdU9TJwI6+iOgo7/asmf6tULCocvwa9sQYrEINnbW2jeugCLvMQpzFRCU2htX8fomMi91JnG7desW+vfvjy+//BIdO3Y0djhEZGDs50pkWvLz8yGVqvfRKvtdoVBUqWxZuYrW5+fnq+27ouOUWbNmDRYsWFDdU6oxpVJAyrlb8GntDSvryvug5ecUwE5mi1m7JtXquIP+3btW2xORaTHZppIKhQKtWrVSawdfWFiI8ePHw8XFBXK5HMuWLVOte//995Gbm4vZs2cjMjISQ4YMMULURGQso0aNwuzZswEAxcXFak/5icjwbG1tNRKnst/t7e2rVLasXEXrbW1t1fZd0XHKTJw4EadPn9b42bJlS3VPs0rO70vE5re2I3772coLA9g6awc+G7kJjwu0N3EkIstkkm/cCgsLMXz4cI1X9xWNTvXFF19Uef+m1kSCiGrPyckJAFT9XD/55BMjR0Rk2by9vZGZmYni4mJYWZXebmRkZMDOzg4uLi4aZTMyMtSWZWRkqJo5VrTe29tb9bu/v7/q3wC0NpMsW17eOn1oFOyDdr2D0KyKc6K1erY5Mps2gJXUJG/TiMhITK5GSExMxPDhw/H0YJd5eXlYt24d9u7di+DgYAQHByMhIQGrVq3SGFa4MoZuIkFEhsF+rkSmo127drC2tlYNPAIAsbGxCAkJ0ZhfLTQ0FEuWLIEgCKpRJY8ePYp33nlHtT42NlY1auStW7dw69YthIaGwsvLC76+voiNjVUlbrGxsfD19TVoclYRF08ZXny7V5XLd365feWFiMjimFxTycOHDyMqKkqjb0p1RqeqjKGbSBCR/pX1c123bh1eeOEFY4dDZPHs7e0xZswYTJo0CSdPnsTOnTuxdOlSvPXWWwBK34oVFBQAKJ3w+uHDh5g6dSoSExMxdepU5OXlYejQoQCAyZMn46uvvsL69evx119/YfTo0ejXrx8aNWqkWj9nzhwcOnQIhw4dwty5c1XHqQuST6bg8rFrxg5DZ4oURfh52R+4fLzunBORIZjcG7fJkydrXV6d0akqY+gmEqaKo0iRKVMoFGjfvj1WrVqlmgOqsLAQ0dHR+OGHH2BnZ4eZM2dixowZANT7uQKAu7s7vv/+e2OFT3VEVepJQ8zfaa5iYmIwefJkREVFwdnZGQsWLMCgQYMAlP4t3rhxI8aOHQuZTIaff/4ZkyZNwhdffIE2bdpgz549cHBwAAB07twZa9aswX/+8x88ePAAPXv2xNq1a1XHmTVrFu7evYuBAwfCysoK48ePx7Rp04xyzvrw4/t7ochT4N0/dJuMGus+4EFaNk7tPI+CnEI069zYKDEQmSOTS9zKU53RqapDoVCots/Nza15gESkM+znSlQ32NvbY/Pmzdi8ebPGuqe7RHTs2BFnzpwpd19jx45VNZV8mkQiQUxMDGJiTDuJ3vvfg1DkPcaL8zSbTQqCgHN7EuDXriHqe7uorRv4zvO4ez0TDzNy4OIpAwD8+WU8rGwk6DKsgyFC1ymPxm54bc0rcPWpZ+xQiMyK2SRu1RmdqjoWL15ssf3dfv31VwBAr15Vb3dPpG/s50qmhPUk6dK1kzdRmKt9Iu3UhHTsWrIfQc82x5AFfdXWebf0xNY5O3DhwCW8/sUIAMCxb07Bxt66VombMa/vhi3Z8omouswmcavO6FTVMW/ePEyfPh0AcO7cOUREROgiXL0qa5Yz/8j0Wu2n7LzZZJJMSVk/10WLFqmaSQHl93NdtGgRlEqlxmAHFZk4cSL69++vsTwpKQkjR46s3QlQncJ6knTptTWvlDsZtncLTzz/ZiQad/DVWGfrKEW30Z3g7ueqWjZh7XCIJbUbqoDXN5F5MZvErTqjU1WHVCpVNbl0dHTUSaxEVHOG7ufK5tJEZChSe/X5JUuKlbh9KQMNW8ohlogROiRYY5vbl+7AydUBUa92UVvu2pDNDIksjcmNKlmeykanqimFQoGcnBzk5OTwpo3IhOmrn+vixYvh7OwMZ2dns3jjTkTm58a5VPz4/l7kPSxQWx6//QzWT/oWf+3X3r82JzMXa1//Gj++v9cQYRKRiTObxA0oHZ2qffv2iIqKQnR0tNroVDXFmzYi86Cvfq7z5s1DdnY2srOzcfjw4VrFSESkzaXYZPy1Pwl3ku+pLW/cwQ+tujeHb+vSScQvxiYj+84j1XqRSISOL7VDhxfbGjReIjJNJp24CYKgGgYc+Gd0qtzcXKSlpWHq1Km1PgZv2ojMw5P9XMvoop+rVCqFTCaDTCZjc2ki0otnJ3TBhC+Go3F79f5rngHuGDy/L+p5OSP9yl18O28X9n16CACgyH+MT4dvxN3kTARFNVNtU5irQJGiyIDRE5GpMJs+bvpiyX3cmjRpYuwQiKpMX/1c2ceNKsJ6knTBWmoN7xaeFZZp0MgNka92RpMQPwCAlY0VAjr6wbNpA1WZ4qISrBy2Hm4+9TH+f8NqHRevbyLzYvGJmyXftO3evdvYIRBV2ZP9XDdu3Ii0tDQsXboUGzdurNV+LXlKEKpcZfWkISbfnh8eU+tRhMn0SazEiBzXWe33oe+/oFZGLBGjUbAv6ns56+SYvA8gMi8m3VTSEMy5j9v88BiD3DQQmQp99HNlc2kiMkXKEiXWT/4WOz/8VbVMLBZh6MJ+6DEp3IiREZGxWPwbN3Ocx01XPvvsMwBAdHS0kSMh0u7pSbjL+rlu3rxZZ8ew5ObSVDnWk2QsgiDg0f1cSB1sKi9cQ7y+icyLxSdulnzTtnr1agCssMmyWXJzaaoc60kylryHBXD1rYdOg5/R2zF4fROZF4tvKklEls2cm0sTUd2VlfYQyfEpuHEu1dihEJGJsPjEjRNwE1k29nEjIlPk17Yh3vz2Vfi29sK37+xGbla+sUMiIiOz+MSNT9uJLBvncSMiU1Xf2wXn9iTg4p9Xcf9WlrHDISIjs/jEjU/biYiIyFSVvWlz8ZQZORIiMjYOTmLBg5P07NnT2CEQGR0HJ6GKsJ4kYxswtxeybj+EcwMnne+7vOu7SFGETW98D9+23ugVzdZIRKbC4hM3XTK3OdWWL19u7BCIjI4TcFNFWE+Ssbn51oObbz297Lu867ukWImHGTlwkfMtH5EpsfimkkRk2dhcmkg/BEHA3Llz4e7ujvr162P27NlQKpXllr9+/Tp69OgBBwcHtGzZEvv371dbv3HjRgQGBsLR0RGdOnXC0aNHVeuysrIgEonUftzc3PR2bnWdrYMU0398HUMW9DN2KET0BIt/42bJzaSmTZsGgE+UybJZcnNpqhzryZqLiYnB119/jR07dqCoqAgjR45EgwYNMHPmTI2ygiDgxRdfROvWrXHq1Cns3LkTAwcORFJSEnx9fbFv3z5ER0dj7dq16NSpEzZv3ow+ffogKSkJXl5eSExMhKurKy5cuKDap1jMZ9OVqej6lljx8yMyNRafuFlyM6mnn2YSEZE61pM1t3LlSixcuBBhYWEAgI8++gjvvvuu1sTt4MGDSE5OxrFjx+Dg4IAWLVrgjz/+wIYNGzB//nxs2rQJY8aMwYgRIwAA77//Pr777jv88ssvmDBhApKSktCsWTN4enoa9BzNHa9vIvNi8Y9T2EyKiIhIt27fvo1bt26hW7duqmVhYWFISUlBenq6Rvm4uDgEBwfDwcFBrfzx48cBALNnz8b06dM1tsvOzgYAJCYmolmzZro+DZOQeTMLivzHxg6DiEyAxb9xYzMpIstmyc2lifSlLDnz8vJSLfPw8AAApKamQi6Xa5R/smxZ+dTUVABAcHCw2rp9+/bh8uXLePbZZwEASUlJKCoqQseOHZGWlobw8HAsX75c4zhPHk9bApmUlFSd09S7rNvZ+GzUJgSGN8HLH/Q3djhEZGQ1fuP27LPP4o8//ih3/U8//YSgoKCa7p6ITFxdqQMWL14MZ2dnODs7IyKCw14TVVVBQQGuXr2q9afsIUjZg9En/132oORJ+fn5amXLymsrm5ycjLFjx2LEiBGqhO7ixYvIycnB8uXLsW3bNty+fRv9+vVDSUmJ1tjXrFmD9u3ba/yMHDmyZh+Gnji62qP1c4Fo0a2psUMhIhNQ5Tdu+fn5yMzMVP1+6NAhDBw4EE2balYmSqUSe/fuxfXr13UTJREZXV2tA+bNm6dqgnXu3Dkmb0RVFB8fj6ioKK3rPv74YwClSZqtra3q3wBgb2+vUd7W1hb3799XW6ZQKDTKXr58GT169ECTJk2wdu1a1fKEhASIRCLY2dkBALZv3w65XI74+Hh06dJF43gTJ05E//6ab7CSkpJMKnmzllpj0Lu9jR0GEZmIKidueXl5aNeunao9uUgkwtSpUzF16lSt5QVBwHPPPaeTIEk/pkyZYuwQyIzU1TqAzaWpIqwnyxcZGQlBELSuu337NmbPno2MjAz4+/sDADIyMgBAa/NFb29vJCQkqC3LyMhQK5uQkIDu3bujcePG2Lt3rypJAzSTwQYNGsDV1RVpaWla45PL5eU2o7QkvL6JzEuVEzd3d3ds3boVJ06cgCAIWLhwIQYOHIg2bdpolJVIJHB3d8ewYcN0Gqw+WHL/lujoaGOHQGakrtYBRBVhPVkzXl5e8PX1RWxsrCpxi42Nha+vr9aEKTQ0FEuWLEFBQYEqIYuNjVWNSJmeno6ePXuiadOm2Lt3r9pDlpycHPj5+eHHH39UvQFMS0tDZmYmAgMD9Xym5o3XN5F5qdbgJL1790bv3qWv7FNSUjBp0iR06tRJL4EZSl2YDmB+eAzmH9EcbYtI1+piHUBE+jF58mTMmTMHDRs2BADMnTsXM2bMUK2/d+8e7Ozs4OjoiIiICPj4+GDcuHH497//jZ9++gknTpzAxo0bAQAzZ85ESUkJ1q9fj9zcXNWDVkdHR8hkMoSHh2PatGlYu3YtJBIJ3nrrLTz//PNo3bq14U/cgC4dvQYnNwd4NfcwdihEZAA1Hpxk48aNdeKGzZKnA+jfv7/WNv5EVVFX6gCiirCerLlZs2bh5ZdfxsCBAzFkyBCMGjVKNeEzAISEhGDp0qUASt/S79q1C+np6Wjfvj22bNmCHTt2wNfXF4IgYMeOHbhz5w6aN2+uauYol8tV22/evBnBwcHo06cPIiMj4e/vj61btxrlvA2l4FEhvpm7E7sW/1rjffD6JjIvtZoOYN++fdi6dSsyMjK0jtwkEokqHHXOFFhy/5bk5GRjh0Bmri7UAZbcXJoqx3qy5iQSCWJiYhATE6N1/Y0bN9R+DwgI0PoAVSQSIT8/v8Jj1atXDxs2bKhxrMYkCAIyrtyDZ1N3iESiKm9n52SLF2Y/h3py5xofm9c3kXmpceK2evVqvPHGGwBK51p5ehhfMqz54aV/GNlkkgylrtQBdaG5NFF5Tp06hXfeeQfHjh2DlZUVunbtirlz56r6jj1p69atGD16dLlD6JN+nPn5An76+Df0m9UDHfpr9hm+fekOvpy6Hb3+FYFn+rZSW9f+hbrdFJSI1NU4cVuxYgXatm2LvXv3qibVJCLLUVfqAE4HQHVVfHw8IiIi4ODggOeeew6ZmZnYu3cv9u3bh3nz5uH99983dogEwKeVF5qHNYFva6/yC4kAVONtHBHVTTVO3G7duoUVK1aY9Q0bEdVcXakDLLm5NNVt7777Lry9vXH8+HE0aNAAAPDXX39h9OjR+PDDD3Hnzh188cUXRo6SGjRyxSuLB5S73qu5B+bu4eiPRFSLwUmaNGmCO3fu6DIWIjIjrAOITFtcXBymTJmiStoAoE2bNjh69CgiIiKwfv16zuNFRGRGavzGbd68eXjrrbfw0ksvISgoSJcxkYGU12GcqCpYB5AlMOd6UqlUwspK88+8g4MDfvnlFzz33HNYs2YNHBwc8MknnxghQjI2c76+iSxRjRO32NhYODo6om3btmjevDnc3d0hFqu/wDOHEeUsWa9evYwdApkx1gFkCcqrJ8sGhDJlwcHB2LBhAyZNmqQxeJCdnR1++eUXREZGIiYmBkqlEm3aaA6MQXUb7wOIzEuNE7d9+/ZBJBLBx8cH+fn5SElJ0WVcBqOLocDN4Q84ka7VlTqAqK5699130adPHwQGBuLll1/G1KlT4enpqVrv7OyM/fv3o2fPnlixYgVcXFyMFywRkR6NHTsWmzdvLne9h4cHMjIyMHbsWBw6dEhjupIykZGRAIBDhw5VuF8HBwc0atQI48aNUw2Apgs1TtyuX7+usyCMyZKHAi9r3paQkGDkSMgc1ZU6gKgi5lxP9urVCwcOHMDUqVOxdOlSTJgwQaOMu7s7jhw5gilTpmDLli3VmkeMzJ85X99E1eXp6YkdO3ZoXWdjY6Oz/QqCgIyMDKxZswYzZsyAnZ0dJk+eXOP9P6lWE3DXBRwKnMiycQJuqssiIiJw9uxZPHz4EDKZTGsZR0dHfPnll3j77be1ToBNRFQXSKVShIaGGmy//fr1Q+PGjbFx40bjJ27PPvtslcodOHCgpocwCA4FTlQzdaUOsOS37mQ+yprkzz9SsyY35TWDvHv3LlxdXSGRSBAYGIjAwMCahkhEpJUgCLj59208SH2I+g1d4NvayyLe7ltbW8PBwUGn51rjxO3atWsagZSUlCAzMxOFhYXw9/dHq1atah0gEZmmulIH8K071XWffvopvvjiC5w9e1ZjlMmpU6fijz/+wNy5czFt2jQjRUhEddXDjBx8NeMHPEzPhsRKgpLiErjInTFq2Utw8dTeCkCfiouLtS6XSCS1SrCe3G9JSQnS09Px3//+F5cuXcKqVatqvN+n1Xgetxs3buD69etqPzdv3sSjR4+wfft2PHz4EDNnztRZoERkWupKHSCVSiGTySCTyfjWneoUQRAwevRovPXWW7h9+7bWAYQaN24MsViMmTNn4pVXXjFClERUVwmCgK9m/IAHaQ9RUqTE44IilBQp8SDtIbbM+BGCIBg0npSUFFhbW2v9WbZsmc72a2tri0aNGuHnn3/G6tWrdTpfps77uEkkEgwaNAjx8fGYM2cOjh8/rutDEJEJYx1AZBrWrVuHLVu2YMqUKVi6dClsbW01ynzwwQd49913MWnSJHz11Vfo3bs3Ro8ebYRoLZsgCEhNSId3C0+IJTV+pk5kUm7+fRsP03MglKgnaEKJgKz0h7j59234tfE2WDxyuRy7d+/Wus7HxwcAqvTW7ekyT+43KysL77//PpKTk7F582Z07ty5llGr09vgJE2bNsWnn36qr91TOarTD2LXrl36DocsGOsAqgvMuZ5ct24dunXrVmkzHVtbW2zYsAF//fUX1qxZw8TNCM7tTcSuxb+i1xuR6Dw02GDHNefrm0zfg9SHkFiJUVJUorFObCXBg9SHBk3cbGxs0KFDhwrLODg4qAYs00ahUMDV1bXC/Xbt2hUdOnRA7969ceLECTRr1qx2gT9BL491FAoFtmzZggYNGuhj96QjAQEBCAgIMHYYVAexDqC6wpzryYSEBLz44otVKisWizF48GD89ddf+g2KtPJr642gZ5ujcQdfgx7XnK9vMn31G7qgpFgzaQMAZXEJ6jd0MWxAVeDp6Yn79++Xm7ylpqaqzYepjb29PTZt2oRHjx5h3LhxOm0SqvNRJRUKBS5duoSsrCyO1Gbirl69CgCstKlGWAeQJTDnetLKyko1anJVuLm5QSxmMz1jqO/tgiEL+hr8uOZ8fZPp823tBRe5Mx6kPVRrLimSiFBPXjq6pKmJjIzEv//9b+zYsQPDhg1TWxcfH4/U1NQqjardsWNHvP766/j888/x5ZdfYsyYMTqJT6ejSgJQDSn8yiuv6LQzHunegAEDAHDiTaoZ1gFkCcy5nmzatClOnTpV5fInT56Er69h3/iQcZnz9U2mTyQSYdSyl1SjSoqtJFAWl6Ce3AUjYwYZfEoAhUKBuLi4cte3adMGYWFh6N+/P1599VVcvHgR4eHhkEgkOHPmDD7++GOEhYXh5ZdfrtLxPvjgA2zbtg1z587FoEGD4OTkVOtzqHHiduPGjVofnIjMF+sAItM2bNgwvPPOO5g+fTqCgoIqLJuQkICtW7fijTfe0NnxBUHAvHnzsH79epSUlOC1117DkiVLyn2rd/36dUyYMAHHjx+Hn58fVqxYgZ49e6rWt23bVqMp599//41WrVpV+1hEZBgunjL8a8tYk5jHLSMjo8LBQs6ePYt27dph+/btiImJwTfffIOYmBiUlJTA398f0dHRmDVrFiQSSZWO5+rqivfffx//+te/sGDBAixdurTW51DrwUlKSkpw6tQppKSkwMbGBr6+vggONlzHWtKubJASoOYTthJVBesAItM0ceJEfPHFF4iMjMSKFSswbNgwjRsOpVKJb7/9FjNnzoSTkxOmTp2qs+PHxMTg66+/xo4dO1BUVISRI0eiQYMGWqcJEQQBL774Ilq3bo1Tp05h586dGDhwIJKSkuDr64uSkhJcvnwZhw8fVuvo7+bmVu1jEZFhiUQi+LXxNuhAJE/btGkTNm3aVKWy1tbWmDNnDubMmVOl/VYkOjoa0dHRVTpuVdQqcfv5558xZcoUpKWlqTreiUQieHl5YfXq1XjhhRd0EqQ+KRQKVQfE3NxcI0dDZF7qQh1AVFc5Ojpi9+7dGDBgAEaPHo0pU6agffv2kMvlKCkpwd27d3H69Gnk5ubC19cXO3bsgFwu19nxV65ciYULFyIsLAwA8NFHH+Hdd9/VmkwdPHgQycnJOHbsGBwcHNCiRQv88ccf2LBhA+bPn4/r16/j8ePH6Nixo9ZpDapzLCIic1XjxO3IkSMYNGgQPDw88OGHH6JFixZQKpW4ePEiVq9ejZdeegmHDh1Cly5ddBmvzi1evJgDKBDVQF2pA4jqsubNm+P8+fP47LPPsG3bNsTGxqK4uBhA6RDWnTt3xqBBg/D6669XayCTyty+fRu3bt1Ct27dVMvCwsKQkpKC9PR0jQQxLi4OwcHBcHBwUCtfNg9kYmIifHx8tCZt1T0WEZG5qnHiNn/+fPj7++PkyZNwdnZWWzdlyhSEhITggw8+wJ49e2odpD7NmzcP06eXNiU8d+4cIiIijBwRkXmoK3UA37pTXScWi/Hss8+iW7duaNGiBQoKCiCRSFCvXj29HTM9PR0A4OX1z6hxHh4eAEqH0346mUpPT1crW1Y+NTUVAJCUlAQbGxv069cPp06dQvPmzfHJJ5+gY8eO1T5W2fHKtntSUlJStc+ViMhQapy4nThxAv/5z380btgAQCaTYfz48ViyZEmtgjMEqVSqesro6Oho5GgMi6NIUW3UlTqAb92pIuZeTy5fvhwLFy5ETk4OgNK/edHR0Vi8eHGt911QUIC0tDSt68oegjz5Fq/s39rmR8rPz9d44yeVSlVlL168iKysLLz22mtYuHAh1q5di+7duyMxMRH5+fnVOhYArFmzht97mP/1TWRpaj04SXlEIhGKior0tXsiMnHmUgfwrTvVVV9++SVmzJgBf39/jB49GmKxGAcPHkRMTAyKi4uxfPnyWu0/Pj4eUVFRWtd9/PHHAEoTp7LmjWVJlL29vUZ5W1tb3L9/X22ZQqFQlV27di3y8/Mhk8kAAKtXr8bRo0fx1Vdf4bnnnqvWsYDSgVv69++vsTwpKQkjR46s4KyJiIynxolbp06dsH79ekyZMkWtTToAPHr0COvWrUNISEitAyT9+fXXXwEAvXr1MnIkZI7qSh1gyW/dqXLmXE+uXr0aoaGhOHDggCqhEQQBw4YNw5o1a/DRRx/BxsamxvuPjIxUDUr0tNu3b2P27NnIyMiAv78/gNKhuAFobbro7e2t8fYnIyNDVdbKykqVtAGlD4YCAwORlpYGb29vVfmqHKtsuSn3fbtxLhX3rt9HyMC2ej2OOV/fRJaoxonbe++9h6ioKAQFBeGNN95QDc9bNjBBamoqPv/8c50FSrpX9paBTSWoJlgHkCV4up58cqoVU5eUlITFixerDeghEokwbdo0bN++HUlJSWjbVj+JgZeXF3x9fREbG6tKpmJjY+Hr66s1YQoNDcWSJUtQUFAAOzs7VfmyUSKjoqIQGRmJ9957D0DpNAZ//fUXoqOjq30sc/D7miNIvZCO5uFNIHPT3wMl3gcQmZcaJ27h4eH48ccf8a9//QuzZs1STaQnCALkcjm+/fbbcptQEJH5Yx1AZNry8vK09kFt1KgRBEHAw4cP9Xr8yZMnY86cOWjYsCEAYO7cuZgxY4Zq/b1792BnZwdHR0dERETAx8cH48aNw7///W/89NNPOHHiBDZu3AgAeOGFF7Bw4UI888wzaN68OVauXImHDx9i7NixVTqWuXlhZg88SHuo16SNiMxPrfq4tWvXDkOHDsXQoUNx48YNCIKAGzdu4O7duwgPD9dVjERkolgHEJkupVKpeqDyJCur0j/9JSUlej3+rFmzcPfuXQwcOBBWVlYYP348pk2bplofEhKCsWPHYv78+ZBIJNi1axfGjx+P9u3bIyAgADt27ICvry8AYNq0aSgsLMQbb7yBO3fuoFOnTvj999/h5ORUpWOZG48m7vBo4m7sMIjIxNQ4cbtw4QIiIyORnZ2NUaNGYejQoQBKn3KtXr0a3377LWJjY9GoUSOdBUtEpoN1ABFVRCKRICYmBjEx2puX3rhxQ+33gIAAHD58WGtZkUiEt99+G2+//XaNjkVEVBeIa7rh3Llz4eTkhMTERLU28kuWLEFCQgJsbGwwZ84cnQRJRKaHdQCR6bt//z5u3ryp9nPr1i0AwN27dzXW3bx508gRExFReWr8xi0uLg7/+c9/0LRpU411jRs3xr/+9S/VcMBkmpo0aWLsEMiMsQ4gS2Du9eTUqVMxdepUretGjBihsUwkEqG4uFjPUZGpMPfrm8jS1DhxKykpQUFBQbnrBUGocD0Z3+7du40dApkx1gFkCZ6sJ81pREkAGDNmjLFDID0peFQIW0ep1j6M1cH7ACLzUuPErXPnzlizZg0mTpwIFxcXtXW5ublYt24dOnXqVNv4iMhEsQ4gMm1lIzJS3ZJx9R6+eG0rurzSAT0mhhk7HCIyoFrN4xYREYFWrVphxIgRCAgIgEgkQnJyMr755htkZGTwj4aJ++yzzwAA0dHRRo6EzBHrALIErCfJ1NjJbOHZzB3ufvVrvS9e30TmpcaJW6dOnfDbb79h5syZ+OSTT9TWtW3bFps2bULnzp1rHWBN7Ny5E7t27eJN4/8ra94z/8h0teWrV68GwAqbasaU6wAiXWE9SabGuYETXv9Cs39iTfD6JjIvtZrHLTw8HPHx8bh37x5SUlJQUlICX19fyOVyXcVXbbNmzcJPP/2E0NBQo8VAZClMsQ4gIiIiqotqlbiVcXd3h7u7aUwU2bFjR/Tp0webN282dihEFsOU6gCAb92JiIio7qnxPG6masiQIbUeZYmIzNesWbMwd+5cCIJg7FCIiIiIdKbOJW5EZNk6duyI//3vf8YOg4iIiEindNJUksxTz549jR0Ckc4NGTIEhw4dMnYYVEewnqS6jNc3kXkx6cRNoVCgffv2WLVqFSIjIwEAhYWFiI6Oxg8//AA7OzvMnDkTM2bMMG6gZmr58uXGDoGIyKSxnqS6jNc3kXkx2cStsLAQw4cPR0JCgtryWbNm4dSpUzhw4ABSUlIwZswY+Pn5YfDgwaoykZGRqkRPm/T0dKSnp2ssT0pK0ln8REREulbe9C5ERFT3mWQft8TERISGhiI5OVlteV5eHtatW4eVK1ciODgYAwcOxOzZs7Fq1apq7X/NmjVo3769xs/IkSN1eRomb9q0aZg2bZqxwyAql0KhQKtWrdSaPhYWFmL8+PFwcXGBXC7HsmXLjBcg1XmsJ6ku4/VNZF5M8o3b4cOHERUVhUWLFsHBwUG1/Pz58ygqKkKXLl1Uy8LCwrBo0SIolUqIxVXLQydOnIj+/ftrLE9KSrKo5G3//v3GDoGoXHzrTqaA9SQZm1Ip4N6N+/Bo7KbzffP6JjIvJpm4TZ48Wevy9PR0uLm5wcbGRrXMw8MDhYWFuH//fpXnkZLL5ZwgmMiEJSYmYvjw4RpD+pe9dd+7dy+Cg4MRHByMhIQErFq1Si1xq8yaNWuwYMECXYdNRKRzcd+dxv7P/kTE2FCc3ZOAAXN7okmIn7HDIiIjMMnErTz5+fmQSqVqy8p+VygUNdqnQqFQbZubm1u7AIlIJ/jWnYiolF/bhmjWtTGc3ByQez8PivzHxg6JiIzEJPu4lcfW1lYjQSv73d7evkb7XLx4MZydneHs7IyIiIhax2jKyjq1E5m6yZMnY/ny5Rrf68reuleVXC5XvbELCgpCQEAAAgIC4OPjo7NzILJ0giBg7ty5cHd3R/369TF79mwolcpyy1+/fh09evSAg4MDWrZsqdaMz9/fHyKRSONn4cKFAICzZ89qrOvQoYPez9EQvFt4YviSF9FhQFu8e+AttIxoauyQiMhIzCpx8/b2RmZmJoqLi1XLMjIyYGdnBxcXlxrtc968ecjOzkZ2djYOHz6so0iJSB/08dbdkh7eUM3cu3GfD75qICYmBl9//TV27NiBH374AVu3bkVMjPbPURAEvPjii/D09MSpU6cwatQoDBw4EDdv3gQAnDx5UtU3NT09HZ9++imcnZ0xZswYAKXNq9u1a6dW5tdffzXYuRqKWCwydghEZERm1VSyXbt2sLa2RlxcHMLCwgAAsbGxCAkJqXITqadJpVLVjZ+jo6POYjUHU6ZMMXYIRNWij7fu8+bNw/TppUOrnzt3jskbqZkyZQp+ifnD2GGYpZUrV2LhwoWqv9cfffQR3n33XcycOVOj7MGDB5GcnIxjx47BwcEBLVq0wB9//IENGzZg/vz5an3Ys7OzsXDhQixbtgx+fqV9vZKSktCiRQt4enoa5uTqCN4HEJkXs0rc7O3tMWbMGEyaNAkbN25EWloali5dio0bN9Z4n5bcxy06OtrYIRBVy5Nv3a2sSquv2r51t+SHN1S56Oho3Pu2Zm9zLdnt27dx69YtdOvWTbUsLCwMKSkpSE9P1xggLC4uDsHBwWp9WsPCwnD8+HGNfS9duhRyuRzjxo1TLUtMTESbNm30cCamLfvOI1jZSOBQr2YPrngfQGRezCpxA0qbXkyePBlRUVFwdnbGggULMGjQoBrvb/HixRY/uhwndCVzoY+37pb88IZIX8qm2/Dy8lIt8/DwAACkpqZqJG7p6elqZcvKp6amqi3Lz8/Hp59+ijVr1qh955OSkqBUKtG6dWtkZ2ejd+/e+OSTTyCTycqNz9ynBCkpVuKz0Zvh7OGE6C/HGDscIjIAk0/cnh4O3N7eHps3b8bmzZt1sn9LbiZVNqpeMCKNGwhRFenjrTsf3lBF+vfvj0uJyXil5Xhjh2JyCgoKkJaWpnVd2UOQJ/ukVtQftbz+q0+X3bZtGxwdHfHSSy+plhUVFSE5ORmNGjXCxo0bkZWVhWnTpmHUqFHYtWuX1vjqwpQgEisxnukbBCdXh8oLl6PsPmD37t26CouI9MjkEzd9s+RmUsnJyQCA4PqRxg2EqBp0/dbdkh/eUOWSk5ORVVj1EUstSXx8PKKiorSu+/jjjwGUJmm2traqfwPa+6Pa2tpqjAyrUCg0ym7fvh0vv/yyqqk0AFhbWyMzMxN2dnawtrYGAGzevBkdOnTA7du3Nd7kAXVnSpDeb2r//Kuq7D6AiMyDxSdubCZFZNr0/dbdkh/eENVGZGSkxvezzO3btzF79mxkZGTA398fQGl/VAAazSSB0v6rCQkJassyMjLUyioUChw6dAhz587V2P7pJpEtWrQAAKSlpWlN3ORyudY4iIhMmVlNB6APljYU+PzwmCoNa13VckTmTqFQICcnBzk5OXx4Q6QjXl5e8PX1RWxsrGpZbGwsfH19tSZMoaGhOHPmDAoKCtTKh4aGqn7/+++/UVRUhI4dO6ptm5iYCCcnJ1y/fl217Ny5c7CyskJAQIAuT4uIyKgsPnHjPG5Els3SHt4QGcrkyZMxZ84cHDp0SPWm7K233lKtv3fvnuphSUREBHx8fDBu3DgkJCRgyZIlOHHiBMaP/6dv4YULF9C4cWONvnCBgYEICAjAhAkTcOHCBcTGxmLChAmYMGEC6tWrZ5iT1ZPDm+Nw+qe/jR0GEZkIi0/cpFIpZDIZZDIZm0kRWSA+vCHSj1mzZuHll1/GwIEDMWTIEIwaNQrTpk1TrQ8JCcHSpUsBABKJBLt27UJ6ejrat2+PLVu2YMeOHfD19VWVv3PnjtZETCwWY/fu3ZDJZAgPD8eAAQPQvXt3LF++XP8nqUclxSX488t4HPv2lLFDISITYfF93CxZTExpU8jjHyRUUpKo7mIfN6pITEwMvprxg7HDMEsSiQQxMTGqvzVPu3HjhtrvAQEBFT48mTNnDubMmaN1nY+PD3788ccax2pKihRFEEskkFhJ8PoXw2FjZ6O3Y5X3f0NEpsniEzdLHpykV69eAJi4ERGVp1evXqwjyWCKFEVYMWQ9PALcMTrmJXg0cdfr8cruA4jIPFh84labOZzqwuAdtT0HTt5N5s6SH94QkWkRSyTwDHBHgyZuxg6FiEyQxfdxs+T+LUFBQVh95mNjh0FkVBychCrCepL0IfdBHpQlSrVl+TkFEAQBo2JeQq/oCCiV2qdaeHQ/D3kPC7Suq66goCAEBQXpZF9EpH8Wn7hxcBIiy2bJD2+IyPDSr9zFsoFf4LfPj6iW5WblY/lLa/HDgj0AgBvnUvFB95U487P6iJKCIGD16M3YMOVbg8ZMRKbB4ptKEpFl4+AkRGRIjvUd4NfWG17NPVTLpPbW8GvbED6tSue4s7GzhpOrA2wd1ac+EIlEaNOrhcZyIrIMTNyIiIiIDMTJ1QFj/ztUbZm11Bojlw5S/e7V3APTtk/Qun3vN6P0Gh8RmS6LT9w4MAEREREREZk6i+/jVtOBCcx9RMn54TG4d+O+scMgMjqFQoGcnBzk5OTw4Q0RGZxSKeDikasoeFRo7FCIyMRZfOJmyQMTDGvxKoa1eNXYYRAZFUeVpIrs2rWL9STp1ZXj1/Dt27txeFOcwY+9a9cu7Nq1y+DHJaKasfjEzZJHlaxv54b6dpwrhiybJT+8ocoFBASwniS98mvbEF2GtcczfQw/LH9AQAACAgIMflwiqhmL7+NmyR4UZAIAb0rIonFUSarI1atX8aAgk/Uk6Y2toxQ9o43ztv/q1asAUKeStyJFEayl1sYOg0gvLP6NmyX7NmkDvk3aYOwwiIhM1oABA1hPUp01YMAADBgwwNhh6EzmzQf4qM9q/Pa/P40dCpFeMHEjIiIiIrNnY28DV9/6cPF0NnYoRHph8U0lOR0AERERkfmTuTli8sZRxg6DSG8s/o0bR5T7x/zwGLOf5oCoujgdABEREZkDi0/cOKIckWWryw9v+DBG05OfR0WfT9k6zndJpqDgUSGUSgHZdx7h4xf+h0+Hb0T89rPGDouIDMziEzdLng6AiPjwhohM272UB/ik/+f49dNDgAgQiUXIup2NG+dSjR0aERmYxSdulmxK8GxMCZ5t7DCIjIoPb6girCdrThAEzJ07F+7u7qhfvz5mz54NpVJZ6XZXr16FnZ2dxvLff/8drVq1gr29PZ599llcu3ZNbf2KFSvg7e0NJycnjB8/Hvn5+To7F2Oyk9nCO9ATDRq7wbmBE2btmoTpO17HS//pXet9JyQkICEhQQdREpEhMHEjIiIinYuJicHXX3+NHTt24IcffsDWrVsRE1Nx091bt26hX79+KCwsVFt+8+ZNvPjiixg3bhxOnjwJd3d3vPjiixAEAQDwww8/YP78+VizZg0OHDiAuLg4zJ5dNxJux3r2GP+/YWj/Qmu1ZVY2NR9f7tzeBGydtQOFuQpdhEhEBsLEzYJdzbqIq1kXjR0GEZHJYj1ZcytXrsTChQsRFhaGqKgofPTRR1i1alW55Xfu3In27dtDKpVqrFu3bh06dOiAGTNmICgoCBs3bsSNGzdUzZtXrlyJqVOnol+/fggJCcGaNWuwYcMGs33rlpuVj+w7j/S2/xvnUpF8MgU/7fwJv/76q96OQ0S6xcTNgu2/vhv7r+82dhhERCaL9WTN3L59G7du3UK3bt1Uy8LCwpCSkoL09HSt2/zyyy94//33sXLlSo11cXFxavuyt7dHcHAwjh8/jpKSEpw8eVJtfWhoKB4/fozz58/r8KwMZ9Mb32H1mM14XFikWnYl/jrivj+jk/33m9kDU7e/hvmL38P06dN1sk8i0j+Ln8eNiIiIdKssOfPy8lIt8/DwAACkpqZCLpdrbLN27VoAwKFDh7Tu78l9le0vNTUVDx8+RGFhodp6KysruLq6IjVV+wAe6enpWhPIpKSkSs7MMNr1bonYrSex9vWvEf3lGADAgbXHkH75Dtr0bAF7Z80+gNVhZS2BzI19eonMDRM3IiIiqraCggKkpaVpXVc2J+KTzR7L/q1QVL9fVX5+vkYTSqlUCoVCoWoOWd56bdasWYMFCxZUOw5DCRvREemX70EsEamWDXr3eeTcy9WatCnyH+P07r/QpmcLONZ3MGSoRGRAFp+4KRQKVcXOyXeJLA/rAKKaiY+PR1RUlNZ1H3/8MYDS75etra3q30BpM8fqsrW11UjCFAoFXFxcNPb/5PryjjVx4kT0799fY3lSUhJGjhxZ7fh0pfhxMfauPIhG7X0xZEFftXXu/q5w93fVul3CgUvY/9mfeFxQhMhxnQ0RKhEZgcUnbosXLzbpp26mjBP7Ul1Q3Tqg7Lqff0SzX8j88Bity3WtohjIcjxdBxv6eoiMjFSN6vi027dvY/bs2cjIyIC/vz8AICMjAwC0NpOsjLe3t2r7MhkZGWjXrh1cXV1ha2uLjIwMBAYGAgCKi4tx//79co8ll8trFIe+PcrMw5mfLyD3QT5aPdu8ytu16h6IIkUxWnUP1GN0RGRsFj84iSVPvlvP1hX1bLU/vSOyFJZcB1DlWE/WjJeXF3x9fREbG6taFhsbC19f3xolTKGhoWr7ys/Px9mzZxEaGgqxWIyQkBC19cePH4e1tTXatm1buxMxsHpezpi0YSRefLtXtbazsbNGp5eegYNL9fq+NWnSBE2aNKnWNkRkPBb/xk0qlaraxVva5LuvtBxv7BCIjM6S6wCqHOvJmps8eTLmzJmDhg0bAgDmzp2LGTNmqNbfu3cPdnZ2Vfrevfrqq/jkk0+wZMkSvPDCC1i4cCEaNWqEyMhIAMCUKVMwceJEtGrVCt7e3pg8eTImTJhQo2aZxubRxN1gx9q9myOmEpkTi0/ciIiISPdmzZqFu3fvYuDAgbCyssL48eMxbdo01fqQkBCMHTsW8+fPr3Rf/v7++PHHHzF16lQsXLgQXbp0wc6dOyESlQ7eMWzYMNy4cQMTJ06EQqHASy+9pOpnR0RUVzBxs2Anbpc2K+noFWbkSIiITBPryZqTSCSIiYlBTIz2/tA3btzQury8vnO9e/dG7969yz3e3LlzMXfu3BrFasqUSgGxW06gYZAcjdv76nTfn332GQAgOjpap/slIv2w+D5uluxUxjGcyjhm7DCIiEwW60kytofp2Tiw9ihit57UWJdw8DKuxt+o8b5Xr16N1atX1yI6IjIkvnEjIiIiMlH1vV0wYulAuPnUV1uuVAr4YcEeOLraY/oPrxspOiIyJCZuRERERCasYUs57Jxs1ZaJxSIMW9wfNnY2RoqKiAyNTSWJiIiITNT5fYn4qM9qJBy8rLGuWefG8G/X0AhREZExMHEjIiIiMlEuchk8Atzh7OFk7FCIyMjYVNKCNXZpbuwQiIhMGutJMja/tg0xeeMovey7Z8+eetkvEekHEzcL9nzjAcYOgYjIpLGepLps+fLlxg6BiKrB4hM3hUIBhUIBAMjNzTVyNERkaKwDiIiIyBxYfOK2ePFiLFiwwNhhGMW+a7sAVP2J8vzw0klU5x+ZXqN1T66fHx6jtawuaDseUXlMvQ548noGDHNN6/P7WV3ViaWq9VBFx3i6XHXrSSJzMm3aNAB880ZkLix+cJJ58+YhOzsb2dnZOHz4sLHDMahrDy/h2sNLxg6DyKgsuQ6gyrGepLps//792L9/v7HDIKIqsvg3blKpFFKpFADg6Oho5GiIyNBYBxAREZE5sPg3bkRERERERKaOiRsREREREZGJY+JGRERERERk4iy+j5sl6+DZxdghEBGZNNaTVJdNmTLF2CEQUTUwcbNgHb3CjB0CEZFJYz1JpuKv/Uk48eM5DH2/H2TuTjrZZ3R0tE72Q0SGwaaSRERERCbuzrVMpF3MQN7DAmOHQkRGwsTNgn2TuB7fJK43dhhERCaL9WTNCYKAuXPnwt3dHfXr18fs2bOhVCor3e7q1auws7PTWL5x40YEBgbC0dERnTp1wtGjR1XrsrKyIBKJ1H7c3Nx0ej7G1v31MMzaPQnypg10ts/+/fujf//+OtsfEekXm0pasKzC+8YOgYjIpLGerLmYmBh8/fXX2LFjB4qKijBy5Eg0aNAAM2fOLHebW7duoV+/figsLFRbvm/fPkRHR2Pt2rXo1KkTNm/ejD59+iApKQleXl5ITEyEq6srLly4oNpGLK5bz6bFYhHsZZoJbW0kJyfrdH9EpF91q1YjIiIik7By5UosXLgQYWFhiIqKwkcffYRVq1aVW37nzp1o3749pFKpxrpNmzZhzJgxGDFiBAICAvD+++/D09MTv/zyCwAgKSkJzZo1g6enp+qnQQPdvZkyd4r8x1AqBWOHQUS1xMSNiIiIdOr27du4desWunXrploWFhaGlJQUpKena93ml19+wfvvv4+VK1dqrJs9ezamT5+usTw7OxsAkJiYiGbNmuko+rol5e80fNR3NX76aL+xQwEA3PwrDd++vQvZdx4ZOxQis8OmkkRERKRTZcmZl5eXapmHhwcAIDU1FXK5XGObtWvXAgAOHTqksS44OFjt93379uHy5ct49tlnAZS+cSsqKkLHjh2RlpaG8PBwLF++XOtxyuLTlkAmJSVV4ezMy9mfL0BZrITYyjSe1V87fRMXjyTjmb6t4eyhm9ExiSxFnUnclEolXnvtNVy6dAlOTk748ssv2UyCyIKwDiAyrIKCAqSlpWldl5ubCwBqzR7L/q1QKGp13OTkZIwdOxYjRoxQJXQXL16Eu7s7li9fDkEQ8Pbbb6Nfv344ceIEJBKJxj7WrFmDBQsW1CoOYynMVWDHon1o1b05WvcIrLR8xJhQyJs2QPsBbQwQXeXCR3VCYHgAPAPcjR0KkdmpM4nbjh07YGdnh6NHj2Lbtm1YvHgxli9fbuywTFrPRhxJiuoO1gGkD6wnyxcfH4+oqCit6z7++GMApUmara2t6t8AYG9vX+NjXr58GT169ECTJk1Ub+gAICEhASKRSDUa5fbt2yGXyxEfH48uXTQnUZ84caLW0RSTkpIwcuTIGsdnCI8yc3H52DXYO9tWKXGr5+WMToOf0bouJiZG1+FVSmIlZtJGVEN1JnE7evQoevbsCQB4/vnnsWTJEiNHZPoC6lVe4ROZC9YBpA+sJ8sXGRkJQdA+4MXt27cxe/ZsZGRkwN/fHwCQkZEBAOU2X6xMQkICunfvjsaNG2Pv3r1qUwY8nQw2aNAArq6u5b4RlMvlNY7D2Nz9XfHmt6/CydWh1vvq1auXDiIiIkMxjQbPOpCTkwOZTAYAcHJywqNH7PRKZElYBxCZDi8vL/j6+iI2Nla1LDY2Fr6+vjVKmNLT09GzZ080bdoU+/fvV33XgdLvfr169XDw4EHVsrS0NGRmZiIwsG4m3vXkzrCyqTPP3omoiurMt14mk6lu1B49egRnZ2cjR2T6Vp8pbcoyJXi2kSMhqj3WAaQPrCdrbvLkyZgzZw4aNmwIAJg7dy5mzJihWn/v3j3Y2dnB0dGx0n3NnDkTJSUlWL9+PXJzc1V96BwdHSGTyRAeHo5p06Zh7dq1kEgkeOutt/D888+jdevW+jm5OiIoKAhA6dtMIjJ9deaNW+fOnbF/f+lQt3v37tXapp2I6i7WAUSmZdasWXj55ZcxcOBADBkyBKNGjcK0adNU60NCQrB06dJK9yMIAnbs2IE7d+6gefPmqmaOcrlctf3mzZsRHByMPn36IDIyEv7+/ti6davezk2fihTFxg6BiEyUySZuCoUCrVq1UhsWuLCwEOPHj4eLiwvkcjmWLVumWjdo0CDk5+eja9eu+Pzzz/HOO+8YIWoi0hXWAUTmTSKRICYmBllZWbh37x6WLFkCkUikWn/jxg3Mnz9fY7un+86JRCLk5+dDEASNn7Lt69Wrhw0bNuDevXvIycnBV199hXr16un7FHXuXsoDLOnzGX5fE6uxLunPq1g5bD3SL981QmREZApMsqlkYWEhhg8frvHqftasWTh16hQOHDiAlJQUjBkzBn5+fhg8eDAkEgk2bNhQpf1b0vwtROaIdQARWSIbO2u4NqwHF0+Zxrr8h/l4mJGDwtzCcrcvKVYidusJNOngh4ZB5jn4ChGVz+QSt8TERAwfPlxjpKq8vDysW7cOe/fuRXBwMIKDg5GQkIBVq1Zh8ODB1TpGbeZvmR9u+KFzDe3Jc6zN+WrbtrJl849Mr3QbbWWqerz5R6arlpXtZ354TLn7rExlsRvC0+dj7ky9DtCmomuvsv8Xrds+cW1W9VhPb/PkcZ9cpuvr/+ltq7q/6hy3snMqr3xdZgnnaImcGzhhyubRWte1798GbXsHwcpac166MneS7+HgumPIuHIPL3/wgr7CJCIjMbnE7fDhw4iKisKiRYvg4PDPULfnz59HUVGRWr+VsLAwLFq0CEqlEmJx1Vt9mvP8LUR1HesAIiLtKkraAMCruQdeXvQC5M09DBQRERmSySVukydP1ro8PT0dbm5usLGxUS3z8PBAYWEh7t+/D3f3qk/maM7zt+jSsBavGjsEIg2sA8iUsJ4kc9OiW9Mql921a5ceIyEiXTO5xK08+fn5kEqlasvKflcoFDXer0KhUG1fNrywpahv52bsEIiqjHUAGQPrSarLAgICjB0CEVWDyY4q+TRbW1uNm7Oy3+3t7Wu838WLF8PZ2RnOzs6IiIioVYzm5kFBJh4UZBo7DKIqYR1AxsB6kuqyq1ev4urVq8YOg4iqyGwSN29vb2RmZqK4+J/5TTIyMmBnZwcXF5ca73fevHnIzs5GdnY2Dh8+rINIzce3SRvwbVLVRuEjMjbWAWQMrCepLhswYAAGDBhg7DCIqIrMJnFr164drK2tERcXp1oWGxuLkJCQag1K8DSpVAqZTAaZTAZHR0ddhEpEesA6gIiIiCyZ2fRxs7e3x5gxYzBp0iRs3LgRaWlpWLp0KTZu3Fir/bJ/C5F5YB1ARERElsxsEjcAiImJweTJkxEVFQVnZ2csWLAAgwYNqtU+Fy9erNP5nIhIf1gHEBERkaUy6aaSgiAgMjJS9bu9vT02b96M3NxcpKWlYerUqbU+Bvu3EJku1gFEREREpczqjZs+SKVS1ZDiEknpxJZJSUnllk/PTTVIXIZQpCwCUP1zOnPmTLnbla2ryn617efJ7cvbj7YyVTle2bZl5Z48fnn7rExlsRvC0+dTkcDAwFqNwFgX1bQOKO9af/q6ruz/paLvUXW+Q9qOq+27U5Xrv6rfCV1vW165J+N/enl16pyn91PR/1uZmtaTxsA6oPYKCgoAVFwH1CVl52usv19keKwDzJtIEATB2EGYiq1bt2LkyJHGDoNIb06fPo3g4GBjh2GyWAdQXcc6oGKsA6iuYx1g3pi4PSEzMxO//vor/P39YWdnZ+xwaiwpKQkjR47Eli1b0KJFC2OHYzT8HEo9+TkMHDiQT9oqYKp1gLley4zbsKoSN5+2V8xU64DqMtdrWNf4OZTifUDdYfFNJZ/k5uaGESNGGDsMnWnRogWfqoCfQ5kWLVqwsq6EqdcB5notM27DMte4TYGp1wHVxWuhFD+HUrwPMH8mPTgJERERERERMXEjIiIiIiIyeUzciIiIiIiITBwTNyIiIiIiIhPHxK0OksvleO+99yCXy40dilHxcyjFz8H8mev/IeM2LHONm3SP10Ipfg6l+DnUHZwOgIiIiIiIyMTxjRsREREREZGJY+JGRERERERk4pi4ERERERERmTgmbmZIEATMnTsX7u7uqF+/PmbPng2lUllu+evXr6NHjx5wcHBAy5YtsX//frX1bdu2hUgkUvu5cOGCvk+jRgoLCzF+/Hi4uLhALpdj2bJl5ZY9e/YsOnXqBHt7e4SEhOD06dNq67/55hs0adIE9vb2GDhwIDIzM/Udvs7o8nNwcXHR+P/Pzc3V9ynQE6r7nS5z9epV2NnZaSz//fff0apVK9jb2+PZZ5/FtWvX9BG2WdVF5lp38LtO2pjTd0/XzPW7rGusGyyUQGZn6dKlgo+Pj3DkyBHhwIEDgpeXl/DJJ59oLatUKoU2bdoII0aMEBITE4UPP/xQsLe3F1JSUgRBEITi4mLB1tZWOHz4sJCenq76KSoqMuQpVdm//vUvoU2bNsLp06eFH3/8UXBychK+//57jXK5ubmCp6enMGPGDCExMVF48803BQ8PDyE3N1cQBEGIj48X7OzshM2bNwvnz58XIiIihL59+xr6dGpMV59DamqqAEBITk5W+/9XKpWGPiWLVp3vdJmbN28KzZs3F56uxlNSUgQHBwdh6dKlwoULF4ShQ4cKrVu31sv/qTnVReZad/C7TtqY03dP18z1u6xrrBssExM3M+Tj4yNs3LhR9ftXX30l+Pn5aS37xx9/CA4ODqovqCAIQvfu3YX33ntPEARBuHLliiAWi4WCggI9Rqwbubm5gq2trXDw4EHVsvfff1+IiIjQKLt+/XqhUaNGqopHqVQKAQEBqs9t1KhRwpgxY1Tlb968KYhEIuHatWt6PAPd0OXn8NtvvwlyudwAUVNFqvOdFgRB2LFjh+Du7i60adNGI3H797//rXYt5OXlCU5OTmrXi66YS11krnUHv+tUHnP57umauX6XdY11g+ViU0kzc/v2bdy6dQvdunVTLQsLC0NKSgrS09M1ysfFxSE4OBgODg5q5Y8fPw4ASExMhI+PD2xtbfUffC2dP38eRUVF6NKli2pZWFgY4uPjNZqIxMXFISwsDCKRCAAgEonQtWtX1XnHxcWpfYY+Pj7w9fVFXFycAc6kdnT5OSQmJqJZs2aGC540VPc7DQC//PIL3n//faxcuVJj3dPXtr29PYKDg1X/58aK25h1kbnWHfyukzbm9N3TNXP9Lusa6wbLxcTNzJRVyl5eXqplHh4eAIDU1FSt5Z8sW1a+rGxSUhJsbGzQr18/eHp6IiIiAidOnNBX+LWSnp4ONzc32NjYqJZ5eHigsLAQ9+/f1yhb0XlXtt6U6fJzSEpKQn5+PiIjIyGXy9GnTx9cvnxZ/ydBKtX9TgPA2rVrMXHixHL3Z4hr25zqInOtO/hdJ23M6buna+b6XdY11g2Wy8rYAZCmgoICpKWlaV1X1llUKpWqlpX9W6FQaJTPz89XK1tWvqzsxYsXkZWVhddeew0LFy7E2rVr0b17d9UTOFNS3rkAmude2XlXtt6U6fJzuHjxIh48eIAPP/wQMpkMH330ker/38nJSY9nYVl0+Z2ujC6v7bpSF5lr3cHvuuWqK989XTPX77KusW6wXEzcTFB8fDyioqK0rvv4448BlH4xy5o1lH357O3tNcrb2tpqPH1RKBSqsmvXrkV+fj5kMhkAYPXq1Th69Ci++uorvP3227o5IR2xtbXVqJDKO/fyypaVq2y9KdPl57Bv3z4UFRXB0dERALB161b4+Pjgp59+wvDhw/V1ChZHl9/pypT3f+7i4lLtfdWVushc6w5+1y1XXfnu6Zq5fpd1jXWD5WJTSRMUGRkJoXTgGI2fESNGAAAyMjJU5cv+LZfLNfbl7e2tVrasfFlZKysrVWUNlLZ9DgwMLPdJnzF5e3sjMzMTxcXFqmUZGRmws7PTuCmt7LwrW2/KdPk5SKVSVWUNlFbwjRo1Msn/f3Omy+90ZXR5bdeVushc6w5+1y1XXfnu6Zq5fpd1jXWD5WLiZma8vLzg6+uL2NhY1bLY2Fj4+vpqrWxCQ0Nx5swZFBQUqJUPDQ0FAERFRWHBggWqdUqlEn/99RcCAwP1eBY1065dO1hbW6t1HI6NjUVISAjEYvVLOTQ0FMeOHYMgCABK57w5evSo6rxDQ0PVPsNbt27h1q1bqvWmTFefgyAIaNKkCTZt2qQqn5eXhytXrpjk/39dVd3vdGWevrbz8/Nx9uxZnV/b5lQXmWvdwe86aWNO3z1dM9fvsq6xbrBgehyxkvRk8eLFgpeXl3Dw4EHh4MGDgpeXl7Bs2TLV+rt37wqPHj0SBKF0fpaWLVsKL7/8snDhwgVh8eLFgqOjo2r+lmXLlgnOzs7Crl27hIsXLwqTJ08WPDw8hJycHKOcW2UmTpwoBAUFCSdOnBB27NghyGQy4YcffhAEQRDS09OF/Px8QRAEITs7W3B3dxfefPNNISEhQXjzzTcFT09P1XDIx44dE2xsbIR169YJ58+fFyIjI4UXXnjBaOdVXbr6HN544w3B19dXOHjwoHDhwgVh4MCBQqtWrYTi4mKjnZslqs53+kkHDx7UmA7g+vXrgq2trbB48WLVPG5t2rTRy5w85lQXmWvdwe86aWNO3z1dM9fvsq6xbrBMTNzMUHFxsTBt2jTBxcVFcHNzE+bMmaN2U+bn56ean0UQSudo6datmyCVSoWgoCDht99+U61TKpXCokWLBF9fX0EqlQrdunUT/v77b0OeTrXk5eUJo0ePFhwcHAQvLy9h+fLlqnUA1Oa1iY+PF5555hnB1tZW6Nixo3DmzBm1fW3cuFHw8fERHBwchIEDBwqZmZkGOova09XnUFBQIEyfPl2Qy+WCvb290K9fP+HmzZsGPBMShOp/p8toS9wEQRD27NkjNGvWTLCzsxO6d++ut3mJzKkuMte6g9910sacvnu6Zq7fZV1j3WCZRILw/+9OiYiIiIiIyCSxjxsREREREZGJY+JGRERERERk4pi4ERERERERmTgmbkRERERERCaOiRsREREREZGJY+JGRERERERk4pi4ERERERERmTgmbkRERERERCaOiRvV2NixYyESidR+JBIJZDIZOnXqhM2bN2ts8+jRIyxbtgwdOnSAs7MzHBwc0LFjR3zxxRdQKpUVHu/1119HZGSkns6GiKrLEHXAr7/+ivDwcNjb28PR0RE9evRAXFycIU6PiCrB+wAiw7IydgBk/pYvXw43NzcAgCAIyM7OxpYtWzB27FhkZmZixowZAIBLly6hf//+uH79OkaMGIFx48ZBoVBg586dmDhxIv7880989dVXEIlEGsdYv3491q5di4iICIOeGxFVTl91wOHDh9G7d28EBQVh0aJFKC4uxurVqxEREYEjR46gY8eORjtnIvoH7wOIDEQgqqExY8YIAITr169rrMvPzxf8/f0FmUwmFBYWCgUFBULz5s2F+vXrC+fPn9coP2XKFAGAsHLlSrXlxcXFwoIFCwSRSCQAECIiIvR0NkRUXfquA9q1ayf4+voKeXl5qmUZGRlCvXr1hB49eujlnIio6ngfQGRYbCpJemFnZ4cXXngBOTk5SEhIwOrVq3Hp0iUsX74cbdq00Si/dOlS1KtXD59//rlqWWFhIYKDg/Hee+9h1KhR8Pb2NuQpEFEt1LYOyMrKwvnz5zF06FDY29urynl4eCAiIgLHjh0z2LkQUfXxPoBI99hUkvRGLC59LlBcXIxvv/0Wjo6OeOWVV7SWtbOzQ3x8PPz8/FTLCgsLkZOTg23btmHo0KHw9/c3RNhEpCO1qQNkMhkuXboEBwcHjbKZmZmwsuKfLyJTx/sAIt3iXz7SC6VSiUOHDkEqlaJFixY4e/YsunbtCmtr63K3adq0qdrvMpkMV65c4Q0akRmqbR0gkUg06gQA+Ouvv3D06FH06tVLL3ETkW7wPoBI9/hNoFrLysqCo6MjgNKnajdu3MDy5ctx/vx5TJs2DYWFhSguLoZcLq/WfsViseppHRGZLn3VAU/Lzc3F6NGjAQBz586tddxEpBu8DyAyDCZuVGvBwcEay6RSKd544w0sWbIEubm5AICSkhJDh0ZEBmCIOiA/Px/9+/fH+fPnMW/ePI4sR2RCeB9AZBhM3KjWtmzZAg8PDwClzZtcXFzQokUL2NraAgDq1asHGxsb3L1715hhEpGe6LsOePjwIfr164ejR4/i1VdfxaJFi3QWOxHVHu8DiAyDiRvVWteuXSvsMCwSidC5c2ecPn0axcXF5bZVf/fdd5GcnIzly5fD09NTT9ESka7psw64e/cuevXqhXPnzuH111/H559/rnWOJyIyHt4HEBkGGw6TQQwaNAi5ubn49ttvta4vKCjAunXr8Pvvv8PV1dXA0RGRvtWkDnj06JEqaZs2bRrWrFnDpI3ITPE+gKj2mLiRQbz++uvw8/PDzJkzceHCBbV1JSUlmDx5Mu7cuYM5c+ZUOOIUEZmnmtQB0dHROHfuHN566y3ExMQYI2wi0hHeBxDVHptKkkHY2tpix44d6NmzJ0JCQjBixAiEhITg/v37+P7773Hu3DkMGTIE06dPN3aoRKQH1a0DkpKS8NVXX8HFxQXt2rXDli1bNPY5cuRIQ58GEdUQ7wOIao+JGxnMM888g3PnzmHFihXYs2cPtm3bBqVSiTZt2mDDhg0YO3Ysm0ER1WHVqQMOHz4MoHRgknHjxmndHxM3IvPC+wCi2hEJgiAYOwgiIiIiIiIqH/u4ERERERERmTgmbkRERERERCaOiRsREREREZGJY+JGRERERERk4pi4ERERERERmTgmbkRERERERCaOiRsREREREZGJY+JGRERERERk4pi4ERERERERmTgmbkRERERERCaOiRsREREREZGJY+JGRERERERk4v4P9LXz7tFxbkMAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 900x300 with 3 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# select european samples\n",
    "fig, axes = plt.subplots(1, 3, figsize = (9, 3), gridspec_kw = {'wspace': 0.4})\n",
    "\n",
    "euro_samples = metadata.query('ancestry_inferred_by_SVs == \"EUR\"')\n",
    "euro_metadata = metadata.copy()\n",
    "\n",
    "cutoffs = []\n",
    "for i, comp in enumerate(['PC1', 'PC2']):\n",
    "    \n",
    "    # determine cutoffs\n",
    "    skew = euro_samples[comp].mean() - euro_samples[comp].median()\n",
    "    \n",
    "    # if left skewed, use 2.5% percentile\n",
    "    if skew < 0:\n",
    "        cut = np.percentile(euro_samples[comp], [1])[0]\n",
    "        euro_metadata = euro_metadata[euro_metadata[comp] > cut]\n",
    "        \n",
    "    else:\n",
    "        cut = np.percentile(euro_samples[comp], [99])[0]\n",
    "        euro_metadata = euro_metadata[euro_metadata[comp] < cut]\n",
    "        \n",
    "    cutoffs.append(cut)\n",
    "    \n",
    "    axes[i].hist(metadata.query('ancestry_inferred_by_SVs == \"EUR\"')[comp], bins = 100);\n",
    "    axes[i].set_yscale('log')\n",
    "    aesthetics.strip_axis(ax = axes[i], x = 'ignore', y = 'ignore')\n",
    "    axes[i].set_ylabel('count')\n",
    "    axes[i].set_xlabel(comp)\n",
    "    \n",
    "    # add cutoff\n",
    "    axes[i].axvline(x = cut, color = 'black', linestyle = 'dashed')\n",
    "    \n",
    "sns.scatterplot(x = 'PC1', y = 'PC2', hue = 'ancestry_inferred_by_SVs', ax = axes[2], data = euro_samples, s = 2, ec = None)\n",
    "axes[2].axhline(cutoffs[1], color = 'black', linestyle = 'dashed')\n",
    "axes[2].axvline(cutoffs[0], color = 'black', linestyle = 'dashed')\n",
    "sns.despine(ax = axes[2])\n",
    "axes[2].legend(bbox_to_anchor = (1, 0.5))\n",
    "\n",
    "fig.savefig(f'data/{folder_name}/european-samples.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "427221f9-034c-4347-a215-5c7faf6f640c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10590 total samples\n",
      "6346 European samples\n"
     ]
    }
   ],
   "source": [
    "print(metadata.shape[0], 'total samples')\n",
    "print(euro_metadata.shape[0], 'European samples')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "873a0d55-da74-4097-bac3-3e16fd65e920",
   "metadata": {},
   "outputs": [],
   "source": [
    "# here, we define the models that will be run\n",
    "covariates = {'no-covariates': [],\n",
    "              'sex': [\"sex\"],\n",
    "              'sex-ancestry': ['sex', 'PC1', 'PC2', 'PC3']}\n",
    "\n",
    "# any analysis without ancestry must be subset to european samples\n",
    "euro_samples = euro_metadata['entity:sample_id'].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2c0c9bba-f061-4759-b7bf-ad863fd09438",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EWSR1 ewing\n",
      "FLI1 ewing\n",
      "MYCN neuroblastoma\n",
      "TERT neuroblastoma\n",
      "ALK neuroblastoma\n",
      "TP53 osteosarcoma\n",
      "TP53 neuroblastoma\n",
      "TP53 ewing\n",
      "RB1 osteosarcoma\n",
      "ATRX neuroblastoma\n",
      "SHANK2 neuroblastoma\n",
      "BRCA1 neuroblastoma\n",
      "CDKN2A ewing\n",
      "DLG2 osteosarcoma\n"
     ]
    }
   ],
   "source": [
    "for index, row in genes_to_test.iterrows():\n",
    "    gene, disease = row[['gene', 'disease']].values\n",
    "    print(gene, disease)\n",
    "    \n",
    "    file_prefix = f'data/{folder_name}/{gene}-{disease}/'\n",
    "    \n",
    "    gene_disease_results = []\n",
    "    for cohort in ['discovery', 'validation']:\n",
    "\n",
    "        # osteosarcoma has no validation datasets, so we skip\n",
    "        if disease == 'osteosarcoma' and cohort == 'validation':\n",
    "            continue\n",
    "            \n",
    "        #######################\n",
    "        #### LOAD OUR DATA ####\n",
    "        #######################\n",
    "        \n",
    "        analysis_dosages = pd.read_csv(file_prefix + f'{cohort}-dosages.csv')\n",
    "        analysis_metadata = pd.read_csv(file_prefix + f'{cohort}-metadata.csv')\n",
    "\n",
    "        # get the cases and controls\n",
    "        analysis_cases = analysis_metadata.query(f'disease == \"{disease}\"')['entity:sample_id'].to_list()\n",
    "        analysis_controls = analysis_metadata.query(f'disease == \"control\"')['entity:sample_id'].to_list()\n",
    "        \n",
    "        # note that we have to drop CNVs here (their AFs are inaccurate)\n",
    "        analysis_dosages = analysis_dosages[~analysis_dosages['name'].str.contains('_CNV_')]\n",
    "        \n",
    "        # load the model data. We next add dosages and burdens to this data.\n",
    "        model_data = metadata.set_index('entity:sample_id').loc[analysis_cases + analysis_controls]\n",
    "        model_data[\"disease\"] = model_data[\"disease\"].replace({disease: 1, \"control\": 0})\n",
    "        \n",
    "        # keep track of all our analyses\n",
    "        exog_cols = []\n",
    "        \n",
    "        ########################\n",
    "        #### ADD SV DOSAGES ####\n",
    "        ########################\n",
    "        \n",
    "        # define an n x m matrix, where n is sv, m is sample, and element is dosage\n",
    "        temp_dosages = analysis_dosages.set_index(\"name\")[analysis_cases + analysis_controls]\n",
    "        model_data.loc[temp_dosages.columns, list(temp_dosages.index)] = temp_dosages.T\n",
    "        \n",
    "        # all these SVs are analyzed independently\n",
    "        exog_cols += list(temp_dosages.index)\n",
    "        \n",
    "        #######################\n",
    "        #### ADD SV BURDEN ####\n",
    "        #######################\n",
    "        \n",
    "        # add three columns to model data: rare, common, and combined\n",
    "        # these describe how many of each SV is present in that sample around this gene\n",
    "        for rare in [True, False]:\n",
    "\n",
    "            # define a label for the analysis\n",
    "            label = {True: 'rare-burden', False: 'common-burden'}[rare]\n",
    "            exog_cols.append(label)\n",
    "            \n",
    "            # subset to rare or common variants\n",
    "            subanalysis_dosages = analysis_dosages[analysis_dosages['rare'] == rare]\n",
    "\n",
    "            # count the number of samples with at least one non-ref allele\n",
    "            sv_counts = (subanalysis_dosages[analysis_cases + analysis_controls] > 0).sum(axis=0)\n",
    "\n",
    "            model_data.loc[sv_counts.index, label] = sv_counts.astype(int)\n",
    "            \n",
    "        # add rare + common burden\n",
    "        model_data['combined-burden'] = model_data['rare-burden'] + model_data['common-burden']\n",
    "        exog_cols.append('combined-burden')\n",
    "        \n",
    "        # we also try a one-hot encoding\n",
    "        for col in ['rare-burden', 'common-burden', 'combined-burden']:\n",
    "            onehot_col = col + '-onehot'\n",
    "            model_data[onehot_col] = (model_data[col] > 0).astype(int)\n",
    "            exog_cols.append(onehot_col)\n",
    "            \n",
    "        model_data.reset_index().to_csv(file_prefix + f'{cohort}-model-data.csv', index=False)\n",
    "\n",
    "        ####################\n",
    "        #### RUN MODELS ####\n",
    "        ####################\n",
    "        \n",
    "        # store model results\n",
    "        modeling_results = []\n",
    "        \n",
    "        # for each set of covariates, run all our models\n",
    "        for covar_label, covar in covariates.items():\n",
    "            \n",
    "            # if ancestry isn't included as a covariate, we subset to euro samples\n",
    "            if 'ancestry' not in covar_label:\n",
    "                parent_covar_data = model_data.loc[set(model_data.index) & set(euro_samples)].copy()\n",
    "            else:\n",
    "                parent_covar_data = model_data.copy()\n",
    "            \n",
    "            # Run each SV and burden analysis\n",
    "            for col in exog_cols:\n",
    "                \n",
    "                analysis_type = 'burden' if 'burden' in col else 'sv'\n",
    "                \n",
    "                # define data\n",
    "                submodel_data = parent_covar_data[['disease', col] + covar]\n",
    "                \n",
    "                # we drop samples that aren't defined for our variable of interest\n",
    "                # this is usually SVs that can't be genotyped\n",
    "                submodel_data_nonna = submodel_data.dropna(subset=[col]).copy()\n",
    "                \n",
    "                # this has the potential to introduce bias, so we quantify what was dropped\n",
    "                old_counts = submodel_data['disease'].value_counts().loc[[0, 1]]\n",
    "                new_counts = submodel_data_nonna['disease'].value_counts().loc[[0, 1]]\n",
    "                dropped = old_counts - new_counts\n",
    "                \n",
    "                # define x and y for the model\n",
    "                Y = submodel_data_nonna[\"disease\"]\n",
    "                x = sm.tools.add_constant(submodel_data_nonna, has_constant = 'add')\n",
    "                x = x.drop(columns = ['disease'])\n",
    "                \n",
    "                # suppress some warnings\n",
    "                with warnings.catch_warnings():\n",
    "                    warnings.simplefilter(\"ignore\")\n",
    "                    logit_model = sm.Logit(\n",
    "                    Y,\n",
    "                    x,\n",
    "                    ).fit(disp=0, method=\"bfgs\", maxiter=1000)\n",
    "                \n",
    "                # store model results as a row\n",
    "                row = [analysis_type, covar_label, col, dropped.loc[0], dropped.loc[1]]\n",
    "                \n",
    "                # extract p and coeff (change coef to nan if p is nan)\n",
    "                # this happens whenever the model is singular\n",
    "                p = logit_model.pvalues.loc[col]\n",
    "                coef = logit_model.params.loc[col] if not np.isnan(p) else np.nan\n",
    "\n",
    "                row += [coef, p]\n",
    "                    \n",
    "                modeling_results.append(row)\n",
    "        \n",
    "        # combine into one dataframe\n",
    "        modeling_results = pd.DataFrame(modeling_results, columns = ['type', 'covar', 'model', 'controls_dropped', 'cases_dropped', 'coef', 'p'])\n",
    "    \n",
    "        ################################\n",
    "        #### BONFERRONI CORRECT SVS ####\n",
    "        ################################\n",
    "        if cohort == 'discovery':\n",
    "            for covar_label in covariates.keys():\n",
    "\n",
    "                covar_results = modeling_results[(modeling_results['type'] == 'sv') & \n",
    "                                                 (modeling_results['covar'] == covar_label)\n",
    "                                                ]\n",
    "\n",
    "                pvals = covar_results['p']\n",
    "                p_bonferroni = statsmodels.stats.multitest.multipletests(pvals, alpha=0.05, method='bonferroni')[1]\n",
    "\n",
    "                modeling_results.loc[covar_results.index, 'p_bonferroni'] = p_bonferroni\n",
    "        \n",
    "        ################\n",
    "        #### EXPORT ####\n",
    "        ################\n",
    "        modeling_results.to_csv(file_prefix + f'{cohort}-model-results.csv', index=False)\n",
    "    \n",
    "    ####################################\n",
    "    #### DEFINE SIGNIFICANT RESULTS ####\n",
    "    ####################################\n",
    "    \n",
    "    discovery = pd.read_csv(file_prefix + 'discovery-model-results.csv')\n",
    "    \n",
    "    if disease == 'osteosarcoma':\n",
    "        cohorts = [discovery]\n",
    "    else:\n",
    "        validation = pd.read_csv(file_prefix + 'validation-model-results.csv')\n",
    "        cohorts = [discovery, validation]\n",
    "        \n",
    "    # subset results in each cohort down to those that are significant\n",
    "    sig_results = []\n",
    "    for i, cohort in enumerate(cohorts):\n",
    "        cohort_sig_results = []\n",
    "        \n",
    "        for index, row in cohort.iterrows():\n",
    "            \n",
    "            # define our p value cutoff\n",
    "            if i == 0 and row['type'] == 'sv':\n",
    "                p_col = 'p_bonferroni'\n",
    "            else:\n",
    "                p_col = 'p'\n",
    "            \n",
    "            if row[p_col] < 0.05:\n",
    "                cohort_sig_results.append(index)\n",
    "            # get all models, which use different covariates\n",
    "                \n",
    "        # subset the cohort to these results\n",
    "        cohort = cohort.loc[cohort_sig_results]\n",
    "        \n",
    "        sig_results.append(cohort)\n",
    "    \n",
    "    # combine results\n",
    "    if len(cohorts) == 1:\n",
    "        sig_results = sig_results[0].copy()\n",
    "    else:\n",
    "        sig_results = sig_results[0].merge(sig_results[1], on = ['type', 'covar', 'model'], \n",
    "                                           suffixes = ['_discovery', '_validation'], how = 'inner')\n",
    "\n",
    "    if len(sig_results):\n",
    "        sig_results.to_csv(file_prefix + 'significant-results.csv', index=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2293f8f-68d9-4ff9-be0d-a4179c9e9ede",
   "metadata": {},
   "source": [
    "# Plot results\n",
    "\n",
    "Here, we define some plots. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e1ad023e-cab8-4a33-9b86-c004e15a3579",
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_colors = {\"DEL\": \"blue\", \"DUP\": \"red\", \"INS\": bold_10[0]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "e6da4011-25ce-4353-86fa-6f6b8960e08f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jettcrowdis/opt/anaconda3/envs/pediatric-germline-svs-3.7.13/lib/python3.7/site-packages/ipykernel_launcher.py:125: UserWarning: FixedFormatter should only be used together with FixedLocator\n"
     ]
    }
   ],
   "source": [
    "for index, row in genes_to_test.iterrows():\n",
    "    gene, disease = row[['gene', 'disease']].values\n",
    "    \n",
    "    file_prefix = f'data/{folder_name}/{gene}-{disease}/'\n",
    "    \n",
    "    # plot discovery and validation, split by common and rare on same plot\n",
    "    num_plots = 2 if disease == 'osteosarcoma' else 4\n",
    "    gene_fig, gene_axes = plt.subplots(num_plots, 1, figsize=(7, num_plots), gridspec_kw = {'hspace': 0.5})\n",
    "    \n",
    "    # figure for gene burden\n",
    "    burden_fig, burden_axes = plt.subplots(num_plots, 1, figsize=(6, num_plots * 3), gridspec_kw = {'hspace': 0.5})\n",
    "    \n",
    "    plot_num = 0\n",
    "    for sv_type, cohort in itertools.product(['common', 'rare'], ['discovery', 'validation']):\n",
    "\n",
    "        # osteosarcoma has no validation datasets, so we skip\n",
    "        if disease == 'osteosarcoma' and cohort == 'validation':\n",
    "            continue\n",
    "            \n",
    "        #######################\n",
    "        #### LOAD OUR DATA ####\n",
    "        #######################\n",
    "        \n",
    "        # load counts, and subset to rare or common\n",
    "        analysis_counts = pd.read_csv(file_prefix + f\"{cohort}-sv-genotypes.csv\")\n",
    "        analysis_counts = analysis_counts[analysis_counts['rare'] == (sv_type == \"rare\")].reset_index()\n",
    "        \n",
    "        # load model data, strictly for burden testing\n",
    "        model_data = pd.read_csv(file_prefix + f\"{cohort}-model-data.csv\")\n",
    "\n",
    "        # significance results - based on the model with both sex and ancestry\n",
    "        model_results = pd.read_csv(file_prefix + f\"{cohort}-model-results.csv\")\n",
    "        model_results = model_results.query('covar == \"sex-ancestry\"').set_index(\"model\")\n",
    "        \n",
    "        # gene-related data\n",
    "        g_gtf = pd.read_csv(file_prefix + f\"{gene}-gtf.csv\")\n",
    "        \n",
    "        ########################\n",
    "        #### PLOT SV BURDEN ####\n",
    "        ########################\n",
    "        ax = burden_axes[plot_num]\n",
    "\n",
    "        max_count = 0\n",
    "        for label in ['cases', 'controls']:\n",
    "            ind = label == \"cases\"\n",
    "            data = model_data.query(f\"disease == {ind}\")[f\"{sv_type}-burden\"].value_counts()\n",
    "            \n",
    "            if max(data.index) > max_count:\n",
    "                max_count = max(data.index)\n",
    "\n",
    "            for i in range(max(data.index)):\n",
    "                if i not in data.index:\n",
    "                    data.loc[i] = 0\n",
    "\n",
    "            data = data.sort_index()\n",
    "            data = data / data.sum()\n",
    "\n",
    "            base = -0.2 if label == 'cases' else 0.2\n",
    "            x = np.arange(base, max(data.index) + 1 + base, 1)\n",
    "\n",
    "            ax.bar(x=x, height=data, width=0.4, label=label)\n",
    "\n",
    "        aesthetics.strip_axis(ax=ax, x='label', y = 'ignore')\n",
    "        \n",
    "        if plot_num == 1:\n",
    "            ax.legend(bbox_to_anchor = (1, 0))\n",
    "\n",
    "        ax.set_xlabel('SV count')\n",
    "        ax.set_ylabel('Density')\n",
    "        cohort_name = 'GMKF' if cohort == 'discovery' else 'St. Jude'\n",
    "        ax.text(x = 0.5, y = 1.05, s = f'{cohort_name} SVs', \n",
    "                transform = ax.transAxes, ha = 'center', va = 'bottom')\n",
    "        ax.set_xticks(range(max_count + 1))\n",
    "        \n",
    "        if sv_type == 'rare':\n",
    "            ax.set_yscale('log')\n",
    "            \n",
    "        # add p value\n",
    "        coef, p = model_results.loc[f'{sv_type}-burden', ['coef', 'p']].values\n",
    "        s = f\" = {round(coef, 2)}\\np = {p:.2}\"\n",
    "        ax.text(x = 1, y = 0.9, s = s, \n",
    "                transform = ax.transAxes, ha = 'right', va = 'top')\n",
    "        \n",
    "        #######################\n",
    "        #### PLOT THE GENE ####\n",
    "        #######################\n",
    "        ax = gene_axes[plot_num]\n",
    "        gene_data = genes_to_test[(genes_to_test['gene'] == gene) &\n",
    "                                  (genes_to_test['disease'] == disease)].iloc[0]\n",
    "        \n",
    "        # we weirdly have to do this now because the code that plots\n",
    "        # the genes does so with relative sizes\n",
    "        ylim = [-0.02, 0.02] if sv_type == 'rare' else [-1, 1]\n",
    "        yticks = [-0.02, 0, 0.02] if sv_type == 'rare' else [-1, 0, 1]\n",
    "\n",
    "        # plot the genetic elements\n",
    "        color = {'neuroblastoma': mc.palette['neuroblastoma']['main'], 'ewing': mc.palette['ewing']['main'], 'osteosarcoma': mc.palette['osteosarcoma']['main']}[disease]\n",
    "        feature_colors = {\"exon\": color, \"CDS\": color}\n",
    "        window = gene_data.loc[['window_start', 'window_end']].values\n",
    "        feature_span = {\"exon\": ylim[-1] / 4, \"CDS\": ylim[-1] / 2}\n",
    "\n",
    "        ax = sv_plot.plot_genomic_elements(\n",
    "            ax,\n",
    "            g_gtf,\n",
    "            window=window,\n",
    "            feature_span=feature_span,\n",
    "            feature_colors=feature_colors,\n",
    "            label=False,\n",
    "        )\n",
    "        aesthetics.strip_axis(ax=ax, x=\"label\", y=\"ignore\")\n",
    "\n",
    "        # plot the range of interest\n",
    "        ax.set_xlim(\n",
    "            window + np.array([-10000, 10000])\n",
    "        )\n",
    "        ax.spines[\"bottom\"].set_bounds(window)\n",
    "\n",
    "        # add labels\n",
    "        ax.text(s=\"cases\", x=-0.03, y=0.75, transform=ax.transAxes, ha=\"right\", va='center')\n",
    "        ax.text(s=\"controls\", x=-0.03, y=0.25, transform=ax.transAxes, ha=\"right\", va='center')\n",
    "        ax.set_ylabel('AF', labelpad = 15)\n",
    "\n",
    "        # switch to Mb\n",
    "        contig = gene_data.loc[\"chrom\"]\n",
    "        ax.set_xticklabels(ax.get_xticks() / 1e6)\n",
    "        ax.set_xlabel(f\"{contig} (Mb)\")\n",
    "        \n",
    "        ax.set_yticks(yticks)\n",
    "        ax.set_yticklabels(np.abs(yticks))\n",
    "        ax.set_ylim(ylim)\n",
    "        \n",
    "        # add title for gene\n",
    "        # ax.text(x = gene_data.loc[['start', 'end']].mean(), y = ylim[-1], s = gene, \n",
    "        #         fontstyle = 'italic', ha = 'center', fontsize = 8)\n",
    "        \n",
    "        # label for plot\n",
    "        cohort_name = 'GMKF' if cohort == 'discovery' else 'St. Jude'\n",
    "        ax.text(x = 0.01, y = 1, s = f'{cohort_name} SVs', transform = ax.transAxes, ha = 'left')\n",
    "        \n",
    "        ######################\n",
    "        #### PLOT THE SVS ####\n",
    "        ######################\n",
    "        # its much easier to plot everything at once with plt.stem, but that function sucks\n",
    "        # we scale the frequency to fill the plot\n",
    "        case_y = analysis_counts['AF_cases']\n",
    "        control_y = analysis_counts['AF_controls']\n",
    "\n",
    "        # add the points\n",
    "        ax.scatter(x=analysis_counts[\"midpoint\"], y=case_y, color=\"black\", s=3, zorder=3)\n",
    "        ax.scatter(x=analysis_counts[\"midpoint\"], y=-control_y, color=\"black\", s=3, zorder=3)\n",
    "        \n",
    "        ################################\n",
    "        #### DENOTE SIGNIFICANT SVS ####\n",
    "        ################################\n",
    "        \n",
    "        # add the line denoting the SV\n",
    "        legend_handles = {}\n",
    "        for i in range(len(analysis_counts)):\n",
    "            \n",
    "            loc = analysis_counts[\"midpoint\"][i]\n",
    "            \n",
    "            # sometimes SVs are shown that are way outside the boundaries\n",
    "            if not (window[0] <= loc <= window[1]):\n",
    "                continue\n",
    "            \n",
    "            name = analysis_counts[\"name\"][i]\n",
    "\n",
    "            case, control = case_y[i], control_y[i]\n",
    "\n",
    "            ax.plot(\n",
    "                [loc, loc],\n",
    "                [case, -control],\n",
    "                color=\"black\",\n",
    "                linewidth=0.5,\n",
    "                zorder=3,\n",
    "            )\n",
    "            \n",
    "            # look up the p value of the sv, if it exists\n",
    "            if name in model_results.index:\n",
    "                p_col = 'p_bonferroni' if cohort == 'discovery' else \"p\"\n",
    "                p_val, coeff = model_results.loc[name, [p_col, \"coef\"]]\n",
    "\n",
    "                if p_val < 0.05:\n",
    "\n",
    "                    y = case if coeff > 0 else -control\n",
    "                    text_y = y + 0.3 * ylim[-1] if coeff > 0 else y - 0.3 * ylim[-1]\n",
    "\n",
    "                    sv_type = name.split('_')[1]\n",
    "                    color = sv_colors.get(sv_type)\n",
    "\n",
    "                    scatter = ax.scatter(\n",
    "                        [loc], [y], color=color, s=10, zorder=3, label=sv_type\n",
    "                    )\n",
    "\n",
    "                    legend_handles[sv_type] = scatter\n",
    "\n",
    "                    # add text for coeff\n",
    "                    ax.text(\n",
    "                        x=loc,\n",
    "                        y=text_y,\n",
    "                        s=f\" = {round(coeff, 2)}, p = {p_val:.1}\",\n",
    "                        ha=\"center\",\n",
    "                        va=\"center\",\n",
    "                    )\n",
    "        \n",
    "        plot_num += 1\n",
    "        \n",
    "    gene_fig.savefig(file_prefix + 'sv-plot.pdf')\n",
    "    plt.close(gene_fig)\n",
    "    \n",
    "    burden_fig.savefig(file_prefix + 'sv-burden.png')\n",
    "    plt.close(burden_fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "7938f150-fa43-4728-992b-d6fcbba7ce8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'neuroblastoma': '#9D4DAD', 'ewing': '#4E7AEB', 'osteosarcoma': '#DBAE42'}"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{'neuroblastoma': mc.palette['neuroblastoma']['main'], 'ewing': mc.palette['ewing']['main'], 'osteosarcoma': mc.palette['osteosarcoma']['main']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "53e59407-8da7-469b-ae43-d232c0375eba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'exon': '#DBAE42', 'CDS': '#DBAE42'}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_colors"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47785e15-2034-4e60-97bd-f42f2dcd0542",
   "metadata": {},
   "source": [
    "# Quality control"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c60af273-bd67-475f-ac7f-627b354043e4",
   "metadata": {},
   "source": [
    "Here, we briefly extract some QC metrics for our \"significant\" SVs, and we also batch these SVs together."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e7f5f448-4819-4afd-9d74-29f41bab2e91",
   "metadata": {},
   "outputs": [],
   "source": [
    "qc_cols = [line.strip() for line in open('ref/qc-columns-order.txt')]\n",
    "\n",
    "sv_cols = ['name', 'svtype', 'ALGORITHMS', 'EVIDENCE', 'SL_MEAN', 'SL_MAX', 'NCR', 'HG38_REF_PATCH_LOCUS']\n",
    "sv_cols += [c for c in svs['discovery'].columns if c.startswith('gnomad') and '_CN_' not in c]\n",
    "\n",
    "ucsc_tracks = [\"simpleRepeat\", \"genomicSuperDups\", \"rmsk\", \"microsat\"]\n",
    "\n",
    "df_cols = qc_cols + sv_cols + ucsc_tracks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5404fd86-cc84-4ba6-9234-ba372e3491c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "significant_results_qc = []\n",
    "for index, row in genes_to_test.iterrows():\n",
    "    gene, disease = row[['gene', 'disease']].values\n",
    "    \n",
    "    file_prefix = f'data/{folder_name}/{gene}-{disease}/'\n",
    "    \n",
    "    ######################\n",
    "    #### READ IN DATA ####\n",
    "    ######################\n",
    "    \n",
    "    # begin by loading the significant results\n",
    "    if os.path.exists(file_prefix + 'significant-results.csv'):\n",
    "        significant_results = pd.read_csv(file_prefix + 'significant-results.csv')\n",
    "    else:\n",
    "        continue\n",
    "        \n",
    "    significant_svs = significant_results.query('type == \"sv\" & covar == \"sex-ancestry\"')['model'].to_list()\n",
    "\n",
    "    if len(significant_svs) == 0:\n",
    "        continue\n",
    "    \n",
    "    #################################################\n",
    "    #### COMBINE SV GENOTYPES WITH MODEL RESULTS ####\n",
    "    #################################################\n",
    "    \n",
    "    # we load in discovery and validation separately, to better enable side by side comparison\n",
    "    cohort_significant_results = []\n",
    "    for cohort in ['discovery', 'validation']:\n",
    "        if disease == 'osteosarcoma' and cohort == 'validation':\n",
    "            continue\n",
    "\n",
    "        cohort_results = pd.read_csv(file_prefix + f'{cohort}-model-results.csv')\n",
    "        cohort_results = cohort_results[(cohort_results['model'].isin(significant_svs)) &\n",
    "                                        (cohort_results['covar'] == \"sex-ancestry\")]\n",
    "        \n",
    "        # some reformating of the model results\n",
    "        cohort_results = cohort_results.drop(columns = ['type', 'covar'])\n",
    "        cohort_results = cohort_results.rename(columns = {'model': 'name'})\n",
    "        cohort_results.insert(1, 'cohort', cohort)\n",
    "        \n",
    "        cohort_genotypes = pd.read_csv(file_prefix + f'{cohort}-sv-genotypes.csv')\n",
    "        cohort_genotypes_condensed = cohort_genotypes[['name', 'gene', '#chr', 'start', 'end']].copy()\n",
    "        cohort_genotypes_condensed[['cases_gt']] = list(zip(cohort_genotypes[['cases_0', 'cases_1', 'cases_2']].values))\n",
    "        cohort_genotypes_condensed[['controls_gt']] = list(zip(cohort_genotypes[['controls_0', 'controls_1', 'controls_2']].values))\n",
    "        cohort_genotypes_condensed[['AF_cases', 'AF_controls']] = cohort_genotypes[['AF_cases', 'AF_controls']]\n",
    "        \n",
    "        merged_sv_data = cohort_results.merge(cohort_genotypes_condensed, on = ['name'])\n",
    "        \n",
    "        cohort_significant_results.append(merged_sv_data)\n",
    "        \n",
    "    cohort_significant_results = pd.concat(cohort_significant_results)\n",
    "    \n",
    "    cohort_significant_results['disease'] = disease\n",
    "    \n",
    "    #############################################\n",
    "    #### ADD SV QUALITY METRICS FROM GATK-SV ####\n",
    "    #############################################\n",
    "    \n",
    "    sv_data = pd.read_csv(file_prefix + 'discovery-svs.csv')\n",
    "    sv_char = sv_data[sv_cols]\n",
    "    \n",
    "    cohort_significant_results = cohort_significant_results.merge(sv_char, on = 'name')\n",
    "    \n",
    "    ##########################################################\n",
    "    #### ADD IN CHARACTERISTICS ABOUT THIS GENOMIC REGION ####\n",
    "    ##########################################################\n",
    "    \n",
    "    for track in ucsc_tracks:\n",
    "        \n",
    "        track_results = []\n",
    "        copy = 1 if disease == 'osteosarcoma' else 2\n",
    "        for index, row in cohort_significant_results.drop_duplicates(subset = 'name').iterrows():\n",
    "            \n",
    "            chrom, start, end = row[['#chr', 'start', 'end']].values\n",
    "            \n",
    "            if start == end:\n",
    "                end += 1\n",
    "            \n",
    "            overlap = helper.fetch_track_overlap(track, (chrom, start, end))\n",
    "            \n",
    "            track_results += [overlap] * copy\n",
    "        \n",
    "        cohort_significant_results[track] = track_results\n",
    "        \n",
    "        \n",
    "    cohort_significant_results = cohort_significant_results[df_cols]\n",
    "\n",
    "    cohort_significant_results.to_csv(file_prefix + 'significant-results-qc.csv', index=False)\n",
    "        \n",
    "    significant_results_qc.append(cohort_significant_results)\n",
    "    \n",
    "significant_results_qc = pd.concat(significant_results_qc).reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4989386-6800-49c4-883c-705ffa7ad01f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pediatric-germline-svs-3.7.13",
   "language": "python",
   "name": "pediatric-germline-svs-3.7.13"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
