{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f653cbc1-3ece-4e0d-8bfd-eaff2126d885",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gseapy as gp\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as stats\n",
    "import glob\n",
    "from biomart import BiomartServer\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "import seaborn as sns\n",
    "import gseapy as gp\n",
    "\n",
    "from sklearn import decomposition\n",
    "from sklearn import preprocessing\n",
    "\n",
    "from scripts import aesthetics\n",
    "\n",
    "aesthetics.activate_paper_rcParams()\n",
    "\n",
    "pd.set_option(\"display.max_columns\", 1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12249b0a-d5b6-45b7-b8f1-e21c7579386a",
   "metadata": {},
   "source": [
    "# RNA SV Impact\n",
    "\n",
    "In `neuroblastoma-cwas`, we've discovered that counterintuitively, the SVs we've discovered lead to _increases_ in expression. This is paradoxical from the impact that we expect from the germline context.\n",
    "\n",
    "I'd like to try examining the impact of _ALL_ SVs on expression. This is going to take a little bit, but it should give us a sense for how some of these SVs are operating"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05ab3a35-0f2c-48bb-af42-bcbe30e09da0",
   "metadata": {},
   "source": [
    "# Read in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8f0dfff5-0eca-49b2-b2bb-c0a88ceabf3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define SVs and dosages\n",
    "sv_path = \"gs://vanallen-pedsv-analysis/beds/PedSV.v2.5.2.full_cohort.analysis_samples.sites.bed.gz\"\n",
    "dosages_path = \"gs://vanallen-pedsv-analysis/beds/PedSV.v2.5.2.full_cohort.analysis_samples.allele_dosages.bed.gz\"\n",
    "\n",
    "# define metadata\n",
    "metadata_path = \"gs://vanallen-pedsv-analysis/sample_info/PedSV.v2.5.2.cohort_metadata.w_control_assignments.tsv.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2a049698-9917-4520-ad06-51ea7c39b4e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# list of genes that are annotated - drop ensembl IDs\n",
    "gene_ref = pd.read_csv(\"ref/gencode_hg38_protein_coding_genes_for_annotation_7_31_23.txt\")\n",
    "gene_ref = gene_ref[~gene_ref['value'].str.startswith('ENSG00')]\n",
    "gene_ref = gene_ref['value'].tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bfad1ed-0377-4af7-9624-1584bee9a350",
   "metadata": {},
   "source": [
    "Load metadata and SVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "706c7c8c-50ba-4fd0-a392-2aa7133562c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(\n",
    "    metadata_path,\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "\n",
    "# add a sex label to metadata\n",
    "metadata[\"sex\"] = (metadata[\"chrX_CopyNumber\"].round() < 2).astype(int)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "712f495b-c514-451d-980f-a339306c0768",
   "metadata": {},
   "source": [
    "Now we load the TPMs to establish what samples are in our dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "51637ca9-a3d6-4162-9c24-f9c6480912ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# generated by a now old notebook - see `old/pgsvs-neuroblastoma-cwas-rna`\n",
    "merged_tpms = pd.read_csv('data/merged_neuroblastoma_tpms.tsv', sep = '\\t')\n",
    "merged_counts = pd.read_csv('data/merged_neuroblastoma_counts.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4fcd892f-b027-41d6-b3d4-615b6fd78315",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "209"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = list(merged_tpms.columns[3:])\n",
    "len(samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "645cfe2d-bc72-4c3a-9a22-6d55b8848c54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "89"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "included_samples = metadata[(metadata['entity:sample_id'].isin(samples)) &\n",
    "                            (metadata['neuroblastoma_case'] == True)]['entity:sample_id'].tolist()\n",
    "len(included_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8c6c6a95-9727-482e-bb0d-42c6d25eb28e",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_tpms = merged_tpms[['gene_id', 'ENSEMBL_ID', 'gene_symbol'] + included_samples]\n",
    "merged_counts = merged_counts[['gene_id', 'ENSEMBL_ID', 'gene_symbol'] + included_samples]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "866f8de4-ff6c-4971-8a44-717508db7641",
   "metadata": {},
   "source": [
    "These are the samples that are in our dosage matrix (others were removed upstream for poor QC). Now we load the SVs. We only keep SVs and dosages that are in our TPM matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "51a5ae5b-27b3-497f-8025-19a7f5620e7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "###############\n",
    "### Dosages ###\n",
    "###############\n",
    "dosage_cols = pd.read_csv(\n",
    "    dosages_path,\n",
    "    sep=\"\\t\",\n",
    "    index_col=False,\n",
    "    nrows = 0\n",
    ")\n",
    "\n",
    "usecols = ['#chr', 'start', 'end', 'ID'] + [s for s in samples if s in dosage_cols.columns]\n",
    "\n",
    "# load in the dosage data for the samples in TPMs\n",
    "dosages = pd.read_csv(\n",
    "    dosages_path,\n",
    "    sep=\"\\t\",\n",
    "    index_col=False,\n",
    "    usecols = usecols\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ac0f9a15-c3c9-4daa-a897-52f35b80c95a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# identify SVs that have non-zero counts in the dosage matrix \n",
    "cohort_dosages = dosages.iloc[:, 4:]\n",
    "\n",
    "# SVs that are poorly genotyped in more than 20% of samples will be excluded\n",
    "nan_svs = np.isnan(cohort_dosages).mean(axis = 1) > 0.20\n",
    "\n",
    "# SVs that have no counts will be excluded\n",
    "nocount_svs = (cohort_dosages.fillna(0) != 0).sum(axis = 1) == 0\n",
    "\n",
    "kept_svs = ~(nan_svs | nocount_svs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "abfa62a1-f656-4638-9f3c-c45855cbc8e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "cohort_dosages = dosages[kept_svs]\n",
    "\n",
    "svs_to_analyze = cohort_dosages['ID'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "40c60881-f88b-490e-9693-d80ce674b47a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jettcrowdis/opt/anaconda3/envs/pediatric-germline-svs-3.7.13/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3457: DtypeWarning: Columns (10,34,35,37,39,41,45,46,47,48,72,73,87,88,106,107,121,122,136,137,155,156,170,171,185,186,204,205,219,220,234,235,253,254,268,269,283,284,302,303,366,367,381,382,400,401,415,416,430,431,449,450,464,465,479,480,498,499,513,514,528,529,547,548,562,563,577,578,596,597,645,646,647,648,649,650,690,691,692,693,694,695,713,714,728,729,743,744,758,759,773,774,788,789,803,804,818,819,833,834,848,849,863,864,878,879,893,894,908,909,923,924,931,937,943) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    }
   ],
   "source": [
    "###############\n",
    "##### SVs #####\n",
    "###############\n",
    "svs = pd.read_csv(\n",
    "    sv_path,\n",
    "    sep=\"\\t\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "48b7d5e0-c915-4910-87fe-5c069de6de4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "svs = svs[svs['name'].isin(svs_to_analyze)].reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "46330d20-ff4c-473f-98e4-ae65f5e41f2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "svs.to_csv('data/svs-for-sv-rna-analysis.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "253e8b96-1c1c-4886-9e88-df5a25431e70",
   "metadata": {},
   "source": [
    "We also do some filtering on the TPM matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a7985d45-25a3-42db-a26f-630ede69bee6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# keep genes that have at least 10 counts in at least 5 samples (following DESEQ's recommendations)\n",
    "expressed = (merged_counts[included_samples] >= 10).sum(axis = 1) >= 5\n",
    "\n",
    "# and fewer than 20% of samples can have zero expression\n",
    "min_greater_than_zero = (merged_tpms[included_samples] == 0).mean(axis = 1) <= 0.20\n",
    "\n",
    "# and we remove genes with duplicate gene symbols, as we won't be able to tell (without annoying difficulty)\n",
    "# which gene is affected by an SV\n",
    "multigene = merged_tpms['gene_symbol'].value_counts()\n",
    "multigene = multigene[multigene > 1].index\n",
    "not_multi = ~(merged_tpms['gene_symbol'].isin(multigene))\n",
    "\n",
    "kept_rows = expressed & min_greater_than_zero & not_multi\n",
    "\n",
    "kept_genes = merged_counts[kept_rows]['gene_symbol'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "84e0ae43-324c-496f-8340-35ef7a27d4ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "23094"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(kept_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3997d348-9e20-449e-874a-468ca6472e5a",
   "metadata": {},
   "source": [
    "And only keep genes in our gene reference, obviously"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e8916101-d255-4003-a6e6-701462aaba2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "kept_genes = sorted(set(kept_genes) & set(gene_ref))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "12c11c1a-b60a-471a-9f0b-1278237653fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14823"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(kept_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7090b047-f11f-4230-84f6-de2bcab7e936",
   "metadata": {},
   "source": [
    "With that, let's get into it!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "31474bce-c20c-4e25-a664-9dde8a3a38a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_tpms = merged_tpms[merged_tpms['gene_symbol'].isin(kept_genes)]\n",
    "merged_counts = merged_counts[merged_counts['gene_symbol'].isin(kept_genes)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e422a96c-b911-44e6-b617-4ab032744c2c",
   "metadata": {},
   "source": [
    "# Think briefly about what information we want to carry forward\n",
    "\n",
    "We want to know how SVs affect the expression of genes around them. Most of our SVs will be singleton or very rare. We need to think about how to systematically examine gene expressions, keeping the info that we want. We have a bit of an expanding problem, in that:\n",
    "\n",
    "1. A given SV can affect multiple genes. I expect the most interesting effects to be on single gene SVs, but we should look at all of them.\n",
    "2. Multiple samples can have an SV. Common SVs can affect multiple SVs and could probably be handled by a MWU, but singleton or rare SVs cannot.\n",
    "3. SVs can have different dosages. We'll be tracking CNVs here, which can vary dramatically in terms of their dosages.\n",
    "\n",
    "It's very difficult to combine all this information into a single dataframe. In addition, we occasionally want different information. For common SVs, we can directy compare expression, but for rare and singleton SVs, a rank-based approach is likely to work better.\n",
    "\n",
    "I'm just going to make this up as I go along."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4364cc2-051b-4963-9236-1da0fd4ad0c4",
   "metadata": {},
   "source": [
    "# Identify SVs that affect genes\n",
    "\n",
    "Here, we'll identify SVs that nominally affect genes. At the end of this process, I want to end up with an SV x gene flat dataframe, where each row contains information about the SV and its relationship to the gene. This will carry forward a lot of redundant information about the SV, but that's ok."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e0ae7102-27ad-471f-a9ef-a75e533ddeca",
   "metadata": {},
   "outputs": [],
   "source": [
    "coding_cols = ['PREDICTED_COPY_GAIN', 'PREDICTED_INTRAGENIC_EXON_DUP', 'PREDICTED_LOF', 'PREDICTED_PARTIAL_EXON_DUP']\n",
    "noncoding_cols = ['PREDICTED_NEAREST_TSS', 'PREDICTED_INTRONIC', 'PREDICTED_PROMOTER', 'PREDICTED_UTR']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8a964498-db00-4074-8bca-5f579fa5521c",
   "metadata": {},
   "outputs": [],
   "source": [
    "svs_that_affect_genes = (~pd.isnull(svs[coding_cols + noncoding_cols])).sum(axis = 1) > 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ad0a1a69-4698-4538-ad90-75e4b56e8beb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(31397, 0.9932616260677001)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_svs = svs[svs_that_affect_genes]\n",
    "gene_svs.shape[0], gene_svs.shape[0] / svs.shape[0] "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "235d3eb0-4b1e-42f7-8fa2-37a143872e03",
   "metadata": {},
   "source": [
    "Somewhat surprisingly, apparently 99% of these SVs are associated with a gene. I suppose these are likely all noncoding effects for distant genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c2356842-af3d-4215-af7a-6b07336b85a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "594"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coding_svs = (~pd.isnull(svs[coding_cols])).sum(axis = 1) > 0\n",
    "coding_svs.sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d055236-d4df-4e94-8f50-c439378344d6",
   "metadata": {},
   "source": [
    "That's better. Alright, let's cobble this together."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "95969be4-be43-44cf-9a78-eba5d6c3a981",
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_gene_df = []\n",
    "for sv_name, row in gene_svs[['name'] + coding_cols + noncoding_cols].set_index('name').iterrows():\n",
    "    row = row[~pd.isnull(row)]\n",
    "    \n",
    "    for genic_rel, gene_list in row.iteritems():\n",
    "        genic_cat = 'coding' if genic_rel in coding_cols else 'non-coding'\n",
    "        \n",
    "        for gene in gene_list.split(','):\n",
    "            sv_gene_df.append([sv_name, genic_cat, genic_rel, gene])\n",
    "        \n",
    "sv_gene_df = pd.DataFrame(sv_gene_df, columns = ['name', 'sv_effect', 'genic_relationship', 'gene'])\n",
    "\n",
    "for sv_effect in ['coding', 'non-coding']:\n",
    "    counts = pd.DataFrame(sv_gene_df.query(f'sv_effect == \"{sv_effect}\"').groupby('name').size().astype(int), \n",
    "                          columns = [f'sv_{sv_effect}_counts']).reset_index()\n",
    "    sv_gene_df = sv_gene_df.merge(counts, how = 'left')\n",
    "    sv_gene_df[f'sv_{sv_effect}_counts'] = sv_gene_df[f'sv_{sv_effect}_counts'].fillna(0)\n",
    "\n",
    "# add some info about the SVs themselves\n",
    "sv_gene_df = sv_gene_df.merge(gene_svs[['#chrom', 'start', 'end', 'name', 'svtype']], on = ['name'], how = 'left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e41f4369-8048-4180-b731-e9ccb4e31a71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>sv_effect</th>\n",
       "      <th>genic_relationship</th>\n",
       "      <th>gene</th>\n",
       "      <th>sv_coding_counts</th>\n",
       "      <th>sv_non-coding_counts</th>\n",
       "      <th>#chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>svtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PedSV.2.5.2_CNV_chr1_1</td>\n",
       "      <td>non-coding</td>\n",
       "      <td>PREDICTED_NEAREST_TSS</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12000</td>\n",
       "      <td>30001</td>\n",
       "      <td>CNV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PedSV.2.5.2_DUP_chr1_1</td>\n",
       "      <td>non-coding</td>\n",
       "      <td>PREDICTED_NEAREST_TSS</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12000</td>\n",
       "      <td>40001</td>\n",
       "      <td>DUP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     name   sv_effect     genic_relationship   gene  \\\n",
       "0  PedSV.2.5.2_CNV_chr1_1  non-coding  PREDICTED_NEAREST_TSS  OR4F5   \n",
       "1  PedSV.2.5.2_DUP_chr1_1  non-coding  PREDICTED_NEAREST_TSS  OR4F5   \n",
       "\n",
       "   sv_coding_counts  sv_non-coding_counts #chrom  start    end svtype  \n",
       "0               0.0                   1.0   chr1  12000  30001    CNV  \n",
       "1               0.0                   1.0   chr1  12000  40001    DUP  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sv_gene_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "7c06a48b-6e2f-4737-941d-c85b33215079",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9909863999745199"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(sv_gene_df.groupby(['name']).size() == 1).sum() / gene_svs.shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79eec55d-2b85-47dc-97bb-10352b3409b7",
   "metadata": {},
   "source": [
    "So we can see that the majority of SVs actually only affect one gene (and this remains true for coding genes too).\n",
    "\n",
    "I'd like to add some info about AF in our cohort subset here. Calculating these are _EXTREMELY_ annoying, since we have mixed CNVs and short SVs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "709041ff-1e72-4b37-b332-461d51977e68",
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_allele_fractions = []\n",
    "\n",
    "for index, row in cohort_dosages.iterrows():\n",
    "    \n",
    "    dosages = row.iloc[4:].dropna()\n",
    "    \n",
    "    if '_CNV' in row['ID']:\n",
    "        average_cn = np.mean(dosages)\n",
    "        af = np.nan\n",
    "        num_samples_affected = (dosages != 0).sum()\n",
    "        \n",
    "    else:\n",
    "        allele_counts = dosages.value_counts().reindex([0, 1, 2]).fillna(0)\n",
    "        alt = (allele_counts * np.array([0, 1, 2])).sum()\n",
    "        ref = len(dosages) * 2\n",
    "        \n",
    "        af = alt/ref\n",
    "        average_cn = np.nan\n",
    "        num_samples_affected = (dosages > 0).sum()\n",
    "        \n",
    "    num_genotyped = len(dosages)\n",
    "        \n",
    "    sv_allele_fractions.append([row['ID'], num_samples_affected, num_genotyped, af, average_cn])\n",
    "    \n",
    "sv_allele_fractions = pd.DataFrame(sv_allele_fractions, columns = ['name', 'num_samples_nonzero', 'num_samples_genotyped', 'af', 'average_cn'])\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1fe9a1d4-b96b-4efa-86c5-0da5b080ef63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>num_samples_nonzero</th>\n",
       "      <th>num_samples_genotyped</th>\n",
       "      <th>af</th>\n",
       "      <th>average_cn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PedSV.2.5.2_CNV_chr1_1</td>\n",
       "      <td>46</td>\n",
       "      <td>89</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.741573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PedSV.2.5.2_DUP_chr1_1</td>\n",
       "      <td>2</td>\n",
       "      <td>81</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     name  num_samples_nonzero  num_samples_genotyped  \\\n",
       "0  PedSV.2.5.2_CNV_chr1_1                   46                     89   \n",
       "1  PedSV.2.5.2_DUP_chr1_1                    2                     81   \n",
       "\n",
       "         af  average_cn  \n",
       "0       NaN    0.741573  \n",
       "1  0.012346         NaN  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sv_allele_fractions.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a0c7ee9d-5811-4ec7-8819-31cf1e4cf014",
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_gene_df = sv_gene_df.merge(sv_allele_fractions, on = 'name', how = 'left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "79cf2da1-94fc-4828-959e-67872381a481",
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_gene_df = sv_gene_df[['name', '#chrom', 'start', 'end', 'svtype', 'sv_coding_counts', 'sv_non-coding_counts', \n",
    "                         'num_samples_nonzero', 'num_samples_genotyped', 'af', 'average_cn', 'sv_effect', 'genic_relationship', 'gene']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "59a4c426-2c09-4d90-bc82-16051f22d3a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>#chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>svtype</th>\n",
       "      <th>sv_coding_counts</th>\n",
       "      <th>sv_non-coding_counts</th>\n",
       "      <th>num_samples_nonzero</th>\n",
       "      <th>num_samples_genotyped</th>\n",
       "      <th>af</th>\n",
       "      <th>average_cn</th>\n",
       "      <th>sv_effect</th>\n",
       "      <th>genic_relationship</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PedSV.2.5.2_CNV_chr1_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12000</td>\n",
       "      <td>30001</td>\n",
       "      <td>CNV</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>46</td>\n",
       "      <td>89</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.741573</td>\n",
       "      <td>non-coding</td>\n",
       "      <td>PREDICTED_NEAREST_TSS</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PedSV.2.5.2_DUP_chr1_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12000</td>\n",
       "      <td>40001</td>\n",
       "      <td>DUP</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>81</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>NaN</td>\n",
       "      <td>non-coding</td>\n",
       "      <td>PREDICTED_NEAREST_TSS</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     name #chrom  start    end svtype  sv_coding_counts  \\\n",
       "0  PedSV.2.5.2_CNV_chr1_1   chr1  12000  30001    CNV               0.0   \n",
       "1  PedSV.2.5.2_DUP_chr1_1   chr1  12000  40001    DUP               0.0   \n",
       "\n",
       "   sv_non-coding_counts  num_samples_nonzero  num_samples_genotyped        af  \\\n",
       "0                   1.0                   46                     89       NaN   \n",
       "1                   1.0                    2                     81  0.012346   \n",
       "\n",
       "   average_cn   sv_effect     genic_relationship   gene  \n",
       "0    0.741573  non-coding  PREDICTED_NEAREST_TSS  OR4F5  \n",
       "1         NaN  non-coding  PREDICTED_NEAREST_TSS  OR4F5  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sv_gene_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c71778f2-cb9f-413c-8f63-1b752fc16062",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(31885, 14)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sv_gene_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77f85384-d07b-43ce-997e-d828600cc954",
   "metadata": {},
   "source": [
    "# Time to look at the RNA\n",
    "\n",
    "I want to do this fully systematically, and then afterwards we can go back and reassess. We have a few different issues that we need to handle:\n",
    "\n",
    "1. Different numbers of samples affected. Some SVs affect many samples, and some affect very few.\n",
    "2. SV dosages. While rare SVs will usually only have `0` or `1` as dosages, others will have `1/2`. \n",
    "3. `CNV`s have way crazier dosages. They should be modelled more holistically.\n",
    "4. Modelling TPMs/counts as an outcome is really annoying (requiring something like DESeq optimally), and non-parametric approaches like the MWU cannot handle multiple covariates.\n",
    "\n",
    "This is a ton to keep track of, as we want to handle these scenarios differently. Here's what we'll do. For each `SV` and `gene` pair, we'll generate the following:\n",
    "\n",
    "1. The average rank of affected samples (non-zero allele). This will be incorrect for CNVs that can have a range of copy numbers that includes negative ones.\n",
    "    * We also include the expression of affected vs. not samples\n",
    "2. MWU test between affected (non-zero allele) and not. This will be incorrect for SVs that have few samples and for SVs that have more dosages.\n",
    "3. An ordinal logistic regression model, incorporating dosage and using the ranks of gene expression as the outcome. This will be broadly incorrect because it does not appropriately model counts.\n",
    "4. The average expression of the gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "d3760d42-a889-45e6-8888-675ce21f7744",
   "metadata": {},
   "outputs": [],
   "source": [
    "from statsmodels.miscmodels.ordinal_model import OrderedModel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "d21b2281-8cf2-4af2-bb2a-a45131495122",
   "metadata": {},
   "outputs": [],
   "source": [
    "# we get to begin by subsetting our gene df to genes that are in our TPM matrix\n",
    "sv_gene_df = sv_gene_df[sv_gene_df['gene'].isin(merged_tpms['gene_symbol'])]\n",
    "\n",
    "# format for analysis\n",
    "analysis_dosages = cohort_dosages.set_index('ID').iloc[:, 3:]\n",
    "analysis_tpms = merged_tpms.set_index('gene_symbol').iloc[:, 2:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "26a8adba-ef15-48a3-b50d-fd756eb34d04",
   "metadata": {},
   "outputs": [],
   "source": [
    "# do some exporting for easy reading in later\n",
    "merged_tpms.to_csv('data/tpms-for-sv-rna-analysis.csv', index=False)\n",
    "cohort_dosages.to_csv('data/dosages-for-sv-rna-analysis.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "f6f8efa9-91cd-4181-a475-37e69e8f2b80",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0, 500, 1000, 1500, 2000, 2500, 3000, 3500, 4000, 4500, 5000, 5500, 6000, 6500, 7000, 7500, 8000, 8500, 9000, 9500, 10000, 10500, 11000, 11500, 12000, 12500, 13000, 13500, 14000, 14500, 15000, 15500, 16000, 16500, 17000, 17500, 18000, 18500, 19000, 19500, 20000, 20500, 21000, 21500, 22000, 22500, 23000, 23500, 24000, 24500, 25000, "
     ]
    }
   ],
   "source": [
    "results = []\n",
    "for i, (index, row) in enumerate(sv_gene_df.iterrows()):\n",
    "    if i % 500 == 0:\n",
    "        print(i, end = ', ')\n",
    "\n",
    "    gene = row['gene']\n",
    "    sv = row['name']\n",
    "    gr = row['genic_relationship']\n",
    "\n",
    "    storage_row = [sv, gene, gr]\n",
    "\n",
    "    # get the dosages\n",
    "    sv_dosages = analysis_dosages.loc[sv].dropna()\n",
    "\n",
    "    # get the expression\n",
    "    gene_expression = analysis_tpms.loc[gene, sv_dosages.index]\n",
    "    mean_expression = gene_expression.mean()\n",
    "\n",
    "    storage_row.append(mean_expression)\n",
    "\n",
    "    ###########################\n",
    "    ### RANK-BASED APPROACH ###\n",
    "    ###########################\n",
    "    affected_samples = sv_dosages[sv_dosages > 0].index\n",
    "    unaffected_samples = sv_dosages[sv_dosages <= 0].index\n",
    "\n",
    "    # rank the expression\n",
    "    expression_ranks = (gene_expression.rank(ascending = False) - 1)\n",
    "    norm_expression_ranks =  expression_ranks / (len(gene_expression) - 1)\n",
    "\n",
    "    avg_affected_rank = norm_expression_ranks.loc[affected_samples].mean()\n",
    "\n",
    "    # store this data\n",
    "    storage_row += [len(affected_samples), avg_affected_rank]\n",
    "\n",
    "    ################\n",
    "    ### MWU TEST ###\n",
    "    ################\n",
    "    gene_exp_affected = gene_expression.loc[affected_samples]\n",
    "    gene_exp_unaffected = gene_expression.loc[unaffected_samples]\n",
    "    try:\n",
    "        p = stats.mannwhitneyu(gene_exp_affected, gene_exp_unaffected)[1]\n",
    "    except:\n",
    "        p = np.nan\n",
    "\n",
    "    storage_row += [gene_exp_affected.mean(), gene_exp_unaffected.mean(), p]\n",
    "\n",
    "    ###################################\n",
    "    ### LOGISTIC ORDINAL REGRESSION ###\n",
    "    ###################################\n",
    "    data = pd.DataFrame([expression_ranks, sv_dosages], index = ['rank', 'dose']).T\n",
    "\n",
    "    # have to convert rank to an ordered variable\n",
    "    data['rank'] = pd.Categorical(data['rank'], categories=sorted(set(data['rank']))[::-1], ordered=True)\n",
    "\n",
    "    try:\n",
    "        mod_log = OrderedModel(data['rank'],\n",
    "                       data[['dose']],\n",
    "                       distr='logit')\n",
    "\n",
    "        res_log = mod_log.fit(method='bfgs', disp=False)\n",
    "\n",
    "        p = res_log.pvalues.loc['dose']\n",
    "        coef = res_log.params.loc['dose']\n",
    "\n",
    "    except:\n",
    "        p, coef = np.nan, np.nan\n",
    "\n",
    "    storage_row += [p, coef]\n",
    "\n",
    "    results.append(storage_row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "4b6bf9b6-c7cf-4ecb-a965-542ab13719cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = ['name', 'gene', 'genic_relationship', 'mean_exp', 'num_greater_0_dosage', 'mean_greater_0_dosage_rank', 'mean_greater_0_exp', 'mean_leq_0_exp', \n",
    "           'mwu_p', 'ordinal_p', 'ordinal_coef']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "893edd38-7c02-42cd-ade5-61db0b9456d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "results = pd.DataFrame(results, columns = columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75a3da08-321d-49f2-a989-ec23b2bd3571",
   "metadata": {},
   "source": [
    "And combine it with some of the SV characteristics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "c67aad73-6613-4be3-a9e0-c87eb1847ed4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "      <th>mean_exp</th>\n",
       "      <th>num_greater_0_dosage</th>\n",
       "      <th>mean_greater_0_dosage_rank</th>\n",
       "      <th>mean_greater_0_exp</th>\n",
       "      <th>mean_leq_0_exp</th>\n",
       "      <th>mwu_p</th>\n",
       "      <th>ordinal_p</th>\n",
       "      <th>ordinal_coef</th>\n",
       "      <th>genic_relationship</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PedSV.2.5.2_CNV_chr1_6</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>0.266629</td>\n",
       "      <td>15</td>\n",
       "      <td>0.507955</td>\n",
       "      <td>0.247333</td>\n",
       "      <td>0.270541</td>\n",
       "      <td>0.912668</td>\n",
       "      <td>0.582179</td>\n",
       "      <td>-0.188116</td>\n",
       "      <td>PREDICTED_NEAREST_TSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PedSV.2.5.2_CNV_chr1_7</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>0.266629</td>\n",
       "      <td>62</td>\n",
       "      <td>0.510264</td>\n",
       "      <td>0.255484</td>\n",
       "      <td>0.292222</td>\n",
       "      <td>0.620135</td>\n",
       "      <td>0.461080</td>\n",
       "      <td>-0.181469</td>\n",
       "      <td>PREDICTED_NEAREST_TSS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     name    gene  mean_exp  num_greater_0_dosage  \\\n",
       "0  PedSV.2.5.2_CNV_chr1_6  OR4F29  0.266629                    15   \n",
       "1  PedSV.2.5.2_CNV_chr1_7  OR4F29  0.266629                    62   \n",
       "\n",
       "   mean_greater_0_dosage_rank  mean_greater_0_exp  mean_leq_0_exp     mwu_p  \\\n",
       "0                    0.507955            0.247333        0.270541  0.912668   \n",
       "1                    0.510264            0.255484        0.292222  0.620135   \n",
       "\n",
       "   ordinal_p  ordinal_coef     genic_relationship  \n",
       "0   0.582179     -0.188116  PREDICTED_NEAREST_TSS  \n",
       "1   0.461080     -0.181469  PREDICTED_NEAREST_TSS  "
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "77328465-917d-4905-baa7-883a51967c1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_gene_df = sv_gene_df.merge(results, on = ['name', 'gene', 'genic_relationship'], how = 'left')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c6794b9-8232-4465-9c22-06b1875cf6b7",
   "metadata": {},
   "source": [
    "Finally, I do want to swap the rank around, such that `0 = low | 1 = high`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "c8cae1ab-4535-4765-a3c5-31a75c87e480",
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_gene_df['mean_greater_0_dosage_rank'] = 1 - sv_gene_df['mean_greater_0_dosage_rank']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "cf88449f-7675-4166-ae63-3eb001b6611d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sv_gene_df.to_csv('data/sv-expression-results/sv-gene-rna-results.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pediatric-germline-svs-3.7.13",
   "language": "python",
   "name": "pediatric-germline-svs-3.7.13"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
