{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d2ac3077-8c01-456c-919b-755f42917cec",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "import warnings\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import palettable\n",
    "import pandas as pd\n",
    "import pybedtools\n",
    "import statsmodels.api as sm\n",
    "from liftover import get_lifter\n",
    "from matplotlib import patches\n",
    "import matplotlib.gridspec as gridspec\n",
    "from matplotlib.lines import Line2D\n",
    "from scipy import stats\n",
    "\n",
    "bold_10 = palettable.cartocolors.qualitative.Bold_10.mpl_colors\n",
    "\n",
    "from scripts import aesthetics, sv_plot\n",
    "\n",
    "aesthetics.activate_paper_rcParams()\n",
    "\n",
    "pd.set_option(\"display.max_columns\", 100)\n",
    "pd.set_option(\"display.max_rows\", 100)\n",
    "\n",
    "%reload_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6e39146-7f7d-4e04-94f7-2245818369b6",
   "metadata": {},
   "source": [
    "# MYCN and 16p Neuroblastoma\n",
    "\n",
    "Similar to EWSR1-FLI1, MYCN is somatically amplified in a subset of neuroblastoma (although its prevalence is significantly lower than Ewing's).\n",
    "\n",
    "The logic here is that perhaps there is something similar happening. There are a few literature notable cases in this context. In particular, there's this paper: https://pubmed.ncbi.nlm.nih.gov/31474320/. Which implicates both a MYCN (chr2) amplification and a 16p11 deletion. We'll take a look at these loci.\n",
    "\n",
    "__Note:__ The paper listed above actually collapsed large CNVs. We'll need to think about whether that is necessary here."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90d1a9c6-0636-4e8e-83ee-e2383a4eb83d",
   "metadata": {},
   "source": [
    "# Load in SVs and samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7e88a526-048c-43b1-883a-0c562c04a5e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jettcrowdis/opt/anaconda3/envs/pediatric-germline-svs-3.7.13/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3457: DtypeWarning: Columns (19,20,21,23,25,27,31,32,33,34,48,49,63,64,82,83,89,98,99,113,114,132,133,138,147,148,162,163,181,182,187,196,197,211,212,230,231,245,246,260,261,279,280,285,294,295,309,310,328,329,336,345,346,351,360,361,375,376,381,390,391,398,407,408,413,422,423,428,437,438,443,452,453,460,469,470,475,484,485,490,499,500,505,514,515,522,531,532,537,546,547,552,561,562,567,576,577,584,593,594,599,608,609,623,624,629,638,639,646,655,656,661,670,671,676,685,686,700,701,707,713,719,725,731) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    }
   ],
   "source": [
    "svs = pd.read_csv(\n",
    "    \"gs://vanallen-pedsv-analysis/beds/PedSV.v1.1.validation_cohort.analysis_samples.wAFs.bed.gz\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "dosages = pd.read_csv(\n",
    "    \"gs://vanallen-pedsv-analysis/beds/PedSV.v1.1.validation_cohort.analysis_samples.wAFs.allele_dosages.bed.gz\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "851b5f43-4cad-4dec-9218-01bba5471ed3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(529, 2098)"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_csv(\n",
    "    \"gs://vanallen-pedsv-analysis/sample_info/gatk_sv_pediatric_cancers_combined_cohort_metadata_3_31_23.w_control_assignments.txt\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "validation_samples = pd.read_csv(\n",
    "    \"gs://vanallen-pedsv-analysis/sample_info/PedSV.v1.validation_cohort_final_samples.list\",\n",
    "    header=None,\n",
    ")[0].to_list()\n",
    "\n",
    "# extract out the samples prsent in the dosage matrix\n",
    "nbl_samples = metadata[metadata[\"disease\"] == \"neuroblastoma\"][\"entity:sample_id\"].to_list()\n",
    "nbl_samples = sorted(set(validation_samples) & set(nbl_samples))\n",
    "\n",
    "control_samples = metadata[metadata[\"neuroblastoma_control\"]][\"entity:sample_id\"].to_list()\n",
    "control_samples = sorted(set(validation_samples) & set(control_samples))\n",
    "\n",
    "# create a new sex column for modeling, XY = 1\n",
    "metadata[\"sex\"] = (metadata[\"chrX_CopyNumber\"].round() < 2).astype(int)\n",
    "\n",
    "len(nbl_samples), len(control_samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa05ebde-0f7d-4811-80df-a1c3cc8e8483",
   "metadata": {},
   "source": [
    "# Identifying regions of interest\n",
    "\n",
    "Unlike EWSR1-FLI1, we aren't simply looking at genes here. The germline amplification noted in that paper is incredibly large, as large as 10 Mb. Similarly, the noted microdeletion on 16p is nearly 10 Mb in size.\n",
    "\n",
    "This is tricky because I don't know how GATK handles CNVs--I assume it uses read depth, in which case these events should easily span MYCN and a narrow region on 16p, but we'll cast a large net either way.\n",
    "\n",
    "Luckily, this paper did us a favor and categorized their large CNVs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "0137ae9c-2fdb-4593-8ae7-d821690866b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "large_nbl_cnvs = pd.read_excel('ref/10.1016-j.ajhg.2019.07.020-supplementary/1-s2.0-S0002929719303039-mmc2.xlsx', skiprows=3)\n",
    "large_nbl_cnvs['start'] = large_nbl_cnvs['Region'].apply(lambda region: region.split(':')[1].split('-')[0].replace(',', '')).astype(int)\n",
    "large_nbl_cnvs['end'] = large_nbl_cnvs['Region'].apply(lambda region: region.split(':')[1].split('-')[1].replace(',', '')).astype(int)\n",
    "large_nbl_cnvs['chrom'] = large_nbl_cnvs['Region'].apply(lambda region: region.split(':')[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a485c17-fc6c-456e-a13b-1c5e4a8bc73d",
   "metadata": {},
   "source": [
    "We'll be taking a look at two: `chr16:p11.2` and `chr2:p25.1-p24.2`. MYCN is located on `chr2:16,080,672-16,087,126`, or `2p24.3`. The chr16 microdeletion has variable location, but it looks as though it primarily appears around 30 Mb along chr16."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a11b3beb-e503-40d1-a983-4d31b81e455d",
   "metadata": {},
   "source": [
    "## MYCN\n",
    "\n",
    "Some digging around in this dataset shows that there actually isn't a reproducible MYCN amplification. Only one sample has one in the paper, and that sample is apparently not given in this dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "8708ee7d-38c6-4f52-adb7-68dd0ab66b48",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chip Barcode</th>\n",
       "      <th>USIa</th>\n",
       "      <th>Group</th>\n",
       "      <th>Region</th>\n",
       "      <th>Event</th>\n",
       "      <th>Length</th>\n",
       "      <th>Cytoband</th>\n",
       "      <th>Probe Median</th>\n",
       "      <th>Probe Count</th>\n",
       "      <th>Minimum Size</th>\n",
       "      <th>Minimum Region</th>\n",
       "      <th>Maximum Size</th>\n",
       "      <th>Maximum Region</th>\n",
       "      <th>Call P-value</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>1562857594_A</td>\n",
       "      <td>PAKHSL</td>\n",
       "      <td>Discovery Cohort Case</td>\n",
       "      <td>chr2:4,114,403-4,651,344</td>\n",
       "      <td>CN Loss</td>\n",
       "      <td>536942</td>\n",
       "      <td>p25.3 - p25.2</td>\n",
       "      <td>-0.636486</td>\n",
       "      <td>74</td>\n",
       "      <td>526241</td>\n",
       "      <td>chr2:4,124,656-4,650,896</td>\n",
       "      <td>547642</td>\n",
       "      <td>chr2:4,104,151-4,651,792</td>\n",
       "      <td>3.970000e-140</td>\n",
       "      <td>4114403</td>\n",
       "      <td>4651344</td>\n",
       "      <td>chr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3190</th>\n",
       "      <td>6205470038_R03C01</td>\n",
       "      <td>Control</td>\n",
       "      <td>Replication Cohort Control</td>\n",
       "      <td>chr2:30,702,846-32,771,540</td>\n",
       "      <td>CN Gain</td>\n",
       "      <td>2068695</td>\n",
       "      <td>p23.1 - p22.3</td>\n",
       "      <td>0.215332</td>\n",
       "      <td>430</td>\n",
       "      <td>2063126</td>\n",
       "      <td>chr2:30,703,971-32,767,096</td>\n",
       "      <td>2074263</td>\n",
       "      <td>chr2:30,701,722-32,775,984</td>\n",
       "      <td>8.800000e-241</td>\n",
       "      <td>30702846</td>\n",
       "      <td>32771540</td>\n",
       "      <td>chr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2917</th>\n",
       "      <td>6092196190_R06C02</td>\n",
       "      <td>Control</td>\n",
       "      <td>Replication Cohort Control</td>\n",
       "      <td>chr2:32,019,597-33,071,570</td>\n",
       "      <td>CN Gain</td>\n",
       "      <td>1051974</td>\n",
       "      <td>p23.1 - p22.3</td>\n",
       "      <td>0.194516</td>\n",
       "      <td>123</td>\n",
       "      <td>1042433</td>\n",
       "      <td>chr2:32,020,049-33,062,481</td>\n",
       "      <td>1061514</td>\n",
       "      <td>chr2:32,019,146-33,080,659</td>\n",
       "      <td>4.850000e-32</td>\n",
       "      <td>32019597</td>\n",
       "      <td>33071570</td>\n",
       "      <td>chr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1299</th>\n",
       "      <td>4067202054_B</td>\n",
       "      <td>Control</td>\n",
       "      <td>Discovery Cohort Control</td>\n",
       "      <td>chr2:32,521,450-33,322,967</td>\n",
       "      <td>CN Gain</td>\n",
       "      <td>801518</td>\n",
       "      <td>p22.3</td>\n",
       "      <td>0.255594</td>\n",
       "      <td>112</td>\n",
       "      <td>780624</td>\n",
       "      <td>chr2:32,540,334-33,320,957</td>\n",
       "      <td>822413</td>\n",
       "      <td>chr2:32,502,566-33,324,978</td>\n",
       "      <td>6.730000e-42</td>\n",
       "      <td>32521450</td>\n",
       "      <td>33322967</td>\n",
       "      <td>chr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3951</th>\n",
       "      <td>7930622149_R05C02</td>\n",
       "      <td>Control</td>\n",
       "      <td>Replication Cohort Control</td>\n",
       "      <td>chr2:32,538,503-33,340,399</td>\n",
       "      <td>CN Gain</td>\n",
       "      <td>801897</td>\n",
       "      <td>p22.3</td>\n",
       "      <td>0.227224</td>\n",
       "      <td>134</td>\n",
       "      <td>798189</td>\n",
       "      <td>chr2:32,541,438-33,339,626</td>\n",
       "      <td>805605</td>\n",
       "      <td>chr2:32,535,568-33,341,172</td>\n",
       "      <td>1.480000e-55</td>\n",
       "      <td>32538503</td>\n",
       "      <td>33340399</td>\n",
       "      <td>chr2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Chip Barcode     USIa                       Group  \\\n",
       "102        1562857594_A   PAKHSL       Discovery Cohort Case   \n",
       "3190  6205470038_R03C01  Control  Replication Cohort Control   \n",
       "2917  6092196190_R06C02  Control  Replication Cohort Control   \n",
       "1299       4067202054_B  Control    Discovery Cohort Control   \n",
       "3951  7930622149_R05C02  Control  Replication Cohort Control   \n",
       "\n",
       "                          Region    Event   Length       Cytoband  \\\n",
       "102     chr2:4,114,403-4,651,344  CN Loss   536942  p25.3 - p25.2   \n",
       "3190  chr2:30,702,846-32,771,540  CN Gain  2068695  p23.1 - p22.3   \n",
       "2917  chr2:32,019,597-33,071,570  CN Gain  1051974  p23.1 - p22.3   \n",
       "1299  chr2:32,521,450-33,322,967  CN Gain   801518          p22.3   \n",
       "3951  chr2:32,538,503-33,340,399  CN Gain   801897          p22.3   \n",
       "\n",
       "      Probe Median  Probe Count  Minimum Size              Minimum Region  \\\n",
       "102      -0.636486           74        526241    chr2:4,124,656-4,650,896   \n",
       "3190      0.215332          430       2063126  chr2:30,703,971-32,767,096   \n",
       "2917      0.194516          123       1042433  chr2:32,020,049-33,062,481   \n",
       "1299      0.255594          112        780624  chr2:32,540,334-33,320,957   \n",
       "3951      0.227224          134        798189  chr2:32,541,438-33,339,626   \n",
       "\n",
       "      Maximum Size              Maximum Region   Call P-value     start  \\\n",
       "102         547642    chr2:4,104,151-4,651,792  3.970000e-140   4114403   \n",
       "3190       2074263  chr2:30,701,722-32,775,984  8.800000e-241  30702846   \n",
       "2917       1061514  chr2:32,019,146-33,080,659   4.850000e-32  32019597   \n",
       "1299        822413  chr2:32,502,566-33,324,978   6.730000e-42  32521450   \n",
       "3951        805605  chr2:32,535,568-33,341,172   1.480000e-55  32538503   \n",
       "\n",
       "           end chrom  \n",
       "102    4651344  chr2  \n",
       "3190  32771540  chr2  \n",
       "2917  33071570  chr2  \n",
       "1299  33322967  chr2  \n",
       "3951  33340399  chr2  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "large_nbl_cnvs.query('chrom == \"chr2\" & start > 1200000').sort_values(by = 'start').iloc[5:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df5d3078-ddff-4f0a-9c1c-106d6796b0ea",
   "metadata": {},
   "source": [
    "We can see that the given CNVs jump from 4 Mb to 30 Mb, essentially skipping MYCN."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "630acdc4-0a70-4cff-a227-f6675c960712",
   "metadata": {},
   "source": [
    "## 16p microdeletion"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a0b8228-120c-4e97-962c-dabf449f7c91",
   "metadata": {},
   "source": [
    "This one is a little harder to wrangle, since we don't have exact genomic coordinates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "3cccdab6-af4d-4c34-9013-fa775d4008d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(47, 17)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nbl_16p_microdel = large_nbl_cnvs[(large_nbl_cnvs['chrom'] == 'chr16') &\n",
    "                                  (large_nbl_cnvs['Cytoband'] == 'p11.2')]\n",
    "nbl_16p_microdel.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "c951e671-1267-43a5-b735-d86f4788f503",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chip Barcode</th>\n",
       "      <th>USIa</th>\n",
       "      <th>Group</th>\n",
       "      <th>Region</th>\n",
       "      <th>Event</th>\n",
       "      <th>Length</th>\n",
       "      <th>Cytoband</th>\n",
       "      <th>Probe Median</th>\n",
       "      <th>Probe Count</th>\n",
       "      <th>Minimum Size</th>\n",
       "      <th>Minimum Region</th>\n",
       "      <th>Maximum Size</th>\n",
       "      <th>Maximum Region</th>\n",
       "      <th>Call P-value</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>chrom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>1552042589_A</td>\n",
       "      <td>PANWZA</td>\n",
       "      <td>Discovery Cohort Case</td>\n",
       "      <td>chr16:29,529,922-30,270,939</td>\n",
       "      <td>CN Loss</td>\n",
       "      <td>741018</td>\n",
       "      <td>p11.2</td>\n",
       "      <td>-0.526034</td>\n",
       "      <td>36</td>\n",
       "      <td>530466</td>\n",
       "      <td>chr16:29,647,342-30,177,807</td>\n",
       "      <td>951569</td>\n",
       "      <td>chr16:29,412,503-30,364,071</td>\n",
       "      <td>1.060000e-94</td>\n",
       "      <td>29529922</td>\n",
       "      <td>30270939</td>\n",
       "      <td>chr16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>1557556242_A</td>\n",
       "      <td>PAMXAW</td>\n",
       "      <td>Discovery Cohort Case</td>\n",
       "      <td>chr16:29,529,922-30,270,939</td>\n",
       "      <td>CN Gain</td>\n",
       "      <td>741018</td>\n",
       "      <td>p11.2</td>\n",
       "      <td>0.241063</td>\n",
       "      <td>36</td>\n",
       "      <td>530466</td>\n",
       "      <td>chr16:29,647,342-30,177,807</td>\n",
       "      <td>951569</td>\n",
       "      <td>chr16:29,412,503-30,364,071</td>\n",
       "      <td>4.540000e-12</td>\n",
       "      <td>29529922</td>\n",
       "      <td>30270939</td>\n",
       "      <td>chr16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chip Barcode    USIa                  Group                       Region  \\\n",
       "42  1552042589_A  PANWZA  Discovery Cohort Case  chr16:29,529,922-30,270,939   \n",
       "63  1557556242_A  PAMXAW  Discovery Cohort Case  chr16:29,529,922-30,270,939   \n",
       "\n",
       "      Event  Length Cytoband  Probe Median  Probe Count  Minimum Size  \\\n",
       "42  CN Loss  741018    p11.2     -0.526034           36        530466   \n",
       "63  CN Gain  741018    p11.2      0.241063           36        530466   \n",
       "\n",
       "                 Minimum Region  Maximum Size               Maximum Region  \\\n",
       "42  chr16:29,647,342-30,177,807        951569  chr16:29,412,503-30,364,071   \n",
       "63  chr16:29,647,342-30,177,807        951569  chr16:29,412,503-30,364,071   \n",
       "\n",
       "    Call P-value     start       end  chrom  \n",
       "42  1.060000e-94  29529922  30270939  chr16  \n",
       "63  4.540000e-12  29529922  30270939  chr16  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nbl_16p_microdel.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cc70536-81ff-481a-9450-387d0a678dee",
   "metadata": {},
   "source": [
    "Some digging into this and the paper suggests that the \"critical region\" is `chr16:29,529,922-30,270,939`. This is the region that we'll define.\n",
    "\n",
    "Unfortunately, we do need to lift this region over, as it was defined by this paper that used hg19."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "d4083809-45ce-43a1-8a2c-bcd438318954",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([('chr16', 29518601, '+')], [('chr16', 30259618, '+')])"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# do the liftover\n",
    "hg19_to_hg38 = get_lifter(\"hg19\", \"hg38\")\n",
    "hg19_to_hg38.query('chr16', 29529922), hg19_to_hg38.query('chr16', 30270939)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7e80c46-cdea-4061-a6aa-3de837524569",
   "metadata": {},
   "source": [
    "So `chr16:295186010-30259618`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "cc777667-d788-4a4d-94cf-eb88c3a4df33",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = {\n",
    "    \"MYCN\": [\"chr2\", 15940550, 15947007],\n",
    "    \"16pmicrodel\": [\"chr16\", 29518601, 30259618],\n",
    "}\n",
    "\n",
    "gene_locs = pd.DataFrame.from_dict(\n",
    "    genes, orient=\"index\", columns=[\"chrom\", \"start\", \"end\"]\n",
    ")\n",
    "gene_locs[\"gene\"] = genes.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "e3e37c32-cecd-45d1-a78f-dfe879545341",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MYCN</th>\n",
       "      <td>chr2</td>\n",
       "      <td>15940550</td>\n",
       "      <td>15947007</td>\n",
       "      <td>MYCN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16pmicrodel</th>\n",
       "      <td>chr16</td>\n",
       "      <td>29518601</td>\n",
       "      <td>30259618</td>\n",
       "      <td>16pmicrodel</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             chrom     start       end         gene\n",
       "MYCN          chr2  15940550  15947007         MYCN\n",
       "16pmicrodel  chr16  29518601  30259618  16pmicrodel"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_locs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8770af1-c1a1-4387-9c04-34850c98fda8",
   "metadata": {},
   "source": [
    "# Isolate the SVs in this region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "00d4b3c1-25c6-42ca-b1fa-a1c4fba9dcf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get 0.5 Mb on either side of the gene\n",
    "window_size = 1e6\n",
    "\n",
    "# expand the gene \"locations\" to include this window\n",
    "expanded_gene_locs = gene_locs.copy()\n",
    "expanded_gene_locs[\"start\"] = (expanded_gene_locs[\"start\"] - window_size / 2).astype(\n",
    "    int\n",
    ")\n",
    "expanded_gene_locs[\"end\"] = (expanded_gene_locs[\"end\"] + window_size / 2).astype(int)\n",
    "\n",
    "# do the intersection with pybedtools. pybedtools requires either a temp file or\n",
    "# a string, and we choose the latter.\n",
    "expanded_gene_bed = pybedtools.BedTool(\n",
    "    expanded_gene_locs.to_csv(sep=\"\\t\", index=False, header=False), from_string=True\n",
    ")\n",
    "sv_bed = pybedtools.BedTool(\n",
    "    svs[[\"#chrom\", \"start\", \"end\", \"name\"]].to_csv(sep=\"\\t\", index=False, header=False),\n",
    "    from_string=True,\n",
    ")\n",
    "\n",
    "svs_in_regions = sv_bed.intersect(expanded_gene_bed, wo=True)\n",
    "svs_in_regions = svs_in_regions.to_dataframe()\n",
    "\n",
    "# drop unnecessary columns\n",
    "svs_in_regions = svs_in_regions.iloc[:, [0, 1, 2, 3, 7]].copy()\n",
    "svs_in_regions.columns = [\"chrom\", \"start\", \"end\", \"name\", \"gene\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "9e762a6c-3fd0-49b4-a0a7-014605183072",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>15443841</td>\n",
       "      <td>15443892</td>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_INS_chr2_343</td>\n",
       "      <td>MYCN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>15453979</td>\n",
       "      <td>15455198</td>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_DUP_chr2_523</td>\n",
       "      <td>MYCN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>15463104</td>\n",
       "      <td>15465416</td>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_DEL_chr2_1106</td>\n",
       "      <td>MYCN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>15481484</td>\n",
       "      <td>15481611</td>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_DEL_chr2_1107</td>\n",
       "      <td>MYCN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>15488285</td>\n",
       "      <td>16373814</td>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_DUP_chr2_524</td>\n",
       "      <td>MYCN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>chr16</td>\n",
       "      <td>30725250</td>\n",
       "      <td>30725312</td>\n",
       "      <td>All_20_Batches.chr16.final_cleanup_DUP_chr16_1201</td>\n",
       "      <td>16pmicrodel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>chr16</td>\n",
       "      <td>30744486</td>\n",
       "      <td>30750487</td>\n",
       "      <td>All_20_Batches.chr16.final_cleanup_DUP_chr16_1203</td>\n",
       "      <td>16pmicrodel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>chr16</td>\n",
       "      <td>30748321</td>\n",
       "      <td>30749044</td>\n",
       "      <td>All_20_Batches.chr16.final_cleanup_DEL_chr16_2142</td>\n",
       "      <td>16pmicrodel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>chr16</td>\n",
       "      <td>30749020</td>\n",
       "      <td>30749182</td>\n",
       "      <td>All_20_Batches.chr16.final_cleanup_DEL_chr16_2143</td>\n",
       "      <td>16pmicrodel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>chr16</td>\n",
       "      <td>30759486</td>\n",
       "      <td>30765487</td>\n",
       "      <td>All_20_Batches.chr16.final_cleanup_DUP_chr16_1204</td>\n",
       "      <td>16pmicrodel</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>357 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end  \\\n",
       "0     chr2  15443841  15443892   \n",
       "1     chr2  15453979  15455198   \n",
       "2     chr2  15463104  15465416   \n",
       "3     chr2  15481484  15481611   \n",
       "4     chr2  15488285  16373814   \n",
       "..     ...       ...       ...   \n",
       "352  chr16  30725250  30725312   \n",
       "353  chr16  30744486  30750487   \n",
       "354  chr16  30748321  30749044   \n",
       "355  chr16  30749020  30749182   \n",
       "356  chr16  30759486  30765487   \n",
       "\n",
       "                                                  name         gene  \n",
       "0       All_20_Batches.chr2.final_cleanup_INS_chr2_343         MYCN  \n",
       "1       All_20_Batches.chr2.final_cleanup_DUP_chr2_523         MYCN  \n",
       "2      All_20_Batches.chr2.final_cleanup_DEL_chr2_1106         MYCN  \n",
       "3      All_20_Batches.chr2.final_cleanup_DEL_chr2_1107         MYCN  \n",
       "4       All_20_Batches.chr2.final_cleanup_DUP_chr2_524         MYCN  \n",
       "..                                                 ...          ...  \n",
       "352  All_20_Batches.chr16.final_cleanup_DUP_chr16_1201  16pmicrodel  \n",
       "353  All_20_Batches.chr16.final_cleanup_DUP_chr16_1203  16pmicrodel  \n",
       "354  All_20_Batches.chr16.final_cleanup_DEL_chr16_2142  16pmicrodel  \n",
       "355  All_20_Batches.chr16.final_cleanup_DEL_chr16_2143  16pmicrodel  \n",
       "356  All_20_Batches.chr16.final_cleanup_DUP_chr16_1204  16pmicrodel  \n",
       "\n",
       "[357 rows x 5 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "svs_in_regions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbb61239-10c9-4bed-a2aa-7e1d3364abb6",
   "metadata": {},
   "source": [
    "We \"quickly\" nab complex SVs as well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "68395f22-4bc5-4217-b44a-f0a5d2d2b82c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get all complex SVs\n",
    "complex_svs = svs[(svs[\"svtype\"] == \"CPX\")][\n",
    "    [\n",
    "        \"name\",\n",
    "        \"#chrom\",\n",
    "        \"start\",\n",
    "        \"end\",\n",
    "        \"svtype\",\n",
    "        \"CHR2\",\n",
    "        \"CPX_INTERVALS\",\n",
    "        \"CPX_TYPE\",\n",
    "        \"END\",\n",
    "    ]\n",
    "].copy()\n",
    "\n",
    "# split each component SV--represented by CPX_INTERVALS--into their own rows\n",
    "complex_svs = complex_svs.assign(\n",
    "    intervals=complex_svs[\"CPX_INTERVALS\"].str.split(\",\")\n",
    ").explode(\"intervals\")\n",
    "\n",
    "# convert this data to a proper bed for intersection\n",
    "complex_sv_bed = complex_svs[[\"name\", \"intervals\"]].copy()\n",
    "complex_sv_bed[\"chrom\"] = complex_sv_bed[\"intervals\"].apply(\n",
    "    lambda interv: interv.split(\":\")[0].split(\"_\")[-1]\n",
    ")\n",
    "complex_sv_bed[\"start\"] = complex_sv_bed[\"intervals\"].apply(\n",
    "    lambda interv: interv.split(\":\")[-1].split(\"-\")[0]\n",
    ")\n",
    "complex_sv_bed[\"end\"] = complex_sv_bed[\"intervals\"].apply(\n",
    "    lambda interv: interv.split(\":\")[-1].split(\"-\")[1]\n",
    ")\n",
    "\n",
    "# convert to a bedtool\n",
    "complex_sv_bed = pybedtools.BedTool(\n",
    "    complex_sv_bed[[\"chrom\", \"start\", \"end\", \"name\"]].to_csv(\n",
    "        sep=\"\\t\", index=False, header=False\n",
    "    ),\n",
    "    from_string=True,\n",
    ")\n",
    "\n",
    "# do our intersection\n",
    "complex_svs_in_regions = complex_sv_bed.intersect(expanded_gene_bed, wo=True)\n",
    "\n",
    "complex_svs_in_regions = complex_svs_in_regions.to_dataframe()\n",
    "\n",
    "# drop unnecessary columns\n",
    "if len(complex_svs_in_regions):\n",
    "    complex_svs_in_regions = complex_svs_in_regions.iloc[:, [0, 1, 2, 3, 7]].copy()\n",
    "    complex_svs_in_regions.columns = [\"chrom\", \"start\", \"end\", \"name\", \"gene\"]\n",
    "else:\n",
    "    complex_svs_in_regions = pd.DataFrame(\n",
    "        columns=[\"chrom\", \"start\", \"end\", \"name\", \"gene\"]\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "5d8503cd-813d-4970-8a19-532309f62b56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>15796015</td>\n",
       "      <td>15930622</td>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_CPX_chr2_34</td>\n",
       "      <td>MYCN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr2</td>\n",
       "      <td>15930622</td>\n",
       "      <td>16087855</td>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_CPX_chr2_34</td>\n",
       "      <td>MYCN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>16087855</td>\n",
       "      <td>16415986</td>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_CPX_chr2_34</td>\n",
       "      <td>MYCN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>15930622</td>\n",
       "      <td>15930835</td>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_CPX_chr2_35</td>\n",
       "      <td>MYCN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>15930622</td>\n",
       "      <td>16415986</td>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_CPX_chr2_35</td>\n",
       "      <td>MYCN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr2</td>\n",
       "      <td>16079849</td>\n",
       "      <td>16081075</td>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_CPX_chr2_36</td>\n",
       "      <td>MYCN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr2</td>\n",
       "      <td>16080529</td>\n",
       "      <td>16081075</td>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_CPX_chr2_36</td>\n",
       "      <td>MYCN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr16</td>\n",
       "      <td>29108850</td>\n",
       "      <td>29109031</td>\n",
       "      <td>All_20_Batches.chr5.final_cleanup_CPX_chr5_198</td>\n",
       "      <td>16pmicrodel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr16</td>\n",
       "      <td>29226774</td>\n",
       "      <td>29226840</td>\n",
       "      <td>All_20_Batches.chr16.final_cleanup_CPX_chr16_59</td>\n",
       "      <td>16pmicrodel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr16</td>\n",
       "      <td>29226774</td>\n",
       "      <td>29227265</td>\n",
       "      <td>All_20_Batches.chr16.final_cleanup_CPX_chr16_59</td>\n",
       "      <td>16pmicrodel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr16</td>\n",
       "      <td>30439094</td>\n",
       "      <td>30439429</td>\n",
       "      <td>All_20_Batches.chr16.final_cleanup_CPX_chr16_61</td>\n",
       "      <td>16pmicrodel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr16</td>\n",
       "      <td>30441230</td>\n",
       "      <td>30442559</td>\n",
       "      <td>All_20_Batches.chr16.final_cleanup_CPX_chr16_61</td>\n",
       "      <td>16pmicrodel</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom     start       end  \\\n",
       "0    chr2  15796015  15930622   \n",
       "1    chr2  15930622  16087855   \n",
       "2    chr2  16087855  16415986   \n",
       "3    chr2  15930622  15930835   \n",
       "4    chr2  15930622  16415986   \n",
       "5    chr2  16079849  16081075   \n",
       "6    chr2  16080529  16081075   \n",
       "7   chr16  29108850  29109031   \n",
       "8   chr16  29226774  29226840   \n",
       "9   chr16  29226774  29227265   \n",
       "10  chr16  30439094  30439429   \n",
       "11  chr16  30441230  30442559   \n",
       "\n",
       "                                               name         gene  \n",
       "0     All_20_Batches.chr2.final_cleanup_CPX_chr2_34         MYCN  \n",
       "1     All_20_Batches.chr2.final_cleanup_CPX_chr2_34         MYCN  \n",
       "2     All_20_Batches.chr2.final_cleanup_CPX_chr2_34         MYCN  \n",
       "3     All_20_Batches.chr2.final_cleanup_CPX_chr2_35         MYCN  \n",
       "4     All_20_Batches.chr2.final_cleanup_CPX_chr2_35         MYCN  \n",
       "5     All_20_Batches.chr2.final_cleanup_CPX_chr2_36         MYCN  \n",
       "6     All_20_Batches.chr2.final_cleanup_CPX_chr2_36         MYCN  \n",
       "7    All_20_Batches.chr5.final_cleanup_CPX_chr5_198  16pmicrodel  \n",
       "8   All_20_Batches.chr16.final_cleanup_CPX_chr16_59  16pmicrodel  \n",
       "9   All_20_Batches.chr16.final_cleanup_CPX_chr16_59  16pmicrodel  \n",
       "10  All_20_Batches.chr16.final_cleanup_CPX_chr16_61  16pmicrodel  \n",
       "11  All_20_Batches.chr16.final_cleanup_CPX_chr16_61  16pmicrodel  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complex_svs_in_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "1cb52e44-0cf5-4cac-87b3-631ad2e5d7d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6, 5)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the number of complex SVs recovered, and the number that were in our original simple overlap\n",
    "len(set(complex_svs_in_regions[\"name\"])), len(\n",
    "    set(complex_svs_in_regions[\"name\"]) & set(svs_in_regions[\"name\"])\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "803dd23d-f056-4979-9ac2-14bd5e450b4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>svtype</th>\n",
       "      <th>AC</th>\n",
       "      <th>AF</th>\n",
       "      <th>ALGORITHMS</th>\n",
       "      <th>AN</th>\n",
       "      <th>BOTHSIDES_SUPPORT</th>\n",
       "      <th>CHR2</th>\n",
       "      <th>CPX_INTERVALS</th>\n",
       "      <th>CPX_TYPE</th>\n",
       "      <th>END</th>\n",
       "      <th>END2</th>\n",
       "      <th>EVIDENCE</th>\n",
       "      <th>HIGH_SR_BACKGROUND</th>\n",
       "      <th>PCRMINUS_NCR</th>\n",
       "      <th>PESR_GT_OVERDISPERSION</th>\n",
       "      <th>PREDICTED_BREAKEND_EXONIC</th>\n",
       "      <th>PREDICTED_COPY_GAIN</th>\n",
       "      <th>PREDICTED_DUP_PARTIAL</th>\n",
       "      <th>PREDICTED_INTERGENIC</th>\n",
       "      <th>PREDICTED_INTRAGENIC_EXON_DUP</th>\n",
       "      <th>PREDICTED_INTRONIC</th>\n",
       "      <th>PREDICTED_INV_SPAN</th>\n",
       "      <th>PREDICTED_LOF</th>\n",
       "      <th>PREDICTED_MSV_EXON_OVERLAP</th>\n",
       "      <th>PREDICTED_NEAREST_TSS</th>\n",
       "      <th>PREDICTED_NONCODING_BREAKPOINT</th>\n",
       "      <th>PREDICTED_NONCODING_SPAN</th>\n",
       "      <th>PREDICTED_PARTIAL_EXON_DUP</th>\n",
       "      <th>PREDICTED_PROMOTER</th>\n",
       "      <th>PREDICTED_TSS_DUP</th>\n",
       "      <th>PREDICTED_UTR</th>\n",
       "      <th>SOURCE</th>\n",
       "      <th>STRANDS</th>\n",
       "      <th>SVLEN</th>\n",
       "      <th>SVTYPE</th>\n",
       "      <th>UNRESOLVED_TYPE</th>\n",
       "      <th>N_BI_GENOS</th>\n",
       "      <th>N_HOMREF</th>\n",
       "      <th>N_HET</th>\n",
       "      <th>N_HOMALT</th>\n",
       "      <th>FREQ_HOMREF</th>\n",
       "      <th>FREQ_HET</th>\n",
       "      <th>FREQ_HOMALT</th>\n",
       "      <th>CN_NUMBER</th>\n",
       "      <th>CN_COUNT</th>\n",
       "      <th>CN_FREQ</th>\n",
       "      <th>...</th>\n",
       "      <th>SAS_neuroblastoma_CN_NUMBER</th>\n",
       "      <th>SAS_neuroblastoma_CN_COUNT</th>\n",
       "      <th>SAS_neuroblastoma_CN_FREQ</th>\n",
       "      <th>SAS_neuroblastoma_CN_NONDIPLOID_COUNT</th>\n",
       "      <th>SAS_neuroblastoma_CN_NONDIPLOID_FREQ</th>\n",
       "      <th>SAS_osteosarcoma_AN</th>\n",
       "      <th>SAS_osteosarcoma_AC</th>\n",
       "      <th>SAS_osteosarcoma_AF</th>\n",
       "      <th>SAS_osteosarcoma_N_BI_GENOS</th>\n",
       "      <th>SAS_osteosarcoma_N_HOMREF</th>\n",
       "      <th>SAS_osteosarcoma_N_HET</th>\n",
       "      <th>SAS_osteosarcoma_N_HOMALT</th>\n",
       "      <th>SAS_osteosarcoma_FREQ_HOMREF</th>\n",
       "      <th>SAS_osteosarcoma_FREQ_HET</th>\n",
       "      <th>SAS_osteosarcoma_FREQ_HOMALT</th>\n",
       "      <th>SAS_osteosarcoma_CN_NUMBER</th>\n",
       "      <th>SAS_osteosarcoma_CN_COUNT</th>\n",
       "      <th>SAS_osteosarcoma_CN_FREQ</th>\n",
       "      <th>SAS_osteosarcoma_CN_NONDIPLOID_COUNT</th>\n",
       "      <th>SAS_osteosarcoma_CN_NONDIPLOID_FREQ</th>\n",
       "      <th>POPMAX_FREQ_HOMREF</th>\n",
       "      <th>POPMAX_FREQ_HET</th>\n",
       "      <th>POPMAX_FREQ_HOMALT</th>\n",
       "      <th>POPMAX_CN_FREQ</th>\n",
       "      <th>POPMAX_CN_NONDIPLOID_FREQ</th>\n",
       "      <th>POPMAX_control_AF</th>\n",
       "      <th>POPMAX_control_FREQ_HOMREF</th>\n",
       "      <th>POPMAX_control_FREQ_HET</th>\n",
       "      <th>POPMAX_control_FREQ_HOMALT</th>\n",
       "      <th>POPMAX_control_CN_FREQ</th>\n",
       "      <th>POPMAX_control_CN_NONDIPLOID_FREQ</th>\n",
       "      <th>POPMAX_ewing_AF</th>\n",
       "      <th>POPMAX_ewing_FREQ_HOMREF</th>\n",
       "      <th>POPMAX_ewing_FREQ_HET</th>\n",
       "      <th>POPMAX_ewing_FREQ_HOMALT</th>\n",
       "      <th>POPMAX_ewing_CN_FREQ</th>\n",
       "      <th>POPMAX_ewing_CN_NONDIPLOID_FREQ</th>\n",
       "      <th>POPMAX_neuroblastoma_AF</th>\n",
       "      <th>POPMAX_neuroblastoma_FREQ_HOMREF</th>\n",
       "      <th>POPMAX_neuroblastoma_FREQ_HET</th>\n",
       "      <th>POPMAX_neuroblastoma_FREQ_HOMALT</th>\n",
       "      <th>POPMAX_neuroblastoma_CN_FREQ</th>\n",
       "      <th>POPMAX_neuroblastoma_CN_NONDIPLOID_FREQ</th>\n",
       "      <th>POPMAX_osteosarcoma_AF</th>\n",
       "      <th>POPMAX_osteosarcoma_FREQ_HOMREF</th>\n",
       "      <th>POPMAX_osteosarcoma_FREQ_HET</th>\n",
       "      <th>POPMAX_osteosarcoma_FREQ_HOMALT</th>\n",
       "      <th>POPMAX_osteosarcoma_CN_FREQ</th>\n",
       "      <th>POPMAX_osteosarcoma_CN_NONDIPLOID_FREQ</th>\n",
       "      <th>FILTER</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24400</th>\n",
       "      <td>chr2</td>\n",
       "      <td>15796014</td>\n",
       "      <td>16415986</td>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_CPX_chr2_34</td>\n",
       "      <td>CPX</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000140</td>\n",
       "      <td>manta</td>\n",
       "      <td>7162</td>\n",
       "      <td>False</td>\n",
       "      <td>chr2</td>\n",
       "      <td>DEL_chr2:15796015-15930622,INV_chr2:15930622-1...</td>\n",
       "      <td>delINVdel</td>\n",
       "      <td>16415986</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PE,SR</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ewing_chromHMM15_Enh</td>\n",
       "      <td>UCNE_conserved,ewing_ABC_MAX_enhancer,ewing_AB...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>619971</td>\n",
       "      <td>CPX</td>\n",
       "      <td>POSTHOC_RD_GT_REJECTION</td>\n",
       "      <td>3581.0</td>\n",
       "      <td>3580.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.999721</td>\n",
       "      <td>0.000279</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.002584</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.012820</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UNRESOLVED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24408</th>\n",
       "      <td>chr2</td>\n",
       "      <td>15930621</td>\n",
       "      <td>16415986</td>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_CPX_chr2_35</td>\n",
       "      <td>CPX</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000140</td>\n",
       "      <td>manta</td>\n",
       "      <td>7162</td>\n",
       "      <td>True</td>\n",
       "      <td>chr2</td>\n",
       "      <td>DUP_chr2:15930622-15930835,INV_chr2:15930622-1...</td>\n",
       "      <td>dupINV</td>\n",
       "      <td>16415986</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SR</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UCNE_conserved,ewing_H3K27Ac_peak,ewing_and_os...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>485364</td>\n",
       "      <td>CPX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3581.0</td>\n",
       "      <td>3580.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.999721</td>\n",
       "      <td>0.000279</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.002584</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.012820</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24427</th>\n",
       "      <td>chr2</td>\n",
       "      <td>16079848</td>\n",
       "      <td>16081075</td>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_CPX_chr2_36</td>\n",
       "      <td>CPX</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000140</td>\n",
       "      <td>manta</td>\n",
       "      <td>7162</td>\n",
       "      <td>False</td>\n",
       "      <td>chr2</td>\n",
       "      <td>INV_chr2:16079849-16081075,DUP_chr2:16080529-1...</td>\n",
       "      <td>INVdup</td>\n",
       "      <td>16081075</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PE,SR</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1226</td>\n",
       "      <td>CPX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3581.0</td>\n",
       "      <td>3580.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.999721</td>\n",
       "      <td>0.000279</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000351</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000299</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000598</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88452</th>\n",
       "      <td>chr5</td>\n",
       "      <td>109577995</td>\n",
       "      <td>109578016</td>\n",
       "      <td>All_20_Batches.chr5.final_cleanup_CPX_chr5_198</td>\n",
       "      <td>CPX</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000140</td>\n",
       "      <td>manta</td>\n",
       "      <td>7162</td>\n",
       "      <td>True</td>\n",
       "      <td>chr16</td>\n",
       "      <td>DUP_chr16:29108850-29109031</td>\n",
       "      <td>dDUP</td>\n",
       "      <td>109578016</td>\n",
       "      <td>109578016.0</td>\n",
       "      <td>PE,SR</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LAT,MAN2A1</td>\n",
       "      <td>osteosarcoma_H3K27Ac_peak</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DUP_chr16:29108850-29109031</td>\n",
       "      <td>NaN</td>\n",
       "      <td>181</td>\n",
       "      <td>CPX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3581.0</td>\n",
       "      <td>3580.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.999721</td>\n",
       "      <td>0.000279</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.002584</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.014706</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214559</th>\n",
       "      <td>chr16</td>\n",
       "      <td>29226773</td>\n",
       "      <td>29227265</td>\n",
       "      <td>All_20_Batches.chr16.final_cleanup_CPX_chr16_59</td>\n",
       "      <td>CPX</td>\n",
       "      <td>3642</td>\n",
       "      <td>0.508517</td>\n",
       "      <td>manta</td>\n",
       "      <td>7162</td>\n",
       "      <td>False</td>\n",
       "      <td>chr16</td>\n",
       "      <td>DUP_chr16:29226774-29226840,INV_chr16:29226774...</td>\n",
       "      <td>dupINV</td>\n",
       "      <td>29227265</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PE,SR</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BOLA2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>491</td>\n",
       "      <td>CPX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3581.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>3494.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>0.003630</td>\n",
       "      <td>0.975705</td>\n",
       "      <td>0.020665</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.006667</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.02584</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.510460</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.020921</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.509375</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.071429</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.538462</td>\n",
       "      <td>0.011601</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.520833</td>\n",
       "      <td>0.029412</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.049242</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214726</th>\n",
       "      <td>chr16</td>\n",
       "      <td>30441230</td>\n",
       "      <td>30442559</td>\n",
       "      <td>All_20_Batches.chr16.final_cleanup_CPX_chr16_61</td>\n",
       "      <td>CPX</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000140</td>\n",
       "      <td>manta</td>\n",
       "      <td>7162</td>\n",
       "      <td>True</td>\n",
       "      <td>chr16</td>\n",
       "      <td>DUP_chr16:30439094-30439429,DEL_chr16:30441230...</td>\n",
       "      <td>dDUP_iDEL</td>\n",
       "      <td>30442559</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PE,SR</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SEPHS2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ewing_chromHMM15_Enh</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DUP_chr16:30439094-30439429</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1664</td>\n",
       "      <td>CPX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3581.0</td>\n",
       "      <td>3580.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.999721</td>\n",
       "      <td>0.000279</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000351</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001894</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.003788</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows Ã— 734 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       #chrom      start        end  \\\n",
       "24400    chr2   15796014   16415986   \n",
       "24408    chr2   15930621   16415986   \n",
       "24427    chr2   16079848   16081075   \n",
       "88452    chr5  109577995  109578016   \n",
       "214559  chr16   29226773   29227265   \n",
       "214726  chr16   30441230   30442559   \n",
       "\n",
       "                                                   name svtype    AC  \\\n",
       "24400     All_20_Batches.chr2.final_cleanup_CPX_chr2_34    CPX     1   \n",
       "24408     All_20_Batches.chr2.final_cleanup_CPX_chr2_35    CPX     1   \n",
       "24427     All_20_Batches.chr2.final_cleanup_CPX_chr2_36    CPX     1   \n",
       "88452    All_20_Batches.chr5.final_cleanup_CPX_chr5_198    CPX     1   \n",
       "214559  All_20_Batches.chr16.final_cleanup_CPX_chr16_59    CPX  3642   \n",
       "214726  All_20_Batches.chr16.final_cleanup_CPX_chr16_61    CPX     1   \n",
       "\n",
       "              AF ALGORITHMS    AN  BOTHSIDES_SUPPORT   CHR2  \\\n",
       "24400   0.000140      manta  7162              False   chr2   \n",
       "24408   0.000140      manta  7162               True   chr2   \n",
       "24427   0.000140      manta  7162              False   chr2   \n",
       "88452   0.000140      manta  7162               True  chr16   \n",
       "214559  0.508517      manta  7162              False  chr16   \n",
       "214726  0.000140      manta  7162               True  chr16   \n",
       "\n",
       "                                            CPX_INTERVALS   CPX_TYPE  \\\n",
       "24400   DEL_chr2:15796015-15930622,INV_chr2:15930622-1...  delINVdel   \n",
       "24408   DUP_chr2:15930622-15930835,INV_chr2:15930622-1...     dupINV   \n",
       "24427   INV_chr2:16079849-16081075,DUP_chr2:16080529-1...     INVdup   \n",
       "88452                         DUP_chr16:29108850-29109031       dDUP   \n",
       "214559  DUP_chr16:29226774-29226840,INV_chr16:29226774...     dupINV   \n",
       "214726  DUP_chr16:30439094-30439429,DEL_chr16:30441230...  dDUP_iDEL   \n",
       "\n",
       "              END         END2 EVIDENCE  HIGH_SR_BACKGROUND  PCRMINUS_NCR  \\\n",
       "24400    16415986          NaN    PE,SR               False           0.0   \n",
       "24408    16415986          NaN       SR               False           0.0   \n",
       "24427    16081075          NaN    PE,SR               False           0.0   \n",
       "88452   109578016  109578016.0    PE,SR               False           0.0   \n",
       "214559   29227265          NaN    PE,SR               False           0.0   \n",
       "214726   30442559          NaN    PE,SR               False           0.0   \n",
       "\n",
       "        PESR_GT_OVERDISPERSION PREDICTED_BREAKEND_EXONIC PREDICTED_COPY_GAIN  \\\n",
       "24400                    False                       NaN                 NaN   \n",
       "24408                    False                       NaN                 NaN   \n",
       "24427                    False                       NaN                 NaN   \n",
       "88452                    False                       NaN                 NaN   \n",
       "214559                   False                       NaN                 NaN   \n",
       "214726                   False                       NaN                 NaN   \n",
       "\n",
       "       PREDICTED_DUP_PARTIAL  PREDICTED_INTERGENIC  \\\n",
       "24400                    NaN                 False   \n",
       "24408                    NaN                 False   \n",
       "24427                    NaN                  True   \n",
       "88452                    NaN                  True   \n",
       "214559                   NaN                  True   \n",
       "214726                   NaN                  True   \n",
       "\n",
       "       PREDICTED_INTRAGENIC_EXON_DUP PREDICTED_INTRONIC PREDICTED_INV_SPAN  \\\n",
       "24400                            NaN                NaN               MYCN   \n",
       "24408                            NaN                NaN               MYCN   \n",
       "24427                            NaN                NaN                NaN   \n",
       "88452                            NaN                NaN                NaN   \n",
       "214559                           NaN                NaN                NaN   \n",
       "214726                           NaN                NaN                NaN   \n",
       "\n",
       "       PREDICTED_LOF PREDICTED_MSV_EXON_OVERLAP PREDICTED_NEAREST_TSS  \\\n",
       "24400            NaN                        NaN                   NaN   \n",
       "24408            NaN                        NaN                   NaN   \n",
       "24427            NaN                        NaN                  MYCN   \n",
       "88452            NaN                        NaN            LAT,MAN2A1   \n",
       "214559           NaN                        NaN                 BOLA2   \n",
       "214726           NaN                        NaN                SEPHS2   \n",
       "\n",
       "       PREDICTED_NONCODING_BREAKPOINT  \\\n",
       "24400            ewing_chromHMM15_Enh   \n",
       "24408                             NaN   \n",
       "24427                             NaN   \n",
       "88452       osteosarcoma_H3K27Ac_peak   \n",
       "214559                            NaN   \n",
       "214726                            NaN   \n",
       "\n",
       "                                 PREDICTED_NONCODING_SPAN  \\\n",
       "24400   UCNE_conserved,ewing_ABC_MAX_enhancer,ewing_AB...   \n",
       "24408   UCNE_conserved,ewing_H3K27Ac_peak,ewing_and_os...   \n",
       "24427                                                 NaN   \n",
       "88452                                                 NaN   \n",
       "214559                                                NaN   \n",
       "214726                               ewing_chromHMM15_Enh   \n",
       "\n",
       "       PREDICTED_PARTIAL_EXON_DUP PREDICTED_PROMOTER PREDICTED_TSS_DUP  \\\n",
       "24400                         NaN                NaN               NaN   \n",
       "24408                         NaN                NaN               NaN   \n",
       "24427                         NaN                NaN               NaN   \n",
       "88452                         NaN                NaN               NaN   \n",
       "214559                        NaN                NaN               NaN   \n",
       "214726                        NaN                NaN               NaN   \n",
       "\n",
       "       PREDICTED_UTR                       SOURCE STRANDS   SVLEN SVTYPE  \\\n",
       "24400            NaN                          NaN     NaN  619971    CPX   \n",
       "24408            NaN                          NaN     NaN  485364    CPX   \n",
       "24427            NaN                          NaN     NaN    1226    CPX   \n",
       "88452            NaN  DUP_chr16:29108850-29109031     NaN     181    CPX   \n",
       "214559           NaN                          NaN     NaN     491    CPX   \n",
       "214726           NaN  DUP_chr16:30439094-30439429     NaN    1664    CPX   \n",
       "\n",
       "                UNRESOLVED_TYPE  N_BI_GENOS  N_HOMREF   N_HET  N_HOMALT  \\\n",
       "24400   POSTHOC_RD_GT_REJECTION      3581.0    3580.0     1.0       0.0   \n",
       "24408                       NaN      3581.0    3580.0     1.0       0.0   \n",
       "24427                       NaN      3581.0    3580.0     1.0       0.0   \n",
       "88452                       NaN      3581.0    3580.0     1.0       0.0   \n",
       "214559                      NaN      3581.0      13.0  3494.0      74.0   \n",
       "214726                      NaN      3581.0    3580.0     1.0       0.0   \n",
       "\n",
       "        FREQ_HOMREF  FREQ_HET  FREQ_HOMALT  CN_NUMBER CN_COUNT CN_FREQ  ...  \\\n",
       "24400      0.999721  0.000279     0.000000        NaN      NaN     NaN  ...   \n",
       "24408      0.999721  0.000279     0.000000        NaN      NaN     NaN  ...   \n",
       "24427      0.999721  0.000279     0.000000        NaN      NaN     NaN  ...   \n",
       "88452      0.999721  0.000279     0.000000        NaN      NaN     NaN  ...   \n",
       "214559     0.003630  0.975705     0.020665        NaN      NaN     NaN  ...   \n",
       "214726     0.999721  0.000279     0.000000        NaN      NaN     NaN  ...   \n",
       "\n",
       "        SAS_neuroblastoma_CN_NUMBER  SAS_neuroblastoma_CN_COUNT  \\\n",
       "24400                           NaN                         NaN   \n",
       "24408                           NaN                         NaN   \n",
       "24427                           NaN                         NaN   \n",
       "88452                           NaN                         NaN   \n",
       "214559                          NaN                         NaN   \n",
       "214726                          NaN                         NaN   \n",
       "\n",
       "        SAS_neuroblastoma_CN_FREQ  SAS_neuroblastoma_CN_NONDIPLOID_COUNT  \\\n",
       "24400                         NaN                                    NaN   \n",
       "24408                         NaN                                    NaN   \n",
       "24427                         NaN                                    NaN   \n",
       "88452                         NaN                                    NaN   \n",
       "214559                        NaN                                    NaN   \n",
       "214726                        NaN                                    NaN   \n",
       "\n",
       "        SAS_neuroblastoma_CN_NONDIPLOID_FREQ  SAS_osteosarcoma_AN  \\\n",
       "24400                                    NaN                  2.0   \n",
       "24408                                    NaN                  2.0   \n",
       "24427                                    NaN                  2.0   \n",
       "88452                                    NaN                  2.0   \n",
       "214559                                   NaN                  2.0   \n",
       "214726                                   NaN                  2.0   \n",
       "\n",
       "        SAS_osteosarcoma_AC  SAS_osteosarcoma_AF  SAS_osteosarcoma_N_BI_GENOS  \\\n",
       "24400                   0.0                  0.0                          1.0   \n",
       "24408                   0.0                  0.0                          1.0   \n",
       "24427                   0.0                  0.0                          1.0   \n",
       "88452                   0.0                  0.0                          1.0   \n",
       "214559                  1.0                  0.5                          1.0   \n",
       "214726                  0.0                  0.0                          1.0   \n",
       "\n",
       "        SAS_osteosarcoma_N_HOMREF  SAS_osteosarcoma_N_HET  \\\n",
       "24400                         1.0                     0.0   \n",
       "24408                         1.0                     0.0   \n",
       "24427                         1.0                     0.0   \n",
       "88452                         1.0                     0.0   \n",
       "214559                        0.0                     1.0   \n",
       "214726                        1.0                     0.0   \n",
       "\n",
       "        SAS_osteosarcoma_N_HOMALT  SAS_osteosarcoma_FREQ_HOMREF  \\\n",
       "24400                         0.0                           1.0   \n",
       "24408                         0.0                           1.0   \n",
       "24427                         0.0                           1.0   \n",
       "88452                         0.0                           1.0   \n",
       "214559                        0.0                           0.0   \n",
       "214726                        0.0                           1.0   \n",
       "\n",
       "       SAS_osteosarcoma_FREQ_HET SAS_osteosarcoma_FREQ_HOMALT  \\\n",
       "24400                        0.0                          0.0   \n",
       "24408                        0.0                          0.0   \n",
       "24427                        0.0                          0.0   \n",
       "88452                        0.0                          0.0   \n",
       "214559                       1.0                          0.0   \n",
       "214726                       0.0                          0.0   \n",
       "\n",
       "        SAS_osteosarcoma_CN_NUMBER  SAS_osteosarcoma_CN_COUNT  \\\n",
       "24400                          NaN                        NaN   \n",
       "24408                          NaN                        NaN   \n",
       "24427                          NaN                        NaN   \n",
       "88452                          NaN                        NaN   \n",
       "214559                         NaN                        NaN   \n",
       "214726                         NaN                        NaN   \n",
       "\n",
       "        SAS_osteosarcoma_CN_FREQ  SAS_osteosarcoma_CN_NONDIPLOID_COUNT  \\\n",
       "24400                        NaN                                   NaN   \n",
       "24408                        NaN                                   NaN   \n",
       "24427                        NaN                                   NaN   \n",
       "88452                        NaN                                   NaN   \n",
       "214559                       NaN                                   NaN   \n",
       "214726                       NaN                                   NaN   \n",
       "\n",
       "        SAS_osteosarcoma_CN_NONDIPLOID_FREQ  POPMAX_FREQ_HOMREF  \\\n",
       "24400                                   NaN            1.000000   \n",
       "24408                                   NaN            1.000000   \n",
       "24427                                   NaN            1.000000   \n",
       "88452                                   NaN            1.000000   \n",
       "214559                                  NaN            0.006667   \n",
       "214726                                  NaN            1.000000   \n",
       "\n",
       "        POPMAX_FREQ_HET  POPMAX_FREQ_HOMALT  POPMAX_CN_FREQ  \\\n",
       "24400          0.002584             0.00000             NaN   \n",
       "24408          0.002584             0.00000             NaN   \n",
       "24427          0.000351             0.00000             NaN   \n",
       "88452          0.002584             0.00000             NaN   \n",
       "214559         1.000000             0.02584             NaN   \n",
       "214726         0.000351             0.00000             NaN   \n",
       "\n",
       "        POPMAX_CN_NONDIPLOID_FREQ  POPMAX_control_AF  \\\n",
       "24400                         NaN           0.000000   \n",
       "24408                         NaN           0.000000   \n",
       "24427                         NaN           0.000299   \n",
       "88452                         NaN           0.000000   \n",
       "214559                        NaN           0.510460   \n",
       "214726                        NaN           0.000000   \n",
       "\n",
       "        POPMAX_control_FREQ_HOMREF  POPMAX_control_FREQ_HET  \\\n",
       "24400                          1.0                 0.000000   \n",
       "24408                          1.0                 0.000000   \n",
       "24427                          1.0                 0.000598   \n",
       "88452                          1.0                 0.000000   \n",
       "214559                         0.0                 1.000000   \n",
       "214726                         1.0                 0.000000   \n",
       "\n",
       "        POPMAX_control_FREQ_HOMALT  POPMAX_control_CN_FREQ  \\\n",
       "24400                     0.000000                     NaN   \n",
       "24408                     0.000000                     NaN   \n",
       "24427                     0.000000                     NaN   \n",
       "88452                     0.000000                     NaN   \n",
       "214559                    0.020921                     NaN   \n",
       "214726                    0.000000                     NaN   \n",
       "\n",
       "        POPMAX_control_CN_NONDIPLOID_FREQ  POPMAX_ewing_AF  \\\n",
       "24400                                 NaN         0.000000   \n",
       "24408                                 NaN         0.000000   \n",
       "24427                                 NaN         0.000000   \n",
       "88452                                 NaN         0.000000   \n",
       "214559                                NaN         0.509375   \n",
       "214726                                NaN         0.000000   \n",
       "\n",
       "       POPMAX_ewing_FREQ_HOMREF POPMAX_ewing_FREQ_HET  \\\n",
       "24400                  1.000000                   0.0   \n",
       "24408                  1.000000                   0.0   \n",
       "24427                  1.000000                   0.0   \n",
       "88452                  1.000000                   0.0   \n",
       "214559                 0.071429                   1.0   \n",
       "214726                 1.000000                   0.0   \n",
       "\n",
       "        POPMAX_ewing_FREQ_HOMALT  POPMAX_ewing_CN_FREQ  \\\n",
       "24400                   0.000000                   NaN   \n",
       "24408                   0.000000                   NaN   \n",
       "24427                   0.000000                   NaN   \n",
       "88452                   0.000000                   NaN   \n",
       "214559                  0.071429                   NaN   \n",
       "214726                  0.000000                   NaN   \n",
       "\n",
       "        POPMAX_ewing_CN_NONDIPLOID_FREQ  POPMAX_neuroblastoma_AF  \\\n",
       "24400                               NaN                 0.012820   \n",
       "24408                               NaN                 0.012820   \n",
       "24427                               NaN                 0.000000   \n",
       "88452                               NaN                 0.000000   \n",
       "214559                              NaN                 0.538462   \n",
       "214726                              NaN                 0.000000   \n",
       "\n",
       "        POPMAX_neuroblastoma_FREQ_HOMREF POPMAX_neuroblastoma_FREQ_HET  \\\n",
       "24400                           1.000000                      0.025641   \n",
       "24408                           1.000000                      0.025641   \n",
       "24427                           1.000000                      0.000000   \n",
       "88452                           1.000000                      0.000000   \n",
       "214559                          0.011601                      1.000000   \n",
       "214726                          1.000000                      0.000000   \n",
       "\n",
       "        POPMAX_neuroblastoma_FREQ_HOMALT  POPMAX_neuroblastoma_CN_FREQ  \\\n",
       "24400                           0.000000                           NaN   \n",
       "24408                           0.000000                           NaN   \n",
       "24427                           0.000000                           NaN   \n",
       "88452                           0.000000                           NaN   \n",
       "214559                          0.076923                           NaN   \n",
       "214726                          0.000000                           NaN   \n",
       "\n",
       "        POPMAX_neuroblastoma_CN_NONDIPLOID_FREQ  POPMAX_osteosarcoma_AF  \\\n",
       "24400                                       NaN                0.000000   \n",
       "24408                                       NaN                0.000000   \n",
       "24427                                       NaN                0.000000   \n",
       "88452                                       NaN                0.014706   \n",
       "214559                                      NaN                0.520833   \n",
       "214726                                      NaN                0.001894   \n",
       "\n",
       "        POPMAX_osteosarcoma_FREQ_HOMREF  POPMAX_osteosarcoma_FREQ_HET  \\\n",
       "24400                          1.000000                      0.000000   \n",
       "24408                          1.000000                      0.000000   \n",
       "24427                          1.000000                      0.000000   \n",
       "88452                          1.000000                      0.029412   \n",
       "214559                         0.029412                      1.000000   \n",
       "214726                         1.000000                      0.003788   \n",
       "\n",
       "        POPMAX_osteosarcoma_FREQ_HOMALT  POPMAX_osteosarcoma_CN_FREQ  \\\n",
       "24400                          0.000000                          NaN   \n",
       "24408                          0.000000                          NaN   \n",
       "24427                          0.000000                          NaN   \n",
       "88452                          0.000000                          NaN   \n",
       "214559                         0.049242                          NaN   \n",
       "214726                         0.000000                          NaN   \n",
       "\n",
       "       POPMAX_osteosarcoma_CN_NONDIPLOID_FREQ      FILTER  \n",
       "24400                                     NaN  UNRESOLVED  \n",
       "24408                                     NaN        PASS  \n",
       "24427                                     NaN        PASS  \n",
       "88452                                     NaN        PASS  \n",
       "214559                                    NaN        PASS  \n",
       "214726                                    NaN        PASS  \n",
       "\n",
       "[6 rows x 734 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "svs[svs[\"name\"].isin(complex_svs_in_regions[\"name\"])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbb8c2a1-90cc-42e4-8b2f-1a3d6993194d",
   "metadata": {},
   "source": [
    "Again, this process picks up only one new SV--a distributed duplication."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33b751df-aa3e-4ee5-8587-96934b00220a",
   "metadata": {},
   "source": [
    "## Subsetting and QC'ing SVs in gene regions\n",
    "\n",
    "To save memory, we subset the big SV and dosage matrices down to just these samples and SVs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "d49dc879-eaff-4496-9586-ed02c565cac4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# keep the gene names\n",
    "svs_in_regions_names = pd.concat(\n",
    "    [svs_in_regions[[\"name\", \"gene\"]], complex_svs_in_regions[[\"name\", \"gene\"]]]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "7b1cf3d7-9ded-4955-adea-0ea7bb610ec6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "369"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(svs_in_regions_names)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e2a164b-3031-4cea-9dd5-5a1d52cd79ea",
   "metadata": {},
   "source": [
    "So a decent number of SVs in this dataset fall within this region. Briefly, Ryan and Riaz wanted to know if there are any MCNVs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "a5fca74e-a42c-43fe-b760-7e14e6316db1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DEL             180\n",
       "DUP              85\n",
       "BND              49\n",
       "INS:ME:ALU       15\n",
       "INS              12\n",
       "CPX               7\n",
       "INS:ME:SVA        6\n",
       "CNV               2\n",
       "INS:ME:LINE1      1\n",
       "INV               1\n",
       "Name: svtype, dtype: int64"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "svs[svs['name'].isin(svs_in_regions_names['name'])]['svtype'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cda81ff-693b-44e0-9dc7-2966c7f0dbd4",
   "metadata": {},
   "source": [
    "So no. We fetch all these SVs from the original dataset now and apply some filters:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "147c0811-45d0-414c-a46e-438a44fe446f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PASS                                    294\n",
      "UNRESOLVED                               49\n",
      "HIGH_PCRMINUS_NOCALL_RATE                20\n",
      "HIGH_PCRMINUS_NOCALL_RATE,UNRESOLVED      4\n",
      "MULTIALLELIC                              2\n",
      "Name: FILTER, dtype: int64\n",
      "\n",
      "Rare:\n",
      " True     296\n",
      "False     73\n",
      "Name: AF, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# subset\n",
    "subset_svs = svs.set_index(\"name\").loc[svs_in_regions_names[\"name\"]].reset_index()\n",
    "subset_svs.insert(0, \"gene\", svs_in_regions_names[\"gene\"].values)\n",
    "\n",
    "# how many are PASS?\n",
    "print(subset_svs[\"FILTER\"].value_counts())\n",
    "\n",
    "# and how many are rare?\n",
    "print(\"\\nRare:\\n\", (subset_svs[\"AF\"] < 0.01).value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "675fd756-75ee-495a-8720-ca07bb816ce0",
   "metadata": {},
   "source": [
    "That's not bad. How many pass the filter, are rare, and are present in the dosage matrix?\n",
    "\n",
    "We'll add in a quick QC method to remove some suspect SVs that Ryan is concerned about."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "2c0ae28d-12c1-42e7-9bc2-391cfdf0c27d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(219, 736)"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# only include SVs that pass the FILTER, are rare, and are in our dosage matrix\n",
    "svs_to_keep = subset_svs[\n",
    "    (\n",
    "        (subset_svs[\"FILTER\"] == \"PASS\")\n",
    "        & (subset_svs[\"name\"].isin(dosages[\"ID\"].to_list()))\n",
    "        & (~subset_svs[\"EVIDENCE\"].isin([\"BAF,RD\", \"RD\"]))\n",
    "        & (~(subset_svs[\"ALGORITHMS\"] == \"depth\"))\n",
    "    )\n",
    "].copy()\n",
    "\n",
    "svs_to_keep[\"rare\"] = svs_to_keep[\"AF\"] < 0.01\n",
    "\n",
    "svs_to_keep.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "30f34365-4cd6-4a2a-a73d-fecea8a227b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DEL             146\n",
       "DUP              25\n",
       "INS:ME:ALU       15\n",
       "CPX              14\n",
       "INS              11\n",
       "INS:ME:SVA        6\n",
       "INS:ME:LINE1      1\n",
       "INV               1\n",
       "Name: svtype, dtype: int64"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "svs_to_keep[\"svtype\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "97238224-ed7a-47e6-8d0b-2d0add992e14",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 0, 'AF')"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXIAAADSCAYAAABXT0tTAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/NK7nSAAAACXBIWXMAAA7EAAAOxAGVKw4bAAAXQElEQVR4nO3df1BU190G8GeBBd1FggLqyK90S4gsmaQJxhhjg6TFwZEETNRM25DEVJNqNNNREkg1KNpNSNRmYg1pZdpQjFHbSJa2VGqdIEPTRlo62LqAhlqrGA1UUFkWWJDz/uHI6wqEvcvu3r13n8/MziRnl3u/Z0kezpx77j0aIYQAEREpVoDcBRAR0fgwyImIFI5BTkSkcAxyIiKFY5ATESkcg5yISOEY5EREChckdwFyuXz5MmpqahAbG4uQkBC5yyEictDX14dz584hNTUV4eHhX/lZvw3ympoaZGdny10GEdFXMpvNyMrK+srP+G2Qx8bGArj+JSUkJMhcDRGRo5aWFmRnZw9l1Vfx2yC/MZ2SkJCA5ORkmashIhqZM1O/vNhJRKRwDHIiIoXzq6mV/v5+DAwMAAB6e3tlroaIyD38akRuMpmg0+mg0+kwa9YsucshInILvwryDRs2wGazwWaz4e9//7tLx9j8zZ8MvYiIfIFfTa1otVpotVoAwIQJE2SuhojIPfxqRE5EpEYMciIihWOQExEpHIOciEjhGORERArHICciUjgGORGRwik2yHt7e/HEE08gNTUVqamp+Pe//y13SUREslBskJeWliIlJQU1NTXYsGEDtm7dKndJRESyUOydnd/73vcQEHD979DAwADv1CQiv6XYIJ80aRIA4OzZs8jLy0NFRYXMFRERyUOxUysA0NjYiMzMTJSUlMBgMMhdDhGRLBQ7Ij937hyWLVuG/fv346677pK7HCIi2Sg2yN944w10dXVhzZo1AICUlBTs2LFjxM+2tbWhvb3doa2lpcXjNRIReYPPBHldXR0WLlyIS5cuDbVZLBa88MILaGhowIwZM/Dmm29i8eLFAIDi4mKnj11cXIzCwkK310xE5At8Yo583759SE9PR19f31Cb3W5HZmYmsrKy0NnZiZ07dyInJwfNzc2Sj7969WqcOHHC4WU2m93YAyIi+cge5Pn5+di+fTsKCgoc2qurq9Hd3Y3c3FxotVpkZGRg0aJFKCsrk3yOqVOnIjk5GYmJiTAYDDAYDIiJiXFXF4iIZCV7kK9duxb19fVISUlxaG9qaoLRaIRGoxlqS0pKgsVicflc3LOTiNRI9iCPjo4esd1qtUKn0zm06XQ6dHd3u3wud+zZSUTka2QP8tHo9Xr09PQ4tNlsNoSGhspUERGRb/LZIDcajTh58qRDW2NjI4xGo8vH5NQKEamRzwZ5WloatFotXn/9ddjtdlRVVaGyshJPPfWUy8fk1AoRqZHPBnlwcDCqqqpw+PBhREZG4qWXXsKePXvGNSInIlIjn7khaP78+bBarQ5tSUlJOHr0qNvOYTKZeGMQEamOz47IPYFTK0SkRj4zIvcGrVYLrVYLAHx+ORGphl8FeX9/PwYGBgBc3yqOiEgN/GpqhcsPiUiN/CrIOUdORGrkV1MrnCMnIjXyqxE5EZEa+dWInBc7iUiN/GpEzoudRKRGfhXkvNhJRGrkV1MrvNhJRGrkVyNyIiI1YpATESmcX02tcNUKEamRX43IuWqFiNTIr4Kcq1aISI38amqFq1aISI38akRORKRGDHIiIoVjkBMRKZxfzZFz+SERqZFfjci5/JCI1EhykJ89exb9/f0jvtfT04NPP/103EV5CpcfEpEaSQ7yr33ta2hoaBjxvbq6OqSnp4+3Jo/RarWYOHEiJk6cyOWHRKQaTs2Rr1y5El9++SUAQAiB9evXIzw8fNjnmpqaRmwnIiLPcSrI586diy1btgAANBoNWlpaEBIS4vCZwMBATJ48GSaTyf1VEhHRqJwK8uXLl2P58uUAgICAAJjNZsyePdujhRERkXMkLz8cHBz0RB1EROQil9aR19fXw2w2o6ura1iwazQavPPOO24pjoiIxiY5yH/5y19i5cqVEEKM+D6DnIjIuyQvP3zrrbeQlpaGM2fO4Nq1axgcHHR4Xbt2zRN1EhHRKCSPyM+cOYNdu3YhLi7OE/V4FG/RJyI1kjwiv/3229HR0eGJWjyOt+gTkRpJDvLc3Fxs3boVX3zxhSfq8Sjeok9EaiR5auWPf/wjzp8/j/j4eMTFxUGv1zu8r9FocPz4cbcV6E7cIYiI1EhykP/vf//DPffc44laiIjIBZKDvLq62hN1EBGRi/zqeeRERGokeUR+9913j/mZf/7zny4VQ0RE0kkO8ilTpkCj0Ti0dXV1wWKx4LbbbsOyZcvcVhwREY1NcpAfPXp0xPZLly4hIyMDM2fOHG9NREQkgdvmyCMiIpCfn4+3337bXYckIiInuPViZ2hoKFpbW915SKe88847KCoq8vp5iYh8geSplZFuzx8cHERrays2b96MO+64wy2FOUMIgVWrVuHIkSNYsWKF185LRORLJAd5ZGTksIudN2g0Ghw4cGDcRTlLCIGMjAzMmTMHFy9e9Np5iYh8ieQgLygoGBbkGo0GYWFheOyxx2AwGNxW3FgCAgKQnZ2N0tJSr52TiMjXSA7yzZs3e6AMIiJylUtbvQ0ODuLXv/41jhw5gitXriAyMhKPPPIIHn/8cQQGBrq7RiIi+gqSg7y3txcLFy5ETU0NAgMDERkZifb2duzevRsPPfQQ/vSnPyEkJMQTtbqsra0N7e3tDm0tLS0yVUNE5F6Slx8WFhbi2LFj+MUvfoGenh5cuHABPT09KCkpQX19PUwmk8vF1NXVISIiwqHNYrFg3rx5CA0NRWJiIj7++ONhP/fss88iPz9/1OMWFxfjrrvucnhlZ2e7XCcRkS+RHOT79+/Hxo0bsXz5cgQFXR/Qa7VaPPfcc9iwYQP27dvnUiH79u1Deno6+vr6htrsdjsyMzORlZWFzs5O7Ny5Ezk5OWhubpZ07NWrV+PEiRMOL7PZ7FKdRES+RnKQX7x4EXPmzBnxvQceeMClG4Ly8/Oxfft2FBQUOLRXV1eju7sbubm50Gq1yMjIwKJFi1BWVibp+FOnTkVycjISExNhMBhgMBgQExMjuU4iIl/k0p6do+0A1NDQgKioKMlFrF27FvX19UhJSXFob2pqgtFodFjumJSUBIvFIvkcAPfsJCJ1khzkS5cuRWFhIX772986tFdUVGDr1q144oknJBcRHR09YrvVaoVOp3No0+l06O7ulnwOgHt2EpE6SV61kpeXh8rKSmRnZyM0NBTTp0/HxYsXYbVace+992LLli1uK06v16Onp8ehzWazITQ01G3nICJSOslBrtfr8de//hWlpaX45JNP0NHRgZiYGKSlpeHll19266bGRqMR27Ztc2hrbGyE0Wh06XgmkwmFhYXuKA0AsPmbP/n/f65d57bjEhFJ4dJDsx577DG0tbXh1KlTAIDKyko8+uijqK2tRXl5udtGzGlpadBqtXj99deRm5uLTz75BJWVlfjb3/7m0vE2bNiAvLw8ANf/IHCenIjUQPIc+Y9+9CN8/vnneOONN4ba0tPTUV5ejn/9619uHfEGBwejqqoKhw8fRmRkJF566SXs2bPH5RE5EZEaSR6R//73v8e2bdscLmoGBwcjOzsbly5dwo9//ONh0yHOmj9/PqxWq0NbUlLSqLsSSeXuqRUiIl8geUTe2dmJ6dOnj/hefHy8Tz9OlqtWiEiNJAf5nXfeifLy8hHfq6ioQEJCwriL8hStVouJEydi4sSJbr0oS0QkJ8lTK6tXr8bzzz+Prq4uLF26FNOmTUNbWxvKy8uxd+9eFBcXe6JOt+jv78fAwACA6w//IiJSA8lBvmLFCpw/fx5FRUXYv38/gOs79YSEhOC1117D888/7/Yi3YVz5ESkRi5tvrxp0ya0t7ejqqoKH3zwASorK/HFF19g06ZN7q7PrThHTkRq5NLGEgAQGhqK9PR0d9bicVqtFlqtFgA4R05EquHSiJyIiHyHyyNyJeLFTiJSI78akfMxtkSkRn4V5LzYSURq5FdTK7zYSURq5FcjciIiNWKQExEpnF9NrXDVChGpkV8FOW/RJ2/gzlHkbX41tcJVK0SkRn41IueqFSJSI78akRMRqRGDnIhI4RjkREQK51dz5Fx+SERq5FdB7u3lh1yGRkTe4FdTK1x+SERq5Fcjci4/JCI18qsRORGRGjHIiYgUjkFORKRwDHIiIoVjkBMRKRyDnIhI4RjkREQK51fryHmLPhGpkV8Fubdu0b/51nwiIk/zq6kV3qJPRGrkVyNy3qJPRGrkVyNyIiI1YpATESkcg5yISOEY5ERECscgJyJSOAY5EZHCMciJiBROsUE+MDCAnJwczJs3D48//ji6u7vlLomISBaKDfKDBw9i6tSp+POf/4y0tDSUlJTIXRIRkSwUG+SfffYZFixYAABYsGABampqZK6IiEgeig3yq1evIiwsDAAwadIkdHV1yVwREZE8FBvkYWFhQ+Hd1dWF2267TeaKiIjkodggnzNnDg4fPgwAOHz4MB588EGZKyIikodin364ZMkS/OEPf8BDDz2E8PBwHDhwYNTPtrW1ob293aGtpaXF0yUSEXmFzwR5XV0dFi5ciEuXLg21WSwWvPDCC2hoaMCMGTPw5ptvYvHixQCAwMBA/OpXv3Lq2MXFxR7fUGKszSSkbDaxuXbdeMtRhZG+M298NzfOe/O5bq7FUzWMdF65eKO/nuJL36O3+MTUyr59+5Ceno6+vr6hNrvdjszMTGRlZaGzsxM7d+5ETk4OmpubJR9/9erVOHHihMPLbDa7sQdERPKRPcjz8/Oxfft2FBQUOLRXV1eju7sbubm50Gq1yMjIwKJFi1BWVib5HFOnTkVycjISExNhMBhgMBgQExPjri4QEclK9iBfu3Yt6uvrkZKS4tDe1NQEo9EIjUYz1JaUlASLxeLyuUwmE3Q6HXQ6HWbNmuXycYiIfInsQR4dHT1iu9VqhU6nc2jT6XTjuhWfe3YSkRrJHuSj0ev16OnpcWiz2WwIDQ2VqSIiIt/kM6tWbmU0GrFt2zaHtsbGRhiNRpePaTKZhq1ekboMsa37osvnd9Z4po/UZKTv2hvfzY3z3nyum2sZqwYpnx3rvHJxtQ++wJe+x/G4kU03LwIZlfAR1dXVQq/XD/17X1+fiIuLEyaTSfT19YlDhw4JnU4nLBaLy+ew2+3CZrMJm80mDhw4IADwxRdffPn0y2w2j5ltPjsiDw4ORlVVFVatWoWioiJMnz4de/bsGdeIXKvVQqvVArj+oC2z2YzY2FiEhIQ49fMtLS3Izs6G2WxGQkKCy3X4IvZNedTaL4B9A66PxM+dO4fU1NQxj+kzQT5//nxYrVaHtqSkJBw9etQj5wsPD0dWVpZLP5uQkIDk5GQ3V+Qb2DflUWu/APbtvvvuc+pYPnuxk4iInMMgJyJSOAY5EZHCMcgliIqKwqZNmxAVFSV3KW7HvimPWvsFsG9SaYQQwm1HIyIir+OInIhI4RjkREQKxyAnIlI4BjkRkcIxyEdgsVgwb948hIaGIjExER9//PGInzt//jwWLlyIsLAwxMXF4Wc/+5mXK5XG2X5ZLBakp6dj8uTJmDFjBn74wx869+AeGTnbtxsGBweRmpqKNWvWeKlC1znbN6vViu9///uIiIjAlClTkJOTM67HPnuas/06deoUvv3tbyM8PBzR0dHYuHEjlLJGo66uDhEREaO+77YMcfkJVCrV19cnbr/9dvHWW28Ju90uDh06JPR6vWhqahr22blz54q1a9eK3t5ecezYMRERESGOHDkiQ9Vjc7Zf/f39Ii4uThQWFgq73S5aW1vF/fffL1555RWZKh+blN/ZDVu2bBEBAQHixRdf9GKl0knp25NPPikWLVokLl++LDo6OsTDDz8s1q1bJ0PVY5PSr1mzZomtW7eKgYEBcfr0aRETEyPKyspkqFqaDz/8UISFhTk8DPBW7soQBvktqqqqRFRUlBgcHBxqW7ZsmXj11VcdPtfc3CwCAwPFlStXhtpeeeUV8Z3vfMdrtUrhbL/OnDkjMjIyxMDAwFDbzp07xYMPPui1WqVytm83/OUvfxFJSUliyZIlPh/kzvbtwoULIigoSFy4cMGhraWlxWu1SiHldxYaGio2bdok+vv7xenTp0VsbKz4zW9+481yJcvLyxP33Xef2L59+6hB7s4M4dTKLZzdYq6pqQkxMTEICwv7ys/5Cmf7FR8fj0OHDiEwMBAAIIRARUUFvvGNb3izXEmkbAt49epVPPPMMygtLYVer/dmmS5xtm//+Mc/EB0djQMHDsBgMCA6OhpFRUWj7sAlNym/s9deew0mkwkTJkyAwWDAggULsGTJEm+WK9loW1jezJ0ZwiC/hbNbzHliKzpPcqXewcFBvPjiizh16tSwzbF9iZS+rVq1Ck8//TRmz57trfLGxdm+dXR0oLW1FfX19WhoaMCxY8fw6aefYuPGjd4s12lSfmdBQUEoKiqC1WqFxWJBbW0tduzY4a1SXeLMH1B3ZgiD/BbObjGntK3opNZ7+fJlPProo6ipqUFtbS2mT5/ujTJd4mzfysrK8N///hevvvqqN8sbF2f7FhISgmvXrmHbtm0ICwtDTEwM8vLycPDgQW+W6zRn+1VfX4+ioiKsW7cOEyZMgNFoREFBAX760596s1yPcGeGMMhvYTQacfLkSYe2kbaYMxqNaG1tdXiG+ni3ovMkZ/sFAP/5z38we/ZsBAUF4bPPPkN8fLy3ynSJs33bu3cvjh8/joiICISHh+PDDz9ESUkJ7r77bm+WK4mzfUtKSgIA2O32obZr16757OoOZ/t19uxZ9Pf3O7QFBwcjKMhntlJwmVszxMW5fNWSssXcAw88IFatWiVsNpuoq6sTERERoqqqSoaqx+Zsv7q6ukRCQoJYsWKFw4UoX+bqtoDPPPOMz1/slPrf45NPPimuXr0qzp8/L+69916Rn58vQ9Vjc7ZfX375pZg8ebLIy8sTdrtdnDlzRtxzzz2ioKBApsqluXULy1u5K0MY5CNobGwUqampYtKkSeKOO+4QBw8eFEII8cEHHzj8UlpbW0VmZqYIDw8XsbGx4r333pOrZKc406+f//znAoDQ6XRCr9cPve6//345Sx+Ts7+zmykhyIVwvm/t7e3iqaeeEtOmTRNTpkwRa9asEb29vXKVPSZn+1VXVydSU1OH/j/buHGjsNvtcpUtya1B7qkM4dMPiYgUjnPkREQKxyAnIlI4BjkRkcIxyImIFI5BTkSkcAxyIiKFY5ATESkcg5zIBTt27IBGo8Ejjzwy4vsajeYrX2az2bsFk6rxhiAiF8ycORMTJkzA8ePHYbFYhj0fQ6PRYMmSJVi/fv2IP3/nnXdi8uTJ3iiV/IDynzxD5GU1NTU4efIkqqursXjxYuzatQvFxcXDPjdt2jTMmTNHhgrJ33BqhUii3bt3Y9q0aXj44Yfx3e9+F3v27MGVK1fkLov8GIOcSIKOjg4cPHgQzz77LAICArBy5UpYrVaUlpYO+6wQAgMDA8NenM0kd+McOZEEb7/9NtavX4/PP/8cX//61wEAs2fPxpUrV9Dc3Dy0ddnNW5jd6r333sMPfvADr9RL/oFBTiSB0WhEREQEfve73w21vf/++1i3bh0OHTqEjIwMANeDfNmyZXj55ZeHHSM+Ph5RUVFeq5nUjxc7iZxUW1uLpqYmABhxxcmuXbuGghwAoqKiMGvWLK/VR/6LQU7kpN27d0Ov16OiogKBgYEO77377rsoLy/H6dOnYTAYZKqQ/BWDnMgJnZ2d+Oijj7B06VJ861vfGvZ+cHAwPvroI7z77rs+v8M7qQ9XrRA5oaysDL29vXj66adHfH/u3LmYOXMm3n//fdhsNi9XR/6OQU7khJKSEkRHR496Sz4APPfcc+js7MTevXu9WBkRV60QESkeR+RERArHICciUjgGORGRwjHIiYgUjkFORKRwDHIiIoVjkBMRKRyDnIhI4RjkREQKxyAnIlI4BjkRkcL9H9vYFW9Ag5yXAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 384x192 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(1, figsize=(4, 2))\n",
    "plt.hist(svs_to_keep[\"AF\"], bins=np.arange(0, 1, 0.01))\n",
    "plt.yscale(\"log\")\n",
    "plt.ylabel(\"count\")\n",
    "plt.xlabel(\"AF\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31037361-2603-42df-9e60-05314ef5ceb1",
   "metadata": {},
   "source": [
    "So there are some SVs in this region, and the vast majority are rare (unsurprisingly). We'll analyze them all together, but at some points keep rare separate."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72b7834b-bc22-49bc-a2e9-42e402a52ac9",
   "metadata": {},
   "source": [
    "## Extract sample dosages\n",
    "\n",
    "Here we pull out the allele counts for each SV, so that we can determine case vs. control."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "37dde93c-0a4c-4239-873e-b8346bec52f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "      <th>AF</th>\n",
       "      <th>rare</th>\n",
       "      <th>#chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ID</th>\n",
       "      <th>PT_2AHXQP58</th>\n",
       "      <th>PT_GMGNDRBJ</th>\n",
       "      <th>PT_MEC48DAY</th>\n",
       "      <th>PT_N8PKC3YS</th>\n",
       "      <th>PT_VEMC27RN</th>\n",
       "      <th>SJ051090</th>\n",
       "      <th>SJ051119</th>\n",
       "      <th>SJ051140</th>\n",
       "      <th>SJ051197</th>\n",
       "      <th>SJ051210</th>\n",
       "      <th>SJ051231</th>\n",
       "      <th>SJ051244</th>\n",
       "      <th>SJ051246</th>\n",
       "      <th>SJ051259</th>\n",
       "      <th>SJ051263</th>\n",
       "      <th>SJ051267</th>\n",
       "      <th>SJ051298</th>\n",
       "      <th>SJ051299</th>\n",
       "      <th>SJ051310</th>\n",
       "      <th>SJ051322</th>\n",
       "      <th>SJ051329</th>\n",
       "      <th>SJ051336</th>\n",
       "      <th>SJ051367</th>\n",
       "      <th>SJ051390</th>\n",
       "      <th>SJ051391</th>\n",
       "      <th>SJ051392</th>\n",
       "      <th>SJ051416</th>\n",
       "      <th>SJ051439</th>\n",
       "      <th>SJ051485</th>\n",
       "      <th>SJ051531</th>\n",
       "      <th>SJ051539</th>\n",
       "      <th>SJ051566</th>\n",
       "      <th>SJ051574</th>\n",
       "      <th>SJ051610</th>\n",
       "      <th>SJ051653</th>\n",
       "      <th>SJ051655</th>\n",
       "      <th>SJ051662</th>\n",
       "      <th>SJ051672</th>\n",
       "      <th>SJ051679</th>\n",
       "      <th>SJ051694</th>\n",
       "      <th>SJ051705</th>\n",
       "      <th>SJ051713</th>\n",
       "      <th>...</th>\n",
       "      <th>SJ054333</th>\n",
       "      <th>SJ055976</th>\n",
       "      <th>SJ055977</th>\n",
       "      <th>SJ055980</th>\n",
       "      <th>SJ055992</th>\n",
       "      <th>SJ056005</th>\n",
       "      <th>SJ056011</th>\n",
       "      <th>SJ056035</th>\n",
       "      <th>SJ056044</th>\n",
       "      <th>SJ056062</th>\n",
       "      <th>SJ056068</th>\n",
       "      <th>SJ056156</th>\n",
       "      <th>SJ056241</th>\n",
       "      <th>SJ056250</th>\n",
       "      <th>SJ056312</th>\n",
       "      <th>SJ056317</th>\n",
       "      <th>SJ056321</th>\n",
       "      <th>SJ056360</th>\n",
       "      <th>SJ056374</th>\n",
       "      <th>SJ056381</th>\n",
       "      <th>SJ056446</th>\n",
       "      <th>SJ056468</th>\n",
       "      <th>SJ056480</th>\n",
       "      <th>SJ056503</th>\n",
       "      <th>SJ056509</th>\n",
       "      <th>SJ056535</th>\n",
       "      <th>SJ056833</th>\n",
       "      <th>SJ056855</th>\n",
       "      <th>SJ056862</th>\n",
       "      <th>SJ056888</th>\n",
       "      <th>SJ056891</th>\n",
       "      <th>SJ056916</th>\n",
       "      <th>SJ056922</th>\n",
       "      <th>SJ056924</th>\n",
       "      <th>SJ058208</th>\n",
       "      <th>SJ058211</th>\n",
       "      <th>SJ058213</th>\n",
       "      <th>SJ058220</th>\n",
       "      <th>SJ058227</th>\n",
       "      <th>SJ058233</th>\n",
       "      <th>SJ058237</th>\n",
       "      <th>SJ058252</th>\n",
       "      <th>SJ058253</th>\n",
       "      <th>SJ058283</th>\n",
       "      <th>SJ063819</th>\n",
       "      <th>SJ063829</th>\n",
       "      <th>SJ063834</th>\n",
       "      <th>SJ071780</th>\n",
       "      <th>SJ071781</th>\n",
       "      <th>ssi_11896</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_DUP_chr2_523</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>0.000140</td>\n",
       "      <td>True</td>\n",
       "      <td>chr2</td>\n",
       "      <td>15453980</td>\n",
       "      <td>15455198</td>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_DUP_chr2_523</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_DEL_chr2_1106</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>0.000279</td>\n",
       "      <td>True</td>\n",
       "      <td>chr2</td>\n",
       "      <td>15463105</td>\n",
       "      <td>15465416</td>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_DEL_chr2_1106</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 3589 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              name  gene        AF  rare  \\\n",
       "1   All_20_Batches.chr2.final_cleanup_DUP_chr2_523  MYCN  0.000140  True   \n",
       "2  All_20_Batches.chr2.final_cleanup_DEL_chr2_1106  MYCN  0.000279  True   \n",
       "\n",
       "   #chr     start       end                                               ID  \\\n",
       "1  chr2  15453980  15455198   All_20_Batches.chr2.final_cleanup_DUP_chr2_523   \n",
       "2  chr2  15463105  15465416  All_20_Batches.chr2.final_cleanup_DEL_chr2_1106   \n",
       "\n",
       "   PT_2AHXQP58  PT_GMGNDRBJ  PT_MEC48DAY  PT_N8PKC3YS  PT_VEMC27RN  SJ051090  \\\n",
       "1          0.0          0.0          0.0          0.0          0.0       0.0   \n",
       "2          0.0          0.0          0.0          0.0          0.0       0.0   \n",
       "\n",
       "   SJ051119  SJ051140  SJ051197  SJ051210  SJ051231  SJ051244  SJ051246  \\\n",
       "1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "   SJ051259  SJ051263  SJ051267  SJ051298  SJ051299  SJ051310  SJ051322  \\\n",
       "1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "   SJ051329  SJ051336  SJ051367  SJ051390  SJ051391  SJ051392  SJ051416  \\\n",
       "1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "   SJ051439  SJ051485  SJ051531  SJ051539  SJ051566  SJ051574  SJ051610  \\\n",
       "1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "   SJ051653  SJ051655  SJ051662  SJ051672  SJ051679  SJ051694  SJ051705  \\\n",
       "1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "   SJ051713  ...  SJ054333  SJ055976  SJ055977  SJ055980  SJ055992  SJ056005  \\\n",
       "1       0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "2       0.0  ...       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "   SJ056011  SJ056035  SJ056044  SJ056062  SJ056068  SJ056156  SJ056241  \\\n",
       "1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "   SJ056250  SJ056312  SJ056317  SJ056321  SJ056360  SJ056374  SJ056381  \\\n",
       "1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "   SJ056446  SJ056468  SJ056480  SJ056503  SJ056509  SJ056535  SJ056833  \\\n",
       "1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "   SJ056855  SJ056862  SJ056888  SJ056891  SJ056916  SJ056922  SJ056924  \\\n",
       "1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "   SJ058208  SJ058211  SJ058213  SJ058220  SJ058227  SJ058233  SJ058237  \\\n",
       "1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "   SJ058252  SJ058253  SJ058283  SJ063819  SJ063829  SJ063834  SJ071780  \\\n",
       "1       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "2       0.0       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "   SJ071781  ssi_11896  \n",
       "1       0.0        0.0  \n",
       "2       0.0        0.0  \n",
       "\n",
       "[2 rows x 3589 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# we begin by subsetting the dosage matrix down to the SVs that we care about. This double counts SVs\n",
    "# that affect more than one gene.\n",
    "subset_dosages = dosages[dosages[\"ID\"].isin(svs_to_keep[\"name\"].to_list())].reset_index(\n",
    "    drop=True\n",
    ")\n",
    "\n",
    "# we add in the information about which gene is affected (the first column is just the SV name, so we drop it)\n",
    "subsetted_gene_dosages = (\n",
    "    svs_to_keep[[\"name\", \"gene\", \"AF\", \"rare\"]]\n",
    "    .merge(subset_dosages, right_on=\"ID\", left_on=\"name\")\n",
    "    .iloc[1:]\n",
    ")\n",
    "\n",
    "subsetted_gene_dosages.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14b74184-633b-4edf-bf7d-1c24b70d4593",
   "metadata": {},
   "source": [
    "Next, we devise two metrics for cases and controls--one listing the allele frequency of each SV and another listing the number of samples with an SV in the gene.\n",
    "\n",
    "We need to convert the dosage (allele count) to allele frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "16e21870-6df1-41b5-9b8e-58aa73c54415",
   "metadata": {},
   "outputs": [],
   "source": [
    "case_control_counts = []\n",
    "case_control_samples_with_svs = []\n",
    "\n",
    "for sample_group, label in zip(\n",
    "    [nbl_samples, control_samples], [\"cases\", \"controls\"]\n",
    "):\n",
    "\n",
    "    # subset to these samples\n",
    "    sample_dosages = subsetted_gene_dosages[\n",
    "        [\"name\", \"gene\", \"#chr\", \"start\", \"end\", \"AF\", \"rare\"] + sample_group\n",
    "    ]\n",
    "\n",
    "    # first, count the SVs. This function counts the number of samples with each genotype.\n",
    "    sv_counts = (\n",
    "        sample_dosages[sample_group]\n",
    "        .apply(lambda row: row.dropna().value_counts(), axis=1)\n",
    "        .fillna(0)\n",
    "        .astype(int)\n",
    "    )\n",
    "\n",
    "    # name columns to genotypes\n",
    "    sv_counts.columns = f\"{label}_\" + sv_counts.columns.astype(int).astype(str)\n",
    "    sv_counts[\"name\"] = sample_dosages[\"name\"]\n",
    "\n",
    "    case_control_counts.append(sv_counts)\n",
    "\n",
    "    # now count the number of samples with at least one SV around each gene.\n",
    "    # we groupby gene, then determine whether each sample has an SV in that gene, then\n",
    "    # count these samples up per gene.\n",
    "    num_samples_with_svs = sample_dosages.groupby([\"gene\"]).apply(\n",
    "        lambda group: (group[sample_group] > 0).any(axis=0).sum()\n",
    "    )\n",
    "\n",
    "    num_samples_with_svs.name = label + \"_all\"\n",
    "    num_samples_with_svs = num_samples_with_svs / len(sample_group)\n",
    "    case_control_samples_with_svs.append(num_samples_with_svs)\n",
    "\n",
    "    # repeat for rare SVs\n",
    "    num_samples_with_svs_rare = (\n",
    "        sample_dosages.query(\"rare == True\")\n",
    "        .groupby([\"gene\"])\n",
    "        .apply(lambda group: (group[sample_group] > 0).any(axis=0).sum())\n",
    "    )\n",
    "\n",
    "    num_samples_with_svs_rare.name = label + \"_rare\"\n",
    "    num_samples_with_svs_rare = num_samples_with_svs_rare / len(sample_group)\n",
    "    case_control_samples_with_svs.append(num_samples_with_svs_rare)\n",
    "\n",
    "# merge\n",
    "case_control_sv_counts = sample_dosages.iloc[:, :7].reset_index(drop=True)\n",
    "for df in case_control_counts:\n",
    "    case_control_sv_counts = case_control_sv_counts.merge(df)\n",
    "\n",
    "case_control_sv_counts[\"midpoint\"] = np.mean(\n",
    "    case_control_sv_counts[[\"start\", \"end\"]], axis=1\n",
    ")\n",
    "\n",
    "# and total up the sample counts\n",
    "case_control_sample_counts = pd.DataFrame(case_control_samples_with_svs).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "9211cb53-58d1-46f8-a68a-0a693e42a4ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "      <th>#chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>AF</th>\n",
       "      <th>rare</th>\n",
       "      <th>cases_0</th>\n",
       "      <th>cases_1</th>\n",
       "      <th>cases_2</th>\n",
       "      <th>controls_0</th>\n",
       "      <th>controls_1</th>\n",
       "      <th>controls_2</th>\n",
       "      <th>midpoint</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_DUP_chr2_523</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>chr2</td>\n",
       "      <td>15453980</td>\n",
       "      <td>15455198</td>\n",
       "      <td>0.000140</td>\n",
       "      <td>True</td>\n",
       "      <td>529</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2097</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>15454589.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_DEL_chr2_1106</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>chr2</td>\n",
       "      <td>15463105</td>\n",
       "      <td>15465416</td>\n",
       "      <td>0.000279</td>\n",
       "      <td>True</td>\n",
       "      <td>529</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2096</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>15464260.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              name  gene  #chr     start  \\\n",
       "0   All_20_Batches.chr2.final_cleanup_DUP_chr2_523  MYCN  chr2  15453980   \n",
       "1  All_20_Batches.chr2.final_cleanup_DEL_chr2_1106  MYCN  chr2  15463105   \n",
       "\n",
       "        end        AF  rare  cases_0  cases_1  cases_2  controls_0  \\\n",
       "0  15455198  0.000140  True      529        0        0        2097   \n",
       "1  15465416  0.000279  True      529        0        0        2096   \n",
       "\n",
       "   controls_1  controls_2    midpoint  \n",
       "0           1           0  15454589.0  \n",
       "1           2           0  15464260.5  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "case_control_sv_counts.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dcc443f-9855-41e8-a315-41f044b56f36",
   "metadata": {},
   "source": [
    "Next we add some columns for allele frequency."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "b6ad6c55-44d8-4d81-b7f5-e80111891291",
   "metadata": {},
   "outputs": [],
   "source": [
    "for label in [\"cases\", \"controls\"]:\n",
    "\n",
    "    # a quick check to see if there are non-0/1/2 genotypes\n",
    "    columns = [col for col in case_control_sv_counts.columns if col.startswith(label)]\n",
    "    assert columns == [f\"{label}_{i}\" for i in range(3)]\n",
    "\n",
    "    alt = (case_control_sv_counts[columns] * np.array([0, 1, 2])).sum(axis=1)\n",
    "    ref = case_control_sv_counts[columns].sum(axis=1) * 2\n",
    "    case_control_sv_counts[f\"AF_{label}\"] = alt / ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "3857c526-8061-4701-96a3-fad2ace72d5b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "      <th>#chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>AF</th>\n",
       "      <th>rare</th>\n",
       "      <th>cases_0</th>\n",
       "      <th>cases_1</th>\n",
       "      <th>cases_2</th>\n",
       "      <th>controls_0</th>\n",
       "      <th>controls_1</th>\n",
       "      <th>controls_2</th>\n",
       "      <th>midpoint</th>\n",
       "      <th>AF_cases</th>\n",
       "      <th>AF_controls</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_DUP_chr2_523</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>chr2</td>\n",
       "      <td>15453980</td>\n",
       "      <td>15455198</td>\n",
       "      <td>0.000140</td>\n",
       "      <td>True</td>\n",
       "      <td>529</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2097</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>15454589.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_DEL_chr2_1106</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>chr2</td>\n",
       "      <td>15463105</td>\n",
       "      <td>15465416</td>\n",
       "      <td>0.000279</td>\n",
       "      <td>True</td>\n",
       "      <td>529</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2096</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>15464260.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000477</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              name  gene  #chr     start  \\\n",
       "0   All_20_Batches.chr2.final_cleanup_DUP_chr2_523  MYCN  chr2  15453980   \n",
       "1  All_20_Batches.chr2.final_cleanup_DEL_chr2_1106  MYCN  chr2  15463105   \n",
       "\n",
       "        end        AF  rare  cases_0  cases_1  cases_2  controls_0  \\\n",
       "0  15455198  0.000140  True      529        0        0        2097   \n",
       "1  15465416  0.000279  True      529        0        0        2096   \n",
       "\n",
       "   controls_1  controls_2    midpoint  AF_cases  AF_controls  \n",
       "0           1           0  15454589.0       0.0     0.000238  \n",
       "1           2           0  15464260.5       0.0     0.000477  "
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "case_control_sv_counts.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "6aa1eb9a-8c92-4f14-a68b-79916d179186",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cases_all</th>\n",
       "      <th>cases_rare</th>\n",
       "      <th>controls_all</th>\n",
       "      <th>controls_rare</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16pmicrodel</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.204159</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.153956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYCN</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.117202</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.071020</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             cases_all  cases_rare  controls_all  controls_rare\n",
       "gene                                                           \n",
       "16pmicrodel        1.0    0.204159           1.0       0.153956\n",
       "MYCN               1.0    0.117202           1.0       0.071020"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "case_control_sample_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22f7d929-06e7-4fc2-a5c8-c67de4621239",
   "metadata": {},
   "source": [
    "Any smoking guns in terms of SVs?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "ffcbf6ff-026e-4156-9782-8ad0eb5dc59c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "      <th>#chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>AF</th>\n",
       "      <th>rare</th>\n",
       "      <th>cases_0</th>\n",
       "      <th>cases_1</th>\n",
       "      <th>cases_2</th>\n",
       "      <th>controls_0</th>\n",
       "      <th>controls_1</th>\n",
       "      <th>controls_2</th>\n",
       "      <th>midpoint</th>\n",
       "      <th>AF_cases</th>\n",
       "      <th>AF_controls</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_DUP_chr2_537</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>chr2</td>\n",
       "      <td>16113712</td>\n",
       "      <td>16113862</td>\n",
       "      <td>0.553337</td>\n",
       "      <td>False</td>\n",
       "      <td>21</td>\n",
       "      <td>347</td>\n",
       "      <td>161</td>\n",
       "      <td>134</td>\n",
       "      <td>1846</td>\n",
       "      <td>118</td>\n",
       "      <td>16113787.0</td>\n",
       "      <td>0.632325</td>\n",
       "      <td>0.496187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_INS_chr2_363</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>chr2</td>\n",
       "      <td>16395357</td>\n",
       "      <td>16395367</td>\n",
       "      <td>0.346272</td>\n",
       "      <td>False</td>\n",
       "      <td>90</td>\n",
       "      <td>438</td>\n",
       "      <td>1</td>\n",
       "      <td>790</td>\n",
       "      <td>1308</td>\n",
       "      <td>0</td>\n",
       "      <td>16395362.0</td>\n",
       "      <td>0.415879</td>\n",
       "      <td>0.311725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_DUP_chr2_541</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>chr2</td>\n",
       "      <td>16225124</td>\n",
       "      <td>16226720</td>\n",
       "      <td>0.858699</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>134</td>\n",
       "      <td>395</td>\n",
       "      <td>0</td>\n",
       "      <td>621</td>\n",
       "      <td>1477</td>\n",
       "      <td>16225922.0</td>\n",
       "      <td>0.873346</td>\n",
       "      <td>0.852002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_DEL_chr2_1123</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>chr2</td>\n",
       "      <td>15787705</td>\n",
       "      <td>15787796</td>\n",
       "      <td>0.043883</td>\n",
       "      <td>False</td>\n",
       "      <td>420</td>\n",
       "      <td>53</td>\n",
       "      <td>0</td>\n",
       "      <td>1892</td>\n",
       "      <td>146</td>\n",
       "      <td>0</td>\n",
       "      <td>15787750.5</td>\n",
       "      <td>0.056025</td>\n",
       "      <td>0.035819</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                name  gene  #chr     start  \\\n",
       "102   All_20_Batches.chr2.final_cleanup_DUP_chr2_537  MYCN  chr2  16113712   \n",
       "128   All_20_Batches.chr2.final_cleanup_INS_chr2_363  MYCN  chr2  16395357   \n",
       "117   All_20_Batches.chr2.final_cleanup_DUP_chr2_541  MYCN  chr2  16225124   \n",
       "22   All_20_Batches.chr2.final_cleanup_DEL_chr2_1123  MYCN  chr2  15787705   \n",
       "\n",
       "          end        AF   rare  cases_0  cases_1  cases_2  controls_0  \\\n",
       "102  16113862  0.553337  False       21      347      161         134   \n",
       "128  16395367  0.346272  False       90      438        1         790   \n",
       "117  16226720  0.858699  False        0      134      395           0   \n",
       "22   15787796  0.043883  False      420       53        0        1892   \n",
       "\n",
       "     controls_1  controls_2    midpoint  AF_cases  AF_controls  \n",
       "102        1846         118  16113787.0  0.632325     0.496187  \n",
       "128        1308           0  16395362.0  0.415879     0.311725  \n",
       "117         621        1477  16225922.0  0.873346     0.852002  \n",
       "22          146           0  15787750.5  0.056025     0.035819  "
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "freq_diff = case_control_sv_counts[\"AF_cases\"] - (case_control_sv_counts[\"AF_controls\"])\n",
    "\n",
    "case_control_sv_counts.iloc[freq_diff.sort_values(ascending=False).index].head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "356eb9dd-9059-476c-afeb-2e96e4d589ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "      <th>#chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>AF</th>\n",
       "      <th>rare</th>\n",
       "      <th>cases_0</th>\n",
       "      <th>cases_1</th>\n",
       "      <th>cases_2</th>\n",
       "      <th>controls_0</th>\n",
       "      <th>controls_1</th>\n",
       "      <th>controls_2</th>\n",
       "      <th>midpoint</th>\n",
       "      <th>AF_cases</th>\n",
       "      <th>AF_controls</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_DEL_chr2_1145</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>chr2</td>\n",
       "      <td>16131510</td>\n",
       "      <td>16133769</td>\n",
       "      <td>0.462937</td>\n",
       "      <td>False</td>\n",
       "      <td>111</td>\n",
       "      <td>361</td>\n",
       "      <td>55</td>\n",
       "      <td>343</td>\n",
       "      <td>1520</td>\n",
       "      <td>233</td>\n",
       "      <td>16132639.5</td>\n",
       "      <td>0.446869</td>\n",
       "      <td>0.473760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>All_20_Batches.chr16.final_cleanup_DUP_chr16_1131</td>\n",
       "      <td>16pmicrodel</td>\n",
       "      <td>chr16</td>\n",
       "      <td>29238380</td>\n",
       "      <td>29238527</td>\n",
       "      <td>0.040233</td>\n",
       "      <td>False</td>\n",
       "      <td>502</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>1848</td>\n",
       "      <td>207</td>\n",
       "      <td>0</td>\n",
       "      <td>29238453.5</td>\n",
       "      <td>0.021905</td>\n",
       "      <td>0.050365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_DUP_chr2_540</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>chr2</td>\n",
       "      <td>16143935</td>\n",
       "      <td>16143990</td>\n",
       "      <td>0.066043</td>\n",
       "      <td>False</td>\n",
       "      <td>512</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>1696</td>\n",
       "      <td>392</td>\n",
       "      <td>10</td>\n",
       "      <td>16143962.5</td>\n",
       "      <td>0.017013</td>\n",
       "      <td>0.098189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>All_20_Batches.chr2.final_cleanup_DEL_chr2_1147</td>\n",
       "      <td>MYCN</td>\n",
       "      <td>chr2</td>\n",
       "      <td>16146439</td>\n",
       "      <td>16146505</td>\n",
       "      <td>0.508936</td>\n",
       "      <td>False</td>\n",
       "      <td>374</td>\n",
       "      <td>134</td>\n",
       "      <td>21</td>\n",
       "      <td>129</td>\n",
       "      <td>887</td>\n",
       "      <td>1082</td>\n",
       "      <td>16146472.0</td>\n",
       "      <td>0.166352</td>\n",
       "      <td>0.727121</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  name         gene   #chr  \\\n",
       "104    All_20_Batches.chr2.final_cleanup_DEL_chr2_1145         MYCN   chr2   \n",
       "180  All_20_Batches.chr16.final_cleanup_DUP_chr16_1131  16pmicrodel  chr16   \n",
       "106     All_20_Batches.chr2.final_cleanup_DUP_chr2_540         MYCN   chr2   \n",
       "111    All_20_Batches.chr2.final_cleanup_DEL_chr2_1147         MYCN   chr2   \n",
       "\n",
       "        start       end        AF   rare  cases_0  cases_1  cases_2  \\\n",
       "104  16131510  16133769  0.462937  False      111      361       55   \n",
       "180  29238380  29238527  0.040233  False      502       23        0   \n",
       "106  16143935  16143990  0.066043  False      512       16        1   \n",
       "111  16146439  16146505  0.508936  False      374      134       21   \n",
       "\n",
       "     controls_0  controls_1  controls_2    midpoint  AF_cases  AF_controls  \n",
       "104         343        1520         233  16132639.5  0.446869     0.473760  \n",
       "180        1848         207           0  29238453.5  0.021905     0.050365  \n",
       "106        1696         392          10  16143962.5  0.017013     0.098189  \n",
       "111         129         887        1082  16146472.0  0.166352     0.727121  "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "case_control_sv_counts.iloc[freq_diff.sort_values(ascending=False).index].tail(4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dd91f9f-b149-42ea-807b-6a80a54029bd",
   "metadata": {},
   "source": [
    "## Where do these fall relative to gene regions?\n",
    "\n",
    "Presumably, the closer they are to the gene, the more we expect them to matter. We'll just map them out on the gene locuses (I like to visualize first).\n",
    "\n",
    "This is a little hamfisted, but we'll split by common SVs and then just rare."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13646e8b-454f-4942-a67b-f3c1d525c97c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pediatric-germline-svs-3.7.13",
   "language": "python",
   "name": "pediatric-germline-svs-3.7.13"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
